<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  <meta name="renderer" content="webkit">
  <meta http-equiv="X-UA-Compatible" content="IE=edge" >
  <link rel="dns-prefetch" href="http://nikolahuang.github.io">
  <title>Anderson Blog</title>
  <meta name="generator" content="hexo-theme-yilia-plus">
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  
    
    <!-- <meta name="keywords" content="">
    <meta name="description" content=""> -->
  
  <meta property="og:type" content="website">
<meta property="og:title" content="Anderson Blog">
<meta property="og:url" content="http://nikolahuang.github.io/page/2/index.html">
<meta property="og:site_name" content="Anderson Blog">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="Anderson">
<meta name="twitter:card" content="summary">
  
    <link rel="alternative" href="/atom.xml" title="Anderson Blog" type="application/atom+xml">
  
  
    <link rel="icon" href="/favicon.ico">
  
  
    <link rel="apple-touch-icon" href="/apple-touch-icon-180x180.png">
  
  <link rel="stylesheet" type="text/css" href="/./main.a5fda8.css">
  <style type="text/css">
    
    #container.show {
      background: linear-gradient(200deg,#a0cfe4,#e8c37e);
    }
  </style>
  

  

  
    
 	  <script src="/lib/clickLove.js"></script>
  
  


  
    <script src="/lib/pace.min.js"></script>
    <link href="//cdn.bootcss.com/pace/1.0.2/themes/blue/pace-theme-loading-bar.css" rel="stylesheet">
  

</head>

<body>
  <div id="container" q-class="show:isCtnShow">
    <canvas id="anm-canvas" class="anm-canvas"></canvas>
    <div class="left-col" q-class="show:isShow">
      


<div class="overlay" style="background: #4d4d4d;background: url('/img/biubiubiu.gif') no-repeat ;"></div>
<div class="intrude-less">
	<header id="header" class="inner">
		<a href="/" class="profilepic">
			<img src="/img/head.jpg" class="js-avatar">
		</a>
		<hgroup>
			<h1 class="header-author"><a href="/">Anderson</a></h1>
		</hgroup>
		

		<nav class="header-menu">
			<ul>
			
				<li><a href="/" target="_blank">主页</a></li>
			
				<li><a href="/null" target="_blank">技术笔记</a></li>
			
				<li><a href="/tags/%E9%9A%8F%E7%AC%94/" target="_blank">随笔</a></li>
			
			</ul>
		</nav>
		<nav class="header-smart-menu">
			
				
					<a q-on="click: openSlider(e, 'innerArchive')" href="javascript:void(0)">所有文章</a>
				
			
				
					<a q-on="click: openSlider(e, 'friends')" href="javascript:void(0)">友链</a>
				
			
				
					<a q-on="click: openSlider(e, 'aboutme')" href="javascript:void(0)">关于我</a>
				
			
		</nav>
		<nav class="header-nav">
			<div class="social">
				
					
						<a class="github" href="https://github.com/nikolahuang/" 
												title="GitHub" target="_blank"><i class="icon-github"></i></a>
					
				
					
						<a class="github" href="#" 
												title="GitHub" target="_blank"><i class="icon-github"></i></a>
					
				
					
						<a class="gitee" href="https://gitee.com/anderson2/" 
												title="码云" target="_blank"><i class="icon-gitee"></i></a>
					
				
					
						<a class="jianshu" href="#" 
												title="简书" target="_blank"><i class="icon-jianshu"></i></a>
					
				
			</div>
		
			
			
			
			
			<div>
				<iframe frameborder="no" border="0" marginwidth="0" marginheight="0" width="240" height="52" src="//music.163.com/outchain/player?type=2&id=1841908846&auto=1&height=32"></iframe>
			</div>
			
				
				<p style="font-size: 12px;">这似乎是首音乐，请尽情的欣赏它吧！<p>
			
		
		</nav>
	</header>		
</div>

    </div>
    <div class="mid-col" q-class="show:isShow,hide:isShow|isFalse">
      
      
      <a class="forkMe" style="position:absolute;z-index:999;top:0;right:0.5em;" 
        href="https://github.com/nikolahuang/" target="_blank">
        <img src="/img/forkme.png"
          class="attachment-full size-full" alt="Fork me on GitHub" data-recalc-dims="1"></a>
      
      
<nav id="mobile-nav">
  	<div class="overlay js-overlay" style="background: #4d4d4d"></div>
	<div class="btnctn js-mobile-btnctn">
  		<div class="slider-trigger list" q-on="click: openSlider(e)"><i class="icon icon-sort"></i></div>
	</div>
	<div class="intrude-less">
		<header id="header" class="inner">
			<div class="profilepic">
				<a href="/">
					<img src="/img/head.jpg" class="js-avatar">
				</a>
			</div>
			<hgroup>
			  <h1 class="header-author js-header-author">Anderson</h1>
			</hgroup>
			
			
			
				
			
				
			
				
			
			
			
			<nav class="header-nav">
				<div class="social">
					
						
						<a class="github" target="_blank" 
							href="https://github.com/nikolahuang/" title="GitHub"><i class="icon-github"></i></a>
						
					
						
						<a class="github" target="_blank" 
							href="#" title="GitHub"><i class="icon-github"></i></a>
						
					
						
						<a class="gitee" target="_blank" 
							href="https://gitee.com/anderson2/" title="码云"><i class="icon-gitee"></i></a>
						
					
						
						<a class="jianshu" target="_blank" 
							href="#" title="简书"><i class="icon-jianshu"></i></a>
						
					
				</div>
			</nav>

			<nav class="header-menu js-header-menu">
				<ul style="width: 70%">
				
				
					<li style="width: 33.333333333333336%"><a href="/">主页</a></li>
		        
					<li style="width: 33.333333333333336%"><a href="/null">技术笔记</a></li>
		        
					<li style="width: 33.333333333333336%"><a href="/tags/%E9%9A%8F%E7%AC%94/">随笔</a></li>
		        
				</ul>
			</nav>
		</header>				
	</div>
	<div class="mobile-mask" style="display:none" q-show="isShow"></div>
</nav>

      <div id="wrapper" class="body-wrap">
        <div class="menu-l">
          <div class="canvas-wrap">
            <canvas data-colors="#eaeaea" data-sectionHeight="100" data-contentId="js-content" id="myCanvas1"
              class="anm-canvas"></canvas>
          </div>
          <div id="js-content" class="content-ll">
            
  
    <script src="/lib/oauthcallback.browser.js"></script>
  
  <script>
    Oauthcallback();
  </script>


  
    <article id="post-表达载体的构建方法及步骤" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/18/%E8%A1%A8%E8%BE%BE%E8%BD%BD%E4%BD%93%E7%9A%84%E6%9E%84%E5%BB%BA%E6%96%B9%E6%B3%95%E5%8F%8A%E6%AD%A5%E9%AA%A4/" target="_blank">表达载体的构建方法及步骤</a>
  
    </h1>
  


  
  
<a href="/2024/03/18/%E8%A1%A8%E8%BE%BE%E8%BD%BD%E4%BD%93%E7%9A%84%E6%9E%84%E5%BB%BA%E6%96%B9%E6%B3%95%E5%8F%8A%E6%AD%A5%E9%AA%A4/" class="archive-article-date">
        <time datetime="2024-03-18T15:08:00.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-18</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <p><strong>一、载体的选择及如何阅读质粒图谱</strong><br>目前，载体主要有病毒和非病毒两大类,其中质粒 DNA 是一种新的非病毒转基因载体。<br>一个合格质粒的组成要素：<br>（1）复制起始位点 Ori 即控制复制起始的位点。原核生物 DNA 分子中只有一个复制起始点。而<br>真核生物 DNA 分子有多个复制起始位点。<br>（2）抗生素抗性基因可以便于加以检测，如 Amp+ ,Kan+<br>（3）多克隆位点 MCS 克隆携带外源基因片段<br>（4） P&#x2F;E 启动子&#x2F;增强子<br>（5）Terms 终止信号<br>（6）加 poly（A）信号可以起到稳定 mRNA 作用<br>选择载体主要依据构建的目的，同时要考虑载体中应有合适的限制酶切位点。如果构建的目<br>的是要表达一个特定的基因，则要选择合适的表达载体。<br>载体选择主要考虑下述3点：<br>【1】.构建 DNA 重组体的目的，克隆扩增&#x2F;基因表达，选择合适的克隆载体&#x2F;表达载体。<br>【2】.载体的类型：<br>（1）克隆载体的克隆能力－据克隆片段大小（大选大，小选小）。如&lt;10kb 选质粒。<br>（2）表达载体据受体细胞类型－原核&#x2F;真核&#x2F;穿梭，E.coli&#x2F;哺乳类细胞表达载体。<br>（3）对原核表达载体应该注意：选择合适的启动子及相应的受体菌，用于表达真核蛋白质时注意克服4个困难和阅读框错位；表达天然蛋白质或融合蛋白作为相应载体的参考。<br>【3】载体 MCS 中的酶切位点数与组成方向因载体不同而异，适应目的基因与载体易于链接，不能产生阅读框架错位。<br>综上所述，选用质粒（最常用）做载体的5点要求：<br>（1）选分子量小的质粒，即小载体（1－1.5kb）→不易损坏，在细菌里面拷贝数也多（也有大载<br>体）；<br>（2）一般使用松弛型质粒在细菌里扩增不受约束，一般 10个以上的拷贝，而严谨型质粒&lt;10个。<br>（3）必需具备一个以上的酶切位点，有选择的余地；<br>（4）必需有易检测的标记，多是抗生素的抗性基因，不特指多位 Ampr（试一试）。<br>（5）满足自己的实验需求，是否需要包装病毒，是否需要加入荧光标记，是否需要加入标签蛋白，是否需要真核抗性（如Puro、G418）等等。<br>无论选用哪种载体，首先都要获得载体分子，然后采用适当的限制酶将载体 DNA 进行切割，获得线性载体分子，以便于与目的基因片段进行连接。<br>如何阅读质粒图谱<br>第一步：首先看 Ori 的位置，了解质粒的类型（原核&#x2F;真核&#x2F;穿梭质粒）<br>第二步：再看筛选标记，如抗性，决定使用什么筛选标记。<br>（1）Ampr 水解β－内酰胺环，解除氨苄的毒性。<br>（2）tetr 可以阻止四环素进入细胞。<br>（3）camr 生成氯霉素羟乙酰基衍生物，使之失去毒性。<br>（4）neor（kanr）氨基糖苷磷酸转移酶使 G418（长那霉素衍生物）失活<br>（5）hygr 使潮霉素β失活。<br>第三步：看多克隆位点（MCS）。它具有多个限制酶的单一切点。便于外源基因的插入。如果在这些酶切位点以外有外源基因的插入，会导致某种标志基因的失活，而便于筛选。决定能不能放目的基因以及如何放置目的基因。<br>第四步：再看外源 DNA 插入片段大小。质粒一般只能容纳小于10Kb 的外源 DNA 片段。一般来说，外源 DNA 片段越长，越难插入，越不稳定，转化效率越低。<br>第五步：是否含有表达系统元件，即启动子－核糖体结合位点－克隆位点－转录终止信号。这是用来区别克隆载体与表达载体。克隆载体中加入一些与表达调控有关的元件即成为表达载体。选用那种载体，还是要以实验目的为准绳。<br>第六步：启动子－核糖体结合位点－克隆位点－转录终止信号<br>（1）启动子－促进 DNA 转录的 DNA 序列，这个 DNA 区域常在基因或操纵子编码序列的上游，是 DNA 分子上可以与 RNApol 特异性结合并使之开始转录的部位，但启动子本身不被转录。<br>（2）增强子&#x2F;沉默子－为真核基因组（包括真核病毒基因组）中的一种具有增强邻近基因转录过程的调控顺序。其作用与增强子所在的位置或方向无关。即在所调控基因上游或下游均可发挥作用。&#x2F;沉默子－负增强子，负调控序列。<br>（3）核糖体结合位点&#x2F;起始密码&#x2F;SD 序列（Rbs&#x2F;AGU&#x2F;SDs）：mRNA 有核糖体的两个结合位点，对于原核而言是 AUG（起始密码）和 SD 序列。<br>（4）转录终止序列（终止子）&#x2F;翻译终止密码子：结构基因的最后一个外显子中有一个 AATAAA的保守序列，此位点 down－stream 有一段 GT 或 T 富丰区，这2部分共同构成 poly（A）加尾信号。结构基因的最后一个外显子中有一个 AATAAA 的保守序列，此位点 down－stream 有一段GT 或 T 富丰区，这2部分共同构成 poly（A）加尾信号。<br>质粒图谱上有的箭头顺时针有的箭头逆时针，那其实是代表两条 DNA 链，即质粒是环状双链DNA，它的启动子等在其中一条链上，而它的抗性基因在另一条链上 .根据表达宿主不同，构建时所选择的载体也会不同。</p>
<p><strong>二、目的基因的获得</strong><br>一般来说，目的基因的获得有三种途径：<br>调取基因：根据目的基因的序列，设计引物从含有目的基因的cDNA中通过PCR的方法调取目的基因，链接到克隆载体挑取单克隆进行测序，以获得想要的基因片段，这种方法相对成本较低，但是调取到的基因往往含有突变，还有不同基因的表达丰度不同，转录本比较复杂，或是基因片段很长，这些情况都很难调取到目的基因。<br>全基因合成：根据目的基因的DNA序列，直接设计合成目的基因。此方法准确性高，相对成本会高一些，个人操作比较困难，需要专业的合成公司完成。优点是可以合成难调取及人工改造的任何基因序列，同时可以进行密码子优化，提高目的基因在宿主内的表达量。</p>
<p><strong>三、克隆构建</strong><br>目前，克隆构建的方法多种多样，除了应用广泛的酶切链接以外，现在还有很多不依赖酶切位点的克隆构建方式。下面具体说一下双酶切方法构建载体的步骤。<br>实验材料</p>
<p>实验试剂</p>
<table>
<thead>
<tr>
<th>试剂名称</th>
<th>生产厂家</th>
</tr>
</thead>
<tbody><tr>
<td>载体pCDNA3.1</td>
<td>Transheep</td>
</tr>
<tr>
<td>大肠杆菌菌株DH5α</td>
<td>Tiangen</td>
</tr>
<tr>
<td>限制性内切酶</td>
<td>Fermentas</td>
</tr>
<tr>
<td>T4连接酶</td>
<td>Fermentas</td>
</tr>
<tr>
<td>质粒DNA小,大量抽提试剂盒</td>
<td>Axygen</td>
</tr>
<tr>
<td>凝胶回收试剂盒</td>
<td>Axygen</td>
</tr>
<tr>
<td>琼脂糖</td>
<td>Biowest</td>
</tr>
<tr>
<td>DNA ladder</td>
<td>Fermentas</td>
</tr>
</tbody></table>
<p>（2）X基因慢病毒载体的构建<br>X基因基因由Transheep全基因合成，构建于载体PUC57中。<br>PUC57-X基因 EcoRI&#x2F;BamHI酶切结果：</p>
<p><img src="https://assets.dxycdn.com/third-party/xheditor/xheditor_skin/blank.gif" alt="img"></p>
<p><img src="https://assets.dxycdn.com/third-party/xheditor/xheditor_skin/blank.gif" alt="img"></p>
<p>酶切完成后进行胶回收<br>\2. 载体用pCDNA3.1双酶切，酶切体系如下。<br>20ul酶切体系 37度3小时<br>4ul pCDNA3.1载体（500 ng&#x2F;ul）<br>1ul BamHI<br>1ul EcoRI<br>2ul 10×buffer<br>12 ul H2O<br>酶切完成后胶回收（见附录）</p>
<p><img src="https://assets.dxycdn.com/third-party/xheditor/xheditor_skin/blank.gif" alt="img"><img src="https://assets.dxycdn.com/third-party/xheditor/xheditor_skin/blank.gif" alt="img"></p>
<p>处理好的目的片段与载体连接反应体系：<br>6ul PCR 酶切回收片段（约50ng&#x2F;ul）<br>1ul 酶切好的载体（约50ng&#x2F;ul）<br>2ul ligase buffer<br>1ul T4ligase<br>10ul H2O<br>以上连接液在16℃过夜。<br>转化 (感受态细胞: DH5a),具体步骤见附录转化部分。<br>抗性: Amp; 37℃，培养过夜<br>转化后X基因分别平板挑菌, 37℃ 250转&#x2F;分钟摇菌14小时，PCR鉴定后，将阳性菌液送上海权阳生物技术有限公司测序。<br>X基因慢病毒载体单克隆菌落PCR鉴定（使用载体通用引物,PCR条带大小应比实际大200bp左右）：</p>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/18/%E8%A1%A8%E8%BE%BE%E8%BD%BD%E4%BD%93%E7%9A%84%E6%9E%84%E5%BB%BA%E6%96%B9%E6%B3%95%E5%8F%8A%E6%AD%A5%E9%AA%A4/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-蛋白质组学数据分析和展示" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/17/%E8%9B%8B%E7%99%BD%E8%B4%A8%E7%BB%84%E5%AD%A6%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E5%92%8C%E5%B1%95%E7%A4%BA/" target="_blank">蛋白质组学数据分析和展示</a>
  
    </h1>
  


  
  
<a href="/2024/03/17/%E8%9B%8B%E7%99%BD%E8%B4%A8%E7%BB%84%E5%AD%A6%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E5%92%8C%E5%B1%95%E7%A4%BA/" class="archive-article-date">
        <time datetime="2024-03-17T09:03:15.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-17</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <p>[TOC]</p>
<p>使用Maxquant对原始谱图搜库完成后，可以得到大量蛋白质的鉴定和定量信息。下一步就需要通过统计分析和生物信息学分析，从搜库得到的大量信息中筛选和寻找关键蛋白和信号通路，以及进行相关功能的验证。本讲我们会为大家介绍如何进行蛋白质组学的定量数据分析以及结果展示，主要包括：蛋白质鉴定和定量数据统计、差异表达蛋白筛选、功能注释和富集分析以及蛋白互作网络分析。</p>
<h3 id="定量数据统计和展示"><a href="#定量数据统计和展示" class="headerlink" title="定量数据统计和展示"></a><strong>定量数据统计和展示</strong></h3><p>在蛋白质组学数据分析过程中，首先会对所有的鉴定和定量数据进行统计和展示，主要关注蛋白鉴定数目、肽段鉴定数目、肽段分布情况以及样品相关性和重复性等内容，展示蛋白质组学数据的定量深度、可信度和重复性等。</p>
<p>如下图所示，蛋白质组定量深度主要体现在各个样品中鉴定及定量到的蛋白和肽段数量，数量越多，表明定量深度越深，意味着能够获得更全面的蛋白质组学信息。同时，不同样品间定量深度的差异，也可以初步反映样品和方法的重复性和稳定性。</p>
<p><img src="https://pic3.zhimg.com/80/v2-6d39941370abb6c9338a6034c662b122_1440w.webp" alt="img"></p>
<p>（肽段数量统计图（上）和蛋白数量统计图（下）[1]）</p>
<p>蛋白质组学数据的可信度，可通过各个蛋白鉴定到的肽段数量以及各条肽段鉴定到的谱图数量分别展示。一般认为鉴定到2条及以上肽段的蛋白或2张及以上谱图的肽段为可信度更高的鉴定结果。</p>
<p><img src="https://pic3.zhimg.com/80/v2-6f67c0ec5eb6fbaf6dbfc9cf5407c916_1440w.webp" alt="img"></p>
<p>（谱图数量分布图（上）&#x2F;肽段数量分布图（下） [1]）</p>
<p>蛋白质组定量数据的相关性和重复性可通过相关系数来体现，组内样品间的蛋白质组定量相关系数往往较高，而组间样品的相关系数则会稍低（如左图）。PCA分析则是通过降维算法，将数百乃至上万个蛋白的定量数据降维为2个或3个反映定量特征的主成分，从而直观展示不同样品之间的定量相关性，在空间位置上接近的点表明其定量特征接近，也就是数据相关性更高。同一组的样品在PCA分析结果中往往会更为聚集（如右图）。</p>
<p><img src="https://pic1.zhimg.com/80/v2-d51216dbd0bc17a7c9b47261d4ac4490_1440w.webp" alt="img"></p>
<p>（相关系数热图（左）和PCA分析图（右）[2]）</p>
<p>定量热图也是蛋白质组学中常用的展示定量数据的结果图形。在定量热图中，通常每一列代表一个样品，每一行代表一个蛋白，以色块颜色表示各个样品中蛋白的定量水平。通过颜色的变化和分布可以直观的展示不同样品中蛋白定量水平的相关性和差异性。</p>
<p><img src="https://pic2.zhimg.com/80/v2-fa5113e8db515018b2c6cb37dbd261b9_1440w.webp" alt="img"></p>
<p>（蛋白定量热图[3]）</p>
<h3 id="差异表达分析"><a href="#差异表达分析" class="headerlink" title="差异表达分析"></a><strong>差异表达分析</strong></h3><p>找出不同处理组间差异表达蛋白是蛋白质组学研究的重要目标之一。统计学上一般选取P值（P-value）和差异倍数（Fold change）两个参数实现差异蛋白的筛选，其中P-value由比较两组样品蛋白定量数据的T检验计算得出，要求各组样品数量不少于3个；Fold change是两样品中同一个蛋白的表达差异倍数，主要根据蛋白定量值计算。如下图所示，根据显著性P&lt;0.05和 Fold Change&gt;1.5或&lt;1&#x2F;1.5设定阈值可挑选出符合标准的差异蛋白质。</p>
<p><img src="https://pic4.zhimg.com/80/v2-391ee040bee2210b184a2c91e62b7893_1440w.webp" alt="img"></p>
<p>（差异蛋白火山图[4]）</p>
<p>筛选出差异表达蛋白后，也可再次通过定量热图与PCA分析图，更为直观地展示这些差异蛋白的定量特征，以及各样本组内及组间的相似和变异程度。</p>
<h3 id="功能注释和富集分析"><a href="#功能注释和富集分析" class="headerlink" title="功能注释和富集分析"></a><strong>功能注释和富集分析</strong></h3><p>以上分析完成后，有些实验还需要对蛋白的功能、分布、参与的通路等方面进行注释和富集，可为后续生物学和功能机制的研究提供参考和线索。</p>
<p>蛋白功能注释是将生物学信息关联到蛋白质序列的过程，使用的数据库主要有GO(gene ontology）、KEGG(Kyoto encyclopedia of genes and genomes)、COG(Clusters of Orthologous Groups of proteins)、HallmarkGeneset、Reactome等。以GO分析为例，可理解为将蛋白对应的基因分门别类地放入一个个功能条目的盒子中去，从功能（Molecular Function）、参与的生物途径（Biological Process）和细胞中的定位（Cellular Component）三个维度来较为全面地描述一个蛋白。如下图GO注释，饼图的每一个分区表示一个GO条目，分区大小表示本条目中蛋白数量在差异蛋白中的占比。</p>
<p><img src="https://pic1.zhimg.com/80/v2-8ca44ab2fd975cfc805711b88cb7c800_1440w.webp" alt="img"></p>
<p>（GO注释饼状图[5] ）</p>
<p>富集分析则是在功能注释的基础上的进一步分析，其主要目的是以背景蛋白为参照，分析差异蛋白较为集中分布的注释条目，从而筛选出与差异蛋白更相关、更可能具有生物学意义的功能注释条目。目前最常用的富集分析算法是基于超几何分布的Fisher精确检验。</p>
<p>富集分析通常以P&lt;0.05作为条目显著富集的筛选阈值。富集得到的注释条目可通过如下柱状图进行展示。其中GO富集分析通常会以不同颜色表示GO条目所属大类。</p>
<p><img src="https://pic1.zhimg.com/80/v2-337b34639e4ba370168261dffd1c8794_1440w.webp" alt="img"></p>
<p><img src="https://pic1.zhimg.com/80/v2-57c2e0b655f6116cb3d0092fc6ad3e98_1440w.webp" alt="img"></p>
<p>（GO富集柱状图（上）和KEGG富集柱状图（下）[6]）</p>
<p>气泡图也是富集分析常见图形展示，相比柱状图可以通过气泡的横坐标、颜色、大小展示更多维度的富集参数，包括E-ratio（富集倍数）、p值和条目中蛋白数量等。其中富集倍数越大、p值越小说明该条目在差异蛋白中富集程度越显著。</p>
<p><img src="https://pic2.zhimg.com/80/v2-b68e4e71aab7f9d68310089014ed8f31_1440w.webp" alt="img"></p>
<p>（GO富集气泡图（左）和KEGG富集气泡图（右）[1]）</p>
<h3 id="互作网络分析"><a href="#互作网络分析" class="headerlink" title="互作网络分析"></a><strong>互作网络分析</strong></h3><p>在生物体内，蛋白质并非单独发挥作用，其功能的行使须依赖于其他蛋白的调控和介导，这种调控和介导要求蛋白质之间存在直接或间接的相互作用。</p>
<p>STRING（Search Tool for the Retrieval of Interaction Gene&#x2F;Proteins）数据库是目前使用最广泛、信息最全面的蛋白互作网络分析工具，既包含了已有报道的蛋白质互作数据，也包含了基于预测的蛋白质互作信息。文章中一般以PPI（Protein‑Protein Interaction）互作网络图进行展示，每个节点为一个差异蛋白，节点之间的连线表示两个蛋白之间存在已知的或预测可能存在的蛋白相互作用，节点颜色表示该蛋白在比较组中的定量信息，如下图所示。PPI互作网络可以直观的展示蛋白间的相互关联，并进一步帮助我们寻找其中的关键节点。</p>
<p><img src="https://pic4.zhimg.com/80/v2-1d8530ad5f6fd4601149dbaa9211b53f_1440w.webp" alt="img"></p>
<p>（蛋白互作网络图[7]）</p>
<p><strong>总结</strong></p>
<p>蛋白质组学的数据分析通常包含以下内容：通过比较不同蛋白质在不同样本中的表达情况，找到差异表达蛋白；通过功能注释和富集分析，找到差异蛋白显著富集的生物学功能和信号通路；对差异表达蛋白进行蛋白互作网络分析，展示差异蛋白间的相互关联，从而寻找网络中的关键节点。这些分析能够帮助我们筛选获得关键蛋白、生物学功能和信号通路，对于深度解析蛋白质组学数据，指导后续功能研究等具有重要意义。</p>
<p><strong>参考文献</strong></p>
<ol>
<li>Shu T, Ning W, Wu D, et al. Plasma Proteomics Identify Biomarkers and Pathogenesis of COVID-19[J]. Immunity, 2020, 53(5): 1108-1122</li>
<li>Lin Y H, Platt M P, Fu H, et al. Global Proteome and Phosphoproteome Characterization of Sepsis-induced Kidney Injury[J]. Mol Cell Proteomics, 2020, 19(12): 2030-2047.</li>
<li>Zhang Q, Lou Y, Yang J, et al. Integrated multiomic analysis reveals comprehensive tumour heterogeneity and novel immunophenotypic classification in hepatocellular carcinomas[J]. Gut, 2019, 68(11): 2019-2031.</li>
<li>Mohanty V, Subbannayya Y, Patil S, et al. Molecular alterations in oral cancer using high-throughput proteomic analysis of formalin-fixed paraffin-embedded tissue[J]. Journal of cell communication and signaling, 2021, 15(3): 447-459.</li>
<li>Wen L, Liu Y F, Jiang C, et al. Comparative Proteomic Profiling and Biomarker Identification of Traditional Chinese Medicine-Based HIV&#x2F;AIDS Syndromes[J]. Sci Rep, 2018, 8(1): 4187-4197.</li>
<li>Zhou F, Luo Q, Han L, et al. Proteomics reveals urine apolipoprotein A-I as a potential biomarker of acute kidney injury following percutaneous coronary intervention in elderly patients[J]. Exp Ther Med, 2021, 22(1): 745-755.</li>
<li>Ji F, Zhou M, Sun Z, et al. Integrative proteomics reveals the role of E3 ubiquitin ligase SYVN1 in hepatocellular carcinoma metastasis[J]. Cancer Commun (Lond), 2021, 41(10): 1007-1023.</li>
</ol>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
	<div class="article-tag tagcloud">
		<i class="icon-price-tags icon"></i>
		<ul class="article-tag-list">
			 
        		<li class="article-tag-list-item">
        			<a href="javascript:void(0)" class="js-tag article-tag-list-link color3">BioInfo</a>
        		</li>
      		
		</ul>
	</div>

    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/17/%E8%9B%8B%E7%99%BD%E8%B4%A8%E7%BB%84%E5%AD%A6%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E5%92%8C%E5%B1%95%E7%A4%BA/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-WES数据分析流程-下" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/17/WES%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B-%E4%B8%8B/" target="_blank">WES数据分析流程(下)--可视化</a>
  
    </h1>
  


  
  
<a href="/2024/03/17/WES%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B-%E4%B8%8B/" class="archive-article-date">
        <time datetime="2024-03-17T06:53:33.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-17</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <p>[TOC]</p>
<p>转自<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/380684876">Die Linke</a></p>
<p>MyNotes：</p>
<p>总体概括为：全外显子测序的标准分析流程通常包括以下步骤：</p>
<p><strong>质量控制（Quality Control，QC）</strong></p>
<p>对原始测序数据进行质量控制，包括去除低质量的序列片段、过滤掉低质量的碱基和低覆盖度的测序片段等。</p>
<p><strong>读取比对（Read Alignment）</strong></p>
<p>将过滤后的测序数据比对到参考基因组上，以确定每个片段的位置和方向。</p>
<p><strong>变异检测（Variant Calling）</strong></p>
<p>通过对比测序数据和参考基因组，识别出样本中存在的单核苷酸多态性（Single Nucleotide Polymorphism，SNP）和插入&#x2F;删除多态性（Insertion&#x2F;Deletion，InDel）等变异。</p>
<p><strong>注释（Annotation）</strong></p>
<p>对检测到的变异进行注释，包括位置信息、功能影响、已知疾病关联等。</p>
<p><strong>过滤（Filtering）</strong></p>
<p>根据预设的过滤标准，过滤掉可能为假阳性的变异，如常见的多态性、测序误差等。</p>
<p><strong>数据解读（Data Interpretation）</strong></p>
<p>将过滤后的变异进行进一步分析和解读，以确定与研究对象相关的变异。</p>
<p><strong>功能验证（Functional Validation）</strong></p>
<p>通过实验验证已识别的变异是否与疾病相关，例如验证变异对基因表达、蛋白质功能的影响等。</p>
<hr>
<h3 id="分析流程"><a href="#分析流程" class="headerlink" title="分析流程"></a>分析流程</h3><ol>
<li><h4 id="准备工作"><a href="#准备工作" class="headerlink" title="准备工作:"></a>准备工作:</h4></li>
<li><h5 id="数据上传"><a href="#数据上传" class="headerlink" title="数据上传:"></a>数据上传:</h5></li>
</ol>
<p>从本地硬盘通过<strong>FileZilla软件</strong>或在云端通过<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=scp%E5%8D%8F%E8%AE%AE&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">scp协议</a>将所需数据传输至进行分析的服务器.</p>
<ul>
<li>FileZilla示例:</li>
</ul>
<p>如下图所示，在窗口拖动数据文件到指定目录即可。</p>
<p><img src="https://pic1.zhimg.com/v2-8111d0db8f7f9825dd9b1e50e8838f78_b.jpg" alt="img"></p>
<ul>
<li>scp协议示例:</li>
</ul>
<p>命令行输入以下命令进行文件传输。</p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">scp -r local_folder remote_user_name@remote_ip:remote_folder</span><br><span class="line">#-r表示递归传出路径下所有文件</span><br><span class="line">#local_folder为传出的本地文件夹</span><br><span class="line">#remote_user_name为接收文件的远程用户名</span><br><span class="line">#remote_ip为接收文件的远程用户所在的ip地址</span><br><span class="line"># remote_folder为远程用户下储存接收文件的文件夹</span><br></pre></td></tr></table></figure>

<ol start="2">
<li><h5 id="软件安装"><a href="#软件安装" class="headerlink" title="软件安装:"></a>软件安装:</h5></li>
</ol>
<ul>
<li>先在服务器上安装conda, conda的安装使用请参照之前的文章<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/174802525">生科院的老张：无root权限conda安装PyRosetta同源建模</a>，安装好后通过conda进一步安装软件<strong>fastqc, multiqc, trimmomatic, bwa, samtools, bcftools, vcftools, gffread</strong>, 参考代码如下:</li>
</ul>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda create -n WES_WGS_WGRS #新建一个名为WES_WGS_WGRS的conda环境</span><br><span class="line">conda activate WES_WGS_WGRS#激活进入新建的WES_WGS_WGRS环境</span><br><span class="line">conda install fastqc multiqc trimmomatic bwa samtools bcftools vcftools gffread #在该环境下安装fastqc等软件</span><br></pre></td></tr></table></figure>

<ul>
<li>如下图所示，<strong>ANNOVAR</strong>软件需要在官网 <a href="https://link.zhihu.com/?target=https://annovar.openbioinformatics.org/en/latest/user-guide/download/">Download ANNOVAR - ANNOVAR Documentation</a> 通过教育邮箱申请下载链接下载。</li>
</ul>
<p><img src="https://pic2.zhimg.com/v2-93f49395e5d0c2147f7287f37e7b7485_b.jpg" alt="img"></p>
<ul>
<li>通过wget下载安装gtfToGenePred:</li>
</ul>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wget http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64.v369/gtfToGenePred</span><br><span class="line">mkdir -p /path/to/save/gtfToGenePred/bin #创建gtfToGenePred的储存路径并创建bin目录</span><br><span class="line">cp gtfToGenePred /path/to/save/gtfToGenePred/bin #将gtfToGenePred移入储存路径下的bin目录</span><br><span class="line">chmod +x /path/to/save/gtfToGenePred/bin/gtfToGenePred #赋予其读写执行全部权限</span><br></pre></td></tr></table></figure>

<ul>
<li>在R中安装stringr, circlize, grid, ComplexHeatmap, 需先安装Bioconductor, 详见之前的文章<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/345896914">生科院的老张：RNAseq转录组差异表达分析教程</a>:</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&#x27;stringr&#x27;</span><span class="punctuation">)</span></span><br><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&#x27;circlize&#x27;</span><span class="punctuation">)</span></span><br><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&#x27;grid&#x27;</span><span class="punctuation">)</span></span><br><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&#x27;ComplexHeatmap&#x27;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<ol start="3">
<li><h5 id="参考基因组和基因组注释文件"><a href="#参考基因组和基因组注释文件" class="headerlink" title="参考基因组和基因组注释文件:"></a>参考基因组和<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%B3%A8%E9%87%8A&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">基因组注释</a>文件:</h5></li>
</ol>
<p>在对应的网站下载参考基因组和基因组注释文件, 如在NCBI的genome搜索<strong>ZM4</strong>能找到我要的物种对应的<strong>reference genome</strong>和<strong>annotation</strong>. </p>
<p>如图所示，点击红色方框内的genome和GFF会分别下载到.fasta格式的参考<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%BA%8F%E5%88%97&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">基因组序列</a>和.gff格式的基因组注释。<strong>下方的基因组全长可以记录下来，在之后的分析中会使用到</strong>。</p>
<p><img src="https://pic4.zhimg.com/v2-f1c0fc836b2974bb1e039a656ad3cc17_b.jpg" alt="img"></p>
<hr>
<ol start="2">
<li><h4 id="质量控制"><a href="#质量控制" class="headerlink" title="质量控制:"></a>质量控制:</h4></li>
</ol>
<p>任何组学分析第一步都是对测序数据进行质量控制，细节可以参考之前的转录组教程 <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/345896914">生科院的老张：RNAseq转录组差异表达分析教程</a> , 和转录组一样，使用md5检查数据完整性，fastqc进行质量检测，multiqc合并报告，trimmomatic进行质量过滤,参考代码如下：</p>
<ol>
<li><h5 id="md5检查数据完整性："><a href="#md5检查数据完整性：" class="headerlink" title="md5检查数据完整性："></a>md5检查数据完整性：</h5></li>
</ol>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">md5sum *gz &gt; md5.txt &amp;&amp; md5sum -c md5.txt</span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">*gz 任意以gz结尾的文件</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">&gt; 运行结果保存至</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">&amp;&amp; 连接符，前一命令完成继续执行下一命令</span></span><br></pre></td></tr></table></figure>

<p>2.2.2 fastqc质量控制与multiqc合并质控报告:</p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">fastqc *gz &amp;&amp; multiqc ./</span><br><span class="line"># 对所有gz结尾文件进行质控并对当前目录下所有质控报告进行合并</span><br></pre></td></tr></table></figure>

<p>2.2.3 trimmomatic质量过滤:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">trimmomatic </span><br><span class="line">PE #双端测序，单端测序为SE</span><br><span class="line">-threads 4 #指定线程数为4</span><br><span class="line">~/WES_WGS_WGRS/data/sample1/sample1_1.fq.gz #输入序列</span><br><span class="line">~/WES_WGS_WGRS/data/sample1/sample1_2.fq.gz</span><br><span class="line">~/WES_WGS_WGRS/data/sample1/sample1_paired_clean_1.fq.gz #输出配对序列和非配对序列</span><br><span class="line">~/WES_WGS_WGRS/data/sample1/sample1_unpair_clean_1.fq.gz </span><br><span class="line">~/WES_WGS_WGRS/data/sample1/sample1_paired_clean_2.fq.gz</span><br><span class="line">~/WES_WGS_WGRS/data/sample1/sample1_unpair_clean_2.fq.gz</span><br><span class="line">ILLUMINACLIP:/data1/guest/yinlei/miniconda3/pkgs/trimmomatic-0.39-1/share/trimmomatic/adapters/TruSeq3-PE-2.fa:2:30:10:1:true </span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">去除ILLUMINA接头，根据质控报告选择trimmomatic文件夹adapters路径下的接头文件</span></span><br><span class="line">LEADING:3 #从reads开头切除质量低于阈值3的碱基</span><br><span class="line">TRAILING:3 #从reads末尾切除质量低于阈值3的碱基</span><br><span class="line">SLIDINGWINDOW:4:20 #从reads 5‘端开始进行长度为4的滑窗过滤，切除碱基质量低于阈值20的碱基</span><br><span class="line">MINLEN:50 #丢弃剪切后长度低于阈值50的reads</span><br><span class="line">TOPHRED33 #将reads的碱基质量值体系转为phred-33，若为phred-64则为TOPHRED64</span><br></pre></td></tr></table></figure>

<p>由于一批测序数据中有多个测序文件，因此需要循环以上过程，参考代码如下图所示：</p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">for dir in A B C D; #循环读取A,B,C,D四个目录</span><br><span class="line">do XXXX; #执行XXXX</span><br><span class="line">done #结束命令</span><br></pre></td></tr></table></figure>

<p><img src="https://pic4.zhimg.com/v2-9703758821f41b86bfa57ed16c315d27_b.jpg" alt="img"></p>
<p>trimmomatic质量过滤</p>
<hr>
<ol start="3">
<li><h4 id="bwa序列比对"><a href="#bwa序列比对" class="headerlink" title="bwa序列比对:"></a>bwa序列比对:</h4></li>
</ol>
<p>将经过质量控制的序列与<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">参考基因组</a>进行比对，参考代码如下：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">#</span><span class="language-bash">建立比对索引</span></span><br><span class="line">bwa index genome.fa</span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">将序列与基因组进行比对</span></span><br><span class="line">bwa mem -t 16 genome.fa sample1_paired_clean_1.fq.gz sample1_paired_clean_2.fq.gz &gt; sample1.sam</span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">参数如下：</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">mem 比对算法</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">-t 16 指定线程数为16</span></span><br></pre></td></tr></table></figure>

<p>同样由于一批测序数据中有多个测序文件，因此需要循环以上过程，参考代码如下图所示：</p>
<p><img src="https://pic1.zhimg.com/v2-b216223c5c2756b90f70edaf521b5f28_b.png" alt="img"></p>
<p>bwa比对基因组</p>
<hr>
<ol start="4">
<li><h4 id="samtools去除PCR重复后排序："><a href="#samtools去除PCR重复后排序：" class="headerlink" title="samtools去除PCR重复后排序："></a>samtools去除PCR重复后排序：</h4></li>
</ol>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">#</span><span class="language-bash">将比对结果sample1.sam转为.bam格式</span></span><br><span class="line">samtools view -bS sample1.sam &gt; sample1.bam</span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">参数如下:</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">-bS 表示输出bam文件，输入为sam文件</span></span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">去除PCR重复</span></span><br><span class="line">samtools rmdup -s sample1.bam sample1.rm.bam</span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">排序</span></span><br><span class="line">samtools sort -@ 16 sample1.rm.bam &gt; sample1.sorted.bam</span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">参数如下：</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">-@ 16 指定线程数为16</span></span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">建立索引</span></span><br><span class="line">samtools index sample1.sorted.bam</span><br></pre></td></tr></table></figure>

<p>同样由于一批测序数据中有多个测序文件，因此需要循环以上过程，参考代码如下图所示：</p>
<p><img src="https://pic4.zhimg.com/v2-4f78fbccb367a64b7a57531bd386ce1b_b.png" alt="img"></p>
<p>samtools去除PCR重复，排序和建立索引</p>
<hr>
<ol start="5">
<li><h4 id="samtools进行SNP-calling："><a href="#samtools进行SNP-calling：" class="headerlink" title="samtools进行SNP calling："></a>samtools进行SNP calling：</h4></li>
</ol>
<p>得到基因组比对的去重复排序结果后，我们将比对信息中SNP的信息提取出来并整理成通用的突变格式VCF.</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">#</span><span class="language-bash">SNP calling</span></span><br><span class="line">samtools mpileup -gSDf genome.fa sample1.sorted.bam sample1.bcf</span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">将SNP鉴定结果sample1.bcf转为vcf格式</span></span><br><span class="line">bcftools view -cvNg sample1.bcf sample1.vcf</span><br></pre></td></tr></table></figure>

<p>同样由于一批测序数据中有多个测序文件，因此需要循环以上过程，参考代码如下图所示：</p>
<p><img src="https://pic3.zhimg.com/v2-3804e06f71efc568efbefce3e3786d9e_b.png" alt="img"></p>
<p>samtools进行SNP calling</p>
<p><img src="https://pic1.zhimg.com/v2-32f56b62fbe940b13b82537ce6542244_b.png" alt="img"></p>
<p>bcftools进行vcf格式转换</p>
<hr>
<ol start="6">
<li><h4 id="vcftools进行SNP质量过滤："><a href="#vcftools进行SNP质量过滤：" class="headerlink" title="vcftools进行SNP质量过滤："></a>vcftools进行SNP质量过滤：</h4></li>
</ol>
<p>得到的VCF文件记录了我们分析得到的外显子(对应WES)或基因组(对应WGS或WGRS)上的全部SNP信息，但是这里边有一些是低质量的噪音信息，我们需要通过设置一系列的筛选标准来将噪音去除，称为SNP质量过滤或数据清洗。我一般选用以下三个标准：</p>
<ul>
<li>–max-missing: 最大缺失值</li>
<li>–minDP: 最小测序深度</li>
<li>–minQ: 最小碱基质量</li>
</ul>
<p>其中比较难以理解的是最大缺失值这个参数，vcftools官方给出的定义是这样的：</p>
<p><img src="https://pic4.zhimg.com/v2-b3f3afce179cbcf85e1a27f1aa631a43_b.png" alt="img"></p>
<p>最大缺失值官方定义</p>
<p>我盯住的是0允许全部缺失而1代表不允许缺失，那就是<strong>说该值是介于0～1之间的浮点数，其大小代表着最少需要多少平行含有该SNP的比例。如0.3就代表至少30%的平行都有该SNP.</strong></p>
<p>参考代码如下：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">vcftools --vcf sample1.vcf --minDP 3 --max-missing 0.3 --minQ 30 --recode --recode-INFO-all --out sample1.filtered</span><br></pre></td></tr></table></figure>

<p>同样由于一批测序数据中有多个测序文件，因此需要循环以上过程，参考代码如下图所示：</p>
<p><img src="https://pic4.zhimg.com/v2-a621573119dd0df2c15dd293d469d9df_b.png" alt="img"></p>
<p>vcftools进行SNP质量过滤</p>
<hr>
<ol start="7">
<li><h4 id="ANNOVAR进行SNP注释"><a href="#ANNOVAR进行SNP注释" class="headerlink" title="ANNOVAR进行SNP注释:"></a>ANNOVAR进行SNP注释:</h4></li>
</ol>
<p>利用前面下载的基因组和注释文件对质量过滤后的SNP进行注释, 【<strong>将基因组和注释文件放在同一路径下】！</strong></p>
<ol>
<li><h5 id="gffread和gtfToGenePred进行注释文件的格式转换："><a href="#gffread和gtfToGenePred进行注释文件的格式转换：" class="headerlink" title="gffread和gtfToGenePred进行注释文件的格式转换："></a>gffread和gtfToGenePred进行注释文件的格式转换：</h5></li>
</ol>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">#</span><span class="language-bash">我们下载的参考基因组注释文件是.gff格式的，需要转换为.gtf格式</span></span><br><span class="line">gffread ZM4_annotation.gff -T -o ZM4_annotation.gtf</span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">将.gtf格式的注释文件转为ANNOVAR需要的输入文件</span></span><br><span class="line">/path/to/bin/gtfToGenePred -genePredExt ZM4_annotation.gtf ZM4_refGene.txt</span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">【此时ZM4_refGene.txt应该和ZM4_annotation.gtf, ZM4_genome.fa是同一路径！！！】</span></span><br></pre></td></tr></table></figure>

<ol start="2">
<li><h5 id="ANNOVAR进行注释："><a href="#ANNOVAR进行注释：" class="headerlink" title="ANNOVAR进行注释："></a>ANNOVAR进行注释：</h5></li>
</ol>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">#</span><span class="language-bash">建立注释索引</span></span><br><span class="line">perl /path/to/annovar/retrieve_seq_from_fasta.pl --format refGene --seqfile ZM4_genome.fa ZM4_refGene.txt --outfile ZM4_refGeneMrna.fa</span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">【此时ZM4_refGeneMrna.fa应该和ZM4_refGene.txt, ZM4_annotation.gtf, ZM4_genome.fa是同一路径！！！】</span></span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">将SNP信息文件.vcf转为ANNOVAR需要的输入文件</span></span><br><span class="line">perl /path/to/annovar/convert2annovar.pl -format vcf4old sample1.filtered.recode.vcf &gt; sample1.annovar</span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">注释</span></span><br><span class="line">perl /path/to/annovar/annotate_variation.pl -buildver ZM4 -outfile sample1.anno sample1.annovar /path/to/ZM4_refGeneMrna.fa and ZM4_refGene.txt/</span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">参数详解：</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">-buildver ZM4 指定注释文件的版本是ZM4, 即注释文件XX_refGene.txt和XX_refGeneMrna.fa的前缀XX</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">-outfile 输出注释文件</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">然后接输入文件</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">最后接注释文件的储存路径，即ZM4_refGeneMrna.fa, ZM4_refGene.txt, ZM4_annotation.gtf, ZM4_genome.fa的储存路径</span></span><br></pre></td></tr></table></figure>

<p>SNP注释结果文件预览结果如下图所示：</p>
<p><img src="https://pic1.zhimg.com/v2-4a973d12b5754a8aafac5ce6e22913cc_b.png" alt="img"></p>
<p>由左至右各列信息为：</p>
<ul>
<li>所在行</li>
<li>同义&#x2F;非同义突变</li>
<li>突变基因&#x2F;外显子&#x2F;核苷酸变化及位置&#x2F;氨基酸变化及位置</li>
<li>染色体</li>
<li>突变起始位置</li>
<li>突变终止位置</li>
<li><a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%8E%9F%E5%A7%8B%E5%BA%8F%E5%88%97&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">原始序列</a></li>
<li>突变序列</li>
<li>同源&#x2F;异源</li>
<li>FQ: Phred probability of all samples being the same</li>
<li>XXX </li>
<li><a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E6%B5%8B%E5%BA%8F%E8%B4%A8%E9%87%8F&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">测序质量</a></li>
</ul>
<p>同样由于一批测序数据中有多个测序文件，因此需要循环注释过程，参考代码如下图所示：</p>
<p><img src="https://pic3.zhimg.com/v2-edb807163059c83846400ed378457922_b.png" alt="img"></p>
<p>ANNOVAR进行SNP注释</p>
<hr>
<ol start="8">
<li><h4 id="可视化补充信息—-samtools统计测序覆盖度和深度"><a href="#可视化补充信息—-samtools统计测序覆盖度和深度" class="headerlink" title="可视化补充信息—-samtools统计测序覆盖度和深度:"></a>可视化补充信息—-samtools<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E7%BB%9F%E8%AE%A1%E6%B5%8B%E5%BA%8F&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">统计测序</a>覆盖度和深度:</h4></li>
</ol>
<p>前方步骤中得到的去除PCR重复排序后的比对结果sample1.sorted.bam, 可通过samtools统计其中包含的测序覆盖度和深度信息用于可视化，参考代码如下：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools depth sample1.sorted.bam sample1.depth</span><br></pre></td></tr></table></figure>

<p>得到的.depth文件预览结果如下：</p>
<ul>
<li>第一列为染色体名称</li>
<li>第二列为染色体上的核苷酸位置</li>
<li>第三列为该位置上的测序深度</li>
</ul>
<p><img src="https://pic2.zhimg.com/v2-a731b92c8c2fe25660cdc6fe1a71b2b1_b.jpg" alt="img"></p>
<p>samtools进行测序覆盖度和深度统计</p>
<hr>
<h3 id="可视化："><a href="#可视化：" class="headerlink" title="可视化："></a>可视化：</h3><ol>
<li><h4 id="测序覆盖度和深度："><a href="#测序覆盖度和深度：" class="headerlink" title="测序覆盖度和深度："></a>测序覆盖度和深度：</h4></li>
</ol>
<p>将上一步得到的.depth文件用于统计测序深度与覆盖度，在Python中的参考代码如下：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="comment">######################################################PARAMETERS#######################################################</span></span><br><span class="line">work_dir=<span class="string">&#x27;/Users/ZYP/Downloads/WES_analysis_on_ZM4_RNAseq_data/3.raw/depth/&#x27;</span> <span class="comment">#输入和输出文件路径</span></span><br><span class="line">files=[<span class="string">&#x27;8P_MUT1&#x27;</span>,<span class="string">&#x27;8P_OE1&#x27;</span>,<span class="string">&#x27;8P_WT1&#x27;</span>,<span class="string">&#x27;G5_MUT1&#x27;</span>,<span class="string">&#x27;G5_OE1&#x27;</span>,<span class="string">&#x27;G5_WT1&#x27;</span>,<span class="string">&#x27;8P_MUT2&#x27;</span>,<span class="string">&#x27;8P_OE2&#x27;</span>,<span class="string">&#x27;8P_WT2&#x27;</span>,<span class="string">&#x27;G5_MUT2&#x27;</span>,<span class="string">&#x27;G5_OE2&#x27;</span>,<span class="string">&#x27;G5_WT2&#x27;</span>] <span class="comment">#输入文件名前缀</span></span><br><span class="line">start_position=<span class="number">254</span><span class="comment">#from depth file 测序起始位置</span></span><br><span class="line">end_position=<span class="number">2056015</span><span class="comment">#from depth file 测序终止位置</span></span><br><span class="line">distance=<span class="number">1000</span><span class="comment">#distance for calculating average depth 用于统计平均深度和覆盖度的片段大小</span></span><br><span class="line">length_genome=<span class="number">2056363</span><span class="comment">#from genome 基因组总长</span></span><br><span class="line"><span class="comment">#######################################################################################################################</span></span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> files:</span><br><span class="line">    list_pos = []</span><br><span class="line">    list_depth = []</span><br><span class="line">    sum_depths = <span class="number">0</span></span><br><span class="line">    distance_pos = <span class="number">1</span></span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(work_dir+name+<span class="string">&#x27;.depth&#x27;</span>):</span><br><span class="line">        sum_depths += <span class="built_in">float</span>(line[:-<span class="number">1</span>].split(<span class="string">&#x27;\t&#x27;</span>)[<span class="number">2</span>])</span><br><span class="line">        <span class="keyword">if</span> start_position &gt;distance*distance_pos:</span><br><span class="line">            list_pos.append(distance_pos)</span><br><span class="line">            list_depth.append(<span class="number">0</span>)</span><br><span class="line">            distance_pos+=<span class="number">1</span></span><br><span class="line">        <span class="keyword">if</span> start_position==distance*distance_pos <span class="keyword">or</span> start_position==length_genome:</span><br><span class="line">            list_pos.append(distance_pos)</span><br><span class="line">            <span class="keyword">if</span> start_position==distance*distance_pos:</span><br><span class="line">                list_depth.append(sum_depths/distance)</span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                list_depth.append(sum_depths/(length_genome%distance))</span><br><span class="line">            sum_depths=<span class="number">0</span></span><br><span class="line">            distance_pos += <span class="number">1</span></span><br><span class="line">        start_position+=<span class="number">1</span></span><br><span class="line">    <span class="keyword">if</span> start_position &lt; length_genome:</span><br><span class="line">        <span class="keyword">if</span> start_position &gt; distance*(distance_pos-<span class="number">1</span>):</span><br><span class="line">            list_pos.append(distance_pos)</span><br><span class="line">            list_depth.append(sum_depths/(start_position-distance*(distance_pos-<span class="number">1</span>)))</span><br><span class="line">            distance_pos+=<span class="number">1</span></span><br><span class="line">        list_pos.append(distance_pos)</span><br><span class="line">        list_depth.append(<span class="number">0</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;data loading finished!&#x27;</span>)</span><br><span class="line">    ax=sns.barplot(x=list_pos,y=list_depth)</span><br><span class="line">    ax.spines[<span class="string">&#x27;top&#x27;</span>].set_visible(<span class="literal">False</span>)</span><br><span class="line">    ax.spines[<span class="string">&#x27;right&#x27;</span>].set_visible(<span class="literal">False</span>)</span><br><span class="line">    ax.spines[<span class="string">&#x27;left&#x27;</span>].set_color(<span class="string">&#x27;gray&#x27;</span>)</span><br><span class="line">    ax.spines[<span class="string">&#x27;bottom&#x27;</span>].set_color(<span class="string">&#x27;gray&#x27;</span>)</span><br><span class="line">    plt.savefig(work_dir+name+<span class="string">&#x27;_depth.pdf&#x27;</span>)</span><br><span class="line">    plt.show()</span><br></pre></td></tr></table></figure>

<p>运行结果如下图所示：</p>
<p><img src="https://pic2.zhimg.com/v2-77bad2eefbf63f30555d5127f8a5fd4d_b.png" alt="img"></p>
<p>全基因组测序深度覆盖图</p>
<hr>
<ol start="2">
<li><h4 id="SNP在基因组浏览器IGV的可视化"><a href="#SNP在基因组浏览器IGV的可视化" class="headerlink" title="SNP在基因组浏览器IGV的可视化:"></a>SNP在<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%B5%8F%E8%A7%88%E5%99%A8&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">基因组浏览器</a>IGV的可视化:</h4></li>
</ol>
<p>前方得到的去重复排序后的比对结果文件.sorted.bam可以在基因组浏览器IGV (下载链接为 <a href="https://link.zhihu.com/?target=http://software.broadinstitute.org/software/igv/download">http://software.broadinstitute.org/software/igv/download</a> ) 中打开，从而在基因组全局和单核苷酸局部进行个性化浏览和分析。</p>
<p><img src="https://pic3.zhimg.com/v2-4ae5313aef5006834c505a4738709c3e_b.jpg" alt="img"></p>
<p>IGV可视化基因组</p>
<hr>
<ol start="3">
<li><h4 id="SNV-单核苷酸变异-和InDel-插入缺失-分析："><a href="#SNV-单核苷酸变异-和InDel-插入缺失-分析：" class="headerlink" title="SNV(单核苷酸变异)和InDel(插入缺失)分析："></a>SNV(单核苷酸变异)和InDel(插入缺失)分析：</h4></li>
</ol>
<p>包括以下部分：</p>
<ul>
<li>SNV(单核苷酸变异，包括同义突变和<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E9%9D%9E%E5%90%8C%E4%B9%89%E7%AA%81%E5%8F%98&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">非同义突变</a>)</li>
<li>Nonsynonymous SNV(非同义突变单核苷酸变异)</li>
<li>Amino Acid Substitution(由非同义突变造成的氨基酸替换)</li>
</ul>
<p>以ANNOVAR的SNP注释结果文件为输入，在Python中的参考代码如下：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="comment">#######################################################PARAMETERS#######################################################</span></span><br><span class="line">path=<span class="string">&#x27;/Users/ZYP/Downloads/WES_analysis_on_ZM4_RNAseq_data/3.raw/annotation/&#x27;</span> <span class="comment">#输入和输出文件路径</span></span><br><span class="line">files=[<span class="string">&#x27;8P_MUT1&#x27;</span>,<span class="string">&#x27;8P_OE1&#x27;</span>,<span class="string">&#x27;8P_WT1&#x27;</span>,<span class="string">&#x27;G5_MUT1&#x27;</span>,<span class="string">&#x27;G5_OE1&#x27;</span>,<span class="string">&#x27;G5_WT1&#x27;</span>,<span class="string">&#x27;8P_MUT2&#x27;</span>,<span class="string">&#x27;8P_OE2&#x27;</span>,<span class="string">&#x27;8P_WT2&#x27;</span>,<span class="string">&#x27;G5_MUT2&#x27;</span>,<span class="string">&#x27;G5_OE2&#x27;</span>,<span class="string">&#x27;G5_WT2&#x27;</span>] <span class="comment">#输入文件名前缀</span></span><br><span class="line"><span class="comment">########################################################################################################################</span></span><br><span class="line">list_sample_SNV=[]</span><br><span class="line">list_para_SNV=[]</span><br><span class="line">list_SNV_name=[]</span><br><span class="line">list_sample_SNV_nonsy=[]</span><br><span class="line">list_para_SNV_nonsy=[]</span><br><span class="line">list_SNV_nonsy_name=[]</span><br><span class="line">list_sample_cAA=[]</span><br><span class="line">list_para_cAA=[]</span><br><span class="line">list_cAA_name=[]</span><br><span class="line">save_file_cAA=<span class="built_in">open</span>(path+<span class="string">&#x27;AA_substitution.txt&#x27;</span>,<span class="string">&#x27;w&#x27;</span>)</span><br><span class="line">save_file_cAA.write(<span class="string">&#x27;Sample&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;SNP type&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;Gene&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;REF NA&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;ALT NA&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;POS NA&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;REF AA&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;ALT AA&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;POS AA&#x27;</span>+<span class="string">&#x27;\n&#x27;</span>)</span><br><span class="line">save_file_InDel=<span class="built_in">open</span>(path+<span class="string">&#x27;InDel.txt&#x27;</span>,<span class="string">&#x27;w&#x27;</span>)</span><br><span class="line">save_file_InDel.write(<span class="string">&#x27;Sample&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;SNP type&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;Hom/Het&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;GeneFrom&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;GeneIn&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;Frame&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+<span class="string">&#x27;Effect&#x27;</span>+<span class="string">&#x27;\n&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> file <span class="keyword">in</span> files:</span><br><span class="line">    sample=file[:-<span class="number">1</span>]</span><br><span class="line">    para=file[-<span class="number">1</span>]</span><br><span class="line">    dict_SNV = &#123;<span class="string">&#x27;A&gt;T|T&gt;A&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;A&gt;C|T&gt;G&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;A&gt;G|T&gt;C&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;C&gt;G|G&gt;C&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;G&gt;A|C&gt;T&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;G&gt;T|C&gt;A&#x27;</span>:<span class="number">0</span>&#125;</span><br><span class="line">    dict_SNV_nonsy = &#123;<span class="string">&#x27;A&gt;T|T&gt;A&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;A&gt;C|T&gt;G&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;A&gt;G|T&gt;C&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;C&gt;G|G&gt;C&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;G&gt;A|C&gt;T&#x27;</span>:<span class="number">0</span>,<span class="string">&#x27;G&gt;T|C&gt;A&#x27;</span>:<span class="number">0</span>&#125;</span><br><span class="line">    dict_cAA = &#123;&#125;</span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(path+<span class="string">&#x27;SNP_annotation_&#x27;</span>+file+<span class="string">&#x27;.txt&#x27;</span>):</span><br><span class="line">        kind=line.split(<span class="string">&#x27;\t&#x27;</span>)[<span class="number">1</span>]</span><br><span class="line">        <span class="keyword">if</span> kind <span class="keyword">in</span> [<span class="string">&#x27;frameshift insertion&#x27;</span>,<span class="string">&#x27;framshift deletion&#x27;</span>]:</span><br><span class="line">            confusion=line.split(<span class="string">&#x27;\t&#x27;</span>)[<span class="number">8</span>]</span><br><span class="line">            geneFrom_geneIn_frame_effect = line.split(<span class="string">&#x27;\t&#x27;</span>)[<span class="number">2</span>]</span><br><span class="line">            geneFrom=geneFrom_geneIn_frame_effect.split(<span class="string">&#x27;:&#x27;</span>)[<span class="number">0</span>]</span><br><span class="line">            geneIn=geneFrom_geneIn_frame_effect.split(<span class="string">&#x27;:&#x27;</span>)[<span class="number">1</span>]</span><br><span class="line">            frame=geneFrom_geneIn_frame_effect.split(<span class="string">&#x27;:&#x27;</span>)[-<span class="number">2</span>]</span><br><span class="line">            effect=geneFrom_geneIn_frame_effect.split(<span class="string">&#x27;:&#x27;</span>)[-<span class="number">1</span>]</span><br><span class="line">            save_file_InDel.write(file+<span class="string">&#x27;\t&#x27;</span>+kind+<span class="string">&#x27;\t&#x27;</span>+confusion+<span class="string">&#x27;\t&#x27;</span>+geneFrom+<span class="string">&#x27;\t&#x27;</span>+geneIn+<span class="string">&#x27;\t&#x27;</span>+frame+<span class="string">&#x27;\t&#x27;</span>+effect+<span class="string">&#x27;\n&#x27;</span>)</span><br><span class="line">        <span class="keyword">if</span> kind <span class="keyword">in</span> [<span class="string">&#x27;synonymous SNV&#x27;</span>,<span class="string">&#x27;nonsynonymous SNV&#x27;</span>]:</span><br><span class="line">            gene_cNA_cAA=line.split(<span class="string">&#x27;\t&#x27;</span>)[<span class="number">2</span>]</span><br><span class="line">            gene=gene_cNA_cAA.split(<span class="string">&#x27;:&#x27;</span>)[<span class="number">1</span>]</span><br><span class="line">            cNA=gene_cNA_cAA.split(<span class="string">&#x27;:&#x27;</span>)[-<span class="number">2</span>][<span class="number">2</span>:]</span><br><span class="line">            cAA=gene_cNA_cAA.split(<span class="string">&#x27;:&#x27;</span>)[-<span class="number">1</span>][<span class="number">2</span>:-<span class="number">1</span>]</span><br><span class="line">            REF_NA=cNA[<span class="number">0</span>]</span><br><span class="line">            ALT_NA=cNA[-<span class="number">1</span>]</span><br><span class="line">            POS_NA=cNA[<span class="number">1</span>:-<span class="number">1</span>]</span><br><span class="line">            REF_AA=cAA[<span class="number">0</span>]</span><br><span class="line">            ALT_AA=cAA[-<span class="number">1</span>]</span><br><span class="line">            POS_AA=cAA[<span class="number">1</span>:-<span class="number">1</span>]</span><br><span class="line">            save_file_cAA.write(file+<span class="string">&#x27;\t&#x27;</span>+kind+<span class="string">&#x27;\t&#x27;</span>+gene+<span class="string">&#x27;\t&#x27;</span>+REF_NA+<span class="string">&#x27;\t&#x27;</span>+ALT_NA+<span class="string">&#x27;\t&#x27;</span>+POS_NA+<span class="string">&#x27;\t&#x27;</span>+REF_AA+<span class="string">&#x27;\t&#x27;</span>+ALT_AA+<span class="string">&#x27;\t&#x27;</span>+POS_AA+<span class="string">&#x27;\n&#x27;</span>)</span><br><span class="line">            <span class="keyword">for</span> key <span class="keyword">in</span> dict_SNV.keys():</span><br><span class="line">                list_changes=key.split(<span class="string">&#x27;|&#x27;</span>)</span><br><span class="line">                <span class="keyword">for</span> change <span class="keyword">in</span> list_changes:</span><br><span class="line">                    <span class="keyword">if</span> REF_NA==change[<span class="number">0</span>] <span class="keyword">and</span> ALT_NA==change[-<span class="number">1</span>]:</span><br><span class="line">                        dict_SNV[key]+=<span class="number">1</span></span><br><span class="line">            <span class="keyword">if</span> kind==<span class="string">&#x27;nonsynonymous SNV&#x27;</span>:</span><br><span class="line">                name_cAA=REF_AA+<span class="string">&#x27;&gt;&#x27;</span>+ALT_AA</span><br><span class="line">                <span class="keyword">for</span> key <span class="keyword">in</span> dict_SNV_nonsy.keys():</span><br><span class="line">                    list_changes = key.split(<span class="string">&#x27;|&#x27;</span>)</span><br><span class="line">                    <span class="keyword">for</span> change <span class="keyword">in</span> list_changes:</span><br><span class="line">                        <span class="keyword">if</span> REF_NA==change[<span class="number">0</span>] <span class="keyword">and</span> ALT_NA==change[-<span class="number">1</span>]:</span><br><span class="line">                            dict_SNV_nonsy[key]+=<span class="number">1</span></span><br><span class="line">                <span class="keyword">if</span> name_cAA <span class="keyword">not</span> <span class="keyword">in</span> dict_cAA.keys():</span><br><span class="line">                    dict_cAA[name_cAA]=<span class="number">1</span></span><br><span class="line">                <span class="keyword">else</span>:</span><br><span class="line">                    dict_cAA[name_cAA]+=<span class="number">1</span></span><br><span class="line">    <span class="keyword">for</span> SNV <span class="keyword">in</span> dict_SNV.keys():</span><br><span class="line">        <span class="keyword">for</span> time <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">0</span>,dict_SNV[SNV]):</span><br><span class="line">            list_sample_SNV.append(sample)</span><br><span class="line">            list_para_SNV.append(para)</span><br><span class="line">            list_SNV_name.append(SNV)</span><br><span class="line">    <span class="keyword">for</span> SNV_nonsy <span class="keyword">in</span> dict_SNV_nonsy.keys():</span><br><span class="line">        <span class="keyword">for</span> time <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">0</span>, dict_SNV[SNV_nonsy]):</span><br><span class="line">            list_sample_SNV_nonsy.append(sample)</span><br><span class="line">            list_para_SNV_nonsy.append(para)</span><br><span class="line">            list_SNV_nonsy_name.append(SNV_nonsy)</span><br><span class="line">    <span class="keyword">for</span> cAA <span class="keyword">in</span> dict_cAA.keys():</span><br><span class="line">        <span class="keyword">for</span> time <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">0</span>, dict_cAA[cAA]):</span><br><span class="line">            list_sample_cAA.append(sample)</span><br><span class="line">            list_para_cAA.append(para)</span><br><span class="line">            list_cAA_name.append(cAA)</span><br><span class="line">data_SNV=&#123;<span class="string">&#x27;sample&#x27;</span>:list_sample_SNV,<span class="string">&#x27;parallel&#x27;</span>:list_para_SNV,<span class="string">&#x27;SNV&#x27;</span>:list_SNV_name&#125;</span><br><span class="line">data_SNV=pd.DataFrame(data_SNV)</span><br><span class="line">ax1=sns.catplot(x=<span class="string">&#x27;SNV&#x27;</span>, hue=<span class="string">&#x27;parallel&#x27;</span>, col=<span class="string">&#x27;sample&#x27;</span>, kind=<span class="string">&#x27;count&#x27;</span>,data=data_SNV,palette=<span class="string">&#x27;Set1&#x27;</span>)</span><br><span class="line">plt.savefig(path+<span class="string">&#x27;SNV.pdf&#x27;</span>)</span><br><span class="line">data_SNV_nonsy=&#123;<span class="string">&#x27;sample&#x27;</span>:list_sample_SNV_nonsy,<span class="string">&#x27;parallel&#x27;</span>:list_para_SNV_nonsy,<span class="string">&#x27;nonsynonymous SNV&#x27;</span>:list_SNV_nonsy_name&#125;</span><br><span class="line">data_SNV_nonsy=pd.DataFrame(data_SNV_nonsy)</span><br><span class="line">ax2=sns.catplot(x=<span class="string">&#x27;nonsynonymous SNV&#x27;</span>, hue=<span class="string">&#x27;parallel&#x27;</span>, col=<span class="string">&#x27;sample&#x27;</span>, kind=<span class="string">&#x27;count&#x27;</span>,data=data_SNV_nonsy,palette=<span class="string">&#x27;Set1&#x27;</span>)</span><br><span class="line">plt.savefig(path+<span class="string">&#x27;Nonsynonymous_SNV.pdf&#x27;</span>)</span><br><span class="line">data_cAA=&#123;<span class="string">&#x27;sample&#x27;</span>:list_sample_cAA,<span class="string">&#x27;parallel&#x27;</span>:list_para_cAA,<span class="string">&#x27;AA substitution&#x27;</span>:list_cAA_name&#125;</span><br><span class="line">data_cAA=pd.DataFrame(data_cAA)</span><br><span class="line">ax3=sns.catplot(x=<span class="string">&#x27;AA substitution&#x27;</span>, hue=<span class="string">&#x27;parallel&#x27;</span>, col=<span class="string">&#x27;sample&#x27;</span>, kind=<span class="string">&#x27;count&#x27;</span>,data=data_cAA,palette=<span class="string">&#x27;Set1&#x27;</span>)</span><br><span class="line">plt.savefig(path+<span class="string">&#x27;AA_substitution.pdf&#x27;</span>)</span><br><span class="line">save_file_cAA.close()</span><br></pre></td></tr></table></figure>

<p>如图所示，运行结束得到可视化结果和记录SNV与InDel的两个文件SNV.txt和InDel.txt .</p>
<p><img src="https://pic4.zhimg.com/v2-0af7b94649de8f5025e4e4b815e2a5c7_b.jpg" alt="img"></p>
<p>SNV and AA substitution</p>
<p><img src="https://pic4.zhimg.com/v2-d2403cc08e122cc8961f30bbedfb7e3f_b.jpg" alt="img"></p>
<p>SNV and InDel</p>
<hr>
<ol start="4">
<li><h4 id="全基因组测序圈图可视化："><a href="#全基因组测序圈图可视化：" class="headerlink" title="全基因组测序圈图可视化："></a>全基因组测序圈图可视化：</h4></li>
</ol>
<p>以前文中提到的测序深度文件.depth, ANNOVAR的SNP注释结果文件, 以及基因组注释文件.gff为输入，在R中的参考代码如下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br><span class="line">229</span><br><span class="line">230</span><br><span class="line">231</span><br><span class="line">232</span><br><span class="line">233</span><br><span class="line">234</span><br><span class="line">235</span><br><span class="line">236</span><br><span class="line">237</span><br><span class="line">238</span><br><span class="line">239</span><br><span class="line">240</span><br><span class="line">241</span><br><span class="line">242</span><br><span class="line">243</span><br><span class="line">244</span><br><span class="line">245</span><br><span class="line">246</span><br><span class="line">247</span><br><span class="line">248</span><br><span class="line">249</span><br><span class="line">250</span><br><span class="line">251</span><br><span class="line">252</span><br><span class="line">253</span><br><span class="line">254</span><br><span class="line">255</span><br><span class="line">256</span><br><span class="line">257</span><br><span class="line">258</span><br><span class="line">259</span><br><span class="line">260</span><br><span class="line">261</span><br><span class="line">262</span><br><span class="line">263</span><br><span class="line">264</span><br><span class="line">265</span><br><span class="line">266</span><br><span class="line">267</span><br><span class="line">268</span><br><span class="line">269</span><br><span class="line">270</span><br><span class="line">271</span><br><span class="line">272</span><br><span class="line">273</span><br><span class="line">274</span><br><span class="line">275</span><br><span class="line">276</span><br><span class="line">277</span><br><span class="line">278</span><br><span class="line">279</span><br><span class="line">280</span><br><span class="line">281</span><br><span class="line">282</span><br><span class="line">283</span><br><span class="line">284</span><br><span class="line">285</span><br><span class="line">286</span><br><span class="line">287</span><br><span class="line">288</span><br><span class="line">289</span><br><span class="line">290</span><br><span class="line">291</span><br><span class="line">292</span><br><span class="line">293</span><br><span class="line">294</span><br><span class="line">295</span><br><span class="line">296</span><br><span class="line">297</span><br><span class="line">298</span><br><span class="line">299</span><br><span class="line">300</span><br><span class="line">301</span><br><span class="line">302</span><br><span class="line">303</span><br><span class="line">304</span><br><span class="line">305</span><br><span class="line">306</span><br><span class="line">307</span><br><span class="line">308</span><br><span class="line">309</span><br><span class="line">310</span><br><span class="line">311</span><br><span class="line">312</span><br><span class="line">313</span><br><span class="line">314</span><br><span class="line">315</span><br><span class="line">316</span><br><span class="line">317</span><br><span class="line">318</span><br><span class="line">319</span><br><span class="line">320</span><br><span class="line">321</span><br><span class="line">322</span><br><span class="line">323</span><br><span class="line">324</span><br><span class="line">325</span><br><span class="line">326</span><br><span class="line">327</span><br><span class="line">328</span><br><span class="line">329</span><br><span class="line">330</span><br><span class="line">331</span><br><span class="line">332</span><br><span class="line">333</span><br><span class="line">334</span><br><span class="line">335</span><br><span class="line">336</span><br><span class="line">337</span><br><span class="line">338</span><br><span class="line">339</span><br><span class="line">340</span><br><span class="line">341</span><br><span class="line">342</span><br><span class="line">343</span><br><span class="line">344</span><br><span class="line">345</span><br><span class="line">346</span><br><span class="line">347</span><br><span class="line">348</span><br><span class="line">349</span><br><span class="line">350</span><br><span class="line">351</span><br><span class="line">352</span><br><span class="line">353</span><br><span class="line">354</span><br><span class="line">355</span><br><span class="line">356</span><br><span class="line">357</span><br><span class="line">358</span><br><span class="line">359</span><br><span class="line">360</span><br><span class="line">361</span><br><span class="line">362</span><br><span class="line">363</span><br><span class="line">364</span><br><span class="line">365</span><br><span class="line">366</span><br><span class="line">367</span><br><span class="line">368</span><br><span class="line">369</span><br><span class="line">370</span><br><span class="line">371</span><br><span class="line">372</span><br><span class="line">373</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>stringr<span class="punctuation">)</span>    <span class="comment">#方便处理字符串</span></span><br><span class="line">library<span class="punctuation">(</span>circlize<span class="punctuation">)</span>    <span class="comment">#绘制圈图</span></span><br><span class="line">library<span class="punctuation">(</span>grid<span class="punctuation">)</span>    <span class="comment">#设置画板</span></span><br><span class="line">library<span class="punctuation">(</span>ComplexHeatmap<span class="punctuation">)</span>    <span class="comment">#绘制图例</span></span><br><span class="line"></span><br><span class="line"><span class="comment">####################################PARAMETERS####################################</span></span><br><span class="line">sample_name <span class="operator">&lt;-</span><span class="string">&#x27;G5_WT2&#x27;</span><span class="comment">#测序样本名称</span></span><br><span class="line">ref_name <span class="operator">&lt;-</span><span class="string">&#x27;ZM4_genome&#x27;</span><span class="comment">#参考基因组名称</span></span><br><span class="line">genome_size <span class="operator">&lt;-</span><span class="number">205636</span> <span class="comment">#参考基因组长度</span></span><br><span class="line">out_dir<span class="operator">=</span> <span class="string">&#x27;/Users/ZYP/Downloads/WES_analysis_on_ZM4_RNAseq_data/results/Analysis/&#x27;</span> <span class="comment">#输出路径</span></span><br><span class="line">genome_gff <span class="operator">&lt;-</span> <span class="string">&#x27;/Users/ZYP/Downloads/WES_analysis_on_ZM4_RNAseq_data/results/IGV visualization/ZM4_annotation.gff&#x27;</span> <span class="comment">#参考基因组gff注释文件</span></span><br><span class="line">SNP_file <span class="operator">&lt;-</span><span class="string">&#x27;/Users/ZYP/Downloads/WES_analysis_on_ZM4_RNAseq_data/raw/annotation/SNP_annotation_G5_WT2.txt&#x27;</span> <span class="comment">#SNP注释文件</span></span><br><span class="line">depth_base_stat <span class="operator">&lt;-</span> <span class="string">&#x27;/Users/ZYP/Downloads/WES_analysis_on_ZM4_RNAseq_data/raw/depth/G5_WT2.depth&#x27;</span> <span class="comment">#测序深度覆盖度文件</span></span><br><span class="line">seq_split <span class="operator">&lt;-</span> 1000  <span class="comment">#滑窗大小用于统计平均测序深度和SNP密度</span></span><br><span class="line"><span class="comment">##################################################################################</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##测序覆盖度、平均深度统计</span></span><br><span class="line">depth_base <span class="operator">&lt;-</span> read.delim<span class="punctuation">(</span>depth_base_stat<span class="punctuation">,</span> stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>depth_base<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;gene&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;pos&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;depth&#x27;</span><span class="punctuation">)</span></span><br><span class="line">depth <span class="operator">&lt;-</span> depth_base<span class="operator">$</span>depth</span><br><span class="line">depth_base <span class="operator">&lt;-</span> depth_base<span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="number">2</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">depth_base<span class="operator">$</span>end <span class="operator">&lt;-</span> depth_base<span class="operator">$</span>pos</span><br><span class="line">depth_base<span class="operator">$</span>depth <span class="operator">&lt;-</span> depth</span><br><span class="line"></span><br><span class="line">depth_exist <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>depth_base<span class="punctuation">,</span> depth <span class="operator">!=</span> <span class="number">0</span><span class="punctuation">)</span></span><br><span class="line">coverage <span class="operator">&lt;-</span> <span class="built_in">round</span><span class="punctuation">(</span><span class="number">100</span> <span class="operator">*</span> <span class="built_in">sum</span><span class="punctuation">(</span>depth_exist<span class="operator">$</span>seq_end <span class="operator">-</span> depth_exist<span class="operator">$</span>seq_start <span class="operator">+</span> <span class="number">1</span><span class="punctuation">)</span> <span class="operator">/</span> genome_size<span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">average_depth <span class="operator">&lt;-</span> <span class="built_in">round</span><span class="punctuation">(</span>mean<span class="punctuation">(</span>depth_base<span class="operator">$</span>depth<span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">depth_stat <span class="operator">&lt;-</span> <span class="literal">NULL</span></span><br><span class="line">gene <span class="operator">&lt;-</span> unique<span class="punctuation">(</span>depth_exist<span class="operator">$</span>gene<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>gene_n <span class="keyword">in</span> gene<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  depth_subset <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>depth_exist<span class="punctuation">,</span> gene <span class="operator">==</span> gene_n<span class="punctuation">)</span></span><br><span class="line">  seq_end <span class="operator">&lt;-</span> seq_split</span><br><span class="line">  depth_num <span class="operator">&lt;-</span> 0</span><br><span class="line">  </span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="number">1</span><span class="operator">:</span>nrow<span class="punctuation">(</span>depth_subset<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    <span class="keyword">if</span> <span class="punctuation">(</span>depth_subset<span class="punctuation">[</span>i<span class="punctuation">,</span><span class="string">&#x27;pos&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;=</span> seq_end<span class="punctuation">)</span> depth_num <span class="operator">&lt;-</span> depth_num <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">    <span class="keyword">else</span> <span class="punctuation">&#123;</span></span><br><span class="line">      depth_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>depth_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>gene_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_end<span class="punctuation">,</span>depth_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">      </span><br><span class="line">      seq_end <span class="operator">&lt;-</span> seq_end <span class="operator">+</span> seq_split</span><br><span class="line">      depth_num <span class="operator">&lt;-</span> 0</span><br><span class="line">      <span class="keyword">while</span> <span class="punctuation">(</span>depth_subset<span class="punctuation">[</span>i<span class="punctuation">,</span><span class="string">&#x27;pos&#x27;</span><span class="punctuation">]</span> <span class="operator">&gt;</span> seq_end<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">        depth_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>depth_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>gene_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_end<span class="punctuation">,</span> depth_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">        seq_end <span class="operator">&lt;-</span> seq_end <span class="operator">+</span> seq_split</span><br><span class="line">      <span class="punctuation">&#125;</span></span><br><span class="line">      depth_num <span class="operator">&lt;-</span> depth_num <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">  <span class="keyword">while</span> <span class="punctuation">(</span>seq_end <span class="operator">&lt;</span> depth_stat<span class="punctuation">[</span>gene_n<span class="punctuation">,</span><span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    depth_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>depth_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>gene_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_end<span class="punctuation">,</span> depth_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    seq_end <span class="operator">&lt;-</span> seq_end <span class="operator">+</span> seq_split</span><br><span class="line">    depth_num <span class="operator">&lt;-</span> 0</span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  depth_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>depth_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>gene_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_stat<span class="punctuation">[</span>gene_n<span class="punctuation">,</span><span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">]</span><span class="punctuation">,</span> depth_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">depth_stat <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>depth_stat<span class="punctuation">,</span> stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line"><span class="built_in">names</span><span class="punctuation">(</span>depth_stat<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;gene&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_start&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;depth_num&#x27;</span><span class="punctuation">)</span></span><br><span class="line">depth_stat<span class="operator">$</span>seq_start <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>depth_stat<span class="operator">$</span>seq_start<span class="punctuation">)</span></span><br><span class="line">depth_stat<span class="operator">$</span>seq_end <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>depth_stat<span class="operator">$</span>seq_end<span class="punctuation">)</span></span><br><span class="line">depth_stat<span class="operator">$</span>depth_num <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>depth_stat<span class="operator">$</span>depth_num<span class="punctuation">)</span></span><br><span class="line">depth_stat</span><br><span class="line"></span><br><span class="line">seq_stat <span class="operator">&lt;-</span> <span class="literal">NULL</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>gene <span class="keyword">in</span> unique<span class="punctuation">(</span>depth_base<span class="operator">$</span>gene<span class="punctuation">)</span><span class="punctuation">)</span> seq_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>seq_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>gene<span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">,</span> <span class="number">2056363</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">seq_stat <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>seq_stat<span class="punctuation">,</span> stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>seq_stat<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;seq_ID&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_start&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">)</span></span><br><span class="line">rownames<span class="punctuation">(</span>seq_stat<span class="punctuation">)</span> <span class="operator">&lt;-</span> seq_stat<span class="operator">$</span>seq_ID</span><br><span class="line">seq_stat<span class="operator">$</span>seq_start <span class="operator">&lt;-</span> <span class="built_in">as.integer</span><span class="punctuation">(</span>seq_stat<span class="operator">$</span>seq_start<span class="punctuation">)</span></span><br><span class="line">seq_stat<span class="operator">$</span>seq_end <span class="operator">&lt;-</span> <span class="built_in">as.integer</span><span class="punctuation">(</span>seq_stat<span class="operator">$</span>seq_end<span class="punctuation">)</span></span><br><span class="line">seq_stat</span><br><span class="line"></span><br><span class="line"><span class="comment">##参考基因组基因信息，CDS &amp; rRNA &amp; tRNA</span></span><br><span class="line">gene <span class="operator">&lt;-</span> read.delim<span class="punctuation">(</span>genome_gff<span class="punctuation">,</span> header <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> comment.char <span class="operator">=</span> <span class="string">&#x27;#&#x27;</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">gene <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>gene<span class="punctuation">,</span> V3 <span class="operator">%in%</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;CDS&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;rRNA&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;tRNA&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line"><span class="built_in">names</span><span class="punctuation">(</span>gene<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;seq_ID&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_start&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;type&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;strand&#x27;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">gene<span class="punctuation">[</span>which<span class="punctuation">(</span>gene<span class="operator">$</span>type <span class="operator">==</span> <span class="string">&#x27;CDS&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&#x27;type&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="number">1</span></span><br><span class="line">gene<span class="punctuation">[</span>which<span class="punctuation">(</span>gene<span class="operator">$</span>type <span class="operator">==</span> <span class="string">&#x27;rRNA&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&#x27;type&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="number">2</span></span><br><span class="line">gene<span class="punctuation">[</span>which<span class="punctuation">(</span>gene<span class="operator">$</span>type <span class="operator">==</span> <span class="string">&#x27;tRNA&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&#x27;type&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="number">3</span></span><br><span class="line">gene<span class="operator">$</span>type <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>gene<span class="operator">$</span>type<span class="punctuation">)</span></span><br><span class="line">gene <span class="operator">&lt;-</span> <span class="built_in">list</span><span class="punctuation">(</span>subset<span class="punctuation">(</span>gene<span class="punctuation">,</span> strand <span class="operator">==</span> <span class="string">&#x27;-&#x27;</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="operator">-</span><span class="number">5</span><span class="punctuation">]</span><span class="punctuation">,</span> subset<span class="punctuation">(</span>gene<span class="punctuation">,</span> strand <span class="operator">==</span> <span class="string">&#x27;+&#x27;</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="operator">-</span><span class="number">5</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">gene</span><br><span class="line"></span><br><span class="line"><span class="comment">#读取 SNP 文件，统计 SNP 类型</span></span><br><span class="line">snp <span class="operator">&lt;-</span> read.delim<span class="punctuation">(</span>SNP_file<span class="punctuation">,</span> header <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> colClasses <span class="operator">=</span> <span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span> comment.char <span class="operator">=</span> <span class="string">&#x27;#&#x27;</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">4</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">,</span> <span class="number">8</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">snp<span class="operator">$</span>V5 <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>snp<span class="operator">$</span>V5<span class="punctuation">)</span></span><br><span class="line">snp<span class="operator">$</span>change <span class="operator">&lt;-</span> str_c<span class="punctuation">(</span>snp<span class="operator">$</span>V7<span class="punctuation">,</span> snp<span class="operator">$</span>V8<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;AT&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;A&gt;T|T&gt;A&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;tv&#x27;</span></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;AG&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;A&gt;G|T&gt;C&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;ti&#x27;</span></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;AC&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;A&gt;C|T&gt;G&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;tv&#x27;</span></span><br><span class="line"></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;TA&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;A&gt;T|T&gt;A&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;tv&#x27;</span></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;TG&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;A&gt;C|T&gt;G&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;tv&#x27;</span></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;TC&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;A&gt;G|T&gt;C&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;ti&#x27;</span></span><br><span class="line"></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;GA&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;G&gt;A|C&gt;T&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;ti&#x27;</span></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;GT&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;G&gt;T|C&gt;A&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;tv&#x27;</span></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;GC&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;G&gt;C|C&gt;G&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;tv&#x27;</span></span><br><span class="line"></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;CA&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;G&gt;T|C&gt;A&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;tv&#x27;</span></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;CT&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;G&gt;A|C&gt;T&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;ti&#x27;</span></span><br><span class="line">change <span class="operator">&lt;-</span> which<span class="punctuation">(</span>snp<span class="operator">$</span>change <span class="operator">==</span> <span class="string">&#x27;CG&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type1&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;G&gt;C|C&gt;G&#x27;</span>; snp<span class="punctuation">[</span>change<span class="punctuation">,</span><span class="string">&#x27;type2&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="string">&#x27;tv&#x27;</span></span><br><span class="line"></span><br><span class="line">change_list <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;A&gt;T|T&gt;A&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;A&gt;G|T&gt;C&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;A&gt;C|T&gt;G&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;G&gt;A|C&gt;T&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;G&gt;T|C&gt;A&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;G&gt;C|C&gt;G&#x27;</span><span class="punctuation">)</span></span><br><span class="line">change_type <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;synonymous SNV&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;nonsynonymous SNV&#x27;</span><span class="punctuation">)</span></span><br><span class="line">anno_list <span class="operator">&lt;-</span> <span class="built_in">list</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">#length(anno_list) &lt;- length(change_list) * length(change_type)</span></span><br><span class="line">l <span class="operator">=</span> <span class="number">0</span></span><br><span class="line">anno <span class="operator">&lt;-</span> read.delim<span class="punctuation">(</span>SNP_file<span class="punctuation">,</span> header <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> colClasses <span class="operator">=</span> <span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span> comment.char <span class="operator">=</span> <span class="string">&#x27;#&#x27;</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">4</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">,</span> <span class="number">12</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">anno<span class="operator">$</span>V7 <span class="operator">&lt;-</span>snp<span class="operator">$</span>type1</span><br><span class="line">anno<span class="operator">$</span>V5 <span class="operator">&lt;-</span> <span class="built_in">as.integer</span><span class="punctuation">(</span>anno<span class="operator">$</span>V5<span class="punctuation">)</span></span><br><span class="line">anno<span class="operator">$</span>V5.1 <span class="operator">&lt;-</span> <span class="built_in">as.integer</span><span class="punctuation">(</span>anno<span class="operator">$</span>V5.1<span class="punctuation">)</span></span><br><span class="line">anno<span class="operator">$</span>V12 <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>anno<span class="operator">$</span>V12<span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>anno<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">6</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;seq_ID&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;seq_start&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;change&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;ratio&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;type&#x27;</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> change_list<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>j <span class="keyword">in</span> change_type<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    select <span class="operator">&lt;-</span> anno<span class="punctuation">[</span>which<span class="punctuation">(</span>anno<span class="operator">$</span>change <span class="operator">==</span> i <span class="operator">&amp;</span> anno<span class="operator">$</span>type <span class="operator">==</span> j<span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">    <span class="keyword">if</span> <span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>row.names<span class="punctuation">(</span>select<span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">!=</span><span class="number">0</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    l <span class="operator">=</span> l <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">    anno_list<span class="punctuation">[[</span>l<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> select</span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">snp_ti <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>snp<span class="operator">$</span>type2 <span class="operator">==</span> <span class="string">&#x27;ti&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">snp_tv <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>snp<span class="operator">$</span>type2 <span class="operator">==</span> <span class="string">&#x27;tv&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">snp_at <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>snp<span class="operator">$</span>type1 <span class="operator">==</span> <span class="string">&#x27;A&gt;T|T&gt;A&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">snp_ag <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>snp<span class="operator">$</span>type1 <span class="operator">==</span> <span class="string">&#x27;A&gt;G|T&gt;C&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">snp_ac <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>snp<span class="operator">$</span>type1 <span class="operator">==</span> <span class="string">&#x27;A&gt;C|T&gt;G&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">snp_ga <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>snp<span class="operator">$</span>type1 <span class="operator">==</span> <span class="string">&#x27;G&gt;A|C&gt;T&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">snp_gt <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>snp<span class="operator">$</span>type1 <span class="operator">==</span> <span class="string">&#x27;G&gt;T|C&gt;A&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">snp_gc <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>snp<span class="operator">$</span>type1 <span class="operator">==</span> <span class="string">&#x27;G&gt;C|C&gt;G&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#统计 SNP 密度</span></span><br><span class="line">snp <span class="operator">&lt;-</span> snp<span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">,</span> <span class="number">6</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">colnames<span class="punctuation">(</span>snp<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">2</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;seq_ID&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_site&#x27;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">snp_stat <span class="operator">&lt;-</span> <span class="literal">NULL</span></span><br><span class="line">seq_ID <span class="operator">&lt;-</span> unique<span class="punctuation">(</span>snp<span class="operator">$</span>seq_ID<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>seq_ID_n <span class="keyword">in</span> seq_ID<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  snp_subset <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>snp<span class="punctuation">,</span> seq_ID <span class="operator">==</span> seq_ID_n<span class="punctuation">)</span></span><br><span class="line">  seq_end <span class="operator">&lt;-</span> seq_split</span><br><span class="line">  snp_num <span class="operator">&lt;-</span> 0</span><br><span class="line">  </span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="number">1</span><span class="operator">:</span>nrow<span class="punctuation">(</span>snp_subset<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    <span class="keyword">if</span> <span class="punctuation">(</span>snp_subset<span class="punctuation">[</span>i<span class="punctuation">,</span><span class="string">&#x27;seq_site&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;=</span> seq_end<span class="punctuation">)</span> snp_num <span class="operator">&lt;-</span> snp_num <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">    <span class="keyword">else</span> <span class="punctuation">&#123;</span></span><br><span class="line">      snp_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>snp_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>seq_ID_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_end<span class="punctuation">,</span> snp_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">      </span><br><span class="line">      seq_end <span class="operator">&lt;-</span> seq_end <span class="operator">+</span> seq_split</span><br><span class="line">      snp_num <span class="operator">&lt;-</span> 0</span><br><span class="line">      <span class="keyword">while</span> <span class="punctuation">(</span>snp_subset<span class="punctuation">[</span>i<span class="punctuation">,</span><span class="string">&#x27;seq_site&#x27;</span><span class="punctuation">]</span> <span class="operator">&gt;</span> seq_end<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">        snp_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>snp_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>seq_ID_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_end<span class="punctuation">,</span> snp_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">        seq_end <span class="operator">&lt;-</span> seq_end <span class="operator">+</span> seq_split</span><br><span class="line">      <span class="punctuation">&#125;</span></span><br><span class="line">      snp_num <span class="operator">&lt;-</span> snp_num <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">  <span class="keyword">while</span> <span class="punctuation">(</span>seq_end <span class="operator">&lt;</span> seq_stat<span class="punctuation">[</span>seq_ID_n<span class="punctuation">,</span><span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    snp_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>snp_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>seq_ID_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_end<span class="punctuation">,</span> snp_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    seq_end <span class="operator">&lt;-</span> seq_end <span class="operator">+</span> seq_split</span><br><span class="line">    snp_num <span class="operator">&lt;-</span> 0</span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  snp_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>snp_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>seq_ID_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_stat<span class="punctuation">[</span>seq_ID_n<span class="punctuation">,</span><span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">]</span><span class="punctuation">,</span> snp_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">snp_stat <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>snp_stat<span class="punctuation">,</span> stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line"><span class="built_in">names</span><span class="punctuation">(</span>snp_stat<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;seq_ID&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_start&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;snp_num&#x27;</span><span class="punctuation">)</span></span><br><span class="line">snp_stat<span class="operator">$</span>seq_start <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>snp_stat<span class="operator">$</span>seq_start<span class="punctuation">)</span></span><br><span class="line">snp_stat<span class="operator">$</span>seq_end <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>snp_stat<span class="operator">$</span>seq_end<span class="punctuation">)</span></span><br><span class="line">snp_stat<span class="operator">$</span>snp_num <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>snp_stat<span class="operator">$</span>snp_num<span class="punctuation">)</span></span><br><span class="line">snp_stat</span><br><span class="line"></span><br><span class="line"><span class="comment">#读取 SNP 文件，统计 InDel 长度</span></span><br><span class="line">snp <span class="operator">&lt;-</span> read.delim<span class="punctuation">(</span>SNP_file<span class="punctuation">,</span> header <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> colClasses <span class="operator">=</span> <span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span> comment.char <span class="operator">=</span> <span class="string">&#x27;#&#x27;</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">4</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">,</span> <span class="number">8</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">snp<span class="operator">$</span>V5 <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>snp<span class="operator">$</span>V5<span class="punctuation">)</span></span><br><span class="line">snp<span class="operator">$</span>change <span class="operator">&lt;-</span> str_c<span class="punctuation">(</span>snp<span class="operator">$</span>V7<span class="punctuation">,</span> snp<span class="operator">$</span>V8<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">indel <span class="operator">&lt;-</span> read.delim<span class="punctuation">(</span>SNP_file<span class="punctuation">,</span> header <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> colClasses <span class="operator">=</span> <span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span> comment.char <span class="operator">=</span> <span class="string">&#x27;#&#x27;</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">4</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">,</span> <span class="number">8</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">indel<span class="operator">$</span>V5 <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>indel<span class="operator">$</span>V5<span class="punctuation">)</span></span><br><span class="line">indel<span class="operator">$</span><span class="built_in">length</span> <span class="operator">&lt;-</span> str_length<span class="punctuation">(</span>indel<span class="punctuation">[</span> <span class="punctuation">,</span><span class="number">4</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="operator">-</span> str_length<span class="punctuation">(</span>indel<span class="punctuation">[</span> <span class="punctuation">,</span><span class="number">3</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">indel_insert <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>indel<span class="operator">$</span><span class="built_in">length</span> <span class="operator">&gt;</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">indel_delet <span class="operator">&lt;-</span> <span class="built_in">length</span><span class="punctuation">(</span>which<span class="punctuation">(</span>indel<span class="operator">$</span><span class="built_in">length</span> <span class="operator">&lt;</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#统计 InDel 密度</span></span><br><span class="line">indel <span class="operator">&lt;-</span> indel<span class="punctuation">[</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">colnames<span class="punctuation">(</span>indel<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">2</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;seq_ID&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_site&#x27;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">indel_stat <span class="operator">&lt;-</span> <span class="literal">NULL</span></span><br><span class="line">seq_ID <span class="operator">&lt;-</span> unique<span class="punctuation">(</span>indel<span class="operator">$</span>seq_ID<span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>seq_ID_n <span class="keyword">in</span> seq_ID<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  indel_subset <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>indel<span class="punctuation">,</span> seq_ID <span class="operator">==</span> seq_ID_n<span class="punctuation">)</span></span><br><span class="line">  seq_end <span class="operator">&lt;-</span> seq_split</span><br><span class="line">  indel_num <span class="operator">&lt;-</span> 0</span><br><span class="line">  </span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="number">1</span><span class="operator">:</span>nrow<span class="punctuation">(</span>indel_subset<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    <span class="keyword">if</span> <span class="punctuation">(</span>indel_subset<span class="punctuation">[</span>i<span class="punctuation">,</span><span class="string">&#x27;seq_site&#x27;</span><span class="punctuation">]</span> <span class="operator">&lt;=</span> seq_end<span class="punctuation">)</span> indel_num <span class="operator">&lt;-</span> indel_num <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">    <span class="keyword">else</span> <span class="punctuation">&#123;</span></span><br><span class="line">      indel_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>indel_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>seq_ID_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_end<span class="punctuation">,</span> indel_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">      </span><br><span class="line">      seq_end <span class="operator">&lt;-</span> seq_end <span class="operator">+</span> seq_split</span><br><span class="line">      indel_num <span class="operator">&lt;-</span> 0</span><br><span class="line">      <span class="keyword">while</span> <span class="punctuation">(</span>indel_subset<span class="punctuation">[</span>i<span class="punctuation">,</span><span class="string">&#x27;seq_site&#x27;</span><span class="punctuation">]</span> <span class="operator">&gt;</span> seq_end<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">        indel_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>indel_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>seq_ID_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_end<span class="punctuation">,</span> indel_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">        seq_end <span class="operator">&lt;-</span> seq_end <span class="operator">+</span> seq_split</span><br><span class="line">      <span class="punctuation">&#125;</span></span><br><span class="line">      indel_num <span class="operator">&lt;-</span> indel_num <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  </span><br><span class="line">  <span class="keyword">while</span> <span class="punctuation">(</span>seq_end <span class="operator">&lt;</span> seq_stat<span class="punctuation">[</span>seq_ID_n<span class="punctuation">,</span><span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    indel_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>indel_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>seq_ID_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_end<span class="punctuation">,</span> indel_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    seq_end <span class="operator">&lt;-</span> seq_end <span class="operator">+</span> seq_split</span><br><span class="line">    indel_num <span class="operator">&lt;-</span> 0</span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  indel_stat <span class="operator">&lt;-</span> rbind<span class="punctuation">(</span>indel_stat<span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>seq_ID_n<span class="punctuation">,</span> seq_end <span class="operator">-</span> seq_split <span class="operator">+</span> <span class="number">1</span><span class="punctuation">,</span> seq_stat<span class="punctuation">[</span>seq_ID_n<span class="punctuation">,</span><span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">]</span><span class="punctuation">,</span> indel_num<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">indel_stat <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>indel_stat<span class="punctuation">,</span> stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line"><span class="built_in">names</span><span class="punctuation">(</span>indel_stat<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;seq_ID&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_start&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;seq_end&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;indel_num&#x27;</span><span class="punctuation">)</span></span><br><span class="line">indel_stat<span class="operator">$</span>seq_start <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>indel_stat<span class="operator">$</span>seq_start<span class="punctuation">)</span></span><br><span class="line">indel_stat<span class="operator">$</span>seq_end <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>indel_stat<span class="operator">$</span>seq_end<span class="punctuation">)</span></span><br><span class="line">indel_stat<span class="operator">$</span>indel_num <span class="operator">&lt;-</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span>indel_stat<span class="operator">$</span>indel_num<span class="punctuation">)</span></span><br><span class="line">indel_stat</span><br><span class="line"></span><br><span class="line">pdf<span class="punctuation">(</span>str_c<span class="punctuation">(</span>out_dir<span class="punctuation">,</span> sample_name<span class="punctuation">,</span> <span class="string">&#x27;.circlize.pdf&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> width <span class="operator">=</span> <span class="number">40</span><span class="punctuation">,</span> height <span class="operator">=</span> <span class="number">28</span><span class="punctuation">)</span></span><br><span class="line">circle_size <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="string">&#x27;snpc&#x27;</span><span class="punctuation">)</span></span><br><span class="line">circos.par<span class="punctuation">(</span>gap.degree <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> start.degree <span class="operator">=</span> <span class="number">90</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#第一圈，染色体区域</span></span><br><span class="line">circos.genomicInitialize<span class="punctuation">(</span>seq_stat<span class="punctuation">,</span> plotType <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;axis&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;labels&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> major.by <span class="operator">=</span> <span class="number">250000</span><span class="punctuation">,</span> track.height <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">)</span><span class="comment">#, gap.degree=1</span></span><br><span class="line"></span><br><span class="line">circos.genomicTrackPlotRegion<span class="punctuation">(</span></span><br><span class="line">  seq_stat<span class="punctuation">,</span> track.height <span class="operator">=</span> <span class="number">0.05</span><span class="punctuation">,</span> stack <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> bg.border <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span></span><br><span class="line">  panel.fun <span class="operator">=</span> <span class="keyword">function</span><span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> ...<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    circos.genomicRect<span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> col <span class="operator">=</span> <span class="string">&#x27;#049a0b&#x27;</span><span class="punctuation">,</span> border <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span> ...<span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span> <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#第二圈，覆盖度 &amp; 深度</span></span><br><span class="line">circos.genomicTrack<span class="punctuation">(</span></span><br><span class="line">  depth_stat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span><span class="punctuation">,</span> track.height <span class="operator">=</span> <span class="number">0.08</span><span class="punctuation">,</span>  ylim <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="built_in">max</span><span class="punctuation">(</span>depth_stat<span class="operator">$</span>depth<span class="punctuation">)</span> <span class="operator">+</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> bg.col <span class="operator">=</span> <span class="string">&#x27;#EEEEEE6E&#x27;</span><span class="punctuation">,</span> bg.border <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span></span><br><span class="line">  panel.fun <span class="operator">=</span> <span class="keyword">function</span><span class="punctuation">(</span>region<span class="punctuation">,</span>value<span class="punctuation">,</span> ...<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    circos.genomicRect<span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> ytop.column <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span> ybottom <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span> border <span class="operator">=</span> <span class="string">&#x27;white&#x27;</span><span class="punctuation">,</span> lwd <span class="operator">=</span> <span class="number">0.2</span><span class="punctuation">,</span> col <span class="operator">=</span> <span class="string">&#x27;red&#x27;</span><span class="punctuation">,</span> ...<span class="punctuation">)</span></span><br><span class="line">    circos.lines<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="built_in">max</span><span class="punctuation">(</span>region<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span>average_depth<span class="punctuation">,</span> average_depth<span class="punctuation">)</span><span class="punctuation">,</span> col <span class="operator">=</span> <span class="string">&#x27;red3&#x27;</span><span class="punctuation">,</span> lwd<span class="operator">=</span><span class="number">0.15</span><span class="punctuation">,</span> lty <span class="operator">=</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">    circos.yaxis<span class="punctuation">(</span> lwd <span class="operator">=</span> <span class="number">0.1</span><span class="punctuation">,</span> tick.length <span class="operator">=</span> convert_x<span class="punctuation">(</span><span class="number">0.15</span><span class="punctuation">,</span> <span class="string">&#x27;mm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span>labels.cex <span class="operator">=</span> <span class="number">0.3</span><span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span> <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#第三圈，CDS &amp; rRNA &amp; tRNA</span></span><br><span class="line">color_assign <span class="operator">&lt;-</span> colorRamp2<span class="punctuation">(</span>breaks <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">)</span><span class="punctuation">,</span> col <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;#00ADFF&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;orange&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;green2&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">circos.genomicTrackPlotRegion<span class="punctuation">(</span></span><br><span class="line">  gene<span class="punctuation">,</span> track.height <span class="operator">=</span> <span class="number">0.16</span><span class="punctuation">,</span> stack <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> bg.border <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span></span><br><span class="line">  panel.fun <span class="operator">=</span> <span class="keyword">function</span><span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> ...<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    circos.genomicRect<span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> col <span class="operator">=</span> color_assign<span class="punctuation">(</span>value<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span> border <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span> ...<span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span> <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#第四圈，SNP 密度</span></span><br><span class="line">value_max <span class="operator">&lt;-</span> <span class="built_in">max</span><span class="punctuation">(</span>snp_stat<span class="operator">$</span>snp_num<span class="punctuation">)</span></span><br><span class="line">colorsChoice <span class="operator">&lt;-</span> colorRampPalette<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;white&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#245B8E&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">color_assign <span class="operator">&lt;-</span> colorRamp2<span class="punctuation">(</span>breaks <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="operator">:</span>value_max<span class="punctuation">)</span><span class="punctuation">,</span> col <span class="operator">=</span> colorsChoice<span class="punctuation">(</span>value_max <span class="operator">+</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">circos.genomicTrackPlotRegion<span class="punctuation">(</span></span><br><span class="line">  snp_stat<span class="punctuation">,</span> track.height <span class="operator">=</span> <span class="number">0.08</span><span class="punctuation">,</span> stack <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> bg.border <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span></span><br><span class="line">  panel.fun <span class="operator">=</span> <span class="keyword">function</span><span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> ...<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    circos.genomicRect<span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> col <span class="operator">=</span> color_assign<span class="punctuation">(</span>value<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span> border <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span> ...<span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span> <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#第五圈，InDel 密度</span></span><br><span class="line">value_max <span class="operator">&lt;-</span> <span class="built_in">max</span><span class="punctuation">(</span>indel_stat<span class="operator">$</span>indel_num<span class="punctuation">)</span></span><br><span class="line">colorsChoice <span class="operator">&lt;-</span> colorRampPalette<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;white&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#7744A4&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">color_assign <span class="operator">&lt;-</span> colorRamp2<span class="punctuation">(</span>breaks <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">0</span><span class="operator">:</span>value_max<span class="punctuation">)</span><span class="punctuation">,</span> col <span class="operator">=</span> colorsChoice<span class="punctuation">(</span>value_max <span class="operator">+</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">circos.genomicTrackPlotRegion<span class="punctuation">(</span></span><br><span class="line">  indel_stat<span class="punctuation">,</span> track.height <span class="operator">=</span> <span class="number">0.08</span><span class="punctuation">,</span> stack <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> bg.border <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span></span><br><span class="line">  panel.fun <span class="operator">=</span> <span class="keyword">function</span><span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> ...<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    circos.genomicRect<span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> col <span class="operator">=</span> color_assign<span class="punctuation">(</span>value<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span> border <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span> ...<span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span> <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#第六圈, SNV分布图</span></span><br><span class="line">color_assign <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;#BC80BD&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#FDB462&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#80B1D3&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#FB8072&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#8DD3C7&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#FFFFB3&#x27;</span><span class="punctuation">)</span></span><br><span class="line">pch_assign <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">16</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">circos.genomicTrackPlotRegion<span class="punctuation">(</span></span><br><span class="line">  anno_list<span class="punctuation">,</span> track.height <span class="operator">=</span> <span class="number">0.08</span><span class="punctuation">,</span> bg.border <span class="operator">=</span> <span class="string">&#x27;black&#x27;</span><span class="punctuation">,</span> bg.lwd <span class="operator">=</span> <span class="number">0.4</span><span class="punctuation">,</span></span><br><span class="line">  panel.fun <span class="operator">=</span> <span class="keyword">function</span><span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> ...<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  n <span class="operator">=</span> getI<span class="punctuation">(</span>...<span class="punctuation">)</span></span><br><span class="line">  circos.genomicPoints<span class="punctuation">(</span>region<span class="punctuation">,</span> value<span class="punctuation">,</span> cex <span class="operator">=</span> <span class="number">1.0</span><span class="punctuation">,</span> pch <span class="operator">=</span> pch_assign<span class="punctuation">[</span>ifelse<span class="punctuation">(</span>n<span class="operator">%%</span><span class="number">2</span><span class="punctuation">,</span> n<span class="operator">%%</span><span class="number">2</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">,</span> col <span class="operator">=</span> color_assign<span class="punctuation">[</span>ifelse<span class="punctuation">(</span>n<span class="operator">%%</span><span class="number">6</span><span class="punctuation">,</span> n<span class="operator">%%</span><span class="number">6</span><span class="punctuation">,</span> <span class="number">6</span><span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">,</span> ...<span class="punctuation">)</span></span><br><span class="line">  circos.yaxis<span class="punctuation">(</span>labels.cex <span class="operator">=</span> <span class="number">0.3</span><span class="punctuation">,</span> lwd <span class="operator">=</span> <span class="number">0.1</span><span class="punctuation">,</span> tick.length <span class="operator">=</span> convert_x<span class="punctuation">(</span><span class="number">0.15</span><span class="punctuation">,</span> <span class="string">&#x27;mm&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  xlim <span class="operator">=</span> CELL_META<span class="operator">$</span>xlim</span><br><span class="line">  ylim <span class="operator">=</span> CELL_META<span class="operator">$</span>ylim</span><br><span class="line">  <span class="punctuation">&#125;</span> <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#draw.sector(start.degree, end.degree, clock.wise = FALSE)#突出显示</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#测序深度、覆盖度图例</span></span><br><span class="line">depth_legend <span class="operator">&lt;-</span> Legend<span class="punctuation">(</span></span><br><span class="line">  at <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span> labels <span class="operator">=</span> str_c<span class="punctuation">(</span><span class="string">&#x27; Depth ( average: &#x27;</span><span class="punctuation">,</span> average_depth<span class="punctuation">,</span> <span class="string">&#x27; X )&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> labels_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">8</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  title <span class="operator">=</span> str_c<span class="punctuation">(</span><span class="string">&#x27;Coverage: &#x27;</span><span class="punctuation">,</span> coverage <span class="punctuation">,</span> <span class="string">&#x27; %&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> title_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">9</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  grid_height <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0.4</span><span class="punctuation">,</span> <span class="string">&#x27;cm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> grid_width <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0.4</span><span class="punctuation">,</span> <span class="string">&#x27;cm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&#x27;points&#x27;</span><span class="punctuation">,</span> pch <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span> background <span class="operator">=</span> <span class="string">&#x27;red&#x27;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#CDS &amp; rRNA &amp; tRNA 图例</span></span><br><span class="line">gene_legend <span class="operator">&lt;-</span> Legend<span class="punctuation">(</span></span><br><span class="line">  at <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">3</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span> labels <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27; CDS&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27; rRNA&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27; tRNA&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> labels_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">8</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  title <span class="operator">=</span> <span class="string">&#x27;CDS | rRNA | tRNA&#x27;</span><span class="punctuation">,</span> title_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">9</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">  grid_height <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0.4</span><span class="punctuation">,</span> <span class="string">&#x27;cm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> grid_width <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0.4</span><span class="punctuation">,</span> <span class="string">&#x27;cm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&#x27;points&#x27;</span><span class="punctuation">,</span> pch <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span> background <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;#00ADFF&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;orange&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;green2&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#SNP 密度图例</span></span><br><span class="line">snp_legend <span class="operator">&lt;-</span> Legend<span class="punctuation">(</span></span><br><span class="line">  at <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>seq<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="built_in">max</span><span class="punctuation">(</span>snp_stat<span class="operator">$</span>snp_num<span class="punctuation">)</span><span class="punctuation">,</span> length.out <span class="operator">=</span> <span class="number">6</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> labels_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">8</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  col_fun <span class="operator">=</span> colorRamp2<span class="punctuation">(</span><span class="built_in">round</span><span class="punctuation">(</span>seq<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="built_in">max</span><span class="punctuation">(</span>snp_stat<span class="operator">$</span>snp_num<span class="punctuation">)</span><span class="punctuation">,</span> length.out <span class="operator">=</span> <span class="number">6</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> colorRampPalette<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;white&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#245B8E&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">(</span><span class="number">6</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  title_position <span class="operator">=</span> <span class="string">&#x27;topleft&#x27;</span><span class="punctuation">,</span> title <span class="operator">=</span> <span class="string">&#x27;SNP density&#x27;</span><span class="punctuation">,</span> legend_height <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">4</span><span class="punctuation">,</span> <span class="string">&#x27;cm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> title_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">9</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#SNV 分布图例</span></span><br><span class="line">snv_legend <span class="operator">&lt;-</span> Legend<span class="punctuation">(</span></span><br><span class="line">  at <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">,</span> <span class="number">6</span><span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">,</span> <span class="number">8</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  labels <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27; SNP: A&gt;T|T&gt;A&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27; SNP: A&gt;G|T&gt;C&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27; SNP: A&gt;C|T&gt;G&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27; SNP: G&gt;A|C&gt;T&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27; SNP: G&gt;T|C&gt;A&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27; SNP: G&gt;C|C&gt;G&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;synonymous SNV&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;nonsynonymous SNV&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  labels_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">5</span><span class="punctuation">)</span><span class="punctuation">,</span> title <span class="operator">=</span> <span class="string">&#x27;variant type\n&#x27;</span><span class="punctuation">,</span> title_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">6</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  grid_height <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0.4</span><span class="punctuation">,</span> <span class="string">&#x27;cm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> grid_width <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0.4</span><span class="punctuation">,</span> <span class="string">&#x27;cm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&#x27;points&#x27;</span><span class="punctuation">,</span> background <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span></span><br><span class="line">  legend_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>col <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&#x27;#BC80BD&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#FDB462&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#80B1D3&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#FB8072&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#8DD3C7&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;#FFFFB3&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;gray&#x27;</span><span class="punctuation">,</span> <span class="string">&#x27;gray&#x27;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  pch <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">20</span><span class="punctuation">,</span> <span class="number">20</span><span class="punctuation">,</span> <span class="number">20</span><span class="punctuation">,</span> <span class="number">20</span><span class="punctuation">,</span> <span class="number">20</span><span class="punctuation">,</span> <span class="number">20</span><span class="punctuation">,</span> <span class="number">16</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span> <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#左侧统计总览（涵括了上文大部分的统计概况，例如 SNP 替换类型统计等）</span></span><br><span class="line">stat_legend <span class="operator">&lt;-</span> Legend<span class="punctuation">(</span></span><br><span class="line">  at <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span> labels <span class="operator">=</span> <span class="string">&#x27;1&#x27;</span><span class="punctuation">,</span> labels_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">0</span><span class="punctuation">)</span><span class="punctuation">,</span> title_gp <span class="operator">=</span> gpar<span class="punctuation">(</span>fontsize <span class="operator">=</span> <span class="number">9</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">  grid_height <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="string">&#x27;cm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> grid_width <span class="operator">=</span> unit<span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span> <span class="string">&#x27;cm&#x27;</span><span class="punctuation">)</span><span class="punctuation">,</span> type <span class="operator">=</span> <span class="string">&#x27;points&#x27;</span><span class="punctuation">,</span> pch <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span> background <span class="operator">=</span> <span class="literal">NA</span><span class="punctuation">,</span> </span><br><span class="line">  title <span class="operator">=</span> str_c<span class="punctuation">(</span><span class="string">&#x27;Sample: &#x27;</span><span class="punctuation">,</span> sample_name<span class="punctuation">,</span> <span class="string">&#x27;\nRefer species: &#x27;</span><span class="punctuation">,</span> ref_name<span class="punctuation">,</span> <span class="string">&#x27;\nRefer size: &#x27;</span><span class="punctuation">,</span> genome_size<span class="punctuation">,</span> <span class="string">&#x27; bp\n\n\nTotal SNP: &#x27;</span><span class="punctuation">,</span> snp_ti <span class="operator">+</span> snp_tv<span class="punctuation">,</span> <span class="string">&#x27;\nTransitions: &#x27;</span><span class="punctuation">,</span> snp_ti<span class="punctuation">,</span> <span class="string">&#x27;\nTransversions: &#x27;</span><span class="punctuation">,</span> snp_tv<span class="punctuation">,</span> <span class="string">&#x27;\nTi/Tv: &#x27;</span><span class="punctuation">,</span> <span class="built_in">round</span><span class="punctuation">(</span>snp_ti <span class="operator">/</span> snp_tv<span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&#x27;\nA&gt;T|T&gt;A: &#x27;</span><span class="punctuation">,</span> snp_at<span class="punctuation">,</span> <span class="string">&#x27;\nA&gt;G|T&gt;C: &#x27;</span><span class="punctuation">,</span> snp_ag<span class="punctuation">,</span> <span class="string">&#x27;\nA&gt;C|T&gt;G: &#x27;</span><span class="punctuation">,</span> snp_ac<span class="punctuation">,</span> <span class="string">&#x27;\nG&gt;A|C&gt;T: &#x27;</span><span class="punctuation">,</span> snp_ga<span class="punctuation">,</span> <span class="string">&#x27;\nG&gt;T|C&gt;A: &#x27;</span><span class="punctuation">,</span> snp_gt<span class="punctuation">,</span> <span class="string">&#x27;\nG&gt;C|C&gt;G: &#x27;</span><span class="punctuation">,</span> snp_gc<span class="punctuation">,</span> <span class="string">&#x27;\n\n\nTotal InDel: &#x27;</span><span class="punctuation">,</span> indel_insert <span class="operator">+</span> indel_delet<span class="punctuation">,</span> <span class="string">&#x27;\nInsert: &#x27;</span><span class="punctuation">,</span> indel_insert<span class="punctuation">,</span> <span class="string">&#x27;\nDelet: &#x27;</span><span class="punctuation">,</span> indel_delet<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#图例存放位置</span></span><br><span class="line">y_coord <span class="operator">&lt;-</span> 1.01</span><br><span class="line">x_coord <span class="operator">&lt;-</span> 0.275</span><br><span class="line"></span><br><span class="line">pushViewport<span class="punctuation">(</span>viewport<span class="punctuation">(</span>x <span class="operator">=</span> x_coord<span class="operator">+</span><span class="number">0.05</span><span class="punctuation">,</span> y <span class="operator">=</span> y_coord<span class="operator">-</span><span class="number">0.208</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">grid.draw<span class="punctuation">(</span>stat_legend<span class="punctuation">)</span></span><br><span class="line">upViewport<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">pushViewport<span class="punctuation">(</span>viewport<span class="punctuation">(</span>x <span class="operator">=</span> x_coord<span class="operator">-</span><span class="number">0.0005</span> <span class="punctuation">,</span> y <span class="operator">=</span> y_coord<span class="operator">-</span><span class="number">0.165</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">grid.draw<span class="punctuation">(</span>depth_legend<span class="punctuation">)</span></span><br><span class="line">upViewport<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">pushViewport<span class="punctuation">(</span>viewport<span class="punctuation">(</span>x <span class="operator">=</span> x_coord<span class="operator">-</span><span class="number">0.005</span> <span class="punctuation">,</span> y <span class="operator">=</span> y_coord<span class="operator">-</span><span class="number">0.197</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">grid.draw<span class="punctuation">(</span>gene_legend<span class="punctuation">)</span></span><br><span class="line">upViewport<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">pushViewport<span class="punctuation">(</span>viewport<span class="punctuation">(</span>x <span class="operator">=</span> x_coord<span class="operator">-</span><span class="number">0.0097</span> <span class="punctuation">,</span> y <span class="operator">=</span> y_coord<span class="operator">-</span><span class="number">0.25</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">grid.draw<span class="punctuation">(</span>snp_legend<span class="punctuation">)</span></span><br><span class="line">y_coord <span class="operator">&lt;-</span> y_coord</span><br><span class="line">upViewport<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">pushViewport<span class="punctuation">(</span>viewport<span class="punctuation">(</span>x <span class="operator">=</span> x_coord<span class="operator">-</span><span class="number">0.0097</span> <span class="punctuation">,</span> y <span class="operator">=</span> y_coord<span class="operator">-</span><span class="number">0.3</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">grid.draw<span class="punctuation">(</span>indel_legend<span class="punctuation">)</span></span><br><span class="line">upViewport<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">pushViewport<span class="punctuation">(</span>viewport<span class="punctuation">(</span>x <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">,</span> y <span class="operator">=</span> <span class="number">0.5</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">grid.draw<span class="punctuation">(</span>snv_legend<span class="punctuation">)</span></span><br><span class="line">upViewport<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">circos.clear<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">dev.off<span class="punctuation">(</span><span class="punctuation">)</span><span class="comment">#保存</span></span><br></pre></td></tr></table></figure>

<p>运行结果如图所示，由外向内依次是：</p>
<ul>
<li>左上方图例和统计信息</li>
<li>基因组</li>
<li>染色体</li>
<li>测序深度与覆盖度</li>
<li>编码区和<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E9%9D%9E%E7%BC%96%E7%A0%81%E5%8C%BA&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">非编码区</a></li>
<li>SNV和InDel密度</li>
<li>SNV分布</li>
<li>SNV类型</li>
</ul>
<p><img src="https://pic1.zhimg.com/v2-015b406c0f4b81f31bfd5c180b32798c_b.jpg" alt="img"></p>
<p>全基因组测序圈图</p>
<p><img src="https://pic4.zhimg.com/v2-bae7e95a80a91430a351d587f163290f_b.jpg" alt="img"></p>
<p>全基因组测序圈图参数</p>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
	<div class="article-tag tagcloud">
		<i class="icon-price-tags icon"></i>
		<ul class="article-tag-list">
			 
        		<li class="article-tag-list-item">
        			<a href="javascript:void(0)" class="js-tag article-tag-list-link color3">BioInfo</a>
        		</li>
      		
		</ul>
	</div>

    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/17/WES%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B-%E4%B8%8B/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-WES数据分析流程-上" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/17/WES%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B-%E4%B8%8A/" target="_blank">WES数据分析流程(上)</a>
  
    </h1>
  


  
  
<a href="/2024/03/17/WES%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B-%E4%B8%8A/" class="archive-article-date">
        <time datetime="2024-03-17T05:09:57.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-17</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <p>[TOC]</p>
<p>来源：<a target="_blank" rel="noopener" href="https://www.zhihu.com/people/ma-pan-13-72">多啦A梦詹</a>；<a target="_blank" rel="noopener" href="https://www.zhihu.com/question/432826287/answer/3080622038">菩年</a>;<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/380684876">Die Linke</a></p>
<hr>
<ol>
<li>WES，WGS和WGRS的区别和联系是什么？</li>
</ol>
<ul>
<li><strong>WES (whole exome sequencing, <a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%85%A8%E5%A4%96%E6%98%BE%E5%AD%90%E7%BB%84%E6%B5%8B%E5%BA%8F&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">全外显子组测序</a>)</strong>: 是指利用序列捕获技术将【全基因组的外显子区域DNA】捕获富集后进行高通量测序，能够直接发现与蛋白质功能变异相关的遗传变异SNP(单核苷酸多态性)。以<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">人类基因组</a>为例，虽然外显子(蛋白质编码区)只占基因组的1%，但人类基因组85%的致病突变都在外显子区域，因此具有重要意义。</li>
<li><strong>WGS (whole genome sequencing, 全基因组测序)</strong>: 是指对基因组整体进行<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E9%AB%98%E9%80%9A%E9%87%8F%E6%B5%8B%E5%BA%8F&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">高通量测序</a>，分析不同个体间的差异，同时完成SNP及基因组结构注释。</li>
<li><strong>WGRS (whole genome re-sequencing, <a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%85%A8%E5%9F%BA%E5%9B%A0%E7%BB%84%E9%87%8D%E6%B5%8B%E5%BA%8F&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">全基因组重测序</a>)</strong>: 是指对已知参考基因组和注释的物种进行不同个体间的全基因组测序(WGS)<strong>，</strong>并在此基础上对个体或群体进行<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%B7%AE%E5%BC%82%E6%80%A7%E5%88%86%E6%9E%90&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">差异性分析</a>，鉴定出与某类表型相关的SNP.</li>
</ul>
<p>通过比较可以发现，这三个测序技术<strong>其实都是在找基因组上的SNP(<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%8D%95%E6%A0%B8%E8%8B%B7%E9%85%B8%E5%A4%9A%E6%80%81%E6%80%A7&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">单核苷酸多态性</a>)</strong>, SNP分为四种：</p>
<ol>
<li>SNV(单核苷酸变异)</li>
<li>InDel(插入缺失)</li>
<li>SV(结构变异)</li>
<li>CNV(<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">拷贝数变异</a>)</li>
</ol>
<p>只是这三个技术的<strong>覆盖程度不同</strong>：</p>
<ul>
<li>WES：覆盖全基因组上的外显子区域</li>
<li>WGS：覆盖全基因组</li>
<li>WGRS：覆盖全基因组，是WGS在不同样本上的重复</li>
</ul>
<p>只要你掌握了其中一种的<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22article%22,%22sourceId%22:%22380684876%22%7D">测序分析</a>与可视化，剩下的两种其实是一样的，本文就以WES测序分析与可视化为例（原因是手头只有RNA-Seq的转录组数据，勉强可以用来做WES数据的替代品吧），进行WES&#x2F;WGS&#x2F;WGRS的分析教学。</p>
<p>外显子是基因组中能够转录组出成熟RNA的部分。一个基因组中所有外显子的集合，即为<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%A4%96%E6%98%BE%E5%AD%90%E7%BB%84&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22answer%22,%22sourceId%22:%223080622038%22%7D">外显子组</a>。值得注意的是，通常所说的全外显子组测序，是指针对蛋白编码基因的外显子，很少涉及非编码基因。</p>
<p>基因(gene)是DNA中含有特定遗传信息的一段<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E6%A0%B8%E8%8B%B7%E9%85%B8%E5%BA%8F%E5%88%97&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22answer%22,%22sourceId%22:%223080622038%22%7D">核苷酸序列</a>的总称，是具有遗传效应的DNA分子片段,是控制生物性状的基本遗传单位。人类基因区间的大小可从数百个bp至超过200万个bp不等。根据<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84%E8%AE%A1%E5%88%92&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22answer%22,%22sourceId%22:%223080622038%22%7D">人类基因组计划</a>(The Human Genome Project)估计，人类拥有20000-25000个蛋白编码基因。</p>
<p>基因组(genome)指一个生物体所包含DNA的全部遗传信息。基因组由基因区域和非编码区域组成。**人类的基因组大小约为30亿个碱基对(bp)(3GB)**，其中非编码区域占到绝大多数，编码蛋白质的区域仅占约2%左右。</p>
<p>外显子组(exome)是基因组中所有外显子的集合。人类拥有约18万个外显子，约占人类基因组的1%，即<strong>约3000万个bp(30MB)。</strong></p>
<p><img src="https://pica.zhimg.com/80/v2-0aa2c0929b7fc072f415fb3e52f9321a_1440w.webp" alt="img"></p>
<p>图：蛋白编码基因由内含子(非编码序列)和外显子(包括编码序列以及UTR区域)组成。要翻译有功能的蛋白，要进行以下步骤：基因从DNA转录为RNA前体，通过剪接形成成熟RNA，成熟RNA序列翻译成氨基酸链，以及蛋白质分子的翻译后修饰。</p>
<p>关于外显子，需要注意的一个特殊情况是非翻译区(UTR)。在mRNA的两侧分别存在5’UTR(前导序列)和3’UTR(尾部序列)，它们的作用分别是调控翻译的启动和终止。它们由外显子序列构成，但不会被翻译成氨基酸。 所以，并非所有外显子序列都会被翻译成氨基酸。</p>
<p><img src="https://pica.zhimg.com/80/v2-d72f78f1b03486e21ddc1bc10ccb5d68_1440w.webp" alt="img"></p>
<p>图：信使RNA前体(pre-mRNA)中的外显子。外显子既包括编码氨基酸的序列(红色)，也包括不被翻译的序列(灰色)。</p>
<p>对外显子组(基因组里的所有外显子)进行测序的方法，即为 全外显子组测序 (Whole-Exome Sequencing，WES)，也称为 外显子组测序、全<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%A4%96%E6%98%BE%E5%AD%90%E6%B5%8B%E5%BA%8F&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22answer%22,%22sourceId%22:%223080622038%22%7D">外显子测序</a>，全外测序 等。</p>
<p>全基因组测序(Whole-Genome Sequencing, WGS)是对整个基因组进行测序。靶向测序(Targeted-sequencing，也称Panel sequencing)是对选定的基因进行测序，通常有几十个至一千个基因不等。因而，从覆盖基因组的范围来说，全基因组测序&gt;全外显子组测序&gt;靶向测序。</p>
<p>全外测序可以视作一种特殊的靶向测序——它靶向的区域是基因组上的所有外显子。</p>
<p><img src="https://pic1.zhimg.com/80/v2-29a4d2c84fbc1e7621f1239b79b8f1ea_1440w.webp" alt="img"></p>
<p>panel测序有两种技术原理：杂交捕获测序和多重扩增子测序。全外是基于序列杂交原理实现的</p>
<p>需要特别说明的是对CNV的检测。使用全外检测CNV时，在杂交捕获过程中，由于各个外显子的杂交效率不同，故不同外显子的覆盖率差异会较大。当出现阳性结果时，无法判断是由于杂交未捕获到，还是由于缺失。 故使用全外检测CNV容易出现假阳性结果。一般情况下，全外测序不用于CNV的检测 ，但在癌症研究中，利用癌组织和癌旁组织对照，可以检测体细胞CNV。</p>
<p>人类全外显子组所占基因组比例不超过2%，但它包含了约85%与疾病相关的变异，因此在研究编码基因变异层面，全外测序是比全基因组测序更为经济高效的替代方法。全外测序适用于<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%AD%9F%E5%BE%B7%E5%B0%94&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22answer%22,%22sourceId%22:%223080622038%22%7D">孟德尔</a>疾病、肿瘤、复杂疾病等多个研究领域。对于表现出异质性的疾病，或者患者表现出多个系统受累的复杂疾病症状时，尤为适合使用全外测序。</p>
<p>例如在肿瘤临床检测中，寻求肺癌靶向治疗的患者通常会先做panel测序，因为与肺癌靶向治疗相关的基因是比较明确的，几十至一百多个基因的panel测序通常就可以满足需求。而对于寻求免疫治疗的患者，通常会使用全外或大panel测序，来评估肿瘤突变负荷(Tumor Mutational Burden, TMB)，TMB高的患者通常对免疫治疗有更好的响应。全外测序是业内公认的评估TMB的金标准。</p>
<p>*<strong>WES检测流程*</strong></p>
<p>一个WES测序的工作流程，大体可以分为这3个部分：文库制备，测序，生信分析。</p>
<p>文库制备通常包含这些步骤：样本处理，DNA提取，定量，建库，杂交捕获，扩增，质控。</p>
<p>测序，目前的仪器包括国外Illumina公司测序平台，以及<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%8D%8E%E5%A4%A7%E6%99%BA%E9%80%A0&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22answer%22,%22sourceId%22:%223080622038%22%7D">华大智造</a>国产测序平台等。</p>
<p>生信分析的流程通常包含这些步骤：质控，拼接比对，去重和重排，变异检测，降噪和过滤，注释等。常用的软件有FastQC，BWA，GATK，ANNOVAR等。</p>
<p>一个完整的全外显子组测序，从样本处理到完成数据分析，通常需要10天左右时间。</p>
<p><em><strong>捕获探针性能的评估指标</strong></em></p>
<p>市面上的厂商会提供现成的全外探针，我们可以根据自己的需要来选择，也可以定制更为符合自己需要的“全外+”探针。</p>
<p>如果选择现成的探针产品，除了关注下面的探针评价指标外，也需要考虑探针大小、探针长度、探针设计等是否符合我们的样品和研究需求。</p>
<p>如果选择定制的产品，则可以在现成的探针产品上加入自己感兴趣的区域，也可以完全自主设计。在设计探针时常用的参考数据库有RefSeq、CCDS、Ensembl、GENCODE和ClinVar等。除了设计捕获区域外，还可以要求对某些特别的区域增加探针密度以提高捕获效率。</p>
<p><strong>中靶率</strong></p>
<p>中靶率(On-target rate)是一个百分数，用来表示测序数据中有多少能够比对到目标区域上。</p>
<p>在基因组上有许多与外显子有同源性的部分(比如内含子和<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E5%9F%BA%E5%9B%A0%E9%97%B4%E5%8C%BA&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22answer%22,%22sourceId%22:%223080622038%22%7D">基因间区</a>)，在实际工作中，这些并不属于目标(外显子)的部分在杂交过程中也会被捕获下来。这种探针捕获到非目标区域片段的情况称为脱靶(off target)。脱靶的数据是无效的，不能用于后续分析，即这部分测序数据被浪费了。</p>
<p>同等情况下，中靶率越高，由于脱靶产生的浪费越少，这款探针越好。</p>
<p><strong>覆盖度</strong></p>
<p>覆盖度(coverage)经常是和深度一起出现的，比如“10X覆盖度”、“30X覆盖度”。比如，“10X覆盖度为90%”指测序数据比对到目标区域后，有90%的区域被覆盖了至少10次，或者说有90%的区域有至少10条reads覆盖。</p>
<p>如果覆盖度没有和深度一起出现，则可以理解为“1X覆盖度”。比如“覆盖度为95%”，指95%的目标区域有至少1条reads覆盖到。换言之，有5%的目标区域连1条覆盖到的reads都没有，它们在这次测序中完全没被测到，被漏掉了。</p>
<p>同等情况下，覆盖度越高，越少比例的目标区域被漏掉，这款探针越好。</p>
<p><strong>均一性</strong></p>
<p>目标区域内不同的位点被覆盖的情况是不同的。比如一次WES测序的平均深度是60X，很有可能有的位点深度为10X，有的为40X，有的为90X这样的情况。均一性(uniformity)越好，即这些位点各自的深度越接近平均深度。</p>
<p>在实际工作中，我们根据期望达到的目标测序深度来分配数据量，即决定了这次测序的平均深度(平均深度&#x3D;数据量&#x2F;探针大小)。当某个区域的实际测序深度高于目标深度时，造成数据的浪费;而当某个区域的实际测序深度低于目标深度时，我们可能会认为这部分数据质量不好而丢弃它，导致这一区域无测序数据。均一性优良的探针可以帮助减少这两种情况的发生。</p>
<p>Fold-80是用来评价均一性的指标。它的定义是，为确保80%的目标碱基达到平均深度所需的额外测序的倍数。Fold-80越低，捕获效率越高，测序浪费越少。理想情况下的Fold-80为1。</p>
<p>Fold-80越低，均一性越好，越能节约测序成本，这款探针越好。</p>
<p><strong>重复率</strong></p>
<p>重复率(Dup rate)指的是<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22answer%22,%22sourceId%22:%223080622038%22%7D">重复序列</a>(Duplicate reads)在总测序序列中的占比。由于这些重复序列不能带来额外信息，相反会影响变异检测结果准确性，因此需要在下游生信分析中去除这些重复序列。Dup rate越高，数据利用率越低，浪费的测序成本也就越多。</p>
<p>同等情况下，重复率越低，越能节省测序成本，这款探针越好</p>
<p><strong>WES对于涉及<a target="_blank" rel="noopener" href="https://www.zhihu.com/search?q=%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82&search_source=Entity&hybrid_search_source=Entity&hybrid_search_extra=%7B%22sourceType%22:%22answer%22,%22sourceId%22:%223080622038%22%7D">拷贝数变异</a>、非编码区变异和结构变异的疾病研究不适用。其次，在对目标区域进行捕获时, 存在捕获不均、捕获偏差等现象, 可以通过增加测序深度, 获得更多的序列信息进行统计分析, 以尽可能弥补这些偏差。</strong></p>
<p>首先创建一个项目目录wes和存放每个步骤生成文件的子目录raw, clean, align, genome, hg19_VCF</p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mkdir &#123;raw,clean,align,genome,hg19_VCF&#125;</span><br></pre></td></tr></table></figure>

<p>此流程包含两个raw外显子测序文件：wes.1.fq.gz，wes.2.fq.gz；存放于raw目录</p>
<h3 id="原始数据质控，用fastp"><a href="#原始数据质控，用fastp" class="headerlink" title="原始数据质控，用fastp"></a>原始数据质控，用fastp</h3><figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># 项目主目录下运行，在clean目录下生成两个过滤的fq文件</span><br><span class="line">nohup fastp -i raw/wes.1.fq.gz -o clean/wes.1.clean.fq.gz -I raw/wes.2.fq.gz -O clean/wes.2.clean.fq.gz &amp;</span><br></pre></td></tr></table></figure>

<h3 id="下载参考基因组（hg19）文件，存放于genome目录，并建立索引："><a href="#下载参考基因组（hg19）文件，存放于genome目录，并建立索引：" class="headerlink" title="下载参考基因组（hg19）文件，存放于genome目录，并建立索引："></a>下载参考基因组（hg19）文件，存放于genome目录，并建立索引：</h3><figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">for i in $(seq 1 22) X Y M;</span><br><span class="line">do </span><br><span class="line">nohup wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/chr$&#123;i&#125;.fa.gz &amp; </span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">for i in $(seq 1 22) X Y M;</span><br><span class="line">do cat chr$&#123;i&#125;.fa &gt;&gt; hg19.fa;</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">nohup bwa index -a bwtsw -p hg19 hg19.fa  &amp;</span><br></pre></td></tr></table></figure>

<h3 id="比对"><a href="#比对" class="headerlink" title="比对"></a>比对</h3><figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"># clean目录下运行此命令，在align目录下生成sam文件</span><br><span class="line">nohup bwa mem -t 4 -M -R &quot;@RG\tID:lane1\tPL:illumina\tLB:library\tSM:wes&quot; ../genome/hg19  wes.1.clean.fq.gz  wes.2.clean.fq.gz &gt;../align/wes.sam 2&gt;wes.bwa.align.log &amp;</span><br><span class="line">-t，线程数；</span><br><span class="line">  -M , -M 将 shorter split hits 标记为次优，以兼容 Picard markDuplicates 软件;</span><br><span class="line">  -R 接的是 Read Group的字符串信息，它是用来将比对的read进行分组的，这个信息对于我们后续对比对数据进行错误率分析和Mark duplicate时非常重要。</span><br><span class="line">  (1) ID，这是Read Group的分组ID，一般设置为测序的lane ID</span><br><span class="line">  (2) PL，指的是所用的测序平台</span><br><span class="line">  (3) SM，样本ID</span><br><span class="line">  (4) LB，测序文库的名字</span><br><span class="line">这些信息设置好之后，在RG字符串中要用制表符（\t）将它们分开</span><br><span class="line">samtools view -b -S wes.sam &gt; wes.bam  #sam2bam,便于存储</span><br><span class="line">samtools sort wes.bam -o wes.sorted.bam    #有顺序的排序，便于后面的操作</span><br><span class="line">samtools flagstat wes.sorted.bam &gt; wes.sorted.bam.flagstat    #统计比对信息</span><br><span class="line">1. QC pass的reads的数量为5002344 ，未通过QC的reads数量为0，意味着一共有5002344条reads；</span><br><span class="line">2.标记为secondary的read</span><br><span class="line">3. 嵌合体reads</span><br><span class="line">4.PCR 重复引起的reads</span><br><span class="line">5.  比对到参考基因组上的reads数量；</span><br><span class="line">6.  paired reads数据数量；</span><br><span class="line">7.  read1的数量；</span><br><span class="line">8.  read2 的数量；</span><br><span class="line">9.  正确地匹配到参考序列的reads数量；</span><br><span class="line">10.  一对reads都比对到了参考序列上的数量，但是并不一定比对到同一条染色体上；</span><br><span class="line">11. 一对reads中只有一条与参考序列相匹配的数量； </span><br><span class="line">12.  一对reads比对到不同染色体的数量；</span><br><span class="line">13.  一对reads比对到不同染色体的且比对质量值大于5的数量。</span><br></pre></td></tr></table></figure>

<h3 id="外显子区域覆盖度"><a href="#外显子区域覆盖度" class="headerlink" title="外显子区域覆盖度"></a>外显子区域覆盖度</h3><p>需要生成外显子interval文件，生成这个文件的前提又需要dict文件和外显子bed文件（此处用的是安捷伦外显子bed文件，也可以去UCSC下载）。</p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">gatk CreateSequenceDictionary -R hg19.fa -O hg19.dict</span><br><span class="line">gatk BedToIntervalList -I S31285117_Regions.bed -O Exon.Interval.bed -SD ./genome/hg19.dict</span><br><span class="line">gatk CollectHsMetrics -BI Exon.Interval.bed -TI Exon.Interval.bed -I wes.sorted.bam -O wes.stat.txt</span><br></pre></td></tr></table></figure>

<h3 id="标记PCR重复序列并建立索引"><a href="#标记PCR重复序列并建立索引" class="headerlink" title="标记PCR重复序列并建立索引"></a>标记PCR重复序列并建立索引</h3><p>以前用picard标记重复序列，现在这个工具全部整合到gatk中了</p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">nohup gatk --java-options &quot;-Xmx10G -Djava.io.tmpdir=./&quot; MarkDuplicates -I wes.sorted.bam -O wes.sorted.MarkDuplicates.bam -M wes.sorted.bam.metrics &gt; log.mark 2&gt;&amp;1 &amp;</span><br><span class="line">samtools view -f 1024 wes.sorted.MarkDuplicates.bam | less</span><br><span class="line">samtools index wes.sorted.MarkDuplicates.bam</span><br></pre></td></tr></table></figure>

<p>wes.sorted.bam.metrics有统计信息 wes.sorted.MarkDuplicates.bam创建索引文件，他的作用能够让我们可以随机访问这个文件中的任意位置，而且后面的步骤也要求这个bam文件一定要有索引。</p>
<h3 id="变异检测"><a href="#变异检测" class="headerlink" title="变异检测"></a>变异检测</h3><h3 id="重新校正碱基质量值（BQSR）"><a href="#重新校正碱基质量值（BQSR）" class="headerlink" title="重新校正碱基质量值（BQSR）"></a>重新校正碱基质量值（BQSR）</h3><p><strong>变异检测是一个极度依赖测序碱基质量值，因为这个质量值是衡量我们测序出来的这个碱基到底有多正确的重要指标</strong>。它来自于测序图像数据的base calling，因此，基本上是由测序仪和测序系统来决定的，计算出来的碱基质量值未必与真实结果统一。 BQSR（Base Quality Score Recalibration）这个步骤主要是通过机器学习的方法构建测序碱基的错误率模型，然后对这些碱基的质量值进行相应的调整。 这里包含了两个步骤： 第一步，BaseRecalibrator，这里计算出了所有需要进行重校正的read和特征值，然后把这些信息输出为一份校准表文件（wes.recal_data.table） 第二步，ApplyBQSR，这一步利用第一步得到的校准表文件（wes.recal_data.table）重新调整原来BAM文件中的碱基质量值，并使用这个新的质量值重新输出一份新的BAM文件。</p>
<h3 id="首先下载变异注释文件"><a href="#首先下载变异注释文件" class="headerlink" title="首先下载变异注释文件"></a>首先下载变异注释文件</h3><figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">wget  -c -r -nd -np -k -L -p ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/hg19</span><br><span class="line">下载文件均为压缩文件，需要解压才能使用</span><br><span class="line">samtools faidx hg19.fa</span><br><span class="line">GENOME=./genome/hg19.fa</span><br><span class="line">hg19_VCF=./hg19_VCF/</span><br><span class="line">gatk --java-options &quot;-Xmx10G -Djava.io.tmpdir=./&quot; BaseRecalibrator \</span><br><span class="line">-R $GENOME -I ./align/wes.sorted.MarkDuplicates.bam \</span><br><span class="line">--known-sites $hg19_VCF/1000G_phase1.indels.hg19.sites.vcf \</span><br><span class="line">--known-sites $hg19_VCF/Mills_and_1000G_gold_standard.indels.hg19.sites.vcf \</span><br><span class="line">--known-sites $hg19_VCF/dbsnp_138.hg19.vcf \</span><br><span class="line">-L S31285117_Regions.bed -O wes.recal_data.table</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">gatk --java-options &quot;-Xmx10G -Djava.io.tmpdir=./&quot; ApplyBQSR \</span><br><span class="line">-R $GENOME -I ./align/wes.sorted.MarkDuplicates.bam \</span><br><span class="line">-bqsr wes.recal_data.table -L S31285117_Regions.bed  \</span><br><span class="line">-O wes.sorted.MarkDuplicates.BQSR.bam</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># gatk AnalyzeCovariates -bqsr wes.recal_data.table -plots wes.recal_data.table.plot</span><br></pre></td></tr></table></figure>

<h3 id="变异检测-1"><a href="#变异检测-1" class="headerlink" title="变异检测"></a>变异检测</h3><p>HaplotypeCaller的应用有两种做法，差别只在于要不要在中间生成一个gVCF： （1）直接进行HaplotypeCaller，这适合于单样本，或者那种固定样本数量的情况，也就是只执行一次HaplotypeCaller。如果增加一个样本就得重新运行这个HaplotypeCaller，而这个时候算法需要重新去读取所有人的BAM文件，浪费大量时间； （2）每个样本先各自生成gVCF，然后再进行群体joint-genotype。gVCF全称是genome VCF，是每个样本用于变异检测的中间文件，格式类似于VCF，它把joint-genotype过程中所需的所有信息都记录在这里面，文件无论是大小还是数据量都远远小于原来的BAM文件。这样一旦新增加样本也不需要再重新去读取所有人的BAM文件了，只需为新样本生成一份gVCF，然后重新执行这个joint-genotype就行了。</p>
<p>推荐使用第二种，<strong>变异检测不是一个样本的事情，有越多的同类样本放在一起joint calling结果将会越准确，而如果样本足够多的话，在低测序深度的情况下也同样可以获得完整并且准确的结果。</strong></p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">第一种方法</span><br><span class="line">gatk --java-options &quot;-Xmx8G -Djava.io.tmpdir=./&quot; HaplotypeCaller \</span><br><span class="line">-R $GENOME -I wes.sorted.MarkDuplicates.BQSR.bam \</span><br><span class="line">-D $hg19_VCF/dbsnp_138.hg19.vcf -L S31285117_Regions.bed \</span><br><span class="line">-O wes.raw1.vcf</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">第二种方法</span><br><span class="line"># 1.生成中间文件gvcf</span><br><span class="line">gatk --java-options &quot;-Xmx8G -Djava.io.tmpdir=./&quot; HaplotypeCaller -R $GENOME \</span><br><span class="line">--emit-ref-confidence GVCF -I wes.sorted.MarkDuplicates.BQSR.bam \</span><br><span class="line">-D $hg19_VCF/dbsnp_138.hg19.vcf -L S31285117_Regions.bed  -O wes.gvcf</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># 2.通过gvcf检测变异</span><br><span class="line">gatk --java-options &quot;-Xmx8G -Djava.io.tmpdir=./&quot; GenotypeGVCFs \</span><br><span class="line">-R $GENOME -V wes.gvcf -L S31285117_Regions.bed  \</span><br><span class="line">-O wes.raw.vcf</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">若有多个样本的gvcf文件，运行gatk CombineGVCFs  \</span><br><span class="line">-V 1.gvcf –V 2.gvcf ……  -O final.gvcf ，再用final.gvcf运行第二步</span><br></pre></td></tr></table></figure>

<h3 id="变异质控与过滤"><a href="#变异质控与过滤" class="headerlink" title="变异质控与过滤"></a>变异质控与过滤</h3><p><strong>质控的含义和目的是指通过一定的标准，最大可能地剔除假阳性的结果，并尽可能地保留最多的正确数据。 第一种方法 GATK VQSR</strong>，它通过机器学习的方法利用多个不同的数据特征训练一个模型（高斯混合模型）对变异数据进行质控，使用VQSR需要具备以下两个条件： 第一，需要一个精心准备的已知变异集，它将作为训练质控模型的真集。比如，Hapmap、OMNI，1000G和dbsnp等这些国际性项目的数据，这些可以作为高质量的已知变异集。 第二，要求新检测的结果中有足够多的变异，不然VQSR在进行模型训练的时候会因为可用的变异位点数目不足而无法进行。</p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">gatk --java-options &quot;-Xmx8G -Djava.io.tmpdir=./&quot; VariantRecalibrator -R $GENOME -V wes.raw.vcf \</span><br><span class="line">-resource hapmap,known=false,training=true,truth=true,prior=15.0:$hg19_VCF/hapmap_3.3.hg19.sites.vcf \</span><br><span class="line">-resource omini,known=false,training=true,truth=false,prior=12.0:$hg19_VCF/1000G_omni2.5.hg19.sites.vcf \ </span><br><span class="line">-resource 1000G,known=false,training=true,truth=false,prior=10.0:$hg19_VCF/1000G_phase1.snps.high_confidence.hg19.sites.vcf \</span><br><span class="line">-resource dbsnp,known=true,training=false,truth=false,prior=6.0:$hg19_VCF/dbsnp_138.hg19.vcf \</span><br><span class="line">-an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR -an DP -mode SNP \</span><br><span class="line">-O wes.snps.recal.vcf --tranches-file wes.snps.tranches --rscript-file wes.snps.plots.R &amp; </span><br><span class="line"></span><br><span class="line"></span><br><span class="line">gatk  ApplyRecalibration  -V wes.raw.vcf -O wes.VQSR.vcf --recal-file wes.snps.recal.vcf  \</span><br><span class="line">--tranches-file wes.snps.tranches  -mode SNP</span><br></pre></td></tr></table></figure>

<p>此方法要求新检测的结果中有足够多的变异，不然VQSR在进行模型训练的时候会因为可用的变异位点数目不足而无法进行。可能很多非人的物种在完成变异检测之后没法使用GATK VQSR的方法进行质控，一些小panel、外显子测序，由于最后的变异位点不够，也无法使用VQSR。全基因组分析或多个样本的全外显子组分析适合用此方法。</p>
<p><strong>第二种方法通过过滤指标过滤。</strong> QualByDepth（QD）：QD是变异质量值（Quality）除以覆盖深度（Depth）得到的比值。 FisherStrand (FS)：FS是一个通过Fisher检验的p-value转换而来的值，它要描述的是测序或者比对时对于只含有变异的read以及只含有参考序列碱基的read是否存在着明显的正负链特异性（Strand bias，或者说是差异性） StrandOddsRatio (SOR)：对链特异（Strand bias）的一种描述. RMSMappingQuality (MQ)：MQ这个值是所有比对至该位点上的read的比对质量值的均方根. MappingQualityRankSumTest (MQRankSum) ReadPosRankSumTest (ReadPosRankSum) <strong>通过过滤指标过滤</strong></p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">使用SelectVariants，选出SNP</span><br><span class="line">gatk SelectVariants -select-type SNP -V wes.raw.vcf -O wes.snp.vcf</span><br><span class="line"># 为SNP作过滤</span><br><span class="line">gatk VariantFiltration -V wes.snp.vcf --filter-expression &quot;QD &lt; 2.0 || MQ &lt; 40.0 || FS &gt; 60.0 || SOR &gt; 3.0 || MQRankSum &lt; -12.5 || ReadPosRankSum &lt; -8.0&quot; --filter-name &quot;PASS&quot; -O wes.snp.filter.vcf</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">使用SelectVariants，选出Indel</span><br><span class="line">gatk SelectVariants -select-type INDEL -V wes.raw.vcf -O wes.indel.vcf</span><br><span class="line"># 为Indel作过滤</span><br><span class="line">gatk VariantFiltration -V wes.indel.vcf --filter-expression &quot;QD &lt; 2.0 || FS &gt; 200.0 || SOR &gt; 10.0 || MQRankSum &lt; -12.5 || ReadPosRankSum &lt; -8.0&quot; --filter-name &quot;PASS&quot; -O wes.indel.filter.vcf</span><br></pre></td></tr></table></figure>

<h3 id="突变注释"><a href="#突变注释" class="headerlink" title="突变注释"></a>突变注释</h3><p>ANNOVAR是一个perl编写的命令行工具，能在安装了perl解释器的多种操作系统上执行。允许多种输入文件格式，包括最常被使用的VCF格式。输出文件也有多种格式，包括注释过的VCF文件、用tab或者逗号分隔的txt文件，ANNOVAR能快速注释遗传变异并预测其功能。这个软件需要edu邮箱注册才能下载。 ANNOVAR website</p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">数据库下载</span><br><span class="line">perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar refGene humandb/</span><br><span class="line"> # -buildver 表示version </span><br><span class="line"> # -downdb 下载数据库的指令</span><br><span class="line"> # -webfrom annovar 从annovar提供的镜像下载，不加此参数将寻找数据库本身的源</span><br><span class="line"> # humandb/ 存放于humandb/目录下</span><br><span class="line">其它数据库下载</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb gwascatalog humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb ljb26_all --webfrom annovar humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb esp6500siv2_ea --webfrom annovar humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb esp6500siv2_all --webfrom annovar humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb 1000g2014oct humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb cytoBand humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb avsift -webfrom annovar humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb snp138 humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb genomicSuperDups humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb phastConsElements46way humandb/ &amp;</span><br><span class="line">perl annotate_variation.pl --buildver hg19 --downdb tfbs humandb/</span><br></pre></td></tr></table></figure>

<h3 id="三种注释方式"><a href="#三种注释方式" class="headerlink" title="三种注释方式"></a>三种注释方式</h3><p>**Gene-based Annotation(基于基因的注释)**：基于基因的注释（gene-based annotation）揭示variant与已知基因直接的关系以及对其产生的功能性影响。 <strong>Region-based Annotation（基于区域的注释）</strong>：基于过滤的注释精确匹配查询变异与数据库中的记录：如果它们有相同的染色体，起始位置，结束位置，REF的等位基因和ALT的等位基因，才能认为匹配。基于区域的注释看起来更像一个区域的查询（这个区域也可以是一个单一的位点），在一个数据库中，它不在乎位置的精确匹配，它不在乎核苷酸的识别。基于区域的注释（region-based annotation）揭示variant与不同基因组特定段的关系。 <strong>Filter-based Annotation（基于过滤的注释）</strong>：filter-based和region-based主要的区别是，filter-based针对mutation（核苷酸的变化）而region-based针对染色体上的位置。如在全基因组数据中的变异频率，可使用1000g2015aug、kaviar_20150923等数据库；在全外显组数据中的变异频率，可使用exac03、esp6500siv2等；在孤立的或者低代表人群中的变异频率，可使用ajews等数据库。 <strong>用table_annovar.pl进行注释，可一次性完成三种类型的注释。</strong></p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">convert2annovar.pl # 将多种格式转为.avinput的程序</span><br><span class="line">convert2annovar.pl -format vcf4   wes.snp.filter.vcf &gt;snp.avinput</span><br><span class="line">格式如下：</span><br><span class="line">chr1    69511   69511   A       G       hom     1736.03 53</span><br><span class="line">chr1    877831  877831  T       C       hom     85.10   4</span><br><span class="line">chr1    881627  881627  G       A       het     399.60  37</span><br><span class="line">chr1    887801  887801  A       G       hom     1623.03 39</span><br><span class="line">chr1    888639  888639  T       C       hom     3991.03 106</span><br><span class="line">chr1    888659  888659  T       C       hom     4114.03 109</span><br><span class="line">chr1    897325  897325  G       C       hom     3998.03 113</span><br><span class="line">chr1    898852  898852  C       T       het     647.60  38</span><br><span class="line">chr1    900505  900505  G       C       het     478.60  35</span><br><span class="line">chr1    906272  906272  A       C       het     540.60  30</span><br></pre></td></tr></table></figure>

<p>avinput文件由tab分割，最重要的地方为前5列，分别是: 1. 染色体(Chromosome) 2. 起始位置(Start) 3. 结束位置(End) 4. 参考等位基因(Reference Allele) 5. 替代等位基因(Alternative Allele) ANNOVAR主要也是利用前五列信息对数据库进行比对，注释变异。</p>
<h3 id="SNP注释："><a href="#SNP注释：" class="headerlink" title="SNP注释："></a>SNP注释：</h3><p><strong>用table_annovar.pl进行注释，可一次性完成三种类型的注释。</strong></p>
<figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">table_annovar.pl snp.avinput $humandb -buildver hg19 -out snpanno \</span><br><span class="line">-remove -protocol refGene,cytoBand,genomicSuperDups,esp6500siv2_all \</span><br><span class="line">-operation g,r,r,f -nastring . -csvout</span><br></pre></td></tr></table></figure>

<p>-buildver hg19 表示使用hg19版本 -out snpanno 表示输出文件的前缀为snpanno -remove 表示删除注释过程中的临时文件 -protocol 表示注释使用的数据库，用逗号隔开，且要注意顺序 -operation 表示对应顺序的数据库的类型（g代表gene-based、r代表region-based、f代表filter-based），用逗号隔开，注意顺序 -nastring . 表示用点号替代缺省的值 -csvout 表示最后输出.csv文件</p>
<h3 id="Indel注释同上"><a href="#Indel注释同上" class="headerlink" title="Indel注释同上"></a>Indel注释同上</h3><p>测试数据：链接：<a href="https://link.zhihu.com/?target=https://pan.baidu.com/s/16tX5GrCLBAov9k_GjYYOyA">https://pan.baidu.com/s/16tX5GrCLBAov9k_GjYYOyA</a> 提取码：9pab</p>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
	<div class="article-tag tagcloud">
		<i class="icon-price-tags icon"></i>
		<ul class="article-tag-list">
			 
        		<li class="article-tag-list-item">
        			<a href="javascript:void(0)" class="js-tag article-tag-list-link color3">BioInfo</a>
        		</li>
      		
		</ul>
	</div>

    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/17/WES%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B-%E4%B8%8A/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-使用CRISPR-Cas9的人类B细胞基因组工程" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/16/%E4%BD%BF%E7%94%A8CRISPR-Cas9%E7%9A%84%E4%BA%BA%E7%B1%BBB%E7%BB%86%E8%83%9E%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%B7%A5%E7%A8%8B/" target="_blank">使用CRISPR/Cas9的人类B细胞基因组工程</a>
  
    </h1>
  


  
  
<a href="/2024/03/16/%E4%BD%BF%E7%94%A8CRISPR-Cas9%E7%9A%84%E4%BA%BA%E7%B1%BBB%E7%BB%86%E8%83%9E%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%B7%A5%E7%A8%8B/" class="archive-article-date">
        <time datetime="2024-03-16T15:09:12.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-16</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <p>[TOC]</p>
<p>First Published: November 3rd, 2020</p>
<p>DOI:<a target="_blank" rel="noopener" href="https://dx.doi.org/10.3791/61855">10.3791&#x2F;61855</a></p>
<iframe src="//player.bilibili.com/player.html?aid=1101771635&bvid=BV14w4m1o7fh&cid=1472242106&p=1" scrolling="no" border="0" frameborder="no" framespacing="0" allowfullscreen="true"> </iframe>

<hr>
<h2 id="MyNotes"><a href="#MyNotes" class="headerlink" title="MyNotes:"></a>MyNotes:</h2><p>本文作者的研究工作集中于运用CRISPR&#x2F;Cas9基因编辑技术在人类B细胞中进行基因敲除（KO）和敲入（KI）操作，以便深入研究B细胞中特定基因的功能，并将其应用于B细胞疗法的开发。这项研究创新性地优化了从<strong>外周血单核细胞</strong>（PBMCs）中分离、扩增B细胞的流程，同时改进了CRISPR&#x2F;Cas9系统的转染效率和安全性。</p>
<p>实验过程中，作者首先采用了一套高效且标准化的步骤分离并大量扩增B细胞，使其能在7天内增加44倍，这对于后续基因编辑的成功至关重要。然后，利用<strong>电穿孔技术</strong>将CRISPR&#x2F;Cas9 mRNA与针对CD19基因的sgRNA导入B细胞，以实现CD19基因的精准敲除。实验结果显示，这种方法成功诱导了预期的双链断裂（DSB），并通过非同源末端连接（NHEJ）机制引起了<strong>CD19基因</strong>的失活，其有效性通过流式细胞术和TIDE（Tracking of Indels by Decomposition）分析得到验证，结果显示CD19蛋白的表达显著减少。</p>
<p>此外，作者还展示了如何利用CRISPR&#x2F;Cas9与重组腺相关病毒（rAAV6）载体结合，将绿色荧光蛋白（EGFP）基因特异整合到AAVS1位点，实现了高效的基因敲入。实验结果显示，经过工程化的B细胞在12天后能稳定表达EGFP，并且通过PCR和流式细胞术证实了基因整合的发生。</p>
<p>尽管如此，研究也指出CRISPR&#x2F;Cas9技术存在潜在的脱靶效应和染色体易位风险，因此作者强调了优化实验条件的重要性，比如选择合适的sgRNA以降低脱靶效应，并探讨了使用碱基编辑器作为更为精确的基因编辑手段的可能性。</p>
<p>具体实验步骤流程概括可以如下：</p>
<ol>
<li><p><strong>B细胞分离与激活</strong>：</p>
<ul>
<li>从外周血单核细胞（PBMCs）中分离B细胞，通过密度梯度离心法和负选择试剂盒进行纯化。</li>
<li>对分离出的B细胞进行体外扩增，每两天更换新鲜的激活因子，并严格控制细胞密度。</li>
</ul>
</li>
<li><p><strong>CRISPR&#x2F;Cas9介导的基因敲除（KO）</strong>：</p>
<ul>
<li>设计并合成针对目标基因（如CD19）的sgRNA。</li>
<li>制备CRISPR&#x2F;Cas9-sgRNA复合物，通过电穿孔技术将复合物导入B细胞。</li>
<li>经过电穿孔后的细胞在特定时间内静置和恢复，随后进行细胞培养。</li>
<li>在转染后的第5天，使用流式细胞术检测CD19表达水平，并通过TIDE分析确定DNA双链断裂后引发的Indel形成。</li>
</ul>
</li>
<li><p><strong>基因敲入（KI）实验</strong>：</p>
<ul>
<li>使用重组腺相关病毒（rAAV6）载体携带目的基因（如EGFP）和同源臂。</li>
<li>将CRISPR&#x2F;Cas9系统与rAAV6载体一同导入B细胞，诱导同源定向修复（HDR）介导的敲入。</li>
<li>工程化后的第12天，通过流式细胞术和PCR分析评估EGFP基因是否成功整合及表达。</li>
</ul>
</li>
</ol>
<p>整个实验流程旨在通过精细调控B细胞的分离、激活和基因编辑过程，高效地实现对B细胞基因组的精准工程改造，进而研究基因功能及其在B细胞治疗中的应用潜力。同时，实验关注了优化细胞健康和基因编辑效率的关键步骤，以及对可能存在的脱靶效应和染色体易位等问题的考量。</p>
<hr>
<h3 id="Summary"><a href="#Summary" class="headerlink" title="Summary"></a>Summary</h3><p>在这里，我们为CRISPR&#x2F;Cas9基础基因组工程提供详细的分步协议，用于基因敲除（KO）和敲击（KI），以研究B细胞中基因的生物功能和B细胞治疗的开发。</p>
<h3 id="Abstract"><a href="#Abstract" class="headerlink" title="Abstract"></a>Abstract</h3><p>B细胞是从造血干细胞中提取的淋巴细胞，是适应性免疫系统幽默手臂的关键组成部分。他们之所以成为细胞疗法的候选者，是因为它们容易与周围血液隔离，体外扩张能力，以及体内寿命。此外，它们的正常生物功能——产生大量的抗体——可以用来表达大量的治疗性蛋白质，如抗感染的重组抗体，或用于治疗酶病变的酶。在这里，我们提供详细的方法，将原人类B细胞从外周血单核细胞（PBMCs）中分离，并在体外激活&#x2F;扩大分离的B细胞。然后，我们演示了使用CRISPR&#x2F;Cas9系统在B细胞中针对内源基因的部位特定KO所涉及的步骤。这种方法可以有效地利用各种基因的KO，用于研究感兴趣的基因的生物功能。然后，我们演示了使用CRISPR&#x2F;Cas9系统以及重组的、与腺相关的病毒（rAAV）载体的步骤，以有效实现B细胞中转基因表达盒的现场特定集成。该协议共同提供了一个分步工程平台，可用于人类B细胞的初级研究基因的生物功能以及B细胞治疗的开发。</p>
<hr>
<h3 id="Introduction"><a href="#Introduction" class="headerlink" title="Introduction"></a>Introduction</h3><p>B细胞是来自造血干细胞的淋巴细胞系的子群。他们在适应性幽默免疫系统中发挥关键作用，产生大量的抗体来应对免疫挑战[1]。B细胞也是记忆B细胞和最终分化的长寿血浆细胞的前体，从而提供持久的幽默免疫力[2]。特别是等离子体细胞，在免疫细胞中独一无二，它们能够产生大量的特定抗体，同时存活多年或几十年。此外，从外周血液隔离的便利性使B细胞血统成为新细胞疗法[4]的极好候选者。</p>
<p>以前，随机整合方法，如那些使用扁桃体载体或<em>睡美人</em>转子，已经被用来工程B细胞的转基因传递和表达[5，6，7，8]。然而，这些方法的非特异性使得研究B细胞中特定基因的生物功能变得困难，并具有在治疗环境中插入突变和可变转基因表达和&#x2F;或沉默的固有风险。</p>
<p>CRISPR&#x2F;Cas9系统是一个强大的基因组工程工具，使研究人员能够精确编辑许多物种中各种细胞的基因组。最近，包括我们自己的两组已经成功地开发出了原始人类B细胞[9,10]的外活扩张和靶向基因组工程方法。我们将描述从白血病样本中净化原人类B细胞的过程。之后，我们将描述我们更新的 B 细胞扩展和激活孤立 B 细胞的协议。然后，我们将描述一个通过电波将CRISPR&#x2F;Cas9 mRNA与CD19 sgRNA一起引入激活的B细胞，从而敲除分化19（CD19）的簇、特定的B细胞受体和B细胞的标志的过程。</p>
<p>Cas9 mRNA 被翻译并绑定到 CD19 sgRNA，以形成 CRISPR&#x2F;Cas9-sgRNA 核糖核蛋白复合物 （RNP）。随后，复合物中的sgRNA导致Cas9在基因exon2上的目标序列中产生双链断裂（DSB）。细胞将通过引入或删除核苷酸来修复DSB，导致帧移位突变并导致基因被消灭。然后，我们将用流细胞测量测量CD19的损失，并通过分解（TIDE）分析对英德尔的形成进行跟踪分析。</p>
<hr>
<h3 id="代表结果解读"><a href="#代表结果解读" class="headerlink" title="代表结果解读"></a>代表结果解读</h3><p>更新的扩增和激活协议能够在7天内将B细胞快速扩增至44倍（图1；n &#x3D; 3个捐赠者）。在KO实验中，使用 Trypan 蓝染色测定的B细胞计数显示，电穿孔后24小时，对照组和CD19 KO样本的存活率均超过80%，但细胞回收略有下降（图2A；p ≥ 0.05，n &#x3D; 3个捐赠者）。这一结果表明，电穿孔对B细胞的整体健康略有影响。在转染后的第5天收集B细胞进行流式细胞术和TIDE分析。对照组和KO样本的代表性散点图显示，分别为14%和95%的CD19阴性细胞（图2B）。流式细胞术图的定量分析显示，与对照组相比，KO样本的CD19表达显著降低（图2B；p ≤ 0.0001，n &#x3D; 3个捐赠者）。</p>
<p>基因组测序的色谱图（见表3中的引物序列）在CD19 KO B细胞中显示双峰，表明在CRISPR&#x2F;Cas9介导的DSB后发生了核苷酸的插入&#x2F;缺失，而在对照组中观察到单峰，表明此样本中未发生DSB（图3A）。KO样本的Indel分析色谱图（使用一个免费的在线TIDE分析工具）在CD19位点上的Indel形成百分比很高（&gt;90%），这与流式细胞术检测到的CD19蛋白损失百分比一致（图3B；p ≥ 0.05，n &#x3D; 3个捐赠者）。这些结果表明，CRISPR&#x2F;Cas9有效地在B细胞中产生了CD19 KO。</p>
<p>KI实验的B细胞在工程化后的第12天收集，用于流式细胞术和连接PCR分析（表4）。接收rAAV6载体的样本的散点图显示EGFP阳性细胞占64%（图4），而在对照组中没有观察到EGFP阳性细胞；仅接收AAV载体的样本中观察到最小的EGFP阳性（图5A）。连接PCR扩增（见表3中的引物序列）在KI样本中显示出1.5 Kbps的扩增子（图5B），而在对照组或仅接收载体的样本中没有观察到PCR产物。细胞计数显示，工程化过程对KI样本的细胞回收影响大于对照组或仅接收载体的样本（图5B）。然而，所有样本在工程化后的3天内迅速恢复（图5B）。这些结果表明，成功在AAVS1位点整合EGFP至少在工程化后的12天内能够稳定表达EGFP。</p>
<hr>
<img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/61855fig01.jpg" alt="Figure 1" style="zoom:50%;">

<p>[^图1：体外B细胞扩张]: B细胞在第0天（零）在1个×106 个细胞播种，密度为5×每mL105， 并在7天内扩大44倍（n&#x3D;3独立捐赠者）。</p>
<img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/61855fig02a.jpg" alt="Figure 2A" style="zoom:50%;">

<p>[^图2A]: </p>
<img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/61855fig02b.jpg" alt="Figure 2B" style="zoom:50%;">

<p>[^图2B]: </p>
<img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/61855fig02c.jpg" alt="Figure 2C" style="zoom:50%;">



<p>[^图2C]: <strong>CRISPR&#x2F;Cas9介质CD19淘汰赛（KO）在B细胞中。</strong>（<strong>A</strong>） 条形图显示 &amp; gt;70% 细胞恢复（左面板）和 &gt;80% 的存活率 （右面板） 细胞在输电后 24 小时在控制和 CD19 KO 样本中观察到。（<strong>B</strong>） 活细胞CD19门控的代表流动图分别显示对照样本和CD19 KO样本中84.3%和3.43%的CD19阳性细胞。（<strong>C</strong>） 条形图显示 CRISPR&#x2F;Cas9 介控 CD19 KO 组 CD19 显著减少（p ≤ 0.0001）。</p>
<p>然后，我们将描述使用CRISPR&#x2F;Cas9与重组AAV6载体（rAAV6，同源定向修复（HDR）的捐赠模板）一起，以调解在腺相关病毒集成站点1（AAVS1）基因中插入增强型绿色荧光蛋白<em>（EGFP）</em> 的现场特定插入过程。AAVS1基因是一种没有已知生物功能的活性球体，也是人类基因组上的AAV病毒整合站点：因此，它被认为是基因组工程的”安全港”。在这里，我们报告说，B细胞的膨胀和激活允许在7天的培养中扩大44倍（<strong>图1）。</strong>B细胞的电扩散显示，在24小时后，整体细胞健康（<strong>图2A）</strong>略有下降。对 CD19 标记（<strong>图 2B）</strong>的分散图分析显示编辑的单元块（<strong>图 2C）减少高达</strong>83%。</p>
<img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/61855fig03a.jpg" alt="Figure 3A" style="zoom:50%;">

<p>[^图3A]: </p>
<img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/61855fig03b.jpg" alt="Figure 3B" style="zoom:50%;">

<p>[^图3B]: <strong>CD19蛋白质损失与英德尔形成。</strong>（<strong>A</strong>） 色谱图描绘了控制和 CD19 淘汰 （KO） 样本的测序峰值。控制峰值上的灰色框突出显示 CD19 gRNA 的目标序列，并带有箭头指示的预测切割位置。CD19 KO 显示”双峰”测序围绕预测的切割点，指示在双链断裂后插入&#x2F;删除核苷酸。（<strong>B</strong>） 条形图显示 CD19 蛋白质损失百分比与 CD19 轨迹（p ≥ 0.05）的英德尔形成之间的一致结果。</p>
<img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/61855fig04.jpg" alt="Figure 4" style="zoom:50%;">

<p>[^图4：rAAV6 AAVS1 MND-GFP矢量构造。]: ****表达盒包含一个强大的合成促进器 （MND） 序列，紧接着是增强的绿色荧光蛋白 （EGFP） 编码序列和聚腺素化 （Poly A） 序列。AAVS1 同源臂侧翼在 MND 发起器上游和聚 A 序列的下游。EGFP 将根据 MND 发起人的法规进行表达。在构造的每个组件上方指示序列长度。</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/61855fig05a.jpg" alt="Figure 5A"></p>
<p>[^图5A]: </p>
<img src="/2024/03/16/%E4%BD%BF%E7%94%A8CRISPR-Cas9%E7%9A%84%E4%BA%BA%E7%B1%BBB%E7%BB%86%E8%83%9E%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%B7%A5%E7%A8%8B/61855fig05b.jpg" alt="Figure 5B" style="zoom:50%;">

<p>[^图5B]: </p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/61855fig05c.jpg" alt="Figure 5C"></p>
<p>[^图5C]: <strong>CRISPR&#x2F;Cas9-和rAAV6在工程后第12天将EGFP记者录音带在B单元中的具体集成。</strong>（<strong>A</strong>） 代表流图显示控制或仅矢量样本中无EGFP阳性 B 细胞，而敲击 （KI） 样本中 EGFP 阳性 B 细胞为 64.4%。（<strong>B</strong>） KI样品的交界聚合酶链反应显示预测的1.5 Kbps波段：在控制或仅矢量样本中未发现带。在 PCR 过程中，水用于确保无污染。（<strong>C</strong>） 条形图描绘了在工程后 3 天内控制、仅矢量和 EGFP-KI 样本的细胞生长。断线表示 1 × 106 单元输入。</p>
<p>TIDE对色谱仪（<strong>图3A）</strong>的分析显示，英德尔的百分比与流动细胞仪（<strong>图3B）</strong>所导致的蛋白质损失相似。KI实验的流动细胞学分析表明，接受AAV载体（<strong>图4）</strong>的细胞与RNP一起表示高达64%的EGFP阳性细胞（<strong>图5A），</strong>随后通过结点聚合酶链反应（PCR）（<strong>图5B）</strong>显示成功整合。细胞计数显示，所有样本在工程后3天内迅速恢复（<strong>图5C）。</strong></p>
<hr>
<h3 id="protocol"><a href="#protocol" class="headerlink" title="protocol"></a><strong>protocol</strong></h3><p>从当地血库获得了健康捐赠者的白细胞分离样本。在此描述的所有实验均被明尼苏达大学的机构审查委员会（IRB）认定为研究豁免，并获得了机构生物安全委员会（IBC）的批准。<br>注意：所有实验均按照血源性病原体的通用预防措施进行，使用无菌&#x2F;无菌技术和适当的生物安全二级设备。</p>
<ol>
<li><h4 id="一、准备B细胞扩增培养基的补充剂"><a href="#一、准备B细胞扩增培养基的补充剂" class="headerlink" title="一、准备B细胞扩增培养基的补充剂"></a>一、准备B细胞扩增培养基的补充剂</h4></li>
</ol>
<p>  将CpG寡核苷酸复溶至1 mg&#x2F;mL的浓度。<br>  将CD40配体（CD40L）复溶至100 μg&#x2F;mL的浓度。<br>  将重组人IL-10（rhIL-10）复溶至50 μg&#x2F;mL的浓度。<br>  将重组人IL-15（rhIL-15）复溶至10 μg&#x2F;mL的浓度。<br>  注意：将每种补充剂分装成小份，在-20°C至-80°C下保存，最多可保存6个月。</p>
<ol start="2">
<li><h4 id="二、准备基础培养基"><a href="#二、准备基础培养基" class="headerlink" title="二、准备基础培养基"></a>二、准备基础培养基</h4></li>
</ol>
<p>  将B细胞基础培养基与5%（体积&#x2F;体积）的体外免疫细胞扩增培养基补充剂（例如，CTS免疫细胞SR）和1%（体积&#x2F;体积）的青霉素和链霉素混合。<br>  使用0.22 µm过滤器适配器对基础培养基进行过滤灭菌，装入已灭菌的瓶子中。<br>  将基础培养基在4°C下保存，最多可保存1个月。</p>
<ol start="3">
<li><h4 id="三、准备B细胞扩增培养基"><a href="#三、准备B细胞扩增培养基" class="headerlink" title="三、准备B细胞扩增培养基"></a>三、准备B细胞扩增培养基</h4></li>
</ol>
<p>  将所需量的基础培养基转移到无菌容器中，以5 × 10^5细胞&#x2F;mL的密度培养B细胞。<br>  用1 μg&#x2F;mL CpG、100 ng&#x2F;mL CD40L、50 ng&#x2F;mL rhIL-10和10 ng&#x2F;mL rhIL-15补充基础培养基。<br>  使用0.22 µm过滤器过滤B细胞扩增培养基。<br>  在使用前，将B细胞扩增培养基在37°C、5% CO2、湿度条件下在组织培养孵箱中平衡至少30分钟。<br>  注意：准备新鲜的B细胞扩增培养基以供当天使用。不要准备用于多天的B细胞扩增培养基。此培养基配方鼓励记忆B细胞和激活的原代人B细胞的增殖。</p>
<ol start="4">
<li><h4 id="四、人B细胞的纯化和扩增"><a href="#四、人B细胞的纯化和扩增" class="headerlink" title="四、人B细胞的纯化和扩增"></a>四、人B细胞的纯化和扩增</h4></li>
</ol>
<p>  注意：按照制造商的说明，在开始第4.1步之前，向温控冷冻容器中加入99-100%异丙醇，并将冷冻容器在4°C下冷却。</p>
<ol>
<li><h5 id="从白细胞分离样本中分离出外周血单核细胞（PBMC）"><a href="#从白细胞分离样本中分离出外周血单核细胞（PBMC）" class="headerlink" title="从白细胞分离样本中分离出外周血单核细胞（PBMC）"></a>从白细胞分离样本中分离出外周血单核细胞（PBMC）</h5><ol>
<li>将大约8-10 mL的白细胞分离样本转移到无菌的50 mL锥形管中。</li>
<li>用无菌的1x磷酸盐缓冲液（PBS）将体积补至35 mL。</li>
<li>小心地将35 mL白细胞分离样本层叠在15 mL密度梯度介质上。</li>
<li>在500 × g下离心25分钟，不使用制动，去除血浆层，不要扰动富含单核细胞的层（PBMC层），从界面收集PBMCs，并转移到新的无菌50 mL锥形管中。</li>
<li>用1x PBS将PBMCs的体积补至50 mL。</li>
<li>在500 × g下离心5分钟，不使用制动。去除上清液，不要扰动可能呈现红色的PBMC沉淀。</li>
<li>加入7 mL的氯化铵-氯化钾溶胞缓冲液，用移液管轻轻吹打混合3次，并在室温（RT）下孵育3分钟。</li>
<li>用1x PBS将体积补至50 mL。</li>
<li>在400 × g、低阻力制动的条件下离心5分钟。去除上清液，不要扰动沉淀。沉淀应呈粉红色或白色。<br>注意：为了继续培养新鲜分离的B细胞，请在开始B细胞分离（第4.2步）之前准备B细胞扩增培养基。</li>
</ol>
</li>
<li><h5 id="使用人原代B细胞负选试剂盒从PBMCs中分离B细胞"><a href="#使用人原代B细胞负选试剂盒从PBMCs中分离B细胞" class="headerlink" title="使用人原代B细胞负选试剂盒从PBMCs中分离B细胞"></a>使用人原代B细胞负选试剂盒从PBMCs中分离B细胞</h5><ol>
<li>用隔离缓冲液将PBMCs重新悬浮至5 × 10^7细胞&#x2F;mL的浓度。<br>注意：如果PBMC总数少于5 × 10^7细胞，则减少隔离缓冲液的体积以保持5 × 107细胞&#x2F;mL。细胞悬浮液的最小和最大体积分别为0.25 mL和8 mL。</li>
<li>将最多8 mL（5 × 10^7细胞&#x2F;mL）转移到带有盖子的无菌、聚丙烯、圆底试管中。</li>
<li>向PBMCs中加入50 μL&#x2F;mL的Cocktail增强剂。</li>
<li>向PBMCs中加入50 μL&#x2F;mL的隔离鸡尾酒，盖上试管盖，倒置2-3次以混合。</li>
<li>在RT下孵育5分钟；在孵育的第4分钟，至少30秒涡旋磁性微球。</li>
<li>每1 mL PBMCs转移50 μL磁性微球，盖上试管盖，倒置2-3次以混合。</li>
<li>用隔离缓冲液补至10 mL，并轻轻上下吸液2-3次。</li>
<li>将试管放在磁性站中，在RT下孵育3分钟。</li>
<li>握住磁铁和试管，将磁铁和试管一起倒置，将细胞悬浮液倒入新试管中。丢弃旧试管。</li>
<li>重复步骤4.2.8（将孵育时间减少至2分钟），并将B细胞悬浮液倒入干净的锥形管中。</li>
<li>富集的B细胞准备使用。如果立即使用细胞，请继续第4.3节（人B细胞扩增）。通过流式细胞术（可选）检查分离的B细胞的纯度。如果在使用前需要冷冻细胞，请继续步骤4.2.12。</li>
<li>为了冷冻B细胞，在400 × g下离心5分钟，去除上清液，不要扰动沉淀。</li>
<li>用冷冻培养基以10^7细胞&#x2F;mL的浓度重新悬浮细胞，并分装到1 mL&#x2F;冻存管中。</li>
<li>将冻存管放入冷却的冷冻容器中，在-80°C下储存过夜；然后，将冷冻的冻存管转移到液氮罐中；可保存冻结细胞长达1年。<br>注意：预计分离的B细胞产量为总PBMC的2%-8%，存活率为95%-99%。</li>
</ol>
</li>
<li><h4 id="人B细胞扩增"><a href="#人B细胞扩增" class="headerlink" title="人B细胞扩增"></a>人B细胞扩增</h4><p>注意：如果使用新鲜分离的B细胞，请跳过步骤4.3.1-4.3.5。计数细胞，并将所需数量的细胞转移到无菌锥形管中，然后继续进行步骤4.3.6。</p>
<ol>
<li>在使用前，将胎牛血清（FBS）在温水浴中预温。准备20 mL的B细胞扩增培养基，转移到T25烧瓶中，并在使用前至少15分钟在组织培养孵箱中预平衡培养基（在37°C、5% CO2、湿度的条件下）。</li>
<li>在37°C水浴中解冻B细胞。等待期间，将2 mL预热的FBS转移到无菌的15 mL锥形管中。</li>
<li>B细胞完全解冻后，立即向样本中逐滴加入1 mL预热的FBS，并在RT下孵育1分钟。</li>
<li>轻轻吸液以重新悬浮样本，并将整个体积逐滴转移到含有2 mL预热的FBS的锥形管中。</li>
<li>用无菌的1x PBS将体积补至15 mL，盖上试管盖，轻轻倒置2-3次。</li>
<li>在400 × g下离心5分钟。</li>
<li>去除上清液，不要扰动细胞沉淀，用1 mL预平衡的B细胞扩增培养基重新悬浮细胞沉淀，并计数细胞。总细胞数应约为10^7细胞。</li>
<li>将细胞转移到含有20 mL预平衡的B细胞扩增培养基的烧瓶中。细胞的最终浓度应约为5 × 10^5细胞&#x2F;mL。</li>
<li>垂直将烧瓶放入组织培养孵箱中孵育。</li>
<li>每2天通过将B细胞的全部体积转移到无菌锥形管中并重复步骤4.3.6-4.3.9来完全更换扩增培养基。<br>注意：T25烧瓶可以容纳10-20 mL的培养基；T75烧瓶可以容纳多达20-60 mL的培养基。</li>
</ol>
</li>
<li><h4 id="五、人原代B细胞的工程化处理"><a href="#五、人原代B细胞的工程化处理" class="headerlink" title="五、人原代B细胞的工程化处理"></a>五、人原代B细胞的工程化处理</h4></li>
</ol>
<p>为了获得最佳结果，在扩增&#x2F;激活后的48 ± 2小时对B细胞进行工程化处理。准备B细胞扩增培养基，将1 mL培养基分装到48孔组织培养板中，并在使用前至少15分钟在组织培养孵箱中预平衡。<br>注意：在设计CRISPR&#x2F;Cas9 sgRNA进行基因编辑时，请遵循以下步骤。</p>
<p>   1）使用在线工具设计sgRNAs。</p>
<p>   2）在所有蛋白质变体的共有外显子上设计sgRNAs。</p>
<p>   3）从信誉良好的公司订购化学修饰的sgRNA。</p>
<p>  4）sgRNA通常以冻干形式提供；用无菌的DNase&#x2F;RNase-free Tris-EDTA（TE）缓冲液复溶sgRNA，浓度为1 μg&#x2F;μL。</p>
<pre><code>1. 通过将每个转染反应中的1 μL（1 μg/μL）化学修饰的sgRNA与1.5 μL（1 μg/μL）化学修饰的链球菌Cas9（S.p. Cas9）核酸酶混合来准备CRISPR/Cas9转染底物。对于对照组，使用1 μL TE缓冲液代替sgRNA。
 注意：当使用CD19 sgRNA进行基因敲除（KO）实验时，请参阅图2的结果。当使用AAVS1 sgRNA进行基因敲入（KI）实验时，请参阅图5的结果。
 **将所有组分保持在冰上。**
2. 轻轻混合，并将每个反应的2.5 μL CRISPR/Cas9转染底物转移到0.2 mL 8管条的管中；在室温下放置备用。
3. 打开核转染仪（电穿孔仪），并准备转染试剂，如表格1所示。
 注意：如果需要，这是一个暂停的好步骤，将所有试剂放在冰上。准备好继续实验时，从冰上取下所有试剂。使用S.p. Cas9蛋白时，研究者必须通过将1 μg sgRNA与5 μg SpCas9蛋白混合，避免产生气泡，并在使用前在室温下孵育至少20分钟以获得最佳结果。
4. 计数并将每10^6个B细胞转移到无菌锥形管中进行转染反应。
5. 用无菌的1x PBS将体积补至15 mL，并在400 × g下离心5分钟。等待时，准备主要细胞转染试剂（表格2）并在室温下放置备用。
6. 去除上清液，不要扰动细胞沉淀。
7. 用10 mL无菌的1x PBS重新悬浮细胞沉淀，并在400 × g下离心5分钟。
8. 完全去除上清液，不要扰动细胞沉淀。
9. 向每10^6个B细胞的细胞沉淀中转移0.5 μg化学修饰的GFP mRNA（作为转染报告物，可选）。
10. 用每10^6个B细胞的20 μL主要细胞转染试剂重新悬浮细胞沉淀；轻轻通过吸液5-6次混合。将每个反应的20.5 μL转移到含有2.5 μL CRISPR/Cas9转染底物的0.2 mL管中。
11. 上下吸液一次以混合，并将整个体积（23 μL）转移到转染槽中。盖上盖子，轻轻在台面上敲打试管，以确保液体覆盖槽的底部。
12. 使用核转染仪的人原代B细胞协议进行转染。
    注意：核转染仪（电穿孔仪）可以放在组织培养罩外使用。盖上盖子以确保无菌。
13. 让电穿孔后的细胞在室温下静置15分钟。
14. 从组织培养板中转移80 μL预平衡的B细胞扩增培养基到转染反应的试管中。将试管放入组织培养孵箱中孵育30分钟。
15. 轻轻吸液几次以混合，并将试管中的样本整体转移到48孔组织培养板中适当孔内，含有1 mL B细胞扩增培养基的孔中。细胞的最终浓度应为10^6细胞/mL。
16. 如果进行基因敲入实验，则在电穿孔后大约45分钟，将rAAV6载体以500,000感染复数转移到含有电穿孔细胞的适当孔中。请参阅图4A的rAAV6载体构建示例。
    注意：例如：对照样本将不使用CRISPR/Cas9或rAAV6载体进行电穿孔。仅载体样本将不使用CRISPR/Cas9进行电穿孔，然后使用rAAV6载体进行转导。敲入（KI）样本将使用CRISPR/Cas9进行电穿孔，然后使用rAAV6载体进行转导。rAAV6必须在其靶向DSB位点上下游含有同源臂，以进行同源定向修复（HDR）。
17. 将培养板放入37°C、5% CO2湿度的组织培养孵箱中。
18. 在工程化处理后的第1天计数细胞并记录存活率。
19. 每2天通过计数细胞并将细胞的全部体积转移到干净的1.5 mL微量离心管中，然后离心以400 ×g离心5分钟，去除上清液，不要扰动沉淀。用100 µL新鲜的B细胞扩增培养基重新悬浮细胞，并转移到24孔组织培养板的孔中。将培养基体积补至最终细胞浓度达到5 × 10^5细胞/mL。
    注意：48孔板每孔最多可容纳1 mL培养基；24孔板每孔最多可容纳2 mL培养基；12孔板每孔最多可容纳4 mL培养基；6孔板每孔最多可容纳8 mL培养基。
20. 在至少扩增5天后再进行下游分析，如流式细胞术分析、TIDE分析和连接PCR。
</code></pre>
<hr>
<h3 id="讨论"><a href="#讨论" class="headerlink" title="讨论"></a>讨论</h3><p>直到最近，对原代人B细胞进行精确的基因组工程仍然是一个挑战。我们之前发表的协议使用了CRISPR&#x2F;Cas9来工程化原代人B细胞。在这里，我们概述了改进的B细胞分离、扩增和工程化协议，以实现对CD19的快速KO或对EGFP的敲入。</p>
<p>我们证明，我们的扩增协议可以在7天内将B细胞快速扩增至44倍（图1）。这个协议的扩增速率比Johnson等人和Hung等人的报告更快。确保原代人B细胞健康快速扩增的关键步骤是每两天用新鲜的激活因子替换培养基，并确保培养中的总B细胞数不超过2 × 106细胞&#x2F;mL。</p>
<p>我们还发现，我们改进的CRISPR&#x2F;Cas9转染协议对CD19 KO结果在效率上很高（图2和图3）。与之前的研究一样，我们观察到电穿孔后24小时细胞回收和存活率略有下降。这表明电穿孔会影响原代人B细胞的整体健康；然而，细胞最终在48小时内恢复（数据未显示）。</p>
<p>使用这种电穿孔器协议的好处是，我们可以以更快的速度（16个反应在1分钟内完成），而之前的协议（9，10）完成16个反应大约需要10-12分钟。此外，这个转染系统消除了之前研究中观察到的电穿孔器可能产生的电弧（9，10）。此外，这个工程化方法可以扩大规模，用于更大数量的B细胞，使用更大型的、商业上可获得的槽（数据未显示）。</p>
<p>为了确保最佳的细胞整体健康和KO效率，有几个关键步骤：首先，确保B细胞扩增培养基制备并预平衡至少15分钟后使用。其次，转染底物的总体积不应超过核转染试剂的20%（体积&#x2F;体积）。第三，必须扩增&#x2F;激活48 ± 2小时以获得最佳结果。第四，在将电穿孔槽放入电穿孔器之前，轻轻敲打电穿孔槽，以确保转染反应溶液覆盖槽底。第五，确保电穿孔后的细胞在电穿孔槽中仅静置15分钟；电穿孔后长时间留在核转染试剂中可能会损害整体细胞存活率。第六，确保将细胞与培养基在电穿孔槽中的孵育时间不超过30分钟。第七，将106个电穿孔细胞放入1 mL中，在最初的48小时内，这有助于细胞更快恢复，而不是在较低的密度下培养（数据未显示）。最后，使用Cas9 mRNA或Cas9蛋白将导致类似的编辑效率（数据未显示）；然而，我们为了方便和成本效益使用了Cas9 mRNA。</p>
<p>在使用CRISPR&#x2F;Cas9时，有两个主要担忧：脱靶效应和染色体易位。由于sgRNA与目标序列不匹配的碱基对引起的脱靶效应，导致基因组上可能存在许多可能的结合位点，并产生多个不需要的基因KO。因此，应同时考虑预测的脱靶评分和on-target评分，以尽量减少这个问题。染色体易位可能由于脱靶效应或在敲除多个基因时发生。这可能在实验和临床上造成灾难性事件。碱基编辑器（12种：两种类型的胞嘧啶碱基编辑器和腺嘌呤碱基编辑器）改变一个核苷酸（破坏剪接元件、创建过早终止密码子或创建点突变），导致目标基因KO而不产生DSB（在其他地方详细回顾）。因此，碱基编辑方法可用于单基因或多基因KO，以规避染色体易位问题。</p>
<p>我们还证明，CRISPR&#x2F;Cas9与rAAV6一起，可以用于高效介导特异性KI，并在AAVS1位点表达EGFP。我们在工程化后的至少12天内观察到EGFP表达。我们还观察到KI样本在第12天有两个EGFP阳性群体：一个高表达群体和一个中表达群体（图5A），而仅接收载体的样本显示出最小比例的细胞中存在中表达EGFP。我们推测这种“高和中表达群体”的现象是由于载体双等位基因整合造成的。进一步使用跨越PCR对中和高表达EGFP细胞进行调查可以确认这个假设。使用AAV作为载体的两个注意事项：首先，AAV载体具有有限的货物容量，高达4.7千字节，当敲入大基因时会引起问题。减少同源臂的大小将允许容纳更大的基因，从而降低KI效率（数据未显示）。或者，可以使用多个基因装载载体的同时或顺序整合。研究报道了在免疫健全的动物模型中AAV转导细胞会产生免疫反应并清除这些细胞。作为替代方案，可以探索非病毒基础的供体HDR模板，以规避这个问题。</p>
<p>总之，我们已经证明了从分离、扩增到B细胞工程化的全面、分步过程，这些过程导致了高基因修改效率。这种工程化方法可以用于基因敲除，并研究B细胞中基因的功能。此外，这种方法可以用于工程化B细胞以表达重组抗体，用于对抗感染。最后，这种方法可以应用于工程化B细胞以表达和分泌治疗酶，这些酶可以作为自体细胞为基础的治疗方法，用于治疗酶病。</p>
<p>在未来的研究中，我们可以进一步优化这些协议，提高基因编辑的效率和特异性，并探索新的方法来减少脱靶效应和染色体易位的风险。此外，我们还可以研究如何将这些工程化的B细胞用于临床应用，例如作为免疫治疗或基因治疗的工具。通过不断改进和探索新的技术，我们有望在不久的将来实现更高效、更安全的人B细胞基因组工程。</p>
<hr>
<table>
<thead>
<tr>
<th><strong>Name</strong></th>
<th><strong>gRNA sequence</strong></th>
</tr>
</thead>
<tbody><tr>
<td>CD19</td>
<td>5’-GCTGTGCTGCAGTGCCTCAA-3’</td>
</tr>
<tr>
<td>AAVS1</td>
<td>5’-GTCACCAATCCTGTCCCTAG-3’</td>
</tr>
</tbody></table>
<p><strong>Table 1: gRNA sequences.</strong></p>
<table>
<thead>
<tr>
<th><strong>Reagents</strong></th>
<th><strong>Volume per 1 reaction</strong></th>
</tr>
</thead>
<tbody><tr>
<td>P3 Primary Cell solution</td>
<td>16.4 mL</td>
</tr>
<tr>
<td>Supplement 1</td>
<td>3.6 mL</td>
</tr>
<tr>
<td><strong>Total</strong></td>
<td>20 mL</td>
</tr>
</tbody></table>
<p><strong>Table 2: Preparation of Nucleofection reagent mix.</strong></p>
<table>
<thead>
<tr>
<th><strong>Description</strong></th>
<th><strong>Sequence</strong></th>
<th><strong>Purpose</strong></th>
</tr>
</thead>
<tbody><tr>
<td>CD19 forward primers</td>
<td>5’-AAATTCAGGAAAGGGTTGGAAG-3’</td>
<td>Amplify the CD19 locus for Indel analysis</td>
</tr>
<tr>
<td>CD19 Reverse primer</td>
<td>5’-GCGGACCTCTTCTGTCCATG-3’</td>
<td>Amplify the CD19 locus for Indel analysis</td>
</tr>
<tr>
<td>Junction PCR forward primer</td>
<td>5’-GGACGAGCTGTACAAGTAACG-3’</td>
<td>Junction PCR</td>
</tr>
<tr>
<td>Junction PCR Reverse primer</td>
<td>5’-GAGACAGTGACCAACCATCC-3</td>
<td>Junction PCR</td>
</tr>
</tbody></table>
<p><strong>Table 3: Primers used.</strong></p>
<table>
<thead>
<tr>
<th><strong>Description</strong></th>
<th><strong>Fluorophore</strong></th>
<th><strong>Clone</strong></th>
</tr>
</thead>
<tbody><tr>
<td>Anti-human CD19</td>
<td>PerCP</td>
<td>HIB19</td>
</tr>
<tr>
<td>Viability Dye</td>
<td>eFlour 780</td>
<td>-</td>
</tr>
</tbody></table>
<p><strong>Table 4: Flow cytometry antibody and viability dye.</strong></p>
<hr>
<h3 id="References"><a href="#References" class="headerlink" title="References"></a>References</h3><ol>
<li>LeBien, T. W., Thomas, T. F. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=B+lymphocytes:+how+they+develop+and+function.">B lymphocytes: how they develop and function.</a> <em>Blood</em>. <strong>112</strong> (5), 1570-1580 (2008).</li>
<li>Nutt, S. L., Hodgkin, P. D., Tarlinton, D. M., Corcoran, L. M. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=The+generation+of+antibody-secreting+plasma+cells.">The generation of antibody-secreting plasma cells.</a> <em>Nature Reviews Immunology</em>. <strong>15</strong>, 160-171 (2015).</li>
<li>Slifka, M. K., Antia, R., Whitmire, J. K., Ahmed, R. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Humoral+immunity+due+to+long-lived+plasma+cells.">Humoral immunity due to long-lived plasma cells.</a> <em>Immunity</em>. <strong>8</strong> (3), 363-372 (1998).</li>
<li>Spriggs, M. K., et al. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Recombinant+human+CD40+ligand+stimulates+B+cell+proliferation+and+immunoglobulin+E+secretion.">Recombinant human CD40 ligand stimulates B cell proliferation and immunoglobulin E secretion.</a> <em>Journal of Experimental Medicines</em>. <strong>176</strong> (6), 1543-1550 (1992).</li>
<li>Fusil, F., et al. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=A+lentiviral+vector+allowing+physiologically+regulated+membrane-anchored+and+secreted+antibody+expression+depending+on+B-cell+maturation+status.">A lentiviral vector allowing physiologically regulated membrane-anchored and secreted antibody expression depending on B-cell maturation status.</a> <em>Molecular Therapy</em>. <strong>23</strong> (11), 1734-1747 (2015).</li>
<li>Luo, X. M., et al. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Engineering+human+hematopoietic+stem/progenitor+cells+to+produce+a+broadly+neutralizing+anti-HIV+antibody+after+in+vitro+maturation+to+human+B+lymphocytes.">Engineering human hematopoietic stem&#x2F;progenitor cells to produce a broadly neutralizing anti-HIV antibody after in vitro maturation to human B lymphocytes.</a> <em>Blood</em>. <strong>113</strong> (7), 1422-1431 (2008).</li>
<li>Mock, U., Thiele, R., Uhde, A., Fehse, B., Horn, S. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Efficient+lentiviral+transduction+and+transgene+expression+in+primary+human+B+cells.">Efficient lentiviral transduction and transgene expression in primary human B cells.</a> <em>Human Gene Therapy Methods</em>. <strong>23</strong> (6), 408-415 (2012).</li>
<li>Heltemes-Harris, L. M., et al. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Sleeping+Beauty+transposon+screen+identifies+signaling+modules+that+cooperate+with+STAT5+activation+to+induce+B-cell+acute+lymphoblastic+leukemia.">Sleeping Beauty transposon screen identifies signaling modules that cooperate with STAT5 activation to induce B-cell acute lymphoblastic leukemia.</a> <em>Oncogene</em>. <strong>35</strong> (26), 3454-3464 (2016).</li>
<li>Johnson, M. J., Laoharawee, K., Lahr, W. S., Webber, B. R., Moriarity, B. S. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Engineering+of+primary+human+B+cells+with+CRISPR/Cas9+targeted+nuclease.">Engineering of primary human B cells with CRISPR&#x2F;Cas9 targeted nuclease.</a> <em>Scientific Reports</em>. <strong>8</strong>, 12144 (2018).</li>
<li>Hung, K. L., et al. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Engineering+protein-secreting+plasma+cells+by+homology-directed+repair+in+primary+human+B+cells.">Engineering protein-secreting plasma cells by homology-directed repair in primary human B cells.</a> <em>Molecular Therapy</em>. <strong>26</strong> (2), 456-467 (2018).</li>
<li>Cui, Y., Xu, J., Cheng, M., Liao, X., Peng, S. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Review+of+CRISPR/Cas9+sgRNA+design+tools.">Review of CRISPR&#x2F;Cas9 sgRNA design tools.</a> <em>Interdisciplinary Sciences</em>. <strong>10</strong> (2), 455-465 (2018).</li>
<li>Rees, H. A., Liu, D. R. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Base+editing:+precision+chemistry+on+the+genome+and+transcriptome+of+living+cells.">Base editing: precision chemistry on the genome and transcriptome of living cells.</a> <em>Nature Reviews Genetics</em>. <strong>19</strong> (12), 770-788 (2018).</li>
<li>Spiegel, A., Bachmann, M., Jurado Jiménez, G., Sarov, M. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=CRISPR/Cas9-based+knockout+pipeline+for+reverse+genetics+in+mammalian+cell+culture.">CRISPR&#x2F;Cas9-based knockout pipeline for reverse genetics in mammalian cell culture.</a> <em>Methods</em>. <strong>164-165</strong>, 49-58 (2019).</li>
<li>Suzuki, T., Kazuki, Y., Oshimura, M., Hara, T. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=A+novel+system+for+simultaneous+or+sequential+integration+of+multiple+gene-loading+vectors+into+a+defined+site+of+a+human+artificial+chromosome.">A novel system for simultaneous or sequential integration of multiple gene-loading vectors into a defined site of a human artificial chromosome.</a> <em>PLoS One</em>. <strong>9</strong>, 110404 (2014).</li>
<li>Bak, R. O., Porteus, M. H. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=CRISPR-mediated+integration+of+large+gene+cassettes+using+AAV+donor+vectors.">CRISPR-mediated integration of large gene cassettes using AAV donor vectors.</a> <em>Cell Reports</em>. <strong>20</strong> (3), 750-756 (2017).</li>
<li>Shirley, J. L., de Jong, Y. P., Terhorst, C., Herzog, R. W. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Immune+responses+to+viral+gene+therapy+vectors.">Immune responses to viral gene therapy vectors.</a> <em>Molecular Therapy</em>. <strong>28</strong> (3), 709-722 (2020).</li>
<li>Martino, A. T., Markusic, D. M. <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed&cmd=Search&doptcmdl=Citation&defaultField=Title+Word&term=Immune+response+mechanisms+against+AAV+vectors+in+animal+models.">Immune response mechanisms against AAV vectors in animal models.</a> <em>Molecular Therapy - Methods and Clinical Development</em>. <strong>17</strong>, 198-208 (2020).</li>
</ol>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
	<div class="article-tag tagcloud">
		<i class="icon-price-tags icon"></i>
		<ul class="article-tag-list">
			 
        		<li class="article-tag-list-item">
        			<a href="javascript:void(0)" class="js-tag article-tag-list-link color2">BioProtocol</a>
        		</li>
      		
		</ul>
	</div>

    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/16/%E4%BD%BF%E7%94%A8CRISPR-Cas9%E7%9A%84%E4%BA%BA%E7%B1%BBB%E7%BB%86%E8%83%9E%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%B7%A5%E7%A8%8B/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-强化学习在生物信息学与基因编辑领域的应用" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/16/%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0%E5%9C%A8%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E4%B8%8E%E5%9F%BA%E5%9B%A0%E7%BC%96%E8%BE%91%E9%A2%86%E5%9F%9F%E7%9A%84%E5%BA%94%E7%94%A8/" target="_blank">强化学习在生物信息学与基因编辑领域的应用</a>
  
    </h1>
  


  
  
<a href="/2024/03/16/%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0%E5%9C%A8%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E4%B8%8E%E5%9F%BA%E5%9B%A0%E7%BC%96%E8%BE%91%E9%A2%86%E5%9F%9F%E7%9A%84%E5%BA%94%E7%94%A8/" class="archive-article-date">
        <time datetime="2024-03-16T13:21:13.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-16</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <h1 id="1-背景介绍"><a href="#1-背景介绍" class="headerlink" title="1.背景介绍"></a>1.背景介绍</h1><p>生物信息学和基因编辑是生物科学和生物技术的重要领域，它们在近年来取得了显著的进展。生物信息学涉及对生物数据的分析和解释，包括基因组学、蛋白质结构和功能等。基因编辑则是对基因组的直接修改，用于治疗疾病和改善生物质量。强化学习（Reinforcement Learning，RL）是一种人工智能技术，它可以帮助机器学习从环境中获取反馈，以优化行为策略。在这篇文章中，我们将探讨如何将强化学习应用于生物信息学和基因编辑领域，以及其潜在的应用和挑战。</p>
<h1 id="2-核心概念与联系"><a href="#2-核心概念与联系" class="headerlink" title="2.核心概念与联系"></a>2.核心概念与联系</h1><h1 id="2-1生物信息学"><a href="#2-1生物信息学" class="headerlink" title="2.1生物信息学"></a>2.1生物信息学</h1><p>生物信息学是一门集生物学、计算机科学、数学、信息论等多学科知识的 intersection 领域。生物信息学家利用计算机科学的方法来分析和解释生物数据，例如基因组数据、蛋白质序列数据、微阵列数据等。生物信formatio学家的任务包括：</p>
<ul>
<li>发现新的基因和蛋白质功能</li>
<li>研究基因组的演化和多样性</li>
<li>研究基因表达和控制机制</li>
<li>研究蛋白质结构和功能</li>
<li>研究生物网络和路径径</li>
</ul>
<h1 id="2-2基因编辑"><a href="#2-2基因编辑" class="headerlink" title="2.2基因编辑"></a>2.2基因编辑</h1><p>基因编辑是一种对基因组进行直接修改的技术，用于治疗疾病和改善生物质量。基因编辑技术包括：</p>
<ul>
<li>CRISPR&#x2F;Cas9：一种基因剪接技术，可以精确地修改基因组。</li>
<li>TALEN：一种基因剪接技术，类似于 CRISPR&#x2F;Cas9。</li>
<li>ZFN：一种基因剪接技术，类似于 CRISPR&#x2F;Cas9 和 TALEN。</li>
</ul>
<p>基因编辑技术的应用包括：</p>
<ul>
<li>治疗遗传性疾病，如患血症、癌症等。</li>
<li>改善农业产品，如增加农作物的抗病毒性、增加粮食产量等。</li>
<li>研究基因组的功能和表达。</li>
</ul>
<h1 id="2-3强化学习"><a href="#2-3强化学习" class="headerlink" title="2.3强化学习"></a>2.3强化学习</h1><p>强化学习是一种人工智能技术，它可以帮助机器学习从环境中获取反馈，以优化行为策略。强化学习的核心概念包括：</p>
<ul>
<li>状态（State）：环境的描述。</li>
<li>动作（Action）：机器可以执行的操作。</li>
<li>奖励（Reward）：环境给予的反馈。</li>
<li>策略（Policy）：机器选择动作的规则。</li>
<li>价值函数（Value Function）：预测状态下策略下的累积奖励。</li>
</ul>
<h1 id="3-核心算法原理和具体操作步骤以及数学模型公式详细讲解"><a href="#3-核心算法原理和具体操作步骤以及数学模型公式详细讲解" class="headerlink" title="3.核心算法原理和具体操作步骤以及数学模型公式详细讲解"></a>3.核心算法原理和具体操作步骤以及数学模型公式详细讲解</h1><h1 id="3-1Q-Learning算法"><a href="#3-1Q-Learning算法" class="headerlink" title="3.1Q-Learning算法"></a>3.1Q-Learning算法</h1><p>Q-Learning是一种常用的强化学习算法，它可以帮助机器学习如何在环境中取得最佳的行为策略。Q-Learning的核心思想是通过迭代地更新价值函数，以优化策略。Q-Learning的具体操作步骤如下：</p>
<ol>
<li>初始化状态值和动作值。</li>
<li>选择一个初始状态。</li>
<li>从初始状态中选择一个动作。</li>
<li>执行动作并得到环境的反馈。</li>
<li>更新动作值。</li>
<li>重复步骤3-5，直到收敛。</li>
</ol>
<p>Q-Learning的数学模型公式如下：</p>
<p>Q(s,a)&#x3D;E[∑t&#x3D;0∞γtRt+1∣s0&#x3D;s,a0&#x3D;a]Q(s, a) &#x3D; E[\sum_{t&#x3D;0}^{\infty} \gamma^t R_{t+1} | s_0 &#x3D; s, a_0 &#x3D; a]Q(s,a)&#x3D;E[t&#x3D;0∑∞γtRt+1∣s0&#x3D;s,a0&#x3D;a]</p>
<p>其中，Q(s,a)Q(s, a)Q(s,a) 表示状态 sss 下执行动作 aaa 的累积奖励，Rt+1R_{t+1}Rt+1 表示时间 t+1t+1t+1 的奖励，γ\gammaγ 是折扣因子。</p>
<h1 id="3-2应用到生物信息学和基因编辑领域"><a href="#3-2应用到生物信息学和基因编辑领域" class="headerlink" title="3.2应用到生物信息学和基因编辑领域"></a>3.2应用到生物信息学和基因编辑领域</h1><p>在生物信息学和基因编辑领域，强化学习可以用于优化多种任务，例如：</p>
<ul>
<li>优化基因剪接技术，以提高治疗疾病的效果。</li>
<li>优化基因编辑任务，以提高农业产品的质量。</li>
<li>优化基因组分析任务，以提高基因组数据的解释率。</li>
</ul>
<p>具体的，强化学习可以用于优化以下任务：</p>
<ul>
<li>优化基因剪接技术的精确性和特异性。</li>
<li>优化基因编辑任务的效率和准确性。</li>
<li>优化基因组分析任务的准确性和速度。</li>
</ul>
<h1 id="4-具体代码实例和详细解释说明"><a href="#4-具体代码实例和详细解释说明" class="headerlink" title="4.具体代码实例和详细解释说明"></a>4.具体代码实例和详细解释说明</h1><h1 id="4-1Python实现Q-Learning算法"><a href="#4-1Python实现Q-Learning算法" class="headerlink" title="4.1Python实现Q-Learning算法"></a>4.1Python实现Q-Learning算法</h1><p>以下是一个简单的 Python 代码实例，用于实现 Q-Learning 算法。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">python复制代码<span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">QLearning</span>:</span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, state_space, action_space, learning_rate, discount_factor</span>):</span><br><span class="line">        self.state_space = state_space</span><br><span class="line">        self.action_space = action_space</span><br><span class="line">        self.learning_rate = learning_rate</span><br><span class="line">        self.discount_factor = discount_factor</span><br><span class="line">        self.q_table = np.zeros((state_space, action_space))</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">choose_action</span>(<span class="params">self, state</span>):</span><br><span class="line">        <span class="comment"># 选择一个动作</span></span><br><span class="line">        <span class="keyword">return</span> np.argmax(self.q_table[state])</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">update_q_value</span>(<span class="params">self, state, action, reward, next_state</span>):</span><br><span class="line">        <span class="comment"># 更新动作值</span></span><br><span class="line">        old_value = self.q_table[state, action]</span><br><span class="line">        new_value = reward + self.discount_factor * np.<span class="built_in">max</span>(self.q_table[next_state])</span><br><span class="line">        self.q_table[state, action] = old_value + self.learning_rate * (new_value - old_value)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">train</span>(<span class="params">self, state_transition_tuples, rewards, learning_epochs</span>):</span><br><span class="line">        <span class="keyword">for</span> epoch <span class="keyword">in</span> <span class="built_in">range</span>(learning_epochs):</span><br><span class="line">            <span class="keyword">for</span> state, action, reward, next_state <span class="keyword">in</span> state_transition_tuples:</span><br><span class="line">                self.update_q_value(state, action, reward, next_state)</span><br></pre></td></tr></table></figure>

<h1 id="4-2应用到生物信息学和基因编辑领域"><a href="#4-2应用到生物信息学和基因编辑领域" class="headerlink" title="4.2应用到生物信息学和基因编辑领域"></a>4.2应用到生物信息学和基因编辑领域</h1><p>在生物信息学和基因编辑领域，可以使用 Q-Learning 算法来优化多种任务。以下是一个简单的 Python 代码实例，用于实现优化基因剪接技术的精确性和特异性的任务。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">python复制代码<span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">GeneEditing</span>:</span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, state_space, action_space, learning_rate, discount_factor</span>):</span><br><span class="line">        self.state_space = state_space</span><br><span class="line">        self.action_space = action_space</span><br><span class="line">        self.learning_rate = learning_rate</span><br><span class="line">        self.discount_factor = discount_factor</span><br><span class="line">        self.q_table = np.zeros((state_space, action_space))</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">choose_action</span>(<span class="params">self, state</span>):</span><br><span class="line">        <span class="comment"># 选择一个动作</span></span><br><span class="line">        <span class="keyword">return</span> np.argmax(self.q_table[state])</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">update_q_value</span>(<span class="params">self, state, action, reward, next_state</span>):</span><br><span class="line">        <span class="comment"># 更新动作值</span></span><br><span class="line">        old_value = self.q_table[state, action]</span><br><span class="line">        new_value = reward + self.discount_factor * np.<span class="built_in">max</span>(self.q_table[next_state])</span><br><span class="line">        self.q_table[state, action] = old_value + self.learning_rate * (new_value - old_value)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">train</span>(<span class="params">self, state_transition_tuples, rewards, learning_epochs</span>):</span><br><span class="line">        <span class="keyword">for</span> epoch <span class="keyword">in</span> <span class="built_in">range</span>(learning_epochs):</span><br><span class="line">            <span class="keyword">for</span> state, action, reward, next_state <span class="keyword">in</span> state_transition_tuples:</span><br><span class="line">                self.update_q_value(state, action, reward, next_state)</span><br></pre></td></tr></table></figure>

<h1 id="5-未来发展趋势与挑战"><a href="#5-未来发展趋势与挑战" class="headerlink" title="5.未来发展趋势与挑战"></a>5.未来发展趋势与挑战</h1><h1 id="5-1未来发展趋势"><a href="#5-1未来发展趋势" class="headerlink" title="5.1未来发展趋势"></a>5.1未来发展趋势</h1><p>未来，强化学习在生物信息学和基因编辑领域的应用将面临以下挑战：</p>
<ul>
<li>数据收集和处理：生物信息学和基因编辑领域产生的数据量巨大，需要开发高效的数据收集和处理方法。</li>
<li>算法优化：需要开发更高效、更准确的强化学习算法，以应对生物信息学和基因编辑领域的复杂性。</li>
<li>多模态学习：需要开发可以处理多种类型数据的强化学习算法，以便在生物信息学和基因编辑领域进行更广泛的应用。</li>
<li>解释性：需要开发可以解释强化学习模型决策的方法，以便在生物信息学和基因编辑领域进行更有说服力的研究。</li>
</ul>
<h1 id="5-2挑战"><a href="#5-2挑战" class="headerlink" title="5.2挑战"></a>5.2挑战</h1><p>在强化学习应用于生物信息学和基因编辑领域面临的挑战包括：</p>
<ul>
<li>数据质量和可靠性：生物信息学和基因编辑领域的数据质量和可靠性可能受到实验条件和技术限制的影响。</li>
<li>算法复杂性：强化学习算法的复杂性可能导致计算成本和时间成本较高。</li>
<li>解释性和可解释性：强化学习模型的决策过程可能难以解释，导致在生物信息学和基因编辑领域的研究难以验证和传播。</li>
</ul>
<h1 id="6-附录常见问题与解答"><a href="#6-附录常见问题与解答" class="headerlink" title="6.附录常见问题与解答"></a>6.附录常见问题与解答</h1><p>Q：强化学习在生物信息学和基因编辑领域的应用有哪些？</p>
<p>A：强化学习可以用于优化多种任务，例如：</p>
<ul>
<li>优化基因剪接技术，以提高治疗疾病的效果。</li>
<li>优化基因编辑任务，以提高农业产品的质量。</li>
<li>优化基因组分析任务，以提高基因组数据的解释率。</li>
</ul>
<p>Q：强化学习在生物信息学和基因编辑领域的挑战有哪些？</p>
<p>A：强化学习在生物信息学和基因编辑领域面临的挑战包括：</p>
<ul>
<li>数据质量和可靠性：生物信息学和基因编辑领域的数据质量和可靠性可能受到实验条件和技术限制的影响。</li>
<li>算法复杂性：强化学习算法的复杂性可能导致计算成本和时间成本较高。</li>
<li>解释性和可解释性：强化学习模型的决策过程可能难以解释，导致在生物信息学和基因编辑领域的研究难以验证和传播。</li>
</ul>
<p>来源：<a target="_blank" rel="noopener" href="https://juejin.cn/post/7315461377258504242">稀土掘金</a></p>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
	<div class="article-tag tagcloud">
		<i class="icon-price-tags icon"></i>
		<ul class="article-tag-list">
			 
        		<li class="article-tag-list-item">
        			<a href="javascript:void(0)" class="js-tag article-tag-list-link color3">BioInfo</a>
        		</li>
      		
		</ul>
	</div>

    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/16/%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0%E5%9C%A8%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6%E4%B8%8E%E5%9F%BA%E5%9B%A0%E7%BC%96%E8%BE%91%E9%A2%86%E5%9F%9F%E7%9A%84%E5%BA%94%E7%94%A8/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-解决Github不能下载，下载速度慢的小妙招" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/15/%E8%A7%A3%E5%86%B3Github%E4%B8%8D%E8%83%BD%E4%B8%8B%E8%BD%BD%EF%BC%8C%E4%B8%8B%E8%BD%BD%E9%80%9F%E5%BA%A6%E6%85%A2%E7%9A%84%E5%B0%8F%E5%A6%99%E6%8B%9B/" target="_blank">解决Github不能下载，下载速度慢的小妙招（转载）</a>
  
    </h1>
  


  
  
<a href="/2024/03/15/%E8%A7%A3%E5%86%B3Github%E4%B8%8D%E8%83%BD%E4%B8%8B%E8%BD%BD%EF%BC%8C%E4%B8%8B%E8%BD%BD%E9%80%9F%E5%BA%A6%E6%85%A2%E7%9A%84%E5%B0%8F%E5%A6%99%E6%8B%9B/" class="archive-article-date">
        <time datetime="2024-03-15T10:19:17.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-15</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <p><a target="_blank" rel="noopener" href="https://developer.aliyun.com/article/888710">原文出处</a></p>
<p><img src="https://ucc.alicdn.com/images/user-upload-01/11509f94c74d46b28a70b884265addf5.png" alt="img"></p>
<h2 id="错误描述"><a href="#错误描述" class="headerlink" title="错误描述"></a>错误描述</h2><p>今天在github上使用 git clone 某个项目代码的时， git clone <a target="_blank" rel="noopener" href="https://github.com/XXXX/xxx-blog.git">https://github.com/XXXX/xxx-blog.git</a> 下载速度很慢，然后下载一段时间后，总是提示下面的错误信息：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">remote: Enumerating objects: 33, done.</span><br><span class="line">remote: Counting objects: 100% (33/33), done.</span><br><span class="line">remote: Compressing objects: 100% (8/8), done.</span><br><span class="line">error: RPC failed; curl 56 OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 10054</span><br><span class="line">fatal: the remote end hung up unexpectedly</span><br><span class="line">fatal: early EOF</span><br><span class="line">fatal: index-pack failed </span><br></pre></td></tr></table></figure>



<h2 id="错误原因"><a href="#错误原因" class="headerlink" title="错误原因"></a>错误原因</h2><p>由于Http协议错误，当 pull 或者 clone 的时候，或者是 github 某个CDN被屏蔽所致。</p>
<h2 id="解决办法1"><a href="#解决办法1" class="headerlink" title="解决办法1"></a>解决办法1</h2><p>协议错误</p>
<p>先执行下列命令</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git config --global http.postBuffer 5242880000</span><br></pre></td></tr></table></figure>

<p>再执行<code>git pull</code>或者 <code>git clone</code> 命令</p>
<h2 id="解决办法2"><a href="#解决办法2" class="headerlink" title="解决办法2"></a>解决办法2</h2><h3 id="修改域名"><a href="#修改域名" class="headerlink" title="修改域名"></a>修改域名</h3><ul>
<li>访问 <a target="_blank" rel="noopener" href="http://github.global.ssl.fastly.net.ipaddress.com/#ipinfo">http://github.global.ssl.fast…</a>获取cdn域名以及IP地址<br><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/5d9c3b2cb1099f4a27c94a53a9c04b8d.png" alt="img"></li>
<li>访问 <a target="_blank" rel="noopener" href="https://github.com.ipaddress.com/#ipinfo">http://github.com.ipaddress.c…</a> 获取cdn域名以及IP地址<br><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/ab03c6248ee9ba8b0e1ac111bad32294.png" alt="img"></li>
<li>或者直接访问<a target="_blank" rel="noopener" href="https://www.ipaddress.com/%E7%BD%91%E5%9D%80%E6%9F%A5%E8%AF%A2%E4%B8%8B%E9%9D%A2%E6%89%80%E9%9C%80%E7%9A%84%E5%9C%B0%E5%9D%80%E5%AF%B9%E5%BA%94%E7%9A%84IP">https://www.ipaddress.com/网址查询下面所需的地址对应的IP</a></li>
</ul>
<h3 id="修改hosts文件"><a href="#修改hosts文件" class="headerlink" title="修改hosts文件"></a>修改hosts文件</h3><p>将上述获取的IP地址添加到hosts文件中</p>
<p>Mac在<code>/etc/hosts</code>中；</p>
<p>windows在<code>C:\Windows\System32\drivers\etc\hosts</code>中；</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># fix git clone github project failed</span><br><span class="line">140.82.113.3 github.com</span><br><span class="line">199.232.5.194 github.global.ssl.fastly.net</span><br><span class="line">192.30.253.120  codeload.github.com</span><br></pre></td></tr></table></figure>

<h3 id="刷新DNS缓存"><a href="#刷新DNS缓存" class="headerlink" title="刷新DNS缓存"></a>刷新DNS缓存</h3><p>Mac刷新DNS缓存</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sudo killall -HUP mDNSResponder</span><br><span class="line">sudo dscacheutil -flushcache</span><br></pre></td></tr></table></figure>

<p>Windows更新DNS缓存</p>
<p>cmd命令窗口执行：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ipconfig /flushdns</span><br></pre></td></tr></table></figure>

<p>使增加的映射生效。</p>
<h2 id="解决办法3"><a href="#解决办法3" class="headerlink" title="解决办法3"></a>解决办法3</h2><p>使用chrome浏览器，安装一个插件【Github 加速器】</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/3c717bdc5fe69e4e44f121b7ceb2c36e.png" alt="img"></p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/watermark%2Ctype_d3F5LXplbmhlaQ%2Cshadow_50%2Ctext_Q1NETiBA55Cm5b2m%2Csize_20%2Ccolor_FFFFFF%2Ct_70%2Cg_se%2Cx_16.png" alt="img"></p>
<p>下载地址： <a target="_blank" rel="noopener" href="https://chrome.zzzmh.cn/info?token=mfnkflidjnladnkldfonnaicljppahpg">https://chrome.zzzmh.cn/info?token=mfnkflidjnladnkldfonnaicljppahpg</a></p>
<p>安装后见crx文件拖到谷歌的扩展功能里面，然后安装成功。<strong>打开gihub会看到如下界面，多了几个加速地址</strong>.</p>
<h2 id="解决办法4"><a href="#解决办法4" class="headerlink" title="解决办法4"></a>解决办法4</h2><p>利用socks5代理，配置好后明显加速。用下面两条命令配置好后，保持代理客户端开启就行了。</p>
<p>设置代理</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># 可以参考以下配置，这里的127.0.0.1代表的就是你本机的IP地址，因为服务就在你本机上运行</span><br><span class="line"># 10809是代理的客户端本地端口,默认是1080</span><br><span class="line">git config --global https.proxy http://127.0.0.1:10800</span><br><span class="line">git config --global https.proxy http://127.0.0.1:10800</span><br><span class="line">git config --global http.proxy &#x27;socks5://127.0.0.1:10800&#x27;</span><br><span class="line">git config --global https.proxy &#x27;socks5://127.0.0.1:10800&#x27;</span><br></pre></td></tr></table></figure>

<p>上述方法挂了全局代理，但是如果要克隆码云、coding等国内仓库，速度就会很慢。更好的方法是<strong>只对github进行代理</strong>，不会影响国内仓库：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">git config --global http.https://github.com.proxy socks5://127.0.0.1:10800</span><br><span class="line">git config --global https.https://github.com.proxy socks5://127.0.0.1:10800</span><br></pre></td></tr></table></figure>

<p>代理后git clone速度真的快了很多~我这边能到1MB&#x2F;s以上</p>
<p>查看代理</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git config --global -l</span><br></pre></td></tr></table></figure>

<p>取消代理</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">git config --global --unset http.proxy</span><br><span class="line">git config --global --unset http.proxy</span><br></pre></td></tr></table></figure>



<h1 id="其他异常问题"><a href="#其他异常问题" class="headerlink" title="其他异常问题"></a>其他异常问题</h1><h2 id="1-解决-curl-7-Failed-to-connect-to-raw-githubusercontent-com-port-443-Connection-refused"><a href="#1-解决-curl-7-Failed-to-connect-to-raw-githubusercontent-com-port-443-Connection-refused" class="headerlink" title="1. 解决 curl: (7) Failed to connect to raw.githubusercontent.com port 443: Connection refused"></a>1. 解决 curl: (7) Failed to connect to raw.githubusercontent.com port 443: Connection refused</h2><p>将下面的IP地址添加到hosts文件中</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">199.232.68.133 raw.githubusercontent.com</span><br><span class="line">199.232.68.133 user-images.githubusercontent.com</span><br><span class="line">199.232.68.133 avatars2.githubusercontent.com</span><br><span class="line">199.232.68.133 avatars1.githubusercontent.com</span><br></pre></td></tr></table></figure>

<h2 id="2-解决curl-56-OpenSSL-SSL-read-Connection-was-reset-errno-10054"><a href="#2-解决curl-56-OpenSSL-SSL-read-Connection-was-reset-errno-10054" class="headerlink" title="2. 解决curl: (56) OpenSSL SSL_read: Connection was reset, errno 10054"></a>2. 解决curl: (56) OpenSSL SSL_read: Connection was reset, errno 10054</h2><p>一、问题：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">error: RPC failed; curl 56 OpenSSL SSL_read: Connection was reset, errno 10054</span><br></pre></td></tr></table></figure>

<p>如图：</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/d87dea031623518fcfe19791939dd606.png" alt="img"></p>
<p>原因：远程仓库的文件过大，需要设置本地仓库大小**</p>
<p><strong>二、解决步骤：</strong></p>
<p>1、首先输入如下命令：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git config http.sslVerify &quot;false&quot;</span><br></pre></td></tr></table></figure>

<p>若出现下列错误：</p>
<p><strong><code>git config http.sslVerify &quot;false&quot; fatal: not in a git directory</code></strong></p>
<p>再继续执行 <strong><code>git config --globle http.sslVerify &quot;false&quot;</code></strong> 问题解决</p>
<p>2、文件大小的上限设置：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git config --global http.postBuffer 524288000</span><br></pre></td></tr></table></figure>

<p>fatal: not in a git directory <code>**再继续执行**</code>git config –globle http.sslVerify “false”&#96;** 问题解决</p>
<p>2、文件大小的上限设置：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git config --global http.postBuffer 524288000</span><br></pre></td></tr></table></figure>

<p>3、如果还是git代码还是下载失败，则需要继续修改git缓存的大小</p>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
	<div class="article-tag tagcloud">
		<i class="icon-price-tags icon"></i>
		<ul class="article-tag-list">
			 
        		<li class="article-tag-list-item">
        			<a href="javascript:void(0)" class="js-tag article-tag-list-link color1">计算机随笔</a>
        		</li>
      		
		</ul>
	</div>

    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/15/%E8%A7%A3%E5%86%B3Github%E4%B8%8D%E8%83%BD%E4%B8%8B%E8%BD%BD%EF%BC%8C%E4%B8%8B%E8%BD%BD%E9%80%9F%E5%BA%A6%E6%85%A2%E7%9A%84%E5%B0%8F%E5%A6%99%E6%8B%9B/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-使用Github-picGo搭建图床，保姆级教程" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/15/%E4%BD%BF%E7%94%A8Github-picGo%E6%90%AD%E5%BB%BA%E5%9B%BE%E5%BA%8A%EF%BC%8C%E4%BF%9D%E5%A7%86%E7%BA%A7%E6%95%99%E7%A8%8B/" target="_blank">使用Github+picGo搭建图床，保姆级教程</a>
  
    </h1>
  


  
  
<a href="/2024/03/15/%E4%BD%BF%E7%94%A8Github-picGo%E6%90%AD%E5%BB%BA%E5%9B%BE%E5%BA%8A%EF%BC%8C%E4%BF%9D%E5%A7%86%E7%BA%A7%E6%95%99%E7%A8%8B/" class="archive-article-date">
        <time datetime="2024-03-15T06:08:47.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-15</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <h1 id="使用Github-picGo搭建图床，保姆级"><a href="#使用Github-picGo搭建图床，保姆级" class="headerlink" title="使用Github+picGo搭建图床，保姆级"></a>使用Github+picGo搭建图床，保姆级</h1><p>转载自<a target="_blank" rel="noopener" href="https://www.zhihu.com/people/xiao-xiao-yu-zi-12-58">小小芋子</a></p>
<h2 id="1-注册一个Github账号"><a href="#1-注册一个Github账号" class="headerlink" title="1. 注册一个Github账号"></a><strong>1. 注册一个Github账号</strong></h2><p>首先你需要一个github账号，如果没有的话，先注册。</p>
<p>github官网地址： <a href="https://link.zhihu.com/?target=https://github.com/">https://github.com/</a></p>
<p>注册过程省略，因为这是保姆都不管的事情。</p>
<p>友情提示：可能在注册过程中会出现 “Unable to verify your captcha response… …”，一直不能正常注册。</p>
<p>网上有各种解决方案：换浏览器、换电脑… …</p>
<h2 id="2-配置Github"><a href="#2-配置Github" class="headerlink" title="2. 配置Github"></a>2. 配置Github</h2><h3 id="2-1-创建一个新仓库，用于存放图片。"><a href="#2-1-创建一个新仓库，用于存放图片。" class="headerlink" title="2.1 创建一个新仓库，用于存放图片。"></a>2.1 创建一个新仓库，用于存放图片。</h3><p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-59b7f87b4fbe21a2492e4b9e529fe87c_1440w.webp" alt="img"></p>
<p>填写仓库名称和描述，且仓库必须是public的，否则存储的图片不能正常访问。</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-d3733c14e8c9b53cb5cf9f7607d51c8a_1440w.webp" alt="img"></p>
<h3 id="2-2-生成一个token，用于picGo访问github"><a href="#2-2-生成一个token，用于picGo访问github" class="headerlink" title="2.2 生成一个token，用于picGo访问github"></a>2.2 生成一个token，用于picGo访问github</h3><p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-4d227062bcfa58dba6c0582c982b737b_1440w.webp" alt="img"></p>
<p>选择左侧菜单的Developer settings</p>
<p><img src="/%E4%BD%BF%E7%94%A8Github-picGo%E6%90%AD%E5%BB%BA%E5%9B%BE%E5%BA%8A%EF%BC%8C%E4%BF%9D%E5%A7%86%E7%BA%A7%E6%95%99%E7%A8%8B.assets/v2-4d227062bcfa58dba6c0582c982b737b_1440w-17104837862607.webp" alt="img"></p>
<p>之后选择左侧Personal access tokens，再点击Generate new token</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-2d5bcabac97557591c4203009d306750_1440w.webp" alt="img"></p>
<p>填写Note，勾选repo.</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-346da4ccf189eb5997abe2fadadca331_1440w.webp" alt="img"></p>
<p>注意，生成的token只会在这里显示一次，所以记得单独保存下来哦。</p>
<p><img src="/%E4%BD%BF%E7%94%A8Github-picGo%E6%90%AD%E5%BB%BA%E5%9B%BE%E5%BA%8A%EF%BC%8C%E4%BF%9D%E5%A7%86%E7%BA%A7%E6%95%99%E7%A8%8B.assets/v2-354ea85a4bd9cfc99157b86cae9a2332_1440w.webp" alt="img"></p>
<p>至此，Github的配置完成。</p>
<h2 id="3-下载picGo，并进行配置"><a href="#3-下载picGo，并进行配置" class="headerlink" title="3. 下载picGo，并进行配置"></a>3. 下载picGo，并进行配置</h2><h3 id="3-1-下载"><a href="#3-1-下载" class="headerlink" title="3.1 下载"></a>3.1 下载</h3><p>网络不好的情况下，PicGo下载可能会多次失败，所以我把下载好的应用放在百度网盘了。</p>
<p>大家按需取用。</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-279f576c0dc0861fd7edcf1d43e5675f_1440w.webp" alt="img"></p>
<p>网盘地址：<a href="https://link.zhihu.com/?target=https://pan.baidu.com/s/1LvkzLJyZcjpflK2Evq3X9A">https://pan.baidu.com/s/1LvkzLJyZcjpflK2Evq3X9A</a> 提取码：s6x7</p>
<p>下载完成，双击启动安装即可。</p>
<p>如果安装成功，picGo不能正常使用，则可以用兼容模式启动。【此问题因电脑而异，也是我在配置过程中踩过的坑。】</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-3f60ccb549e01c95a1702ceca9bda95d_1440w.webp" alt="img"></p>
<h3 id="3-2-配置"><a href="#3-2-配置" class="headerlink" title="3.2 配置"></a>3.2 配置</h3><p>仓库名：[github用户名]&#x2F;[第一步新建的仓库名称]</p>
<p>分支：默认master，从2020.10.01开始，github的默认分支名变更为main</p>
<p>设定token：第一步创建的token</p>
<p>指定存储路径：可填可不填，如果填写了，图片就会存储在img文件夹下</p>
<p>设定自定义域名：<a target="_blank" rel="noopener" href="https://cdn.jsdelivr.net/gh/[github%E7%94%A8%E6%88%B7%E5%90%8D]/[%E4%BB%93%E5%BA%93%E5%90%8D]@main%EF%BC%8C%E6%B3%A8%E6%84%8F%EF%BC%8C%E6%AD%A4%E5%A4%84%E7%9A%84%E5%88%86%E6%94%AF%E4%B8%80%E5%AE%9A%E8%A6%81%E5%A1%AB%E5%86%99@main%EF%BC%8C%E5%90%A6%E5%88%99%E9%BB%98%E8%AE%A4%E4%BD%BF%E7%94%A8master%E5%88%86%E6%94%AF%E3%80%82%E8%80%8C%E7%8E%B0%E5%9C%A8github%E5%88%9B%E5%BB%BA%E7%9A%84%E9%BB%98%E8%AE%A4%E5%88%86%E6%94%AF%E5%90%8D%E4%B8%BAmain%EF%BC%8C%E5%A6%82%E6%9E%9C%E4%B8%8D%E6%8C%87%E5%AE%9A%EF%BC%8C%E5%88%99%E4%BC%9A%E5%87%BA%E7%8E%B0%E5%9B%BE%E7%89%87%E4%B8%8D%E8%83%BD%E4%B8%8A%E4%BC%A0%E7%9A%84%E6%83%85%E5%86%B5%E3%80%82%E3%80%90%E8%B8%A9%E5%9D%91%E4%B8%A4%E5%B0%8F%E6%97%B6%E3%80%91">https://cdn.jsdelivr.net/gh/[github用户名]/[仓库名]@main，注意，此处的分支一定要填写@main，否则默认使用master分支。而现在github创建的默认分支名为main，如果不指定，则会出现图片不能上传的情况。【踩坑两小时】</a></p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-62e4faaa7999d1d32fa80aec44b4034e_1440w.webp" alt="img"></p>
<p>至此，github+picGo的配置完成，可以在上传区进行图片上传了。</p>
<h3 id="3-3-补充几点"><a href="#3-3-补充几点" class="headerlink" title="3.3 补充几点"></a>3.3 补充几点</h3><p>1.原本github的自定义域名应该是：[<a href="https://raw.githubusercontent.com/[username/]/[%E4%BB%93%E5%BA%93%E5%90%8D]](https://link.zhihu.com/?target=https://raw.githubusercontent.com/%5Busername%5D/%5B%E4%BB%93%E5%BA%93%E5%90%8D%5D)">https://raw.githubusercontent.com/[username\]/[仓库名]](https://link.zhihu.com/?target=https%3A//raw.githubusercontent.com/%5Busername%5D/%5B%E4%BB%93%E5%BA%93%E5%90%8D%5D)</a></p>
<p>但是使用这种方式访问图片巨慢，所以教程中使用了jsdelivr作为cdn加速。改变域名即可，不需要任何其他配置。</p>
<p>2.配置完成，可能出现不能上传的情况，请大家耐心检查，也许某一步的配置出现了问题，就像检查bug一样耐心、细心。想象一下图床搭建成功后，写起文章来的丝滑感，是不是又有动力了？</p>
<h2 id="4-PicGo集成Typora"><a href="#4-PicGo集成Typora" class="headerlink" title="4. PicGo集成Typora"></a>4. PicGo集成Typora</h2><p>使用MarkDown语法写文章，我使用的编辑器是typora，一款超级好用的编辑器。在一个编辑框内，编辑完成后，可以马上显示MarkDown语法效果。</p>
<p>举个栗子：</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-ce0b207c19d6ca32ac22937f1dc6cfcd_1440w.webp" alt="img"></p>
<p>并且typora可以和picGo联合在一起使用，当你在编辑器中插入、粘贴图片时，typora+picGo的组合方式，可以自动把图片上传至github保存。</p>
<h3 id="4-1-下载安装"><a href="#4-1-下载安装" class="headerlink" title="4.1 下载安装"></a>4.1 下载安装</h3><p>Typora新版本是收费的，所以建议大家使用旧版本就可以了，完全能满足写作需求。</p>
<p>旧版本安装包和picGo放在一起了（下载地址见上文），大家按需取用即可。</p>
<p>下载完成后正常步骤安装即可。</p>
<h3 id="4-2-配置"><a href="#4-2-配置" class="headerlink" title="4.2 配置"></a>4.2 配置</h3><p>在typora顶部菜单界面，选择“文件” - &gt; “偏好设置”，设置图片存储方式。</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-f595cf5c8a0e3b62944c68deb8fb5878_1440w.webp" alt="img"></p>
<p>选择图片存储方式：上传图片。</p>
<p>上传服务：PicGo(app)</p>
<p>PicGo路径：picGo安装的地址</p>
<p>设置完成，点击“验证图片上传”，提示上传成功，即代表配置成功。</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-e27e9f198aef7bdea365549b2257025d_1440w.webp" alt="img"></p>
<p>注意一点，typora图片验证中的端口，需要与picGo中的server设置内的端口一致，否则typora中不能正常上传。</p>
<h3 id="4-3-picGo监听端口设置"><a href="#4-3-picGo监听端口设置" class="headerlink" title="4.3 picGo监听端口设置"></a>4.3 picGo监听端口设置</h3><p>选择“PicGo设置”–&gt;“设置server”</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-8ff622935a6b0597a21e68e7575e16ef_1440w.webp" alt="img"></p>
<p>【此处有坑】如果监听端口不是36677，则需要修改为36677。否则会出现picGo能正常上传 图片，而typora上传图片失败的情况。</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage2@main/img/v2-304589efaed9a1ba3caf5e8827e77582_1440w.webp" alt="img"></p>
<p>最后，在苹果电脑中可以使用<strong>Typora+Ipic+Github</strong>的方式，有兴趣的话可以自行尝试</p>
<p><strong>总结：github+picGo+typora他们三个是好基友。</strong></p>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
	<div class="article-tag tagcloud">
		<i class="icon-price-tags icon"></i>
		<ul class="article-tag-list">
			 
        		<li class="article-tag-list-item">
        			<a href="javascript:void(0)" class="js-tag article-tag-list-link color1">计算机随笔</a>
        		</li>
      		
		</ul>
	</div>

    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/15/%E4%BD%BF%E7%94%A8Github-picGo%E6%90%AD%E5%BB%BA%E5%9B%BE%E5%BA%8A%EF%BC%8C%E4%BF%9D%E5%A7%86%E7%BA%A7%E6%95%99%E7%A8%8B/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-小鼠体外受精" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/15/%E5%B0%8F%E9%BC%A0%E4%BD%93%E5%A4%96%E5%8F%97%E7%B2%BE/" target="_blank">小鼠体外受精2</a>
  
    </h1>
  


  
  
<a href="/2024/03/15/%E5%B0%8F%E9%BC%A0%E4%BD%93%E5%A4%96%E5%8F%97%E7%B2%BE/" class="archive-article-date">
        <time datetime="2024-03-14T16:08:45.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-15</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <p>[TOC]</p>
<p><a href="javascript:;">杨伟伟</a>，<a href="javascript:;">顾美儿</a>，<a href="javascript:;">桂飞</a>，<a href="javascript:;">宋晓明</a>，<a href="javascript:;">孙筱品</a>，<a href="javascript:;">唐蔚</a>，<a href="javascript:;">吴宝金 </a></p>
<p>来源DOI:  <a target="_blank" rel="noopener" href="https://doi.org/10.21769/BioProtoc.1010947">10.21769&#x2F;BioProtoc.1010947</a></p>
<p>杨伟伟, 顾美儿, 桂飞, 宋晓明, 孙筱品, 唐蔚, 吴宝金. (2022). 小鼠体外受精. &#x2F;&#x2F; 实验动物胚胎操作实验手册. <em>Bio-101</em>: e1010947. DOI: <a target="_blank" rel="noopener" href="https://doi.org/10.21769/BioProtoc.1010947">10.21769&#x2F;BioProtoc.1010947</a>.</p>
<p>第一次发表时间:  2022&#x2F;09&#x2F;20</p>
<p><strong>摘要:</strong> 小鼠体外受精实验是基因工程小鼠的辅助生殖及遗传工程小鼠精子库的重要支撑技术。在常规In Vitro Fertilization(IVF)操作的基础上，Nakagata改良法在modified human tubal fluid (mHTF)中加入Reduced glutathione(GSH)作为受精液（mHTF-GSH），在modified Krebs-Ringer’s bicarbonate medium中加入Methyl-β-cyclodextrin作为获能液（TYH with MβCD），显著提升了体外受精效率。该法适用于新鲜及冷冻复苏精子的体外受精操作。</p>
<p>关键词: 小鼠, mHTF-GSH, TYH with MβCD, 体外受精</p>
<hr>
<h3 id="材料与试剂"><a href="#材料与试剂" class="headerlink" title="材料与试剂"></a>材料与试剂</h3><h4 id="材料"><a href="#材料" class="headerlink" title="材料"></a><strong>材料</strong></h4><ol>
<li>培养皿35mm×10mm（Corning，货号：430588）</li>
<li>胚胎移植口吸管（北京吉田生物科技有限公司，规格：1.0×500mm）</li>
<li>吸头（Axygen，货号：T-300-R-S、T-1000-B-R-S、T-200-Y-R-S）</li>
<li>无菌注射器（上海米沙瓦医科工业有限公司，规格：1mL）</li>
<li>50mL离心管（BD，货号：352070）</li>
<li>滤纸（宝威德，规格：60 × 60 cm）</li>
</ol>
<p><strong>动物：</strong>3-6月龄成年雄鼠，1-2只&#x2F;品系。4-5周龄雌性供体小鼠，4-5只&#x2F;品系（根据雄性小鼠背景品系选用)。本实验室所用动物来自杭州师范大学实验动物中心（生产许可证号：SCXK（浙）2016-0004，使用许可证号：SYXK（浙）2016-0014），饲养在独立换气笼盒系统，动物自由饮水、自由采食经Co60辐射的灭菌饲料。</p>
<h4 id="试剂"><a href="#试剂" class="headerlink" title="试剂"></a><strong>试剂</strong></h4><ol>
<li>注射用血促性素（PMSG, Pregnant Mare Serum Gonadotropin）(宁波市三生药业有限公司, 规格:1000IU×10瓶)和注射用绒促性素（hCG, human Chorionic Gonadotropin,）(宁波市三生药业有限公司, 规格:1000IU×10支)</li>
<li>矿物油（Sigma，货号：M8410）</li>
<li>mHTF：按表1配制。</li>
<li>mHTF-GSH：按表2配制。</li>
<li>TYH with MβCD（TYH with Methyl-β-cyclodextrin）按表3配制。</li>
</ol>
<h4 id="仪器设备"><a href="#仪器设备" class="headerlink" title="仪器设备"></a>仪器设备</h4><ol>
<li>100mL烧杯（蜀牛，货号：GG-17）</li>
<li>100mL容量瓶（天玻，货号：TB9）</li>
<li>生物安全柜（Thermo Scientific，型号：1384 ）</li>
<li>超净工作台（苏净安泰，型号：HS-840-U）</li>
<li>CO2培养箱（Thermo Scientific，型号：3111 ）</li>
<li>体视显微镜（Nikon，型号： SMZ745T）</li>
<li>电子天平（Sartorius，型号：MSA125P-CE）</li>
<li>电热恒温水浴锅（上海精宏实验设备有限公司，型号：DK-S22型）</li>
<li>微量移液器 （eppendorf，货号：3120000224、3120000240）</li>
<li>手术剪（上海金钟，型号：J21010）</li>
<li>医用镊（上海金钟，型号：J40120）</li>
<li>维纳斯剪（上海金钟，型号：YC0040）</li>
<li>显微镊（上海金钟，型号：WA3030）</li>
<li>眼科直剪（上海金钟，型号：Y00030）</li>
<li>眼科直镊（上海金钟，型号：JD1050）</li>
</ol>
<hr>
<h3 id="实验步骤"><a href="#实验步骤" class="headerlink" title="实验步骤"></a>实验步骤</h3><h4 id="一、准备工作"><a href="#一、准备工作" class="headerlink" title="一、准备工作"></a>一、准备工作</h4><ol>
<li><p>超数排卵：供体小鼠腹腔注射PMSG，每只5IU（0.1mL）。46-48h后，腹腔注射hCG，每只5IU（0.1mL）。</p>
</li>
<li><p>精子获能滴（TYH with MβCD，如图1-A）：用TYH with MβCD液在培养皿内做一个100µL液滴（冻存精液滴为90µL），加3-4mL矿物油覆盖，在培养箱（37℃、5%CO2）孵育至少30min。</p>
</li>
<li><p>受精滴（mHTF-GSH，如图1-B）：用mHTF-GSH液在培养皿内做两个200µL液滴（冻存精液滴为90µL），加3-4mL矿物油覆盖，在培养箱（37℃、5%CO2）孵育至少10min。</p>
</li>
<li><p>清洗滴（mHTF，如图1-C）：用mHTF液在培养皿内做3个80µL液滴和1个160µL液滴，加3-4mL矿物油覆盖，在培养箱（37℃、5%CO2）孵育至少30min。</p>
</li>
<li><p>水浴准备：恒温水浴锅开启37℃水浴。</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage@main/img/20220811202317_0252.jpg" alt="img"><br><strong>图1. 小鼠体外受精实验培养滴</strong> A. 精子获能滴（TYH with MβCD） B. 受精滴（mHTF-GSH） C. 清洗滴（mHTF），胚胎按照图中白色箭头方向清洗转移</p>
</li>
</ol>
<hr>
<h4 id="二、操作流程"><a href="#二、操作流程" class="headerlink" title="二、操作流程"></a>二、操作流程</h4><ol>
<li><p>精子的采集及获能</p>
</li>
<li><p>卵母细胞采集（如图2-D,E,F）：将注射hCG的供体小鼠（15-17h后）断颈处死，酒精消毒。剪开背部皮肤，快速找到输卵管，在滤纸上吸干，放入受精滴培养皿的矿物油中。在显微镜下找到膨大部，牵引卵丘团流出，并将其拖入mHTF-GSH滴中（取至4-6簇卵丘团）。注意：为保证质量，在处死动物到卵丘团放入液滴的过程中，务必在30sec内完成。</p>
</li>
<li><p>体外受精（如图2-G,H）：精液孵育50-60min后，观察精子活力，并沿着TYH with MβCD滴的外围用移液器吸取4-5µL精液（冷冻复苏精液可根据精子活力及浓度增加量至10-20µL），放入含有卵母细胞的mHTF-GSH滴中，并将其放入培养箱（37℃、5%CO2）培养3h。</p>
</li>
<li><p>清洗：受精后培养3h，将细胞先转入80µL的mHTF滴中，3次清洗后，转入160µL的mHTF滴培养6h，去除未受精的卵母细胞和含有单原核的孤雌生殖卵母细胞，保留含有第二极体和两个原核的受精卵细胞（如图2-I,J）。</p>
</li>
<li><p>受精卵过夜培养：次日挑选发育正常的2-细胞（如图2-K），用胚胎移植口吸管转移至mHTF的第4个滴中，移植或者胚胎冻存备用。同时记录2-细胞、单细胞和异型细胞的数量，计算受精率。</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage@main/img/20220811203426_9863.jpg" alt="img"><br><strong>图2. 小鼠体外受精过程.</strong> A. 附睾的解剖位置 A-1. 附睾组织结构图 B. 取出的附睾尾 C. 冷冻精子复苏精液 D. 输卵管解剖位置 E. 输卵管壶腹部 F.卵母细胞-卵丘团 G. 获能完成的精子悬液 H. 受精 I、J. 受精6小时后细胞形态 K. 受精24小时后细胞形态</p>
</li>
</ol>
<p>溶液配方</p>
<ol>
<li><p>mHTF配制表</p>
</li>
<li><p>mHTF-GSH配制表</p>
</li>
<li><p>TYH with MβCD配制表</p>
<p>表1：mHTF配制表（100mL体积）</p>
<table>
<thead>
<tr>
<th>名称</th>
<th>重量（mg）</th>
<th>品牌</th>
<th>货号</th>
</tr>
</thead>
<tbody><tr>
<td>NaCl</td>
<td>593.8</td>
<td>Sigma</td>
<td>S-5886</td>
</tr>
<tr>
<td>KCl</td>
<td>35.0</td>
<td>Sigma</td>
<td>P-5405</td>
</tr>
<tr>
<td>CaCl2</td>
<td>57.0</td>
<td>Sigma</td>
<td>G-5670</td>
</tr>
<tr>
<td>Glucose</td>
<td>50.0</td>
<td>Sigma</td>
<td>G-6152</td>
</tr>
<tr>
<td>MgSO4</td>
<td>2.4</td>
<td>Sigma</td>
<td>M2643</td>
</tr>
<tr>
<td>KH2PO4</td>
<td>5.4</td>
<td>Sigma</td>
<td>P-5655</td>
</tr>
<tr>
<td>Sodium pyruvate</td>
<td>3.5</td>
<td>Sigma</td>
<td>P-4562</td>
</tr>
<tr>
<td>60%sodium lacate</td>
<td>0.34mL</td>
<td>Sigma</td>
<td>L-7900</td>
</tr>
<tr>
<td>NaHCO3</td>
<td>210.0</td>
<td>Sigma</td>
<td>S-5761</td>
</tr>
<tr>
<td>Penicillin G</td>
<td>7.5</td>
<td>BD</td>
<td>C15935000</td>
</tr>
<tr>
<td>Streptomycin</td>
<td>5.0</td>
<td>Sigma</td>
<td>S-1277</td>
</tr>
<tr>
<td>Bovine serum albumin</td>
<td>400.0</td>
<td>Sigma</td>
<td>B2064</td>
</tr>
<tr>
<td>0.5%phenol red</td>
<td>0.04ml</td>
<td>Sigma</td>
<td>P-0290</td>
</tr>
</tbody></table>
<p>配制方法：在100mL的烧杯中，加入80mL的灭菌水(Sigma,W1503)，将上述药品依次加入混匀，容量瓶定容至100mL。0.22µM针式过滤器过滤分装在1.5mL的离心管内。4℃保存可用30天。</p>
<p>表2：mHTF-GSH配制（1mL体积）</p>
<table>
<thead>
<tr>
<th>名称</th>
<th>重量（mg）</th>
<th>品牌</th>
<th>货号</th>
</tr>
</thead>
<tbody><tr>
<td>GSH</td>
<td>1.5</td>
<td>Sigma</td>
<td>G4251</td>
</tr>
<tr>
<td>mHTF</td>
<td>1（mL）</td>
<td>-</td>
<td>无</td>
</tr>
</tbody></table>
<p>注：“-” 为混合试剂，配方参见表1</p>
<p>配制方法：称取1.5mg(冷冻精液取3.1mg)的GSH，溶解在装有mHTF(1mL)的1.5mL离心管内，再分次将100µL上述溶液转移至装有mHTF（900µL）的1.5mL离心管内，混匀密封，-20℃冷冻保存备用，有效期约30天。</p>
<p>表3：TYH with MβCD配制（100mL体积）</p>
<table>
<thead>
<tr>
<th>名称</th>
<th>重量（mg）</th>
<th>品牌</th>
<th>货号</th>
</tr>
</thead>
<tbody><tr>
<td>NaCl</td>
<td>697.6</td>
<td>Sigma</td>
<td>S-5886</td>
</tr>
<tr>
<td>KCl</td>
<td>35.6</td>
<td>Sigma</td>
<td>P-5405</td>
</tr>
<tr>
<td>CaCl2.2H2O</td>
<td>25.1</td>
<td>Sigma</td>
<td>G-5670</td>
</tr>
<tr>
<td>Glucose</td>
<td>100.0</td>
<td>Sigma</td>
<td>G-6152</td>
</tr>
<tr>
<td>MgSO4</td>
<td>14.3</td>
<td>Sigma</td>
<td>M2643</td>
</tr>
<tr>
<td>KH2PO4</td>
<td>16.2</td>
<td>Sigma</td>
<td>P-5655</td>
</tr>
<tr>
<td>Sodium pyruvate</td>
<td>5.5</td>
<td>Sigma</td>
<td>P-4562</td>
</tr>
<tr>
<td>Polyvinyl alcohol</td>
<td>100.0</td>
<td>Sigma</td>
<td>P-8136</td>
</tr>
<tr>
<td>Penicillin G</td>
<td>7.5</td>
<td>BD</td>
<td>C15935000</td>
</tr>
<tr>
<td>Streptomycin</td>
<td>5.0</td>
<td>Sigma</td>
<td>S-1277</td>
</tr>
<tr>
<td>Methyl-β-cyclodextrin</td>
<td>98.3</td>
<td>Sigma</td>
<td>C-4555</td>
</tr>
<tr>
<td>NaHCO3</td>
<td>210.6</td>
<td>Sigma</td>
<td>S-5761</td>
</tr>
</tbody></table>
<p>配制方法：在100mL烧杯中加入70mL的灭菌水（Sigma,W1503），将上述药品依次加入，混匀；同时，100mg（Polyvinyl alcohol）放入含有10mL灭菌水的离心管中，60℃水浴10min，溶解后加入烧杯中，再依次加入其它试剂溶解，最后定容至100mL。0.22µM针式过滤器过滤分装在1.5mL的离心管内。4℃保存可用30天。</p>
<p>PMSG配制方法：将1000IU &#x2F;支的PMSG稀释在20mL灭菌水中，制成50IU&#x2F;mL的备用液。-20℃冷冻保存可使用14天。</p>
<p>hCG配制方法：将1000IU&#x2F;支的hCG稀释在20mL灭菌水中，制成50IU&#x2F;mL的备用液。-20℃冷冻保存可使用14天。</p>
</li>
</ol>
<hr>
<h3 id="经验体会"><a href="#经验体会" class="headerlink" title="经验体会"></a>经验体会</h3><ol>
<li>供体小鼠年龄的要求，常用4-5周龄背景品系雌鼠，超排效果较好，获取的卵母细胞数量较多。6-8周龄小鼠的超排效果相对较差。8周龄以上的雌鼠，在超排时，根据体重适当加大超排剂量，超排效果尚可。</li>
<li>体外受精对雄鼠的要求不高，健康雄鼠精液质量较好，受精率较高一般均能达到90%上。若雄鼠年龄偏大，精液质量欠佳，受精率会相对较低一般在50-80%左右。个别品系及个体差异出现过30%左右的受精率。可根据雄鼠状态相应准备足够数量的供体雌鼠。</li>
<li>卵母细胞的采集，尽可能在小鼠断颈后30sec内完成，如果无人配合，可单只处死快速采集，可减少异形细胞或孤雌生殖卵母细胞的出现。</li>
<li>体外受精时精液的用量一般为3-5μL，可根据精液的浓度和精子活力适当调整加液量。冷冻复苏时精液应加倍。</li>
<li>在受精后6h，可观察到含有第二极体和两个原核的受精卵细胞，同时含有单原核的孤雌生殖卵母细胞，这种细胞需剔除。</li>
<li>移卵针的挑选。如果是自行制作的，应保证移卵针前端圆润，以免伤到细胞，同时应注意孔径大小，若过大，在吸放卵细胞时不好控制，导致清洗不干净；若过小，有可能会损伤细胞，吸放比较费力。目前有商品化的移卵针，可根据情况自行选用。</li>
<li>在TYH with MβCD配制过程中，因聚乙烯醇可溶性不好，需要先溶于水，涡旋混匀，在60℃水浴锅内温浴10min，之后再加入总的液体中。</li>
<li>加入还原性谷胱甘肽的HTF作为受精液，可有效提高受精率，尤其对冷冻复苏精子尤为明显。</li>
</ol>
<hr>
<h3 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h3><ol>
<li>Takeo, T. and Nakagata, N. (2011). <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/21778138">Reduced glutathione enhances fertility of frozen&#x2F;thawed C57BL&#x2F;6 mouse sperm after exposure to methyl-beta-cyclodextrin.</a> <em>Biol Reprod</em> 85(5): 1066-1072.</li>
<li>Takeo, T., Sztein, J. and Nakagata, N. (2019). <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/30353518">The CARD Method for Mouse Sperm Cryopreservation and In Vitro Fertilization Using Frozen-Thawed Sperm.</a> <em>Methods Mol Biol</em> 1874: 243-256.</li>
<li>Guan, M., Bogani, D., Marschall, S., Raspa, M., Takeo, T., Nakagata, N. and Fray, M. (2014). <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/25723919">In vitro fertilization in mice using the MBCD-GSH protocol.</a> <em>Curr Protoc Mouse Biol</em> 4(2): 67-83.</li>
<li>Nakagata, N., Takeo, T., Fukumoto, K., Haruguchi, Y., Kondo, T., Takeshita, Y., Nakamuta, Y., Umeno, T. and Tsuchiyama, S. (2014). <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/24492659">Rescue in vitro fertilization method for legacy stock of frozen mouse sperm.</a> <em>J Reprod Dev</em> 60(2): 168-171.</li>
<li>Takeo, T. and Nakagata, N. (2010). <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/20357317">Combination medium of cryoprotective agents containing L-glutamine and methyl-{beta}-cyclodextrin in a preincubation medium yields a high fertilization rate for cryopreserved C57BL&#x2F;6J mouse sperm.</a> <em>Lab Anim</em> 44(2): 132-137.</li>
<li>Takeo, T., Hoshii, T., Kondo, Y., Toyodome, H., Arima, H., Yamamura, K., Irie, T. and Nakagata, N. (2008). <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/18046011">Methyl-beta-cyclodextrin improves fertilizing ability of C57BL&#x2F;6 mouse sperm after freezing and thawing by facilitating cholesterol efflux from the cells.</a> <em>Biol Reprod</em> 78(3): 546-551.</li>
</ol>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
	<div class="article-tag tagcloud">
		<i class="icon-price-tags icon"></i>
		<ul class="article-tag-list">
			 
        		<li class="article-tag-list-item">
        			<a href="javascript:void(0)" class="js-tag article-tag-list-link color2">BioProtocol</a>
        		</li>
      		
		</ul>
	</div>

    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/15/%E5%B0%8F%E9%BC%A0%E4%BD%93%E5%A4%96%E5%8F%97%E7%B2%BE/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-小鼠受精卵原核注射及胞浆注射" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
  
  
    <h1 itemprop="name">
	
	  <a class="article-title" href="/2024/03/15/%E5%B0%8F%E9%BC%A0%E5%8F%97%E7%B2%BE%E5%8D%B5%E5%8E%9F%E6%A0%B8%E6%B3%A8%E5%B0%84%E5%8F%8A%E8%83%9E%E6%B5%86%E6%B3%A8%E5%B0%84/" target="_blank">小鼠受精卵原核注射及胞浆注射</a>
  
    </h1>
  


  
  
<a href="/2024/03/15/%E5%B0%8F%E9%BC%A0%E5%8F%97%E7%B2%BE%E5%8D%B5%E5%8E%9F%E6%A0%B8%E6%B3%A8%E5%B0%84%E5%8F%8A%E8%83%9E%E6%B5%86%E6%B3%A8%E5%B0%84/" class="archive-article-date">
        <time datetime="2024-03-14T16:04:32.000Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2024-03-15</time>
</a>

  
  
    

  
  </header>
  
  <div class="article-entry" itemprop="articleBody">
    
    <p>[TOC]</p>
<p><a href="javascript:;">唐蔚</a>，<a href="javascript:;">李周结</a>，<a href="javascript:;">俞骞</a>，<a href="javascript:;">陈子尘</a>，<a href="javascript:;">王倩雯</a>，<a href="javascript:;">陈国元</a>，<a href="javascript:;">丁一夫</a>，<a href="javascript:;">吴宝金 </a></p>
<p>来源DOI:  <a target="_blank" rel="noopener" href="https://doi.org/10.21769/BioProtoc.1010948">10.21769&#x2F;BioProtoc.1010948</a></p>
<p>第一次发表时间:  2022&#x2F;09&#x2F;20</p>
<p><strong>摘要:</strong> 将外源基因DNA序列直接注射到原核期的受精卵雄性原核中，或将sgRNA及Cas9等基因打靶构建注入受精卵的胞浆内，经过显微注射的受精卵再移植到受体动物的输卵管或子宫内继续发育，进而可得到转基因&#x2F;基因敲除动物。</p>
<h3 id="材料与试剂"><a href="#材料与试剂" class="headerlink" title="材料与试剂"></a>材料与试剂</h3><h4 id="（一）材料"><a href="#（一）材料" class="headerlink" title="（一）材料"></a>（一）材料</h4><ol>
<li><h5 id="手术器械"><a href="#手术器械" class="headerlink" title="手术器械"></a>手术器械</h5><p>系线镊（协和, MR-F201A）<br>弯嘴眼科镊（施强, JYN1030）<br>小血管剪（金钟, JA2405）<br>止血夹（灵涛, LW0080）<br>缝合钳（三友, SY11Cr）<br>带线缝合针（双箭, YY0166-2002）</p>
</li>
<li><p>注射针（带芯毛细玻璃管，SUTTER INSTRUMENT BF100-78-10）</p>
</li>
<li><p>微量注射枪头（Eppendorf Microloader货号：5242956003）</p>
</li>
<li><p>固定针（无芯毛细玻璃管，SUTTER INSTRUMENT B100-75-10）</p>
</li>
<li><p>吸卵针（国产毛细管，规格0.9-1.1×120 mm）</p>
</li>
<li><p>培养皿（FALCON 353001、353002）</p>
</li>
</ol>
<h4 id="（二）试剂"><a href="#（二）试剂" class="headerlink" title="（二）试剂"></a>（二）试剂</h4><ol>
<li>FHM小鼠胚胎培养液（MILLIPORE公司，产品目录号：MR-025-D）</li>
<li>KSOM小鼠胚胎培养液（MILLIPORE公司，产品目录号：MR-121-D）</li>
<li>矿物油（SIGMA M8410）</li>
<li>透明质酸酶（SIGMA H3506 ）</li>
<li>孕马血清PMSG（宁波三生）</li>
<li>人绒毛膜促性腺激素HCG（宁波三生）</li>
<li>细胞松弛素CB（SIGMA C6762）</li>
<li>二甲基亚砜细胞冻存液DMSO（SIGMA D2650）</li>
<li>2,2,2-Tribromoethanol（SIGMA-ALDRICH T48402）</li>
<li>2-Methyl-2-butanol（SIGMA 240486）</li>
<li>ULTRA PURE WATER（MILLIPORE TMS-006-A）</li>
<li>PMSG（见溶液配方）</li>
<li>HCG（见溶液配方）</li>
<li>2.5%阿佛丁麻醉剂（见溶液配方）</li>
<li>CB-HCZB（见溶液配方）</li>
</ol>
<h4 id="（三）供受体小鼠"><a href="#（三）供受体小鼠" class="headerlink" title="（三）供受体小鼠"></a>（三）供受体小鼠</h4><ol>
<li>供体雌鼠（C57BL&#x2F;6J、3-4周）</li>
<li>供体雄鼠（C57BL&#x2F;6J、大于8周）</li>
<li>假孕雌鼠（ICR、大于6周）</li>
<li>结扎雄鼠（ICR、大于6周）</li>
</ol>
<h3 id="仪器设备"><a href="#仪器设备" class="headerlink" title="仪器设备"></a>仪器设备</h3><ol>
<li>拉针仪（SUTTER INSTRUMENT P-97）</li>
<li>煅针仪（NARISHIGE MF-900）</li>
<li>实体镜（NIKON SMZ800N）</li>
<li>倒置显微镜（OLYMPUS IX73）</li>
<li>显微操作系统（NARISHIGE IM-9C）</li>
<li>离心机（eppendorf 5430）</li>
<li>CO2恒温培养箱（Thermo i160）</li>
<li>超洁净工作台（万里 SW-CJ-2FX）</li>
<li>热台</li>
<li>磁力搅拌器（CORNING PC-420D）</li>
</ol>
<h3 id="实验步骤"><a href="#实验步骤" class="headerlink" title="实验步骤"></a>实验步骤</h3><h4 id="（一）实验动物的准备"><a href="#（一）实验动物的准备" class="headerlink" title="（一）实验动物的准备"></a>（一）实验动物的准备</h4><ol>
<li>受精卵供应鼠：选用3到4周龄的C57BL&#x2F;6J雌鼠，或根据研究需要选择特定的品系。</li>
<li>可育雄鼠：选用8周龄以上的C57BL&#x2F;6J雄鼠，最好有过交配记录，或根据研究需要选择特定的品系。</li>
<li>结扎雄鼠（需提前一个月准备，见相关章节）。</li>
<li>假孕雌鼠：由6周龄以上ICR雌鼠与结扎雄鼠交配所成，其性周期与受精卵供应鼠保持一致。</li>
</ol>
<h4 id="（二）受精卵的制备"><a href="#（二）受精卵的制备" class="headerlink" title="（二）受精卵的制备"></a>（二）受精卵的制备</h4><p>若选用C57BL&#x2F;6J背景鼠作为实验小鼠，选取10只3-4周龄的雌性小鼠，于下午19:00-20:00按5IU&#x2F;只腹腔注射孕马血清(PMSG)，46-48小时之后注射按5IU&#x2F;只腹腔注射人绒毛膜促性腺激素(HCG)，然后与可育雄鼠合笼交配，次日上午选出有阴道栓的雌鼠取受精卵。查栓当天10:00-12:00将有阴道栓的小鼠脱颈椎处死，打开腹腔后将完整的输卵管带一小段子宫剪下，置于FHM操作液中，在解剖镜下找到输卵管的膨大部，用1 ml注射器针头撕开，带有卵丘细胞的卵团自行流出。随后将卵团移入含有lmg&#x2F; ml透明质酸酶(Hyaluronidase)的FHM操作液内1 min，溶去卵丘细胞，立即用预先拉制的吸管将卵子吸至新鲜的FHM操作液中洗涤2-3次，然后再将受精卵吸至另外新鲜FHM操作液中，如此进行4-5次，直至清洗干净。最后将受精卵吸至含多个KSOM培养液滴的35 mm培养皿中，此KSOM培养皿需提前放入二氧化碳恒温培养箱内平衡超过10 min，并用矿物油覆盖。待卵原核发育清晰饱满后即可做显微注射。</p>
<h4 id="（三）假孕雌鼠的制备（见相关章节）"><a href="#（三）假孕雌鼠的制备（见相关章节）" class="headerlink" title="（三）假孕雌鼠的制备（见相关章节）"></a>（三）假孕雌鼠的制备（见相关章节）</h4><p>在供卵鼠与可育雄鼠合笼当天，挑选处于发情期的ICR雌鼠与结扎雄鼠合笼，第二天早上检查交配情况，选择有阴栓的假孕雌鼠作为受体使用。</p>
<h4 id="（四）显微注射系统的建立"><a href="#（四）显微注射系统的建立" class="headerlink" title="（四）显微注射系统的建立"></a>（四）显微注射系统的建立</h4><ol>
<li><p>注射针的拉制<br>选用外径为1 mm，内径0.75 mm-0.8 mm带芯毛细玻璃管，装在水平拉针仪上，拧紧左右两侧的夹子，做好固定，并使其中间部位对准加热丝。设定好电流、拉力和时间。然后按开始键，等待毛细管被拉成两个所需的微注射针头。</p>
</li>
<li><p>持卵针（Holder）的拉制<br>选用外径为1 mm，内径0.75 mm无芯毛细玻璃管作为制作持卵针的材料，装在拉针仪上，拧紧左右两侧的夹子，做好固定，并使其中间部位对准加热丝。设定好电流、拉力和时间。按键拉出针坯，用细砂轮在直径为80-100 μm处将玻璃针切断，尽量使断面平齐，然后在煅针仪上处理断面，将断面烧圆。最后用煅针仪铂金丝在距针尖2-3 mm处，将针坯弯成20-30°。</p>
<p><img src="https://cdn.jsdelivr.net/gh/nikolahuang/blogimage@main/img/20221017015809_4598.jpg" alt="img"><br>注射针安装<br>左图：注射针角度过大；右图：注射针角度合适</p>
</li>
</ol>
<h4 id="（五）原核注射操作过程"><a href="#（五）原核注射操作过程" class="headerlink" title="（五）原核注射操作过程"></a>（五）原核注射操作过程</h4><ol>
<li>注射针头装液<br>将制备好的DNA溶液（浓度通常为20-40ng&#x2F;ul）12,000 rpm离心5 min，使用无菌的微量加样器从注射针头的后部加入0.5～1 μl的注射样品。</li>
<li>将注射针和持卵针安装在显微操作臂上。</li>
<li>原核注射的操作：<ol>
<li>在100 mm的培养皿中央制作两排20 μl的CB-HCZB操作液，加入5ml矿物油覆盖。并将培养皿液滴调整到显微镜视野中央，然后吸取10-20枚细胞核清晰的受精卵吹到液滴中央，在吹卵过程中尽量不要吹出气泡以免影响视野。</li>
<li>用低倍物镜对准受精卵调焦，轻轻的落下针尖，并将注射针尖推入视野中心，通过微操作系统的微调调整注射针位置，直到清晰见到针尖为止。用同样的方法调整持卵针的位置，使持卵针开口端正对注射针尖。</li>
<li>使用持卵针吸取边缘清晰、形状饱满且细胞核明显的受精卵，调整受精卵的位置，使受精卵体积较大的雄性原核正对注射针尖且两者之间的距离最小。将放大倍数调至工作倍数（目镜20×—40×），进一步调整显微镜焦距和注射针及持卵针的位置，使细胞核和注射针尖均达到最佳清晰程度。推动操纵杆，小心进针，使注射针针尖进入细胞核。</li>
<li>加大注射压，直到细胞核明显胀大为止。</li>
<li>轻而迅速往外拔针头，直到离开细胞。</li>
<li>移动显微镜载物台，找到下一颗受精卵，重复以上步骤。注射完毕，将受精卵放回KSOM培养液中，在恒温培养箱中继续培养，吸取下一批受精卵继续进行，直到注完为止。将注射过受精卵培养过夜。</li>
<li>经一夜培养，选择发育良好的二细胞期的胚胎用于移植。</li>
</ol>
</li>
</ol>
<h4 id="（六）胞浆注射"><a href="#（六）胞浆注射" class="headerlink" title="（六）胞浆注射"></a>（六）胞浆注射</h4><ol>
<li>将离心完的核酸或其他溶液上清通过配套的微量注射枪头经底部加入胞浆注射针（微量注射时，2 μl溶液就足以完成数小时的注射），将针头向下静置数分钟（也可以用胶带纸粘在实验台上），可以明显观察到针尖端的空气慢慢被液体替换，同时可能观察到微小气泡上浮。基因敲除时注射物的浓度一般是：sgRNA 12.5ng&#x2F;μl, Cas9 mRNA 25 ng&#x2F;μl；条件性基因敲除注射物的浓度通常是：sgRNA 5ng&#x2F;μl, Cas9 mRNA 10ng&#x2F;μl, Donor质粒10ng&#x2F;μl 。</li>
<li>将持卵针与充分静置后的注射针安装上显微操作仪，注射针的角度应尽可能的平（因为胞浆注射针无法像其他显微操作的注射针一样弯折角度，为了确保注射后受精卵的存活率，应尽可能降低注射针与载物台的夹角，详情见图1）。</li>
<li>在10厘米细胞培养皿盖或其他专用的注射皿上制作数个CB-HCZB液滴，其上覆盖以矿物油，用手持式移卵管转移适量的受精卵至CB-HCZB液滴中。一般情况下，可以明显的分辨出受精卵的两个原核，而未受精的卵母细胞只能观察到1个或无法观察到原核。</li>
<li>持卵针和注射针调整到视野中心，将微量注射仪程序调至FUNC 4，即持续注射模式，初始压力值可以设置高一些，约在600左右，轻轻地将注射针的针头“撞”在持卵针上或在持卵针上拨动，将针头部封闭处开放。判断是否撞开的标准是，如果将针头靠近受精卵能观察到受精卵开始旋转，则证明针头已经开放。此时第一次调整注射压力，最适值是针头靠近受精卵，受精卵能缓缓的开始旋转，但不会发生剧烈的前后位移。</li>
<li>在20倍镜下，先用一些状态差的受精卵测试压力，将注射针轻轻扎入卵胞质内，此时可根据情况第二次调整压力，合适的注射压力是根据针头的粗细与是否堵针决定的，理想的注射压力是受精卵被注射的局部区域出现明显膨胀，但受精卵本身又不能产生明显的膨胀。注射压力在10到1000之间都属正常，但针头越细越容易扎入卵胞质，但同时也越容易堵针，如果发现注射时受精卵膨胀的程度显著性降低或者完全没有膨胀，可以使用机器自带的Clean程序将注射针头以较高压力吹洗，如果多次冲洗仍然堵针，可重复4中撞针的步骤，将针头撞的更粗一些，如果针头粗到难以扎破受精卵，则需要重新装液换针。</li>
<li>测试调整完压力后，就可以开始注射，为了最大化存活率，应尽可能的减少针停留在受精卵内部的时间，以避免受精卵无法封口导致死亡。重复操作直至所有受精卵注射完毕。</li>
<li>将注射后的受精卵移回KSOM培养液中继续培养，刚注射完的受精卵会残留CB-HCZB，因此在过夜培养前需要反复洗涤几次。</li>
</ol>
<h4 id="（七）胚胎的移植"><a href="#（七）胚胎的移植" class="headerlink" title="（七）胚胎的移植"></a>（七）胚胎的移植</h4><p>将经过显微注射，发育到二细胞期的胚胎移植入0.5天的假孕雌鼠体内（具体操作见相关章节）。</p>
<p>溶液配方</p>
<ol>
<li>PMSG配置<br>将PMSG 冻干粉溶于0.9%NaCl 无菌生理盐水,浓度50IU&#x2F;ml，根据实验需要的量分装到1.5 ml EP 管中。-20 °C至少可保存一个月。</li>
<li>HCG配置<br>将冻干粉溶于0.9%NaCl 无菌生理盐水,浓度50IU&#x2F; ml，根据实验需要的量分装到1.5 ml EP 管中。-20 °C至少可保存一个月</li>
<li>2.5%阿佛丁麻醉剂配制<br>称取10g 2,2,2-Tribromoethanol后加入10 ml 2-Methyl-2-butanol，240转&#x2F;25℃避光振荡至沉淀完全消失，配制成浓储液。使用时按照2.5%浓度将浓缩液加入相应的ULTRA PURE WATER（2.5 ml浓缩液：97.5 ml超纯水），280转&#x2F;25℃避光振荡至沉淀完全消失，0.25nm过滤器过滤后即为工作液。</li>
<li>CB-HCZB 配制方法<br>CB先用2.5 ml DMSO充分溶解，配制成浓储液（每管100 μl分装）。使用时每管浓储加400 μl DMSO配成2mg&#x2F; ml的次级浓储液，再每管2.5 μl分装到Ep管中，使用时每管加入1 ml FHM即为工作液。</li>
</ol>
<h3 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h3><ol>
<li>1.Andras, N., Marina, G., Kristina, V. and Richard, B.(2003). Manipulating the Mouse Embryo. Cold Spring Harbor Laboratory Press. Cold Spring Harbor, New York.</li>
</ol>

    

    

    

  </div>
  <div class="article-info article-info-index">
    
    
	<div class="article-tag tagcloud">
		<i class="icon-price-tags icon"></i>
		<ul class="article-tag-list">
			 
        		<li class="article-tag-list-item">
        			<a href="javascript:void(0)" class="js-tag article-tag-list-link color2">BioProtocol</a>
        		</li>
      		
		</ul>
	</div>

    

    
    <p class="article-more-link">
      <a class="article-more-a" href="/2024/03/15/%E5%B0%8F%E9%BC%A0%E5%8F%97%E7%B2%BE%E5%8D%B5%E5%8E%9F%E6%A0%B8%E6%B3%A8%E5%B0%84%E5%8F%8A%E8%83%9E%E6%B5%86%E6%B3%A8%E5%B0%84/">展开全文 >></a>
    </p>
    

    
    <div class="clearfix"></div>
  </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
  
    <nav id="page-nav">
      <a class="extend prev" rel="prev" href="/">&laquo; Prev</a><a class="page-number" href="/">1</a><span class="page-number current">2</span><a class="page-number" href="/page/3/">3</a><a class="page-number" href="/page/4/">4</a><span class="space">&hellip;</span><a class="page-number" href="/page/7/">7</a><a class="extend next" rel="next" href="/page/3/">Next &raquo;</a>
    </nav>
  


          </div>
        </div>
      </div>
      <footer id="footer">
  <div class="outer">
    <div id="footer-info">
    	<div class="footer-left">
    		&copy; 2023-2024 <a href="http://nikolahuang.github.io/" target="_blank">Anderson</a>
    	</div>
      	<div class="footer-right">
			
			
      		GitHub:<a href="https://github.com/JoeyBling/hexo-theme-yilia-plus" target="_blank">hexo-theme-yilia-plus</a> by Litten
      	</div>
    </div>
  </div>
  
  
	<script src="/lib/busuanzi.pure.js"></script>
	
  
  
	
	<span id="busuanzi_container_site_pv" style="display:none">
		本站总访问量<span id="busuanzi_value_site_pv"></span>次	
	        <span class="post-meta-divider" >|</span>
	</span>
  	<span id="busuanzi_container_site_uv" style='display:none'>
  		本站访客数<span id="busuanzi_value_site_uv"></span>人
  	</span>
  
</footer>

    </div>
    <script>
	var yiliaConfig = {
		mathjax: false,
		isHome: true,
		isPost: false,
		isArchive: false,
		isTag: false,
		isCategory: false,
		open_in_new: true,
		toc_hide_index: true,
		root: "/",
		innerArchive: true,
		showTags: true
	}
</script>

<script>!function(r){function e(t){if(i[t])return i[t].exports;var n=i[t]={exports:{},id:t,loaded:!1};return r[t].call(n.exports,n,n.exports,e),n.loaded=!0,n.exports}var i={};e.m=r,e.c=i,e.p="./",e(0)}([function(t,n,r){r(208),t.exports=r(205)},function(t,n,r){var d=r(3),y=r(46),g=r(26),b=r(27),x=r(47),m="prototype",S=function(t,n,r){var e,i,o,u,c=t&S.F,f=t&S.G,a=t&S.S,s=t&S.P,l=t&S.B,h=f?d:a?d[n]||(d[n]={}):(d[n]||{})[m],v=f?y:y[n]||(y[n]={}),p=v[m]||(v[m]={});for(e in f&&(r=n),r)o=((i=!c&&h&&void 0!==h[e])?h:r)[e],u=l&&i?x(o,d):s&&"function"==typeof o?x(Function.call,o):o,h&&b(h,e,o,t&S.U),v[e]!=o&&g(v,e,u),s&&p[e]!=o&&(p[e]=o)};d.core=y,S.F=1,S.G=2,S.S=4,S.P=8,S.B=16,S.W=32,S.U=64,S.R=128,t.exports=S},function(t,n,r){var e=r(5);t.exports=function(t){if(!e(t))throw TypeError(t+" is not an object!");return t}},function(t,n){var r=t.exports="undefined"!=typeof window&&window.Math==Math?window:"undefined"!=typeof self&&self.Math==Math?self:Function("return this")();"number"==typeof __g&&(__g=r)},function(t,n){t.exports=function(t){try{return!!t()}catch(t){return!0}}},function(t,n){t.exports=function(t){return"object"==typeof t?null!==t:"function"==typeof t}},function(t,n){var r=t.exports="undefined"!=typeof window&&window.Math==Math?window:"undefined"!=typeof self&&self.Math==Math?self:Function("return this")();"number"==typeof __g&&(__g=r)},function(t,n,r){var e=r(118)("wks"),i=r(79),o=r(3).Symbol,u="function"==typeof o;(t.exports=function(t){return e[t]||(e[t]=u&&o[t]||(u?o:i)("Symbol."+t))}).store=e},function(t,n,r){var e=r(49),i=Math.min;t.exports=function(t){return 0<t?i(e(t),9007199254740991):0}},function(t,n){var r={}.hasOwnProperty;t.exports=function(t,n){return r.call(t,n)}},function(t,n,r){t.exports=!r(4)(function(){return 7!=Object.defineProperty({},"a",{get:function(){return 7}}).a})},function(t,n,r){var e=r(2),i=r(174),o=r(53),u=Object.defineProperty;n.f=r(10)?Object.defineProperty:function(t,n,r){if(e(t),n=o(n,!0),e(r),i)try{return u(t,n,r)}catch(t){}if("get"in r||"set"in r)throw TypeError("Accessors not supported!");return"value"in r&&(t[n]=r.value),t}},function(t,n,r){t.exports=!r(20)(function(){return 7!=Object.defineProperty({},"a",{get:function(){return 7}}).a})},function(t,n,r){var e=r(14),i=r(24);t.exports=r(12)?function(t,n,r){return e.f(t,n,i(1,r))}:function(t,n,r){return t[n]=r,t}},function(t,n,r){var e=r(22),i=r(60),o=r(42),u=Object.defineProperty;n.f=r(12)?Object.defineProperty:function(t,n,r){if(e(t),n=o(n,!0),e(r),i)try{return u(t,n,r)}catch(t){}if("get"in r||"set"in r)throw TypeError("Accessors not supported!");return"value"in r&&(t[n]=r.value),t}},function(t,n,r){var e=r(96),i=r(34);t.exports=function(t){return e(i(t))}},function(t,n,r){var e=r(40)("wks"),i=r(25),o=r(6).Symbol,u="function"==typeof o;(t.exports=function(t){return e[t]||(e[t]=u&&o[t]||(u?o:i)("Symbol."+t))}).store=e},function(t,n,r){var e=r(51);t.exports=function(t){return Object(e(t))}},function(t,n){t.exports=function(t){return"object"==typeof t?null!==t:"function"==typeof t}},function(t,n){var r=t.exports={version:"2.6.9"};"number"==typeof __e&&(__e=r)},function(t,n){t.exports=function(t){try{return!!t()}catch(t){return!0}}},function(t,n){t.exports=function(t){if("function"!=typeof t)throw TypeError(t+" is not a function!");return t}},function(t,n,r){var e=r(18);t.exports=function(t){if(!e(t))throw TypeError(t+" is not an object!");return t}},function(t,n){t.exports=!0},function(t,n){t.exports=function(t,n){return{enumerable:!(1&t),configurable:!(2&t),writable:!(4&t),value:n}}},function(t,n){var r=0,e=Math.random();t.exports=function(t){return"Symbol(".concat(void 0===t?"":t,")_",(++r+e).toString(36))}},function(t,n,r){var e=r(11),i=r(75);t.exports=r(10)?function(t,n,r){return e.f(t,n,i(1,r))}:function(t,n,r){return t[n]=r,t}},function(t,n,r){var o=r(3),u=r(26),c=r(30),f=r(79)("src"),e=r(219),i="toString",a=(""+e).split(i);r(46).inspectSource=function(t){return e.call(t)},(t.exports=function(t,n,r,e){var i="function"==typeof r;i&&(c(r,"name")||u(r,"name",n)),t[n]!==r&&(i&&(c(r,f)||u(r,f,t[n]?""+t[n]:a.join(String(n)))),t===o?t[n]=r:e?t[n]?t[n]=r:u(t,n,r):(delete t[n],u(t,n,r)))})(Function.prototype,i,function(){return"function"==typeof this&&this[f]||e.call(this)})},function(t,n,r){var e=r(1),i=r(4),u=r(51),c=/"/g,o=function(t,n,r,e){var i=String(u(t)),o="<"+n;return""!==r&&(o+=" "+r+'="'+String(e).replace(c,"&quot;")+'"'),o+">"+i+"</"+n+">"};t.exports=function(n,t){var r={};r[n]=t(o),e(e.P+e.F*i(function(){var t=""[n]('"');return t!==t.toLowerCase()||3<t.split('"').length}),"String",r)}},function(t,n,r){var e=r(65),i=r(35);t.exports=Object.keys||function(t){return e(t,i)}},function(t,n){var r={}.hasOwnProperty;t.exports=function(t,n){return r.call(t,n)}},function(t,n,r){var e=r(117),i=r(75),o=r(33),u=r(53),c=r(30),f=r(174),a=Object.getOwnPropertyDescriptor;n.f=r(10)?a:function(t,n){if(t=o(t),n=u(n,!0),f)try{return a(t,n)}catch(t){}if(c(t,n))return i(!e.f.call(t,n),t[n])}},function(t,n,r){var e=r(30),i=r(17),o=r(154)("IE_PROTO"),u=Object.prototype;t.exports=Object.getPrototypeOf||function(t){return t=i(t),e(t,o)?t[o]:"function"==typeof t.constructor&&t instanceof t.constructor?t.constructor.prototype:t instanceof Object?u:null}},function(t,n,r){var e=r(116),i=r(51);t.exports=function(t){return e(i(t))}},function(t,n){t.exports=function(t){if(null==t)throw TypeError("Can't call method on  "+t);return t}},function(t,n){t.exports="constructor,hasOwnProperty,isPrototypeOf,propertyIsEnumerable,toLocaleString,toString,valueOf".split(",")},function(t,n){t.exports={}},function(t,n){n.f={}.propertyIsEnumerable},function(t,n,r){var e=r(14).f,i=r(9),o=r(16)("toStringTag");t.exports=function(t,n,r){t&&!i(t=r?t:t.prototype,o)&&e(t,o,{configurable:!0,value:n})}},function(t,n,r){var e=r(40)("keys"),i=r(25);t.exports=function(t){return e[t]||(e[t]=i(t))}},function(t,n,r){var e=r(19),i=r(6),o="__core-js_shared__",u=i[o]||(i[o]={});(t.exports=function(t,n){return u[t]||(u[t]=void 0!==n?n:{})})("versions",[]).push({version:e.version,mode:r(23)?"pure":"global",copyright:"© 2019 Denis Pushkarev (zloirock.ru)"})},function(t,n){var r=Math.ceil,e=Math.floor;t.exports=function(t){return isNaN(t=+t)?0:(0<t?e:r)(t)}},function(t,n,r){var i=r(18);t.exports=function(t,n){if(!i(t))return t;var r,e;if(n&&"function"==typeof(r=t.toString)&&!i(e=r.call(t)))return e;if("function"==typeof(r=t.valueOf)&&!i(e=r.call(t)))return e;if(!n&&"function"==typeof(r=t.toString)&&!i(e=r.call(t)))return e;throw TypeError("Can't convert object to primitive value")}},function(t,n,r){var e=r(6),i=r(19),o=r(23),u=r(44),c=r(14).f;t.exports=function(t){var n=i.Symbol||(i.Symbol=o?{}:e.Symbol||{});"_"==t.charAt(0)||t in n||c(n,t,{value:u.f(t)})}},function(t,n,r){n.f=r(16)},function(t,n){var r={}.toString;t.exports=function(t){return r.call(t).slice(8,-1)}},function(t,n){var r=t.exports={version:"2.6.9"};"number"==typeof __e&&(__e=r)},function(t,n,r){var o=r(21);t.exports=function(e,i,t){if(o(e),void 0===i)return e;switch(t){case 1:return function(t){return e.call(i,t)};case 2:return function(t,n){return e.call(i,t,n)};case 3:return function(t,n,r){return e.call(i,t,n,r)}}return function(){return e.apply(i,arguments)}}},function(t,n,r){"use strict";var e=r(4);t.exports=function(t,n){return!!t&&e(function(){n?t.call(null,function(){},1):t.call(null)})}},function(t,n){var r=Math.ceil,e=Math.floor;t.exports=function(t){return isNaN(t=+t)?0:(0<t?e:r)(t)}},function(t,n,r){var x=r(47),m=r(116),S=r(17),w=r(8),e=r(138);t.exports=function(l,t){var h=1==l,v=2==l,p=3==l,d=4==l,y=6==l,g=5==l||y,b=t||e;return function(t,n,r){for(var e,i,o=S(t),u=m(o),c=x(n,r,3),f=w(u.length),a=0,s=h?b(t,f):v?b(t,0):void 0;a<f;a++)if((g||a in u)&&(i=c(e=u[a],a,o),l))if(h)s[a]=i;else if(i)switch(l){case 3:return!0;case 5:return e;case 6:return a;case 2:s.push(e)}else if(d)return!1;return y?-1:p||d?d:s}}},function(t,n){t.exports=function(t){if(null==t)throw TypeError("Can't call method on  "+t);return t}},function(t,n,r){var i=r(1),o=r(46),u=r(4);t.exports=function(t,n){var r=(o.Object||{})[t]||Object[t],e={};e[t]=n(r),i(i.S+i.F*u(function(){r(1)}),"Object",e)}},function(t,n,r){var i=r(5);t.exports=function(t,n){if(!i(t))return t;var r,e;if(n&&"function"==typeof(r=t.toString)&&!i(e=r.call(t)))return e;if("function"==typeof(r=t.valueOf)&&!i(e=r.call(t)))return e;if(!n&&"function"==typeof(r=t.toString)&&!i(e=r.call(t)))return e;throw TypeError("Can't convert object to primitive value")}},function(t,n,r){var d=r(6),y=r(19),g=r(93),b=r(13),x=r(9),m="prototype",S=function(t,n,r){var e,i,o,u=t&S.F,c=t&S.G,f=t&S.S,a=t&S.P,s=t&S.B,l=t&S.W,h=c?y:y[n]||(y[n]={}),v=h[m],p=c?d:f?d[n]:(d[n]||{})[m];for(e in c&&(r=n),r)(i=!u&&p&&void 0!==p[e])&&x(h,e)||(o=i?p[e]:r[e],h[e]=c&&"function"!=typeof p[e]?r[e]:s&&i?g(o,d):l&&p[e]==o?function(e){var t=function(t,n,r){if(this instanceof e){switch(arguments.length){case 0:return new e;case 1:return new e(t);case 2:return new e(t,n)}return new e(t,n,r)}return e.apply(this,arguments)};return t[m]=e[m],t}(o):a&&"function"==typeof o?g(Function.call,o):o,a&&((h.virtual||(h.virtual={}))[e]=o,t&S.R&&v&&!v[e]&&b(v,e,o)))};S.F=1,S.G=2,S.S=4,S.P=8,S.B=16,S.W=32,S.U=64,S.R=128,t.exports=S},function(t,n,r){var e=r(34);t.exports=function(t){return Object(e(t))}},function(t,n,r){var o=r(196),e=r(1),i=r(118)("metadata"),u=i.store||(i.store=new(r(200))),c=function(t,n,r){var e=u.get(t);if(!e){if(!r)return;u.set(t,e=new o)}var i=e.get(n);if(!i){if(!r)return;e.set(n,i=new o)}return i};t.exports={store:u,map:c,has:function(t,n,r){var e=c(n,r,!1);return void 0!==e&&e.has(t)},get:function(t,n,r){var e=c(n,r,!1);return void 0===e?void 0:e.get(t)},set:function(t,n,r,e){c(r,e,!0).set(t,n)},keys:function(t,n){var r=c(t,n,!1),e=[];return r&&r.forEach(function(t,n){e.push(n)}),e},key:function(t){return void 0===t||"symbol"==typeof t?t:String(t)},exp:function(t){e(e.S,"Reflect",t)}}},function(t,n,r){"use strict";if(r(10)){var g=r(68),b=r(3),x=r(4),m=r(1),S=r(132),e=r(159),h=r(47),w=r(70),i=r(75),_=r(26),o=r(76),u=r(49),O=r(8),E=r(194),c=r(78),f=r(53),a=r(30),M=r(81),P=r(5),v=r(17),p=r(145),j=r(72),F=r(32),A=r(73).f,d=r(161),s=r(79),l=r(7),y=r(50),I=r(120),L=r(119),N=r(162),T=r(82),k=r(125),R=r(77),C=r(137),D=r(166),G=r(11),W=r(31),U=G.f,V=W.f,B=b.RangeError,q=b.TypeError,z=b.Uint8Array,K="ArrayBuffer",H="Shared"+K,J="BYTES_PER_ELEMENT",$="prototype",Y=Array[$],X=e.ArrayBuffer,Q=e.DataView,Z=y(0),tt=y(2),nt=y(3),rt=y(4),et=y(5),it=y(6),ot=I(!0),ut=I(!1),ct=N.values,ft=N.keys,at=N.entries,st=Y.lastIndexOf,lt=Y.reduce,ht=Y.reduceRight,vt=Y.join,pt=Y.sort,dt=Y.slice,yt=Y.toString,gt=Y.toLocaleString,bt=l("iterator"),xt=l("toStringTag"),mt=s("typed_constructor"),St=s("def_constructor"),wt=S.CONSTR,_t=S.TYPED,Ot=S.VIEW,Et="Wrong length!",Mt=y(1,function(t,n){return It(L(t,t[St]),n)}),Pt=x(function(){return 1===new z(new Uint16Array([1]).buffer)[0]}),jt=!!z&&!!z[$].set&&x(function(){new z(1).set({})}),Ft=function(t,n){var r=u(t);if(r<0||r%n)throw B("Wrong offset!");return r},At=function(t){if(P(t)&&_t in t)return t;throw q(t+" is not a typed array!")},It=function(t,n){if(!(P(t)&&mt in t))throw q("It is not a typed array constructor!");return new t(n)},Lt=function(t,n){return Nt(L(t,t[St]),n)},Nt=function(t,n){for(var r=0,e=n.length,i=It(t,e);r<e;)i[r]=n[r++];return i},Tt=function(t,n,r){U(t,n,{get:function(){return this._d[r]}})},kt=function(t){var n,r,e,i,o,u,c=v(t),f=arguments.length,a=1<f?arguments[1]:void 0,s=void 0!==a,l=d(c);if(null!=l&&!p(l)){for(u=l.call(c),e=[],n=0;!(o=u.next()).done;n++)e.push(o.value);c=e}for(s&&2<f&&(a=h(a,arguments[2],2)),n=0,r=O(c.length),i=It(this,r);n<r;n++)i[n]=s?a(c[n],n):c[n];return i},Rt=function(){for(var t=0,n=arguments.length,r=It(this,n);t<n;)r[t]=arguments[t++];return r},Ct=!!z&&x(function(){gt.call(new z(1))}),Dt=function(){return gt.apply(Ct?dt.call(At(this)):At(this),arguments)},Gt={copyWithin:function(t,n){return D.call(At(this),t,n,2<arguments.length?arguments[2]:void 0)},every:function(t){return rt(At(this),t,1<arguments.length?arguments[1]:void 0)},fill:function(t){return C.apply(At(this),arguments)},filter:function(t){return Lt(this,tt(At(this),t,1<arguments.length?arguments[1]:void 0))},find:function(t){return et(At(this),t,1<arguments.length?arguments[1]:void 0)},findIndex:function(t){return it(At(this),t,1<arguments.length?arguments[1]:void 0)},forEach:function(t){Z(At(this),t,1<arguments.length?arguments[1]:void 0)},indexOf:function(t){return ut(At(this),t,1<arguments.length?arguments[1]:void 0)},includes:function(t){return ot(At(this),t,1<arguments.length?arguments[1]:void 0)},join:function(t){return vt.apply(At(this),arguments)},lastIndexOf:function(t){return st.apply(At(this),arguments)},map:function(t){return Mt(At(this),t,1<arguments.length?arguments[1]:void 0)},reduce:function(t){return lt.apply(At(this),arguments)},reduceRight:function(t){return ht.apply(At(this),arguments)},reverse:function(){for(var t,n=this,r=At(n).length,e=Math.floor(r/2),i=0;i<e;)t=n[i],n[i++]=n[--r],n[r]=t;return n},some:function(t){return nt(At(this),t,1<arguments.length?arguments[1]:void 0)},sort:function(t){return pt.call(At(this),t)},subarray:function(t,n){var r=At(this),e=r.length,i=c(t,e);return new(L(r,r[St]))(r.buffer,r.byteOffset+i*r.BYTES_PER_ELEMENT,O((void 0===n?e:c(n,e))-i))}},Wt=function(t,n){return Lt(this,dt.call(At(this),t,n))},Ut=function(t){At(this);var n=Ft(arguments[1],1),r=this.length,e=v(t),i=O(e.length),o=0;if(r<i+n)throw B(Et);for(;o<i;)this[n+o]=e[o++]},Vt={entries:function(){return at.call(At(this))},keys:function(){return ft.call(At(this))},values:function(){return ct.call(At(this))}},Bt=function(t,n){return P(t)&&t[_t]&&"symbol"!=typeof n&&n in t&&String(+n)==String(n)},qt=function(t,n){return Bt(t,n=f(n,!0))?i(2,t[n]):V(t,n)},zt=function(t,n,r){return!(Bt(t,n=f(n,!0))&&P(r)&&a(r,"value"))||a(r,"get")||a(r,"set")||r.configurable||a(r,"writable")&&!r.writable||a(r,"enumerable")&&!r.enumerable?U(t,n,r):(t[n]=r.value,t)};wt||(W.f=qt,G.f=zt),m(m.S+m.F*!wt,"Object",{getOwnPropertyDescriptor:qt,defineProperty:zt}),x(function(){yt.call({})})&&(yt=gt=function(){return vt.call(this)});var Kt=o({},Gt);o(Kt,Vt),_(Kt,bt,Vt.values),o(Kt,{slice:Wt,set:Ut,constructor:function(){},toString:yt,toLocaleString:Dt}),Tt(Kt,"buffer","b"),Tt(Kt,"byteOffset","o"),Tt(Kt,"byteLength","l"),Tt(Kt,"length","e"),U(Kt,xt,{get:function(){return this[_t]}}),t.exports=function(t,l,n,o){var h=t+((o=!!o)?"Clamped":"")+"Array",r="get"+t,u="set"+t,v=b[h],c=v||{},e=v&&F(v),i=!v||!S.ABV,f={},a=v&&v[$],p=function(t,i){U(t,i,{get:function(){return t=i,(n=this._d).v[r](t*l+n.o,Pt);var t,n},set:function(t){return n=i,r=t,e=this._d,o&&(r=(r=Math.round(r))<0?0:255<r?255:255&r),void e.v[u](n*l+e.o,r,Pt);var n,r,e},enumerable:!0})};i?(v=n(function(t,n,r,e){w(t,v,h,"_d");var i,o,u,c,f=0,a=0;if(P(n)){if(!(n instanceof X||(c=M(n))==K||c==H))return _t in n?Nt(v,n):kt.call(v,n);i=n,a=Ft(r,l);var s=n.byteLength;if(void 0===e){if(s%l)throw B(Et);if((o=s-a)<0)throw B(Et)}else if(s<(o=O(e)*l)+a)throw B(Et);u=o/l}else u=E(n),i=new X(o=u*l);for(_(t,"_d",{b:i,o:a,l:o,e:u,v:new Q(i)});f<u;)p(t,f++)}),a=v[$]=j(Kt),_(a,"constructor",v)):x(function(){v(1)})&&x(function(){new v(-1)})&&k(function(t){new v,new v(null),new v(1.5),new v(t)},!0)||(v=n(function(t,n,r,e){var i;return w(t,v,h),P(n)?n instanceof X||(i=M(n))==K||i==H?void 0!==e?new c(n,Ft(r,l),e):void 0!==r?new c(n,Ft(r,l)):new c(n):_t in n?Nt(v,n):kt.call(v,n):new c(E(n))}),Z(e!==Function.prototype?A(c).concat(A(e)):A(c),function(t){t in v||_(v,t,c[t])}),v[$]=a,g||(a.constructor=v));var s=a[bt],d=!!s&&("values"==s.name||null==s.name),y=Vt.values;_(v,mt,!0),_(a,_t,h),_(a,Ot,!0),_(a,St,v),(o?new v(1)[xt]==h:xt in a)||U(a,xt,{get:function(){return h}}),f[h]=v,m(m.G+m.W+m.F*(v!=c),f),m(m.S,h,{BYTES_PER_ELEMENT:l}),m(m.S+m.F*x(function(){c.of.call(v,1)}),h,{from:kt,of:Rt}),J in a||_(a,J,l),m(m.P,h,Gt),R(h),m(m.P+m.F*jt,h,{set:Ut}),m(m.P+m.F*!d,h,Vt),g||a.toString==yt||(a.toString=yt),m(m.P+m.F*x(function(){new v(1).slice()}),h,{slice:Wt}),m(m.P+m.F*(x(function(){return[1,2].toLocaleString()!=new v([1,2]).toLocaleString()})||!x(function(){a.toLocaleString.call([1,2])})),h,{toLocaleString:Dt}),T[h]=d?s:y,g||d||_(a,bt,y)}}else t.exports=function(){}},function(t,n){var r={}.toString;t.exports=function(t){return r.call(t).slice(8,-1)}},function(t,n,r){var e=r(18),i=r(6).document,o=e(i)&&e(i.createElement);t.exports=function(t){return o?i.createElement(t):{}}},function(t,n,r){t.exports=!r(12)&&!r(20)(function(){return 7!=Object.defineProperty(r(59)("div"),"a",{get:function(){return 7}}).a})},function(t,n,r){"use strict";var x=r(23),m=r(54),S=r(66),w=r(13),_=r(36),O=r(98),E=r(38),M=r(104),P=r(16)("iterator"),j=!([].keys&&"next"in[].keys()),F="values",A=function(){return this};t.exports=function(t,n,r,e,i,o,u){O(r,n,e);var c,f,a,s=function(t){if(!j&&t in p)return p[t];switch(t){case"keys":case F:return function(){return new r(this,t)}}return function(){return new r(this,t)}},l=n+" Iterator",h=i==F,v=!1,p=t.prototype,d=p[P]||p["@@iterator"]||i&&p[i],y=d||s(i),g=i?h?s("entries"):y:void 0,b="Array"==n&&p.entries||d;if(b&&((a=M(b.call(new t)))!==Object.prototype&&a.next&&(E(a,l,!0),x||"function"==typeof a[P]||w(a,P,A))),h&&d&&d.name!==F&&(v=!0,y=function(){return d.call(this)}),x&&!u||!j&&!v&&p[P]||w(p,P,y),_[n]=y,_[l]=A,i)if(c={values:h?y:s(F),keys:o?y:s("keys"),entries:g},u)for(f in c)f in p||S(p,f,c[f]);else m(m.P+m.F*(j||v),n,c);return c}},function(t,n,e){var i=e(22),o=e(101),u=e(35),c=e(39)("IE_PROTO"),f=function(){},a="prototype",s=function(){var t,n=e(59)("iframe"),r=u.length;for(n.style.display="none",e(95).appendChild(n),n.src="javascript:",(t=n.contentWindow.document).open(),t.write("<script>document.F=Object<\/script>"),t.close(),s=t.F;r--;)delete s[a][u[r]];return s()};t.exports=Object.create||function(t,n){var r;return null!==t?(f[a]=i(t),r=new f,f[a]=null,r[c]=t):r=s(),void 0===n?r:o(r,n)}},function(t,n,r){var e=r(65),i=r(35).concat("length","prototype");n.f=Object.getOwnPropertyNames||function(t){return e(t,i)}},function(t,n){n.f=Object.getOwnPropertySymbols},function(t,n,r){var u=r(9),c=r(15),f=r(92)(!1),a=r(39)("IE_PROTO");t.exports=function(t,n){var r,e=c(t),i=0,o=[];for(r in e)r!=a&&u(e,r)&&o.push(r);for(;n.length>i;)u(e,r=n[i++])&&(~f(o,r)||o.push(r));return o}},function(t,n,r){t.exports=r(13)},function(t,n,r){var e=r(7)("unscopables"),i=Array.prototype;null==i[e]&&r(26)(i,e,{}),t.exports=function(t){i[e][t]=!0}},function(t,n){t.exports=!1},function(t,n,r){var e=r(79)("meta"),i=r(5),o=r(30),u=r(11).f,c=0,f=Object.isExtensible||function(){return!0},a=!r(4)(function(){return f(Object.preventExtensions({}))}),s=function(t){u(t,e,{value:{i:"O"+ ++c,w:{}}})},l=t.exports={KEY:e,NEED:!1,fastKey:function(t,n){if(!i(t))return"symbol"==typeof t?t:("string"==typeof t?"S":"P")+t;if(!o(t,e)){if(!f(t))return"F";if(!n)return"E";s(t)}return t[e].i},getWeak:function(t,n){if(!o(t,e)){if(!f(t))return!0;if(!n)return!1;s(t)}return t[e].w},onFreeze:function(t){return a&&l.NEED&&f(t)&&!o(t,e)&&s(t),t}}},function(t,n){t.exports=function(t,n,r,e){if(!(t instanceof n)||void 0!==e&&e in t)throw TypeError(r+": incorrect invocation!");return t}},function(t,n,r){var h=r(47),v=r(177),p=r(145),d=r(2),y=r(8),g=r(161),b={},x={};(n=t.exports=function(t,n,r,e,i){var o,u,c,f,a=i?function(){return t}:g(t),s=h(r,e,n?2:1),l=0;if("function"!=typeof a)throw TypeError(t+" is not iterable!");if(p(a)){for(o=y(t.length);l<o;l++)if((f=n?s(d(u=t[l])[0],u[1]):s(t[l]))===b||f===x)return f}else for(c=a.call(t);!(u=c.next()).done;)if((f=v(c,s,u.value,n))===b||f===x)return f}).BREAK=b,n.RETURN=x},function(t,n,e){var i=e(2),o=e(183),u=e(141),c=e(154)("IE_PROTO"),f=function(){},a="prototype",s=function(){var t,n=e(140)("iframe"),r=u.length;for(n.style.display="none",e(143).appendChild(n),n.src="javascript:",(t=n.contentWindow.document).open(),t.write("<script>document.F=Object<\/script>"),t.close(),s=t.F;r--;)delete s[a][u[r]];return s()};t.exports=Object.create||function(t,n){var r;return null!==t?(f[a]=i(t),r=new f,f[a]=null,r[c]=t):r=s(),void 0===n?r:o(r,n)}},function(t,n,r){var e=r(185),i=r(141).concat("length","prototype");n.f=Object.getOwnPropertyNames||function(t){return e(t,i)}},function(t,n,r){var e=r(185),i=r(141);t.exports=Object.keys||function(t){return e(t,i)}},function(t,n){t.exports=function(t,n){return{enumerable:!(1&t),configurable:!(2&t),writable:!(4&t),value:n}}},function(t,n,r){var i=r(27);t.exports=function(t,n,r){for(var e in n)i(t,e,n[e],r);return t}},function(t,n,r){"use strict";var e=r(3),i=r(11),o=r(10),u=r(7)("species");t.exports=function(t){var n=e[t];o&&n&&!n[u]&&i.f(n,u,{configurable:!0,get:function(){return this}})}},function(t,n,r){var e=r(49),i=Math.max,o=Math.min;t.exports=function(t,n){return(t=e(t))<0?i(t+n,0):o(t,n)}},function(t,n){var r=0,e=Math.random();t.exports=function(t){return"Symbol(".concat(void 0===t?"":t,")_",(++r+e).toString(36))}},function(t,n,r){var e=r(5);t.exports=function(t,n){if(!e(t)||t._t!==n)throw TypeError("Incompatible receiver, "+n+" required!");return t}},function(t,n,r){var i=r(45),o=r(7)("toStringTag"),u="Arguments"==i(function(){return arguments}());t.exports=function(t){var n,r,e;return void 0===t?"Undefined":null===t?"Null":"string"==typeof(r=function(t,n){try{return t[n]}catch(t){}}(n=Object(t),o))?r:u?i(n):"Object"==(e=i(n))&&"function"==typeof n.callee?"Arguments":e}},function(t,n){t.exports={}},function(t,n,r){var e=r(11).f,i=r(30),o=r(7)("toStringTag");t.exports=function(t,n,r){t&&!i(t=r?t:t.prototype,o)&&e(t,o,{configurable:!0,value:n})}},function(t,n,r){var u=r(1),e=r(51),c=r(4),f=r(157),i="["+f+"]",o=RegExp("^"+i+i+"*"),a=RegExp(i+i+"*$"),s=function(t,n,r){var e={},i=c(function(){return!!f[t]()||"​"!="​"[t]()}),o=e[t]=i?n(l):f[t];r&&(e[r]=o),u(u.P+u.F*i,"String",e)},l=s.trim=function(t,n){return t=String(e(t)),1&n&&(t=t.replace(o,"")),2&n&&(t=t.replace(a,"")),t};t.exports=s},function(t,n,r){t.exports={default:r(88),__esModule:!0}},function(t,n,r){t.exports={default:r(89),__esModule:!0}},function(t,n,r){"use strict";function e(t){return t&&t.__esModule?t:{default:t}}n.__esModule=!0;var i=e(r(86)),o=e(r(85)),u="function"==typeof o.default&&"symbol"==typeof i.default?function(t){return typeof t}:function(t){return t&&"function"==typeof o.default&&t.constructor===o.default&&t!==o.default.prototype?"symbol":typeof t};n.default="function"==typeof o.default&&"symbol"===u(i.default)?function(t){return void 0===t?"undefined":u(t)}:function(t){return t&&"function"==typeof o.default&&t.constructor===o.default&&t!==o.default.prototype?"symbol":void 0===t?"undefined":u(t)}},function(t,n,r){r(111),r(109),r(112),r(113),t.exports=r(19).Symbol},function(t,n,r){r(110),r(114),t.exports=r(44).f("iterator")},function(t,n){t.exports=function(t){if("function"!=typeof t)throw TypeError(t+" is not a function!");return t}},function(t,n){t.exports=function(){}},function(t,n,r){var f=r(15),a=r(107),s=r(106);t.exports=function(c){return function(t,n,r){var e,i=f(t),o=a(i.length),u=s(r,o);if(c&&n!=n){for(;u<o;)if((e=i[u++])!=e)return!0}else for(;u<o;u++)if((c||u in i)&&i[u]===n)return c||u||0;return!c&&-1}}},function(t,n,r){var o=r(90);t.exports=function(e,i,t){if(o(e),void 0===i)return e;switch(t){case 1:return function(t){return e.call(i,t)};case 2:return function(t,n){return e.call(i,t,n)};case 3:return function(t,n,r){return e.call(i,t,n,r)}}return function(){return e.apply(i,arguments)}}},function(t,n,r){var c=r(29),f=r(64),a=r(37);t.exports=function(t){var n=c(t),r=f.f;if(r)for(var e,i=r(t),o=a.f,u=0;i.length>u;)o.call(t,e=i[u++])&&n.push(e);return n}},function(t,n,r){var e=r(6).document;t.exports=e&&e.documentElement},function(t,n,r){var e=r(58);t.exports=Object("z").propertyIsEnumerable(0)?Object:function(t){return"String"==e(t)?t.split(""):Object(t)}},function(t,n,r){var e=r(58);t.exports=Array.isArray||function(t){return"Array"==e(t)}},function(t,n,r){"use strict";var e=r(62),i=r(24),o=r(38),u={};r(13)(u,r(16)("iterator"),function(){return this}),t.exports=function(t,n,r){t.prototype=e(u,{next:i(1,r)}),o(t,n+" Iterator")}},function(t,n){t.exports=function(t,n){return{value:n,done:!!t}}},function(t,n,r){var e=r(25)("meta"),i=r(18),o=r(9),u=r(14).f,c=0,f=Object.isExtensible||function(){return!0},a=!r(20)(function(){return f(Object.preventExtensions({}))}),s=function(t){u(t,e,{value:{i:"O"+ ++c,w:{}}})},l=t.exports={KEY:e,NEED:!1,fastKey:function(t,n){if(!i(t))return"symbol"==typeof t?t:("string"==typeof t?"S":"P")+t;if(!o(t,e)){if(!f(t))return"F";if(!n)return"E";s(t)}return t[e].i},getWeak:function(t,n){if(!o(t,e)){if(!f(t))return!0;if(!n)return!1;s(t)}return t[e].w},onFreeze:function(t){return a&&l.NEED&&f(t)&&!o(t,e)&&s(t),t}}},function(t,n,r){var u=r(14),c=r(22),f=r(29);t.exports=r(12)?Object.defineProperties:function(t,n){c(t);for(var r,e=f(n),i=e.length,o=0;o<i;)u.f(t,r=e[o++],n[r]);return t}},function(t,n,r){var e=r(37),i=r(24),o=r(15),u=r(42),c=r(9),f=r(60),a=Object.getOwnPropertyDescriptor;n.f=r(12)?a:function(t,n){if(t=o(t),n=u(n,!0),f)try{return a(t,n)}catch(t){}if(c(t,n))return i(!e.f.call(t,n),t[n])}},function(t,n,r){var e=r(15),i=r(63).f,o={}.toString,u="object"==typeof window&&window&&Object.getOwnPropertyNames?Object.getOwnPropertyNames(window):[];t.exports.f=function(t){return u&&"[object Window]"==o.call(t)?function(t){try{return i(t)}catch(t){return u.slice()}}(t):i(e(t))}},function(t,n,r){var e=r(9),i=r(55),o=r(39)("IE_PROTO"),u=Object.prototype;t.exports=Object.getPrototypeOf||function(t){return t=i(t),e(t,o)?t[o]:"function"==typeof t.constructor&&t instanceof t.constructor?t.constructor.prototype:t instanceof Object?u:null}},function(t,n,r){var f=r(41),a=r(34);t.exports=function(c){return function(t,n){var r,e,i=String(a(t)),o=f(n),u=i.length;return o<0||u<=o?c?"":void 0:(r=i.charCodeAt(o))<55296||56319<r||o+1===u||(e=i.charCodeAt(o+1))<56320||57343<e?c?i.charAt(o):r:c?i.slice(o,o+2):e-56320+(r-55296<<10)+65536}}},function(t,n,r){var e=r(41),i=Math.max,o=Math.min;t.exports=function(t,n){return(t=e(t))<0?i(t+n,0):o(t,n)}},function(t,n,r){var e=r(41),i=Math.min;t.exports=function(t){return 0<t?i(e(t),9007199254740991):0}},function(t,n,r){"use strict";var e=r(91),i=r(99),o=r(36),u=r(15);t.exports=r(61)(Array,"Array",function(t,n){this._t=u(t),this._i=0,this._k=n},function(){var t=this._t,n=this._k,r=this._i++;return!t||r>=t.length?(this._t=void 0,i(1)):i(0,"keys"==n?r:"values"==n?t[r]:[r,t[r]])},"values"),o.Arguments=o.Array,e("keys"),e("values"),e("entries")},function(t,n){},function(t,n,r){"use strict";var e=r(105)(!0);r(61)(String,"String",function(t){this._t=String(t),this._i=0},function(){var t,n=this._t,r=this._i;return r>=n.length?{value:void 0,done:!0}:(t=e(n,r),this._i+=t.length,{value:t,done:!1})})},function(t,n,r){"use strict";var e=r(6),u=r(9),i=r(12),o=r(54),c=r(66),f=r(100).KEY,a=r(20),s=r(40),l=r(38),h=r(25),v=r(16),p=r(44),d=r(43),y=r(94),g=r(97),b=r(22),x=r(18),m=r(55),S=r(15),w=r(42),_=r(24),O=r(62),E=r(103),M=r(102),P=r(64),j=r(14),F=r(29),A=M.f,I=j.f,L=E.f,N=e.Symbol,T=e.JSON,k=T&&T.stringify,R="prototype",C=v("_hidden"),D=v("toPrimitive"),G={}.propertyIsEnumerable,W=s("symbol-registry"),U=s("symbols"),V=s("op-symbols"),B=Object[R],q="function"==typeof N&&!!P.f,z=e.QObject,K=!z||!z[R]||!z[R].findChild,H=i&&a(function(){return 7!=O(I({},"a",{get:function(){return I(this,"a",{value:7}).a}})).a})?function(t,n,r){var e=A(B,n);e&&delete B[n],I(t,n,r),e&&t!==B&&I(B,n,e)}:I,J=function(t){var n=U[t]=O(N[R]);return n._k=t,n},$=q&&"symbol"==typeof N.iterator?function(t){return"symbol"==typeof t}:function(t){return t instanceof N},Y=function(t,n,r){return t===B&&Y(V,n,r),b(t),n=w(n,!0),b(r),u(U,n)?(r.enumerable?(u(t,C)&&t[C][n]&&(t[C][n]=!1),r=O(r,{enumerable:_(0,!1)})):(u(t,C)||I(t,C,_(1,{})),t[C][n]=!0),H(t,n,r)):I(t,n,r)},X=function(t,n){b(t);for(var r,e=y(n=S(n)),i=0,o=e.length;i<o;)Y(t,r=e[i++],n[r]);return t},Q=function(t){var n=G.call(this,t=w(t,!0));return!(this===B&&u(U,t)&&!u(V,t))&&(!(n||!u(this,t)||!u(U,t)||u(this,C)&&this[C][t])||n)},Z=function(t,n){if(t=S(t),n=w(n,!0),t!==B||!u(U,n)||u(V,n)){var r=A(t,n);return!r||!u(U,n)||u(t,C)&&t[C][n]||(r.enumerable=!0),r}},tt=function(t){for(var n,r=L(S(t)),e=[],i=0;r.length>i;)u(U,n=r[i++])||n==C||n==f||e.push(n);return e},nt=function(t){for(var n,r=t===B,e=L(r?V:S(t)),i=[],o=0;e.length>o;)!u(U,n=e[o++])||r&&!u(B,n)||i.push(U[n]);return i};q||(c((N=function(){if(this instanceof N)throw TypeError("Symbol is not a constructor!");var n=h(0<arguments.length?arguments[0]:void 0),r=function(t){this===B&&r.call(V,t),u(this,C)&&u(this[C],n)&&(this[C][n]=!1),H(this,n,_(1,t))};return i&&K&&H(B,n,{configurable:!0,set:r}),J(n)})[R],"toString",function(){return this._k}),M.f=Z,j.f=Y,r(63).f=E.f=tt,r(37).f=Q,P.f=nt,i&&!r(23)&&c(B,"propertyIsEnumerable",Q,!0),p.f=function(t){return J(v(t))}),o(o.G+o.W+o.F*!q,{Symbol:N});for(var rt="hasInstance,isConcatSpreadable,iterator,match,replace,search,species,split,toPrimitive,toStringTag,unscopables".split(","),et=0;rt.length>et;)v(rt[et++]);for(var it=F(v.store),ot=0;it.length>ot;)d(it[ot++]);o(o.S+o.F*!q,"Symbol",{for:function(t){return u(W,t+="")?W[t]:W[t]=N(t)},keyFor:function(t){if(!$(t))throw TypeError(t+" is not a symbol!");for(var n in W)if(W[n]===t)return n},useSetter:function(){K=!0},useSimple:function(){K=!1}}),o(o.S+o.F*!q,"Object",{create:function(t,n){return void 0===n?O(t):X(O(t),n)},defineProperty:Y,defineProperties:X,getOwnPropertyDescriptor:Z,getOwnPropertyNames:tt,getOwnPropertySymbols:nt});var ut=a(function(){P.f(1)});o(o.S+o.F*ut,"Object",{getOwnPropertySymbols:function(t){return P.f(m(t))}}),T&&o(o.S+o.F*(!q||a(function(){var t=N();return"[null]"!=k([t])||"{}"!=k({a:t})||"{}"!=k(Object(t))})),"JSON",{stringify:function(t){for(var n,r,e=[t],i=1;arguments.length>i;)e.push(arguments[i++]);if(r=n=e[1],(x(n)||void 0!==t)&&!$(t))return g(n)||(n=function(t,n){if("function"==typeof r&&(n=r.call(this,t,n)),!$(n))return n}),e[1]=n,k.apply(T,e)}}),N[R][D]||r(13)(N[R],D,N[R].valueOf),l(N,"Symbol"),l(Math,"Math",!0),l(e.JSON,"JSON",!0)},function(t,n,r){r(43)("asyncIterator")},function(t,n,r){r(43)("observable")},function(t,n,r){r(108);for(var e=r(6),i=r(13),o=r(36),u=r(16)("toStringTag"),c="CSSRuleList,CSSStyleDeclaration,CSSValueList,ClientRectList,DOMRectList,DOMStringList,DOMTokenList,DataTransferItemList,FileList,HTMLAllCollection,HTMLCollection,HTMLFormElement,HTMLSelectElement,MediaList,MimeTypeArray,NamedNodeMap,NodeList,PaintRequestList,Plugin,PluginArray,SVGLengthList,SVGNumberList,SVGPathSegList,SVGPointList,SVGStringList,SVGTransformList,SourceBufferList,StyleSheetList,TextTrackCueList,TextTrackList,TouchList".split(","),f=0;f<c.length;f++){var a=c[f],s=e[a],l=s&&s.prototype;l&&!l[u]&&i(l,u,a),o[a]=o.Array}},function(t,n,r){"use strict";var e=r(2);t.exports=function(){var t=e(this),n="";return t.global&&(n+="g"),t.ignoreCase&&(n+="i"),t.multiline&&(n+="m"),t.unicode&&(n+="u"),t.sticky&&(n+="y"),n}},function(t,n,r){var e=r(45);t.exports=Object("z").propertyIsEnumerable(0)?Object:function(t){return"String"==e(t)?t.split(""):Object(t)}},function(t,n){n.f={}.propertyIsEnumerable},function(t,n,r){var e=r(46),i=r(3),o="__core-js_shared__",u=i[o]||(i[o]={});(t.exports=function(t,n){return u[t]||(u[t]=void 0!==n?n:{})})("versions",[]).push({version:e.version,mode:r(68)?"pure":"global",copyright:"© 2019 Denis Pushkarev (zloirock.ru)"})},function(t,n,r){var i=r(2),o=r(21),u=r(7)("species");t.exports=function(t,n){var r,e=i(t).constructor;return void 0===e||null==(r=i(e)[u])?n:o(r)}},function(t,n,r){var f=r(33),a=r(8),s=r(78);t.exports=function(c){return function(t,n,r){var e,i=f(t),o=a(i.length),u=s(r,o);if(c&&n!=n){for(;u<o;)if((e=i[u++])!=e)return!0}else for(;u<o;u++)if((c||u in i)&&i[u]===n)return c||u||0;return!c&&-1}}},function(t,n,r){"use strict";var g=r(3),b=r(1),x=r(27),m=r(76),S=r(69),w=r(71),_=r(70),O=r(5),E=r(4),M=r(125),P=r(83),j=r(144);t.exports=function(e,t,n,r,i,o){var u=g[e],c=u,f=i?"set":"add",a=c&&c.prototype,s={},l=function(t){var r=a[t];x(a,t,"delete"==t?function(t){return!(o&&!O(t))&&r.call(this,0===t?0:t)}:"has"==t?function(t){return!(o&&!O(t))&&r.call(this,0===t?0:t)}:"get"==t?function(t){return o&&!O(t)?void 0:r.call(this,0===t?0:t)}:"add"==t?function(t){return r.call(this,0===t?0:t),this}:function(t,n){return r.call(this,0===t?0:t,n),this})};if("function"==typeof c&&(o||a.forEach&&!E(function(){(new c).entries().next()}))){var h=new c,v=h[f](o?{}:-0,1)!=h,p=E(function(){h.has(1)}),d=M(function(t){new c(t)}),y=!o&&E(function(){for(var t=new c,n=5;n--;)t[f](n,n);return!t.has(-0)});d||(((c=t(function(t,n){_(t,c,e);var r=j(new u,t,c);return null!=n&&w(n,i,r[f],r),r})).prototype=a).constructor=c),(p||y)&&(l("delete"),l("has"),i&&l("get")),(y||v)&&l(f),o&&a.clear&&delete a.clear}else c=r.getConstructor(t,e,i,f),m(c.prototype,n),S.NEED=!0;return P(c,e),s[e]=c,b(b.G+b.W+b.F*(c!=u),s),o||r.setStrong(c,e,i),c}},function(t,n,r){"use strict";r(197);var s=r(27),l=r(26),h=r(4),v=r(51),p=r(7),d=r(152),y=p("species"),g=!h(function(){var t=/./;return t.exec=function(){var t=[];return t.groups={a:"7"},t},"7"!=="".replace(t,"$<a>")}),b=function(){var t=/(?:)/,n=t.exec;t.exec=function(){return n.apply(this,arguments)};var r="ab".split(t);return 2===r.length&&"a"===r[0]&&"b"===r[1]}();t.exports=function(r,t,n){var e=p(r),o=!h(function(){var t={};return t[e]=function(){return 7},7!=""[r](t)}),i=o?!h(function(){var t=!1,n=/a/;return n.exec=function(){return t=!0,null},"split"===r&&(n.constructor={},n.constructor[y]=function(){return n}),n[e](""),!t}):void 0;if(!o||!i||"replace"===r&&!g||"split"===r&&!b){var u=/./[e],c=n(v,e,""[r],function(t,n,r,e,i){return n.exec===d?o&&!i?{done:!0,value:u.call(n,r,e)}:{done:!0,value:t.call(r,n,e)}:{done:!1}}),f=c[0],a=c[1];s(String.prototype,r,f),l(RegExp.prototype,e,2==t?function(t,n){return a.call(t,this,n)}:function(t){return a.call(t,this)})}}},function(t,n,r){var e=r(45);t.exports=Array.isArray||function(t){return"Array"==e(t)}},function(t,n,r){var e=r(5),i=r(45),o=r(7)("match");t.exports=function(t){var n;return e(t)&&(void 0!==(n=t[o])?!!n:"RegExp"==i(t))}},function(t,n,r){var o=r(7)("iterator"),u=!1;try{var e=[7][o]();e.return=function(){u=!0},Array.from(e,function(){throw 2})}catch(t){}t.exports=function(t,n){if(!n&&!u)return!1;var r=!1;try{var e=[7],i=e[o]();i.next=function(){return{done:r=!0}},e[o]=function(){return i},t(e)}catch(t){}return r}},function(t,n,r){"use strict";t.exports=r(68)||!r(4)(function(){var t=Math.random();__defineSetter__.call(null,t,function(){}),delete r(3)[t]})},function(t,n){n.f=Object.getOwnPropertySymbols},function(t,n,r){"use strict";var i=r(81),o=RegExp.prototype.exec;t.exports=function(t,n){var r=t.exec;if("function"==typeof r){var e=r.call(t,n);if("object"!=typeof e)throw new TypeError("RegExp exec method returned something other than an Object or null");return e}if("RegExp"!==i(t))throw new TypeError("RegExp#exec called on incompatible receiver");return o.call(t,n)}},function(t,n,r){"use strict";var e=r(1),u=r(21),c=r(47),f=r(71);t.exports=function(t){e(e.S,t,{from:function(t){var n,r,e,i,o=arguments[1];return u(this),(n=void 0!==o)&&u(o),null==t?new this:(r=[],n?(e=0,i=c(o,arguments[2],2),f(t,!1,function(t){r.push(i(t,e++))})):f(t,!1,r.push,r),new this(r))}})}},function(t,n,r){"use strict";var e=r(1);t.exports=function(t){e(e.S,t,{of:function(){for(var t=arguments.length,n=new Array(t);t--;)n[t]=arguments[t];return new this(n)}})}},function(t,n,r){var f=r(49),a=r(51);t.exports=function(c){return function(t,n){var r,e,i=String(a(t)),o=f(n),u=i.length;return o<0||u<=o?c?"":void 0:(r=i.charCodeAt(o))<55296||56319<r||o+1===u||(e=i.charCodeAt(o+1))<56320||57343<e?c?i.charAt(o):r:c?i.slice(o,o+2):e-56320+(r-55296<<10)+65536}}},function(t,n,r){for(var e,i=r(3),o=r(26),u=r(79),c=u("typed_array"),f=u("view"),a=!(!i.ArrayBuffer||!i.DataView),s=a,l=0,h="Int8Array,Uint8Array,Uint8ClampedArray,Int16Array,Uint16Array,Int32Array,Uint32Array,Float32Array,Float64Array".split(",");l<9;)(e=i[h[l++]])?(o(e.prototype,c,!0),o(e.prototype,f,!0)):s=!1;t.exports={ABV:a,CONSTR:s,TYPED:c,VIEW:f}},function(t,n,r){var e=r(3).navigator;t.exports=e&&e.userAgent||""},function(t,n){"use strict";var r,e={versions:(r=window.navigator.userAgent,{trident:-1<r.indexOf("Trident"),presto:-1<r.indexOf("Presto"),webKit:-1<r.indexOf("AppleWebKit"),gecko:-1<r.indexOf("Gecko")&&-1==r.indexOf("KHTML"),mobile:!!r.match(/AppleWebKit.*Mobile.*/),ios:!!r.match(/\(i[^;]+;( U;)? CPU.+Mac OS X/),android:-1<r.indexOf("Android")||-1<r.indexOf("Linux"),iPhone:-1<r.indexOf("iPhone")||-1<r.indexOf("Mac"),iPad:-1<r.indexOf("iPad"),webApp:-1==r.indexOf("Safari"),weixin:-1==r.indexOf("MicroMessenger")})};t.exports=e},function(t,n,r){"use strict";var e,i=r(87),l=(e=i)&&e.__esModule?e:{default:e},h=function(){function n(t,n,r){return n||r?String.fromCharCode(n||r):o[t]||t}function r(t){return s[t]}var e=/&quot;|&lt;|&gt;|&amp;|&nbsp;|&apos;|&#(\d+);|&#(\d+)/g,i=/['<> "&]/g,o={"&quot;":'"',"&lt;":"<","&gt;":">","&amp;":"&","&nbsp;":" "},u=/\u00a0/g,c=/<br\s*\/?>/gi,f=/\r?\n/g,a=/\s/g,s={};for(var t in o)s[o[t]]=t;return o["&apos;"]="'",s["'"]="&#39;",{encode:function(t){return t?(""+t).replace(i,r).replace(f,"<br/>").replace(a,"&nbsp;"):""},decode:function(t){return t?(""+t).replace(c,"\n").replace(e,n).replace(u," "):""},encodeBase16:function(t){if(!t)return t;for(var n=[],r=0,e=(t+="").length;r<e;r++)n.push(t.charCodeAt(r).toString(16).toUpperCase());return n.join("")},encodeBase16forJSON:function(t){if(!t)return t;for(var n=[],r=0,e=(t=t.replace(/[\u4E00-\u9FBF]/gi,function(t){return escape(t).replace("%u","\\u")})).length;r<e;r++)n.push(t.charCodeAt(r).toString(16).toUpperCase());return n.join("")},decodeBase16:function(t){if(!t)return t;for(var n=[],r=0,e=(t+="").length;r<e;r+=2)n.push(String.fromCharCode("0x"+t.slice(r,r+2)));return n.join("")},encodeObject:function(t){if(t instanceof Array)for(var n=0,r=t.length;n<r;n++)t[n]=h.encodeObject(t[n]);else if("object"==(void 0===t?"undefined":(0,l.default)(t)))for(var e in t)t[e]=h.encodeObject(t[e]);else if("string"==typeof t)return h.encode(t);return t},loadScript:function(t){var n=document.createElement("script");document.getElementsByTagName("body")[0].appendChild(n),n.setAttribute("src",t)},addLoadEvent:function(t){var n=window.onload;"function"!=typeof window.onload?window.onload=t:window.onload=function(){n(),t()}}}}();t.exports=h},function(t,n,r){"use strict";var e=r(131)(!0);t.exports=function(t,n,r){return n+(r?e(t,n).length:1)}},function(t,n,r){"use strict";var c=r(17),f=r(78),a=r(8);t.exports=function(t){for(var n=c(this),r=a(n.length),e=arguments.length,i=f(1<e?arguments[1]:void 0,r),o=2<e?arguments[2]:void 0,u=void 0===o?r:f(o,r);i<u;)n[i++]=t;return n}},function(t,n,r){var e=r(215);t.exports=function(t,n){return new(e(t))(n)}},function(t,n,r){"use strict";var e=r(11),i=r(75);t.exports=function(t,n,r){n in t?e.f(t,n,i(0,r)):t[n]=r}},function(t,n,r){var e=r(5),i=r(3).document,o=e(i)&&e(i.createElement);t.exports=function(t){return o?i.createElement(t):{}}},function(t,n){t.exports="constructor,hasOwnProperty,isPrototypeOf,propertyIsEnumerable,toLocaleString,toString,valueOf".split(",")},function(t,n,r){var e=r(7)("match");t.exports=function(n){var r=/./;try{"/./"[n](r)}catch(t){try{return r[e]=!1,!"/./"[n](r)}catch(n){}}return!0}},function(t,n,r){var e=r(3).document;t.exports=e&&e.documentElement},function(t,n,r){var o=r(5),u=r(153).set;t.exports=function(t,n,r){var e,i=n.constructor;return i!==r&&"function"==typeof i&&(e=i.prototype)!==r.prototype&&o(e)&&u&&u(t,e),t}},function(t,n,r){var e=r(82),i=r(7)("iterator"),o=Array.prototype;t.exports=function(t){return void 0!==t&&(e.Array===t||o[i]===t)}},function(t,n,r){"use strict";var e=r(72),i=r(75),o=r(83),u={};r(26)(u,r(7)("iterator"),function(){return this}),t.exports=function(t,n,r){t.prototype=e(u,{next:i(1,r)}),o(t,n+" Iterator")}},function(t,n,r){"use strict";var x=r(68),m=r(1),S=r(27),w=r(26),_=r(82),O=r(146),E=r(83),M=r(32),P=r(7)("iterator"),j=!([].keys&&"next"in[].keys()),F="values",A=function(){return this};t.exports=function(t,n,r,e,i,o,u){O(r,n,e);var c,f,a,s=function(t){if(!j&&t in p)return p[t];switch(t){case"keys":case F:return function(){return new r(this,t)}}return function(){return new r(this,t)}},l=n+" Iterator",h=i==F,v=!1,p=t.prototype,d=p[P]||p["@@iterator"]||i&&p[i],y=d||s(i),g=i?h?s("entries"):y:void 0,b="Array"==n&&p.entries||d;if(b&&((a=M(b.call(new t)))!==Object.prototype&&a.next&&(E(a,l,!0),x||"function"==typeof a[P]||w(a,P,A))),h&&d&&d.name!==F&&(v=!0,y=function(){return d.call(this)}),x&&!u||!j&&!v&&p[P]||w(p,P,y),_[n]=y,_[l]=A,i)if(c={values:h?y:s(F),keys:o?y:s("keys"),entries:g},u)for(f in c)f in p||S(p,f,c[f]);else m(m.P+m.F*(j||v),n,c);return c}},function(t,n){var r=Math.expm1;t.exports=!r||22025.465794806718<r(10)||r(10)<22025.465794806718||-2e-17!=r(-2e-17)?function(t){return 0==(t=+t)?t:-1e-6<t&&t<1e-6?t+t*t/2:Math.exp(t)-1}:r},function(t,n){t.exports=Math.sign||function(t){return 0==(t=+t)||t!=t?t:t<0?-1:1}},function(t,n,r){var c=r(3),f=r(158).set,a=c.MutationObserver||c.WebKitMutationObserver,s=c.process,l=c.Promise,h="process"==r(45)(s);t.exports=function(){var r,e,i,t=function(){var t,n;for(h&&(t=s.domain)&&t.exit();r;){n=r.fn,r=r.next;try{n()}catch(t){throw r?i():e=void 0,t}}e=void 0,t&&t.enter()};if(h)i=function(){s.nextTick(t)};else if(!a||c.navigator&&c.navigator.standalone)if(l&&l.resolve){var n=l.resolve(void 0);i=function(){n.then(t)}}else i=function(){f.call(c,t)};else{var o=!0,u=document.createTextNode("");new a(t).observe(u,{characterData:!0}),i=function(){u.data=o=!o}}return function(t){var n={fn:t,next:void 0};e&&(e.next=n),r||(r=n,i()),e=n}}},function(t,n,r){"use strict";function e(t){var r,e;this.promise=new t(function(t,n){if(void 0!==r||void 0!==e)throw TypeError("Bad Promise constructor");r=t,e=n}),this.resolve=i(r),this.reject=i(e)}var i=r(21);t.exports.f=function(t){return new e(t)}},function(t,n,r){"use strict";var e,i,u=r(115),c=RegExp.prototype.exec,f=String.prototype.replace,o=c,a="lastIndex",s=(e=/a/,i=/b*/g,c.call(e,"a"),c.call(i,"a"),0!==e[a]||0!==i[a]),l=void 0!==/()??/.exec("")[1];(s||l)&&(o=function(t){var n,r,e,i,o=this;return l&&(r=new RegExp("^"+o.source+"$(?!\\s)",u.call(o))),s&&(n=o[a]),e=c.call(o,t),s&&e&&(o[a]=o.global?e.index+e[0].length:n),l&&e&&1<e.length&&f.call(e[0],r,function(){for(i=1;i<arguments.length-2;i++)void 0===arguments[i]&&(e[i]=void 0)}),e}),t.exports=o},function(t,n,i){var r=i(5),e=i(2),o=function(t,n){if(e(t),!r(n)&&null!==n)throw TypeError(n+": can't set as prototype!")};t.exports={set:Object.setPrototypeOf||("__proto__"in{}?function(t,r,e){try{(e=i(47)(Function.call,i(31).f(Object.prototype,"__proto__").set,2))(t,[]),r=!(t instanceof Array)}catch(t){r=!0}return function(t,n){return o(t,n),r?t.__proto__=n:e(t,n),t}}({},!1):void 0),check:o}},function(t,n,r){var e=r(118)("keys"),i=r(79);t.exports=function(t){return e[t]||(e[t]=i(t))}},function(t,n,r){var e=r(124),i=r(51);t.exports=function(t,n,r){if(e(n))throw TypeError("String#"+r+" doesn't accept regex!");return String(i(t))}},function(t,n,r){"use strict";var i=r(49),o=r(51);t.exports=function(t){var n=String(o(this)),r="",e=i(t);if(e<0||e==1/0)throw RangeError("Count can't be negative");for(;0<e;(e>>>=1)&&(n+=n))1&e&&(r+=n);return r}},function(t,n){t.exports="\t\n\v\f\r   ᠎             　\u2028\u2029\ufeff"},function(t,n,r){var e,i,o,u=r(47),c=r(175),f=r(143),a=r(140),s=r(3),l=s.process,h=s.setImmediate,v=s.clearImmediate,p=s.MessageChannel,d=s.Dispatch,y=0,g={},b="onreadystatechange",x=function(){var t=+this;if(g.hasOwnProperty(t)){var n=g[t];delete g[t],n()}},m=function(t){x.call(t.data)};h&&v||(h=function(t){for(var n=[],r=1;arguments.length>r;)n.push(arguments[r++]);return g[++y]=function(){c("function"==typeof t?t:Function(t),n)},e(y),y},v=function(t){delete g[t]},"process"==r(45)(l)?e=function(t){l.nextTick(u(x,t,1))}:d&&d.now?e=function(t){d.now(u(x,t,1))}:p?(o=(i=new p).port2,i.port1.onmessage=m,e=u(o.postMessage,o,1)):s.addEventListener&&"function"==typeof postMessage&&!s.importScripts?(e=function(t){s.postMessage(t+"","*")},s.addEventListener("message",m,!1)):e=b in a("script")?function(t){f.appendChild(a("script"))[b]=function(){f.removeChild(this),x.call(t)}}:function(t){setTimeout(u(x,t,1),0)}),t.exports={set:h,clear:v}},function(t,n,r){"use strict";function e(t,n,r){var e,i,o,u=new Array(r),c=8*r-n-1,f=(1<<c)-1,a=f>>1,s=23===n?W(2,-24)-W(2,-77):0,l=0,h=t<0||0===t&&1/t<0?1:0;for((t=G(t))!=t||t===C?(i=t!=t?1:0,e=f):(e=U(V(t)/B),t*(o=W(2,-e))<1&&(e--,o*=2),2<=(t+=1<=e+a?s/o:s*W(2,1-a))*o&&(e++,o/=2),f<=e+a?(i=0,e=f):1<=e+a?(i=(t*o-1)*W(2,n),e+=a):(i=t*W(2,a-1)*W(2,n),e=0));8<=n;u[l++]=255&i,i/=256,n-=8);for(e=e<<n|i,c+=n;0<c;u[l++]=255&e,e/=256,c-=8);return u[--l]|=128*h,u}function i(t,n,r){var e,i=8*r-n-1,o=(1<<i)-1,u=o>>1,c=i-7,f=r-1,a=t[f--],s=127&a;for(a>>=7;0<c;s=256*s+t[f],f--,c-=8);for(e=s&(1<<-c)-1,s>>=-c,c+=n;0<c;e=256*e+t[f],f--,c-=8);if(0===s)s=1-u;else{if(s===o)return e?NaN:a?-C:C;e+=W(2,n),s-=u}return(a?-1:1)*e*W(2,s-n)}function o(t){return t[3]<<24|t[2]<<16|t[1]<<8|t[0]}function u(t){return[255&t]}function c(t){return[255&t,t>>8&255]}function f(t){return[255&t,t>>8&255,t>>16&255,t>>24&255]}function a(t){return e(t,52,8)}function s(t){return e(t,23,4)}function l(t,n,r){M(t[I],n,{get:function(){return this[r]}})}function h(t,n,r,e){var i=O(+r);if(i+n>t[H])throw R(L);var o=t[K]._b,u=i+t[J],c=o.slice(u,u+n);return e?c:c.reverse()}function v(t,n,r,e,i,o){var u=O(+r);if(u+n>t[H])throw R(L);for(var c=t[K]._b,f=u+t[J],a=e(+i),s=0;s<n;s++)c[f+s]=a[o?s:n-s-1]}var p=r(3),d=r(10),y=r(68),g=r(132),b=r(26),x=r(76),m=r(4),S=r(70),w=r(49),_=r(8),O=r(194),E=r(73).f,M=r(11).f,P=r(137),j=r(83),F="ArrayBuffer",A="DataView",I="prototype",L="Wrong index!",N=p[F],T=p[A],k=p.Math,R=p.RangeError,C=p.Infinity,D=N,G=k.abs,W=k.pow,U=k.floor,V=k.log,B=k.LN2,q="byteLength",z="byteOffset",K=d?"_b":"buffer",H=d?"_l":q,J=d?"_o":z;if(g.ABV){if(!m(function(){N(1)})||!m(function(){new N(-1)})||m(function(){return new N,new N(1.5),new N(NaN),N.name!=F})){for(var $,Y=(N=function(t){return S(this,N),new D(O(t))})[I]=D[I],X=E(D),Q=0;X.length>Q;)($=X[Q++])in N||b(N,$,D[$]);y||(Y.constructor=N)}var Z=new T(new N(2)),tt=T[I].setInt8;Z.setInt8(0,2147483648),Z.setInt8(1,2147483649),!Z.getInt8(0)&&Z.getInt8(1)||x(T[I],{setInt8:function(t,n){tt.call(this,t,n<<24>>24)},setUint8:function(t,n){tt.call(this,t,n<<24>>24)}},!0)}else N=function(t){S(this,N,F);var n=O(t);this._b=P.call(new Array(n),0),this[H]=n},T=function(t,n,r){S(this,T,A),S(t,N,A);var e=t[H],i=w(n);if(i<0||e<i)throw R("Wrong offset!");if(e<i+(r=void 0===r?e-i:_(r)))throw R("Wrong length!");this[K]=t,this[J]=i,this[H]=r},d&&(l(N,q,"_l"),l(T,"buffer","_b"),l(T,q,"_l"),l(T,z,"_o")),x(T[I],{getInt8:function(t){return h(this,1,t)[0]<<24>>24},getUint8:function(t){return h(this,1,t)[0]},getInt16:function(t){var n=h(this,2,t,arguments[1]);return(n[1]<<8|n[0])<<16>>16},getUint16:function(t){var n=h(this,2,t,arguments[1]);return n[1]<<8|n[0]},getInt32:function(t){return o(h(this,4,t,arguments[1]))},getUint32:function(t){return o(h(this,4,t,arguments[1]))>>>0},getFloat32:function(t){return i(h(this,4,t,arguments[1]),23,4)},getFloat64:function(t){return i(h(this,8,t,arguments[1]),52,8)},setInt8:function(t,n){v(this,1,t,u,n)},setUint8:function(t,n){v(this,1,t,u,n)},setInt16:function(t,n){v(this,2,t,c,n,arguments[2])},setUint16:function(t,n){v(this,2,t,c,n,arguments[2])},setInt32:function(t,n){v(this,4,t,f,n,arguments[2])},setUint32:function(t,n){v(this,4,t,f,n,arguments[2])},setFloat32:function(t,n){v(this,4,t,s,n,arguments[2])},setFloat64:function(t,n){v(this,8,t,a,n,arguments[2])}});j(N,F),j(T,A),b(T[I],g.VIEW,!0),n[F]=N,n[A]=T},function(t,n,r){var e=r(3),i=r(46),o=r(68),u=r(195),c=r(11).f;t.exports=function(t){var n=i.Symbol||(i.Symbol=o?{}:e.Symbol||{});"_"==t.charAt(0)||t in n||c(n,t,{value:u.f(t)})}},function(t,n,r){var e=r(81),i=r(7)("iterator"),o=r(82);t.exports=r(46).getIteratorMethod=function(t){if(null!=t)return t[i]||t["@@iterator"]||o[e(t)]}},function(t,n,r){"use strict";var e=r(67),i=r(178),o=r(82),u=r(33);t.exports=r(147)(Array,"Array",function(t,n){this._t=u(t),this._i=0,this._k=n},function(){var t=this._t,n=this._k,r=this._i++;return!t||r>=t.length?(this._t=void 0,i(1)):i(0,"keys"==n?r:"values"==n?t[r]:[r,t[r]])},"values"),o.Arguments=o.Array,e("keys"),e("values"),e("entries")},function(t,n){t.exports=function(t,n){t.classList?t.classList.add(n):t.className+=" "+n}},function(t,n){t.exports=function(t,n){if(t.classList)t.classList.remove(n);else{var r=new RegExp("(^|\\b)"+n.split(" ").join("|")+"(\\b|$)","gi");t.className=t.className.replace(r," ")}}},function(t,n,r){var e=r(45);t.exports=function(t,n){if("number"!=typeof t&&"Number"!=e(t))throw TypeError(n);return+t}},function(t,n,r){"use strict";var a=r(17),s=r(78),l=r(8);t.exports=[].copyWithin||function(t,n){var r=a(this),e=l(r.length),i=s(t,e),o=s(n,e),u=2<arguments.length?arguments[2]:void 0,c=Math.min((void 0===u?e:s(u,e))-o,e-i),f=1;for(o<i&&i<o+c&&(f=-1,o+=c-1,i+=c-1);0<c--;)o in r?r[i]=r[o]:delete r[i],i+=f,o+=f;return r}},function(t,n,r){var e=r(71);t.exports=function(t,n){var r=[];return e(t,!1,r.push,r,n),r}},function(t,n,r){var s=r(21),l=r(17),h=r(116),v=r(8);t.exports=function(t,n,r,e,i){s(n);var o=l(t),u=h(o),c=v(o.length),f=i?c-1:0,a=i?-1:1;if(r<2)for(;;){if(f in u){e=u[f],f+=a;break}if(f+=a,i?f<0:c<=f)throw TypeError("Reduce of empty array with no initial value")}for(;i?0<=f:f<c;f+=a)f in u&&(e=n(e,u[f],f,o));return e}},function(t,n,r){"use strict";var o=r(21),u=r(5),c=r(175),f=[].slice,a={};t.exports=Function.bind||function(n){var r=o(this),e=f.call(arguments,1),i=function(){var t=e.concat(f.call(arguments));return this instanceof i?function(t,n,r){if(!(n in a)){for(var e=[],i=0;i<n;i++)e[i]="a["+i+"]";a[n]=Function("F,a","return new F("+e.join(",")+")")}return a[n](t,r)}(r,t.length,t):c(r,t,n)};return u(r.prototype)&&(i.prototype=r.prototype),i}},function(t,n,r){"use strict";var u=r(11).f,c=r(72),f=r(76),a=r(47),s=r(70),l=r(71),e=r(147),i=r(178),o=r(77),h=r(10),v=r(69).fastKey,p=r(80),d=h?"_s":"size",y=function(t,n){var r,e=v(n);if("F"!==e)return t._i[e];for(r=t._f;r;r=r.n)if(r.k==n)return r};t.exports={getConstructor:function(t,o,r,e){var i=t(function(t,n){s(t,i,o,"_i"),t._t=o,t._i=c(null),t._f=void 0,t._l=void 0,t[d]=0,null!=n&&l(n,r,t[e],t)});return f(i.prototype,{clear:function(){for(var t=p(this,o),n=t._i,r=t._f;r;r=r.n)r.r=!0,r.p&&(r.p=r.p.n=void 0),delete n[r.i];t._f=t._l=void 0,t[d]=0},delete:function(t){var n=p(this,o),r=y(n,t);if(r){var e=r.n,i=r.p;delete n._i[r.i],r.r=!0,i&&(i.n=e),e&&(e.p=i),n._f==r&&(n._f=e),n._l==r&&(n._l=i),n[d]--}return!!r},forEach:function(t){p(this,o);for(var n,r=a(t,1<arguments.length?arguments[1]:void 0,3);n=n?n.n:this._f;)for(r(n.v,n.k,this);n&&n.r;)n=n.p},has:function(t){return!!y(p(this,o),t)}}),h&&u(i.prototype,"size",{get:function(){return p(this,o)[d]}}),i},def:function(t,n,r){var e,i,o=y(t,n);return o?o.v=r:(t._l=o={i:i=v(n,!0),k:n,v:r,p:e=t._l,n:void 0,r:!1},t._f||(t._f=o),e&&(e.n=o),t[d]++,"F"!==i&&(t._i[i]=o)),t},getEntry:y,setStrong:function(t,r,n){e(t,r,function(t,n){this._t=p(t,r),this._k=n,this._l=void 0},function(){for(var t=this,n=t._k,r=t._l;r&&r.r;)r=r.p;return t._t&&(t._l=r=r?r.n:t._t._f)?i(0,"keys"==n?r.k:"values"==n?r.v:[r.k,r.v]):(t._t=void 0,i(1))},n?"entries":"values",!n,!0),o(r)}}},function(t,n,r){var e=r(81),i=r(167);t.exports=function(t){return function(){if(e(this)!=t)throw TypeError(t+"#toJSON isn't generic");return i(this)}}},function(t,n,r){"use strict";var u=r(76),c=r(69).getWeak,i=r(2),f=r(5),a=r(70),s=r(71),e=r(50),l=r(30),h=r(80),o=e(5),v=e(6),p=0,d=function(t){return t._l||(t._l=new y)},y=function(){this.a=[]},g=function(t,n){return o(t.a,function(t){return t[0]===n})};y.prototype={get:function(t){var n=g(this,t);if(n)return n[1]},has:function(t){return!!g(this,t)},set:function(t,n){var r=g(this,t);r?r[1]=n:this.a.push([t,n])},delete:function(n){var t=v(this.a,function(t){return t[0]===n});return~t&&this.a.splice(t,1),!!~t}},t.exports={getConstructor:function(t,r,e,i){var o=t(function(t,n){a(t,o,r,"_i"),t._t=r,t._i=p++,t._l=void 0,null!=n&&s(n,e,t[i],t)});return u(o.prototype,{delete:function(t){if(!f(t))return!1;var n=c(t);return!0===n?d(h(this,r)).delete(t):n&&l(n,this._i)&&delete n[this._i]},has:function(t){if(!f(t))return!1;var n=c(t);return!0===n?d(h(this,r)).has(t):n&&l(n,this._i)}}),o},def:function(t,n,r){var e=c(i(n),!0);return!0===e?d(t).set(n,r):e[t._i]=r,t},ufstore:d}},function(t,n,r){"use strict";var p=r(123),d=r(5),y=r(8),g=r(47),b=r(7)("isConcatSpreadable");t.exports=function t(n,r,e,i,o,u,c,f){for(var a,s,l=o,h=0,v=!!c&&g(c,f,3);h<i;){if(h in e){if(a=v?v(e[h],h,r):e[h],s=!1,d(a)&&(s=void 0!==(s=a[b])?!!s:p(a)),s&&0<u)l=t(n,r,a,y(a.length),l,u-1)-1;else{if(9007199254740991<=l)throw TypeError();n[l]=a}l++}h++}return l}},function(t,n,r){t.exports=!r(10)&&!r(4)(function(){return 7!=Object.defineProperty(r(140)("div"),"a",{get:function(){return 7}}).a})},function(t,n){t.exports=function(t,n,r){var e=void 0===r;switch(n.length){case 0:return e?t():t.call(r);case 1:return e?t(n[0]):t.call(r,n[0]);case 2:return e?t(n[0],n[1]):t.call(r,n[0],n[1]);case 3:return e?t(n[0],n[1],n[2]):t.call(r,n[0],n[1],n[2]);case 4:return e?t(n[0],n[1],n[2],n[3]):t.call(r,n[0],n[1],n[2],n[3])}return t.apply(r,n)}},function(t,n,r){var e=r(5),i=Math.floor;t.exports=function(t){return!e(t)&&isFinite(t)&&i(t)===t}},function(t,n,r){var o=r(2);t.exports=function(t,n,r,e){try{return e?n(o(r)[0],r[1]):n(r)}catch(n){var i=t.return;throw void 0!==i&&o(i.call(t)),n}}},function(t,n){t.exports=function(t,n){return{value:n,done:!!t}}},function(t,n,r){var o=r(149),e=Math.pow,u=e(2,-52),c=e(2,-23),f=e(2,127)*(2-c),a=e(2,-126);t.exports=Math.fround||function(t){var n,r,e=Math.abs(t),i=o(t);return e<a?i*(e/a/c+1/u-1/u)*a*c:f<(r=(n=(1+c/u)*e)-(n-e))||r!=r?i*(1/0):i*r}},function(t,n){t.exports=Math.log1p||function(t){return-1e-8<(t=+t)&&t<1e-8?t-t*t/2:Math.log(1+t)}},function(t,n){t.exports=Math.scale||function(t,n,r,e,i){return 0===arguments.length||t!=t||n!=n||r!=r||e!=e||i!=i?NaN:t===1/0||t===-1/0?t:(t-n)*(i-e)/(r-n)+e}},function(t,n,r){"use strict";var h=r(10),v=r(74),p=r(127),d=r(117),y=r(17),g=r(116),i=Object.assign;t.exports=!i||r(4)(function(){var t={},n={},r=Symbol(),e="abcdefghijklmnopqrst";return t[r]=7,e.split("").forEach(function(t){n[t]=t}),7!=i({},t)[r]||Object.keys(i({},n)).join("")!=e})?function(t,n){for(var r=y(t),e=arguments.length,i=1,o=p.f,u=d.f;i<e;)for(var c,f=g(arguments[i++]),a=o?v(f).concat(o(f)):v(f),s=a.length,l=0;l<s;)c=a[l++],h&&!u.call(f,c)||(r[c]=f[c]);return r}:i},function(t,n,r){var u=r(11),c=r(2),f=r(74);t.exports=r(10)?Object.defineProperties:function(t,n){c(t);for(var r,e=f(n),i=e.length,o=0;o<i;)u.f(t,r=e[o++],n[r]);return t}},function(t,n,r){var e=r(33),i=r(73).f,o={}.toString,u="object"==typeof window&&window&&Object.getOwnPropertyNames?Object.getOwnPropertyNames(window):[];t.exports.f=function(t){return u&&"[object Window]"==o.call(t)?function(t){try{return i(t)}catch(t){return u.slice()}}(t):i(e(t))}},function(t,n,r){var u=r(30),c=r(33),f=r(120)(!1),a=r(154)("IE_PROTO");t.exports=function(t,n){var r,e=c(t),i=0,o=[];for(r in e)r!=a&&u(e,r)&&o.push(r);for(;n.length>i;)u(e,r=n[i++])&&(~f(o,r)||o.push(r));return o}},function(t,n,r){var f=r(10),a=r(74),s=r(33),l=r(117).f;t.exports=function(c){return function(t){for(var n,r=s(t),e=a(r),i=e.length,o=0,u=[];o<i;)n=e[o++],f&&!l.call(r,n)||u.push(c?[n,r[n]]:r[n]);return u}}},function(t,n,r){var e=r(73),i=r(127),o=r(2),u=r(3).Reflect;t.exports=u&&u.ownKeys||function(t){var n=e.f(o(t)),r=i.f;return r?n.concat(r(t)):n}},function(t,n,r){var e=r(3).parseFloat,i=r(84).trim;t.exports=1/e(r(157)+"-0")!=-1/0?function(t){var n=i(String(t),3),r=e(n);return 0===r&&"-"==n.charAt(0)?-0:r}:e},function(t,n,r){var e=r(3).parseInt,i=r(84).trim,o=r(157),u=/^[-+]?0[xX]/;t.exports=8!==e(o+"08")||22!==e(o+"0x16")?function(t,n){var r=i(String(t),3);return e(r,n>>>0||(u.test(r)?16:10))}:e},function(t,n){t.exports=function(t){try{return{e:!1,v:t()}}catch(t){return{e:!0,v:t}}}},function(t,n,r){var e=r(2),i=r(5),o=r(151);t.exports=function(t,n){if(e(t),i(n)&&n.constructor===t)return n;var r=o.f(t);return(0,r.resolve)(n),r.promise}},function(t,n){t.exports=Object.is||function(t,n){return t===n?0!==t||1/t==1/n:t!=t&&n!=n}},function(t,n,r){var s=r(8),l=r(156),h=r(51);t.exports=function(t,n,r,e){var i=String(h(t)),o=i.length,u=void 0===r?" ":String(r),c=s(n);if(c<=o||""==u)return i;var f=c-o,a=l.call(u,Math.ceil(f/u.length));return a.length>f&&(a=a.slice(0,f)),e?a+i:i+a}},function(t,n,r){var e=r(49),i=r(8);t.exports=function(t){if(void 0===t)return 0;var n=e(t),r=i(n);if(n!==r)throw RangeError("Wrong length!");return r}},function(t,n,r){n.f=r(7)},function(t,n,r){"use strict";var e=r(170),i=r(80);t.exports=r(121)("Map",function(t){return function(){return t(this,0<arguments.length?arguments[0]:void 0)}},{get:function(t){var n=e.getEntry(i(this,"Map"),t);return n&&n.v},set:function(t,n){return e.def(i(this,"Map"),0===t?0:t,n)}},e,!0)},function(t,n,r){"use strict";var e=r(152);r(1)({target:"RegExp",proto:!0,forced:e!==/./.exec},{exec:e})},function(t,n,r){r(10)&&"g"!=/./g.flags&&r(11).f(RegExp.prototype,"flags",{configurable:!0,get:r(115)})},function(t,n,r){"use strict";var e=r(170),i=r(80);t.exports=r(121)("Set",function(t){return function(){return t(this,0<arguments.length?arguments[0]:void 0)}},{add:function(t){return e.def(i(this,"Set"),t=0===t?0:t,t)}},e)},function(t,n,r){"use strict";var o,e=r(3),i=r(50)(0),u=r(27),c=r(69),f=r(182),a=r(172),s=r(5),l=r(80),h=r(80),v=!e.ActiveXObject&&"ActiveXObject"in e,p="WeakMap",d=c.getWeak,y=Object.isExtensible,g=a.ufstore,b=function(t){return function(){return t(this,0<arguments.length?arguments[0]:void 0)}},x={get:function(t){if(s(t)){var n=d(t);return!0===n?g(l(this,p)).get(t):n?n[this._i]:void 0}},set:function(t,n){return a.def(l(this,p),t,n)}},m=t.exports=r(121)(p,b,x,a,!0,!0);h&&v&&(f((o=a.getConstructor(b,p)).prototype,x),c.NEED=!0,i(["delete","has","get","set"],function(e){var t=m.prototype,i=t[e];u(t,e,function(t,n){if(!s(t)||y(t))return i.call(this,t,n);this._f||(this._f=new o);var r=this._f[e](t,n);return"set"==e?this:r})}))},,,,function(t,n){"use strict";t.exports={init:function(){var t=document.querySelector("#page-nav");t&&!document.querySelector("#page-nav .extend.prev")&&(t.innerHTML='<a class="extend prev disabled" rel="prev">&laquo; Prev</a>'+t.innerHTML),t&&!document.querySelector("#page-nav .extend.next")&&(t.innerHTML=t.innerHTML+'<a class="extend next disabled" rel="next">Next &raquo;</a>'),yiliaConfig&&yiliaConfig.open_in_new&&document.querySelectorAll(".article-entry a:not(.article-more-a)").forEach(function(t){var n=t.getAttribute("target");n&&""!==n||t.setAttribute("target","_blank")}),yiliaConfig&&yiliaConfig.toc_hide_index&&document.querySelectorAll(".toc-number").forEach(function(t){t.style.display="none"})}}},function(t,n,r){"use strict";function e(t){return t&&t.__esModule?t:{default:t}}function i(t,n,r,e,i){var o=function(t){for(var n=t.offsetLeft,r=t.offsetParent;null!==r;)n+=r.offsetLeft,r=r.offsetParent;return n}(t),u=function(t){for(var n=t.offsetTop,r=t.offsetParent;null!==r;)n+=r.offsetTop,r=r.offsetParent;return n}(t)-n;if(u-r<=i){var c=t.$newDom;c||(c=t.cloneNode(!0),(0,a.default)(t,c),(t.$newDom=c).style.position="fixed",c.style.top=(r||u)+"px",c.style.left=o+"px",c.style.zIndex=e||2,c.style.width="100%",c.style.color="#fff"),c.style.visibility="visible",t.style.visibility="hidden"}else{t.style.visibility="visible";var f=t.$newDom;f&&(f.style.visibility="hidden")}}function o(){var t=document.querySelector(".js-overlay"),n=document.querySelector(".js-header-menu");i(t,document.body.scrollTop,-63,2,0),i(n,document.body.scrollTop,1,3,0)}var f=e(r(163)),a=e((e(r(164)),r(414))),u=e(r(134)),c=e(r(204)),s=r(135);u.default.versions.mobile&&window.screen.width<800&&(function(){for(var t=document.querySelectorAll(".js-header-menu li a"),n=window.location.pathname,r=0,e=t.length;r<e;r++){var i=t[r];o=n,u=i.getAttribute("href"),c=/\/|index.html/g,o.replace(c,"")===u.replace(c,"")&&(0,f.default)(i,"active")}var o,u,c}(),document.querySelector("#container").addEventListener("scroll",function(t){o()}),window.addEventListener("scroll",function(t){o()}),o()),(0,s.addLoadEvent)(function(){c.default.init()}),t.exports={}},,,function(t,n,r){(function(t){"use strict";function n(t,n,r){t[n]||Object.defineProperty(t,n,{writable:!0,configurable:!0,value:r})}if(r(413),r(209),r(211),t._babelPolyfill)throw new Error("only one instance of babel-polyfill is allowed");t._babelPolyfill=!0;n(String.prototype,"padLeft","".padStart),n(String.prototype,"padRight","".padEnd),"pop,reverse,shift,keys,values,entries,indexOf,every,some,forEach,map,filter,find,findIndex,includes,join,slice,concat,push,splice,unshift,sort,lastIndexOf,reduce,reduceRight,copyWithin,fill".split(",").forEach(function(t){[][t]&&n(Array,t,Function.call.bind([][t]))})}).call(n,function(){return this}())},function(L,t){(function(t){!function(t){"use strict";function o(t,n,r,e){var o,u,c,f,i=n&&n.prototype instanceof h?n:h,a=Object.create(i.prototype),s=new p(e||[]);return a._invoke=(o=t,u=r,c=s,f=_,function(t,n){if(f===E)throw new Error("Generator is already running");if(f===M){if("throw"===t)throw n;return d()}for(c.method=t,c.arg=n;;){var r=c.delegate;if(r){var e=v(r,c);if(e){if(e===P)continue;return e}}if("next"===c.method)c.sent=c._sent=c.arg;else if("throw"===c.method){if(f===_)throw f=M,c.arg;c.dispatchException(c.arg)}else"return"===c.method&&c.abrupt("return",c.arg);f=E;var i=l(o,u,c);if("normal"===i.type){if(f=c.done?M:O,i.arg===P)continue;return{value:i.arg,done:c.done}}"throw"===i.type&&(f=M,c.method="throw",c.arg=i.arg)}}),a}function l(t,n,r){try{return{type:"normal",arg:t.call(n,r)}}catch(t){return{type:"throw",arg:t}}}function h(){}function r(){}function n(){}function e(t){["next","throw","return"].forEach(function(n){t[n]=function(t){return this._invoke(n,t)}})}function u(c){function f(t,n,r,e){var i=l(c[t],c,n);if("throw"!==i.type){var o=i.arg,u=o.value;return u&&"object"==typeof u&&y.call(u,"__await")?Promise.resolve(u.__await).then(function(t){f("next",t,r,e)},function(t){f("throw",t,r,e)}):Promise.resolve(u).then(function(t){o.value=t,r(o)},e)}e(i.arg)}var n;"object"==typeof t.process&&t.process.domain&&(f=t.process.domain.bind(f)),this._invoke=function(r,e){function t(){return new Promise(function(t,n){f(r,e,t,n)})}return n=n?n.then(t,t):t()}}function v(t,n){var r=t.iterator[n.method];if(r===a){if(n.delegate=null,"throw"===n.method){if(t.iterator.return&&(n.method="return",n.arg=a,v(t,n),"throw"===n.method))return P;n.method="throw",n.arg=new TypeError("The iterator does not provide a 'throw' method")}return P}var e=l(r,t.iterator,n.arg);if("throw"===e.type)return n.method="throw",n.arg=e.arg,n.delegate=null,P;var i=e.arg;return i?i.done?(n[t.resultName]=i.value,n.next=t.nextLoc,"return"!==n.method&&(n.method="next",n.arg=a),n.delegate=null,P):i:(n.method="throw",n.arg=new TypeError("iterator result is not an object"),n.delegate=null,P)}function i(t){var n={tryLoc:t[0]};1 in t&&(n.catchLoc=t[1]),2 in t&&(n.finallyLoc=t[2],n.afterLoc=t[3]),this.tryEntries.push(n)}function c(t){var n=t.completion||{};n.type="normal",delete n.arg,t.completion=n}function p(t){this.tryEntries=[{tryLoc:"root"}],t.forEach(i,this),this.reset(!0)}function f(n){if(n){var t=n[b];if(t)return t.call(n);if("function"==typeof n.next)return n;if(!isNaN(n.length)){var r=-1,e=function t(){for(;++r<n.length;)if(y.call(n,r))return t.value=n[r],t.done=!1,t;return t.value=a,t.done=!0,t};return e.next=e}}return{next:d}}function d(){return{value:a,done:!0}}var a,s=Object.prototype,y=s.hasOwnProperty,g="function"==typeof Symbol?Symbol:{},b=g.iterator||"@@iterator",x=g.asyncIterator||"@@asyncIterator",m=g.toStringTag||"@@toStringTag",S="object"==typeof L,w=t.regeneratorRuntime;if(w)S&&(L.exports=w);else{(w=t.regeneratorRuntime=S?L.exports:{}).wrap=o;var _="suspendedStart",O="suspendedYield",E="executing",M="completed",P={},j={};j[b]=function(){return this};var F=Object.getPrototypeOf,A=F&&F(F(f([])));A&&A!==s&&y.call(A,b)&&(j=A);var I=n.prototype=h.prototype=Object.create(j);r.prototype=I.constructor=n,n.constructor=r,n[m]=r.displayName="GeneratorFunction",w.isGeneratorFunction=function(t){var n="function"==typeof t&&t.constructor;return!!n&&(n===r||"GeneratorFunction"===(n.displayName||n.name))},w.mark=function(t){return Object.setPrototypeOf?Object.setPrototypeOf(t,n):(t.__proto__=n,m in t||(t[m]="GeneratorFunction")),t.prototype=Object.create(I),t},w.awrap=function(t){return{__await:t}},e(u.prototype),u.prototype[x]=function(){return this},w.AsyncIterator=u,w.async=function(t,n,r,e){var i=new u(o(t,n,r,e));return w.isGeneratorFunction(n)?i:i.next().then(function(t){return t.done?t.value:i.next()})},e(I),I[m]="Generator",I[b]=function(){return this},I.toString=function(){return"[object Generator]"},w.keys=function(r){var e=[];for(var t in r)e.push(t);return e.reverse(),function t(){for(;e.length;){var n=e.pop();if(n in r)return t.value=n,t.done=!1,t}return t.done=!0,t}},w.values=f,p.prototype={constructor:p,reset:function(t){if(this.prev=0,this.next=0,this.sent=this._sent=a,this.done=!1,this.delegate=null,this.method="next",this.arg=a,this.tryEntries.forEach(c),!t)for(var n in this)"t"===n.charAt(0)&&y.call(this,n)&&!isNaN(+n.slice(1))&&(this[n]=a)},stop:function(){this.done=!0;var t=this.tryEntries[0].completion;if("throw"===t.type)throw t.arg;return this.rval},dispatchException:function(r){function t(t,n){return o.type="throw",o.arg=r,e.next=t,n&&(e.method="next",e.arg=a),!!n}if(this.done)throw r;for(var e=this,n=this.tryEntries.length-1;0<=n;--n){var i=this.tryEntries[n],o=i.completion;if("root"===i.tryLoc)return t("end");if(i.tryLoc<=this.prev){var u=y.call(i,"catchLoc"),c=y.call(i,"finallyLoc");if(u&&c){if(this.prev<i.catchLoc)return t(i.catchLoc,!0);if(this.prev<i.finallyLoc)return t(i.finallyLoc)}else if(u){if(this.prev<i.catchLoc)return t(i.catchLoc,!0)}else{if(!c)throw new Error("try statement without catch or finally");if(this.prev<i.finallyLoc)return t(i.finallyLoc)}}}},abrupt:function(t,n){for(var r=this.tryEntries.length-1;0<=r;--r){var e=this.tryEntries[r];if(e.tryLoc<=this.prev&&y.call(e,"finallyLoc")&&this.prev<e.finallyLoc){var i=e;break}}i&&("break"===t||"continue"===t)&&i.tryLoc<=n&&n<=i.finallyLoc&&(i=null);var o=i?i.completion:{};return o.type=t,o.arg=n,i?(this.method="next",this.next=i.finallyLoc,P):this.complete(o)},complete:function(t,n){if("throw"===t.type)throw t.arg;return"break"===t.type||"continue"===t.type?this.next=t.arg:"return"===t.type?(this.rval=this.arg=t.arg,this.method="return",this.next="end"):"normal"===t.type&&n&&(this.next=n),P},finish:function(t){for(var n=this.tryEntries.length-1;0<=n;--n){var r=this.tryEntries[n];if(r.finallyLoc===t)return this.complete(r.completion,r.afterLoc),c(r),P}},catch:function(t){for(var n=this.tryEntries.length-1;0<=n;--n){var r=this.tryEntries[n];if(r.tryLoc===t){var e=r.completion;if("throw"===e.type){var i=e.arg;c(r)}return i}}throw new Error("illegal catch attempt")},delegateYield:function(t,n,r){return this.delegate={iterator:f(t),resultName:n,nextLoc:r},"next"===this.method&&(this.arg=a),P}}}}("object"==typeof t?t:"object"==typeof window?window:"object"==typeof self?self:this)}).call(t,function(){return this}())},,function(t,n,r){r(221),t.exports=r(46).RegExp.escape},,,,function(t,n,r){var e=r(5),i=r(123),o=r(7)("species");t.exports=function(t){var n;return i(t)&&("function"!=typeof(n=t.constructor)||n!==Array&&!i(n.prototype)||(n=void 0),e(n)&&(null===(n=n[o])&&(n=void 0))),void 0===n?Array:n}},function(t,n,r){"use strict";var e=r(4),i=Date.prototype.getTime,o=Date.prototype.toISOString,u=function(t){return 9<t?t:"0"+t};t.exports=e(function(){return"0385-07-25T07:06:39.999Z"!=o.call(new Date(-5e13-1))})||!e(function(){o.call(new Date(NaN))})?function(){if(!isFinite(i.call(this)))throw RangeError("Invalid time value");var t=this,n=t.getUTCFullYear(),r=t.getUTCMilliseconds(),e=n<0?"-":9999<n?"+":"";return e+("00000"+Math.abs(n)).slice(e?-6:-4)+"-"+u(t.getUTCMonth()+1)+"-"+u(t.getUTCDate())+"T"+u(t.getUTCHours())+":"+u(t.getUTCMinutes())+":"+u(t.getUTCSeconds())+"."+(99<r?r:"0"+u(r))+"Z"}:o},function(t,n,r){"use strict";var e=r(2),i=r(53);t.exports=function(t){if("string"!==t&&"number"!==t&&"default"!==t)throw TypeError("Incorrect hint");return i(e(this),"number"!=t)}},function(t,n,r){var c=r(74),f=r(127),a=r(117);t.exports=function(t){var n=c(t),r=f.f;if(r)for(var e,i=r(t),o=a.f,u=0;i.length>u;)o.call(t,e=i[u++])&&n.push(e);return n}},function(t,n,r){t.exports=r(118)("native-function-to-string",Function.toString)},function(t,n){t.exports=function(n,r){var e=r===Object(r)?function(t){return r[t]}:r;return function(t){return String(t).replace(n,e)}}},function(t,n,r){var e=r(1),i=r(220)(/[\\^$*+?.()|[\]{}]/g,"\\$&");e(e.S,"RegExp",{escape:function(t){return i(t)}})},function(t,n,r){var e=r(1);e(e.P,"Array",{copyWithin:r(166)}),r(67)("copyWithin")},function(t,n,r){"use strict";var e=r(1),i=r(50)(4);e(e.P+e.F*!r(48)([].every,!0),"Array",{every:function(t){return i(this,t,arguments[1])}})},function(t,n,r){var e=r(1);e(e.P,"Array",{fill:r(137)}),r(67)("fill")},function(t,n,r){"use strict";var e=r(1),i=r(50)(2);e(e.P+e.F*!r(48)([].filter,!0),"Array",{filter:function(t){return i(this,t,arguments[1])}})},function(t,n,r){"use strict";var e=r(1),i=r(50)(6),o="findIndex",u=!0;o in[]&&Array(1)[o](function(){u=!1}),e(e.P+e.F*u,"Array",{findIndex:function(t){return i(this,t,1<arguments.length?arguments[1]:void 0)}}),r(67)(o)},function(t,n,r){"use strict";var e=r(1),i=r(50)(5),o="find",u=!0;o in[]&&Array(1)[o](function(){u=!1}),e(e.P+e.F*u,"Array",{find:function(t){return i(this,t,1<arguments.length?arguments[1]:void 0)}}),r(67)(o)},function(t,n,r){"use strict";var e=r(1),i=r(50)(0),o=r(48)([].forEach,!0);e(e.P+e.F*!o,"Array",{forEach:function(t){return i(this,t,arguments[1])}})},function(t,n,r){"use strict";var h=r(47),e=r(1),v=r(17),p=r(177),d=r(145),y=r(8),g=r(139),b=r(161);e(e.S+e.F*!r(125)(function(t){Array.from(t)}),"Array",{from:function(t){var n,r,e,i,o=v(t),u="function"==typeof this?this:Array,c=arguments.length,f=1<c?arguments[1]:void 0,a=void 0!==f,s=0,l=b(o);if(a&&(f=h(f,2<c?arguments[2]:void 0,2)),null==l||u==Array&&d(l))for(r=new u(n=y(o.length));s<n;s++)g(r,s,a?f(o[s],s):o[s]);else for(i=l.call(o),r=new u;!(e=i.next()).done;s++)g(r,s,a?p(i,f,[e.value,s],!0):e.value);return r.length=s,r}})},function(t,n,r){"use strict";var e=r(1),i=r(120)(!1),o=[].indexOf,u=!!o&&1/[1].indexOf(1,-0)<0;e(e.P+e.F*(u||!r(48)(o)),"Array",{indexOf:function(t){return u?o.apply(this,arguments)||0:i(this,t,arguments[1])}})},function(t,n,r){var e=r(1);e(e.S,"Array",{isArray:r(123)})},function(t,n,r){"use strict";var e=r(1),i=r(33),o=[].join;e(e.P+e.F*(r(116)!=Object||!r(48)(o)),"Array",{join:function(t){return o.call(i(this),void 0===t?",":t)}})},function(t,n,r){"use strict";var e=r(1),i=r(33),o=r(49),u=r(8),c=[].lastIndexOf,f=!!c&&1/[1].lastIndexOf(1,-0)<0;e(e.P+e.F*(f||!r(48)(c)),"Array",{lastIndexOf:function(t){if(f)return c.apply(this,arguments)||0;var n=i(this),r=u(n.length),e=r-1;for(1<arguments.length&&(e=Math.min(e,o(arguments[1]))),e<0&&(e=r+e);0<=e;e--)if(e in n&&n[e]===t)return e||0;return-1}})},function(t,n,r){"use strict";var e=r(1),i=r(50)(1);e(e.P+e.F*!r(48)([].map,!0),"Array",{map:function(t){return i(this,t,arguments[1])}})},function(t,n,r){"use strict";var e=r(1),i=r(139);e(e.S+e.F*r(4)(function(){function t(){}return!(Array.of.call(t)instanceof t)}),"Array",{of:function(){for(var t=0,n=arguments.length,r=new("function"==typeof this?this:Array)(n);t<n;)i(r,t,arguments[t++]);return r.length=n,r}})},function(t,n,r){"use strict";var e=r(1),i=r(168);e(e.P+e.F*!r(48)([].reduceRight,!0),"Array",{reduceRight:function(t){return i(this,t,arguments.length,arguments[1],!0)}})},function(t,n,r){"use strict";var e=r(1),i=r(168);e(e.P+e.F*!r(48)([].reduce,!0),"Array",{reduce:function(t){return i(this,t,arguments.length,arguments[1],!1)}})},function(t,n,r){"use strict";var e=r(1),i=r(143),a=r(45),s=r(78),l=r(8),h=[].slice;e(e.P+e.F*r(4)(function(){i&&h.call(i)}),"Array",{slice:function(t,n){var r=l(this.length),e=a(this);if(n=void 0===n?r:n,"Array"==e)return h.call(this,t,n);for(var i=s(t,r),o=s(n,r),u=l(o-i),c=new Array(u),f=0;f<u;f++)c[f]="String"==e?this.charAt(i+f):this[i+f];return c}})},function(t,n,r){"use strict";var e=r(1),i=r(50)(3);e(e.P+e.F*!r(48)([].some,!0),"Array",{some:function(t){return i(this,t,arguments[1])}})},function(t,n,r){"use strict";var e=r(1),i=r(21),o=r(17),u=r(4),c=[].sort,f=[1,2,3];e(e.P+e.F*(u(function(){f.sort(void 0)})||!u(function(){f.sort(null)})||!r(48)(c)),"Array",{sort:function(t){return void 0===t?c.call(o(this)):c.call(o(this),i(t))}})},function(t,n,r){r(77)("Array")},function(t,n,r){var e=r(1);e(e.S,"Date",{now:function(){return(new Date).getTime()}})},function(t,n,r){var e=r(1),i=r(216);e(e.P+e.F*(Date.prototype.toISOString!==i),"Date",{toISOString:i})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(53);e(e.P+e.F*r(4)(function(){return null!==new Date(NaN).toJSON()||1!==Date.prototype.toJSON.call({toISOString:function(){return 1}})}),"Date",{toJSON:function(t){var n=i(this),r=o(n);return"number"!=typeof r||isFinite(r)?n.toISOString():null}})},function(t,n,r){var e=r(7)("toPrimitive"),i=Date.prototype;e in i||r(26)(i,e,r(217))},function(t,n,r){var e=Date.prototype,i="Invalid Date",o="toString",u=e[o],c=e.getTime;new Date(NaN)+""!=i&&r(27)(e,o,function(){var t=c.call(this);return t==t?u.call(this):i})},function(t,n,r){var e=r(1);e(e.P,"Function",{bind:r(169)})},function(t,n,r){"use strict";var e=r(5),i=r(32),o=r(7)("hasInstance"),u=Function.prototype;o in u||r(11).f(u,o,{value:function(t){if("function"!=typeof this||!e(t))return!1;if(!e(this.prototype))return t instanceof this;for(;t=i(t);)if(this.prototype===t)return!0;return!1}})},function(t,n,r){var e=r(11).f,i=Function.prototype,o=/^\s*function ([^ (]*)/;"name"in i||r(10)&&e(i,"name",{configurable:!0,get:function(){try{return(""+this).match(o)[1]}catch(t){return""}}})},function(t,n,r){var e=r(1),i=r(180),o=Math.sqrt,u=Math.acosh;e(e.S+e.F*!(u&&710==Math.floor(u(Number.MAX_VALUE))&&u(1/0)==1/0),"Math",{acosh:function(t){return(t=+t)<1?NaN:94906265.62425156<t?Math.log(t)+Math.LN2:i(t-1+o(t-1)*o(t+1))}})},function(t,n,r){var e=r(1),i=Math.asinh;e(e.S+e.F*!(i&&0<1/i(0)),"Math",{asinh:function t(n){return isFinite(n=+n)&&0!=n?n<0?-t(-n):Math.log(n+Math.sqrt(n*n+1)):n}})},function(t,n,r){var e=r(1),i=Math.atanh;e(e.S+e.F*!(i&&1/i(-0)<0),"Math",{atanh:function(t){return 0==(t=+t)?t:Math.log((1+t)/(1-t))/2}})},function(t,n,r){var e=r(1),i=r(149);e(e.S,"Math",{cbrt:function(t){return i(t=+t)*Math.pow(Math.abs(t),1/3)}})},function(t,n,r){var e=r(1);e(e.S,"Math",{clz32:function(t){return(t>>>=0)?31-Math.floor(Math.log(t+.5)*Math.LOG2E):32}})},function(t,n,r){var e=r(1),i=Math.exp;e(e.S,"Math",{cosh:function(t){return(i(t=+t)+i(-t))/2}})},function(t,n,r){var e=r(1),i=r(148);e(e.S+e.F*(i!=Math.expm1),"Math",{expm1:i})},function(t,n,r){var e=r(1);e(e.S,"Math",{fround:r(179)})},function(t,n,r){var e=r(1),f=Math.abs;e(e.S,"Math",{hypot:function(t,n){for(var r,e,i=0,o=0,u=arguments.length,c=0;o<u;)c<(r=f(arguments[o++]))?(i=i*(e=c/r)*e+1,c=r):0<r?i+=(e=r/c)*e:i+=r;return c===1/0?1/0:c*Math.sqrt(i)}})},function(t,n,r){var e=r(1),i=Math.imul;e(e.S+e.F*r(4)(function(){return-5!=i(4294967295,5)||2!=i.length}),"Math",{imul:function(t,n){var r=65535,e=+t,i=+n,o=r&e,u=r&i;return 0|o*u+((r&e>>>16)*u+o*(r&i>>>16)<<16>>>0)}})},function(t,n,r){var e=r(1);e(e.S,"Math",{log10:function(t){return Math.log(t)*Math.LOG10E}})},function(t,n,r){var e=r(1);e(e.S,"Math",{log1p:r(180)})},function(t,n,r){var e=r(1);e(e.S,"Math",{log2:function(t){return Math.log(t)/Math.LN2}})},function(t,n,r){var e=r(1);e(e.S,"Math",{sign:r(149)})},function(t,n,r){var e=r(1),i=r(148),o=Math.exp;e(e.S+e.F*r(4)(function(){return-2e-17!=!Math.sinh(-2e-17)}),"Math",{sinh:function(t){return Math.abs(t=+t)<1?(i(t)-i(-t))/2:(o(t-1)-o(-t-1))*(Math.E/2)}})},function(t,n,r){var e=r(1),i=r(148),o=Math.exp;e(e.S,"Math",{tanh:function(t){var n=i(t=+t),r=i(-t);return n==1/0?1:r==1/0?-1:(n-r)/(o(t)+o(-t))}})},function(t,n,r){var e=r(1);e(e.S,"Math",{trunc:function(t){return(0<t?Math.floor:Math.ceil)(t)}})},function(t,n,r){"use strict";var e=r(3),i=r(30),o=r(45),u=r(144),s=r(53),c=r(4),f=r(73).f,a=r(31).f,l=r(11).f,h=r(84).trim,v="Number",p=e[v],d=p,y=p.prototype,g=o(r(72)(y))==v,b="trim"in String.prototype,x=function(t){var n=s(t,!1);if("string"==typeof n&&2<n.length){var r,e,i,o=(n=b?n.trim():h(n,3)).charCodeAt(0);if(43===o||45===o){if(88===(r=n.charCodeAt(2))||120===r)return NaN}else if(48===o){switch(n.charCodeAt(1)){case 66:case 98:e=2,i=49;break;case 79:case 111:e=8,i=55;break;default:return+n}for(var u,c=n.slice(2),f=0,a=c.length;f<a;f++)if((u=c.charCodeAt(f))<48||i<u)return NaN;return parseInt(c,e)}}return+n};if(!p(" 0o1")||!p("0b1")||p("+0x1")){p=function(t){var n=arguments.length<1?0:t,r=this;return r instanceof p&&(g?c(function(){y.valueOf.call(r)}):o(r)!=v)?u(new d(x(n)),r,p):x(n)};for(var m,S=r(10)?f(d):"MAX_VALUE,MIN_VALUE,NaN,NEGATIVE_INFINITY,POSITIVE_INFINITY,EPSILON,isFinite,isInteger,isNaN,isSafeInteger,MAX_SAFE_INTEGER,MIN_SAFE_INTEGER,parseFloat,parseInt,isInteger".split(","),w=0;S.length>w;w++)i(d,m=S[w])&&!i(p,m)&&l(p,m,a(d,m));(p.prototype=y).constructor=p,r(27)(e,v,p)}},function(t,n,r){var e=r(1);e(e.S,"Number",{EPSILON:Math.pow(2,-52)})},function(t,n,r){var e=r(1),i=r(3).isFinite;e(e.S,"Number",{isFinite:function(t){return"number"==typeof t&&i(t)}})},function(t,n,r){var e=r(1);e(e.S,"Number",{isInteger:r(176)})},function(t,n,r){var e=r(1);e(e.S,"Number",{isNaN:function(t){return t!=t}})},function(t,n,r){var e=r(1),i=r(176),o=Math.abs;e(e.S,"Number",{isSafeInteger:function(t){return i(t)&&o(t)<=9007199254740991}})},function(t,n,r){var e=r(1);e(e.S,"Number",{MAX_SAFE_INTEGER:9007199254740991})},function(t,n,r){var e=r(1);e(e.S,"Number",{MIN_SAFE_INTEGER:-9007199254740991})},function(t,n,r){var e=r(1),i=r(188);e(e.S+e.F*(Number.parseFloat!=i),"Number",{parseFloat:i})},function(t,n,r){var e=r(1),i=r(189);e(e.S+e.F*(Number.parseInt!=i),"Number",{parseInt:i})},function(t,n,r){"use strict";var e=r(1),a=r(49),s=r(165),l=r(156),i=1..toFixed,o=Math.floor,u=[0,0,0,0,0,0],h="Number.toFixed: incorrect invocation!",v=function(t,n){for(var r=-1,e=n;++r<6;)e+=t*u[r],u[r]=e%1e7,e=o(e/1e7)},p=function(t){for(var n=6,r=0;0<=--n;)r+=u[n],u[n]=o(r/t),r=r%t*1e7},d=function(){for(var t=6,n="";0<=--t;)if(""!==n||0===t||0!==u[t]){var r=String(u[t]);n=""===n?r:n+l.call("0",7-r.length)+r}return n},y=function(t,n,r){return 0===n?r:n%2==1?y(t,n-1,r*t):y(t*t,n/2,r)};e(e.P+e.F*(!!i&&("0.000"!==8e-5.toFixed(3)||"1"!==.9.toFixed(0)||"1.25"!==1.255.toFixed(2)||"1000000000000000128"!==(0xde0b6b3a7640080).toFixed(0))||!r(4)(function(){i.call({})})),"Number",{toFixed:function(t){var n,r,e,i,o=s(this,h),u=a(t),c="",f="0";if(u<0||20<u)throw RangeError(h);if(o!=o)return"NaN";if(o<=-1e21||1e21<=o)return String(o);if(o<0&&(c="-",o=-o),1e-21<o)if(r=(n=function(t){for(var n=0,r=t;4096<=r;)n+=12,r/=4096;for(;2<=r;)n+=1,r/=2;return n}(o*y(2,69,1))-69)<0?o*y(2,-n,1):o/y(2,n,1),r*=4503599627370496,0<(n=52-n)){for(v(0,r),e=u;7<=e;)v(1e7,0),e-=7;for(v(y(10,e,1),0),e=n-1;23<=e;)p(1<<23),e-=23;p(1<<e),v(1,1),p(2),f=d()}else v(0,r),v(1<<-n,0),f=d()+l.call("0",u);return f=0<u?c+((i=f.length)<=u?"0."+l.call("0",u-i)+f:f.slice(0,i-u)+"."+f.slice(i-u)):c+f}})},function(t,n,r){"use strict";var e=r(1),i=r(4),o=r(165),u=1..toPrecision;e(e.P+e.F*(i(function(){return"1"!==u.call(1,void 0)})||!i(function(){u.call({})})),"Number",{toPrecision:function(t){var n=o(this,"Number#toPrecision: incorrect invocation!");return void 0===t?u.call(n):u.call(n,t)}})},function(t,n,r){var e=r(1);e(e.S+e.F,"Object",{assign:r(182)})},function(t,n,r){var e=r(1);e(e.S,"Object",{create:r(72)})},function(t,n,r){var e=r(1);e(e.S+e.F*!r(10),"Object",{defineProperties:r(183)})},function(t,n,r){var e=r(1);e(e.S+e.F*!r(10),"Object",{defineProperty:r(11).f})},function(t,n,r){var e=r(5),i=r(69).onFreeze;r(52)("freeze",function(n){return function(t){return n&&e(t)?n(i(t)):t}})},function(t,n,r){var e=r(33),i=r(31).f;r(52)("getOwnPropertyDescriptor",function(){return function(t,n){return i(e(t),n)}})},function(t,n,r){r(52)("getOwnPropertyNames",function(){return r(184).f})},function(t,n,r){var e=r(17),i=r(32);r(52)("getPrototypeOf",function(){return function(t){return i(e(t))}})},function(t,n,r){var e=r(5);r(52)("isExtensible",function(n){return function(t){return!!e(t)&&(!n||n(t))}})},function(t,n,r){var e=r(5);r(52)("isFrozen",function(n){return function(t){return!e(t)||!!n&&n(t)}})},function(t,n,r){var e=r(5);r(52)("isSealed",function(n){return function(t){return!e(t)||!!n&&n(t)}})},function(t,n,r){var e=r(1);e(e.S,"Object",{is:r(192)})},function(t,n,r){var e=r(17),i=r(74);r(52)("keys",function(){return function(t){return i(e(t))}})},function(t,n,r){var e=r(5),i=r(69).onFreeze;r(52)("preventExtensions",function(n){return function(t){return n&&e(t)?n(i(t)):t}})},function(t,n,r){var e=r(5),i=r(69).onFreeze;r(52)("seal",function(n){return function(t){return n&&e(t)?n(i(t)):t}})},function(t,n,r){var e=r(1);e(e.S,"Object",{setPrototypeOf:r(153).set})},function(t,n,r){"use strict";var e=r(81),i={};i[r(7)("toStringTag")]="z",i+""!="[object z]"&&r(27)(Object.prototype,"toString",function(){return"[object "+e(this)+"]"},!0)},function(t,n,r){var e=r(1),i=r(188);e(e.G+e.F*(parseFloat!=i),{parseFloat:i})},function(t,n,r){var e=r(1),i=r(189);e(e.G+e.F*(parseInt!=i),{parseInt:i})},function(t,n,r){"use strict";var e,i,o,u,c=r(68),f=r(3),a=r(47),s=r(81),l=r(1),h=r(5),v=r(21),p=r(70),d=r(71),y=r(119),g=r(158).set,b=r(150)(),x=r(151),m=r(190),S=r(133),w=r(191),_="Promise",O=f.TypeError,E=f.process,M=E&&E.versions,P=M&&M.v8||"",j=f[_],F="process"==s(E),A=function(){},I=i=x.f,L=!!function(){try{var t=j.resolve(1),n=(t.constructor={})[r(7)("species")]=function(t){t(A,A)};return(F||"function"==typeof PromiseRejectionEvent)&&t.then(A)instanceof n&&0!==P.indexOf("6.6")&&-1===S.indexOf("Chrome/66")}catch(t){}}(),N=function(t){var n;return!(!h(t)||"function"!=typeof(n=t.then))&&n},T=function(s,r){if(!s._n){s._n=!0;var e=s._c;b(function(){for(var f=s._v,a=1==s._s,t=0,n=function(t){var n,r,e,i=a?t.ok:t.fail,o=t.resolve,u=t.reject,c=t.domain;try{i?(a||(2==s._h&&C(s),s._h=1),!0===i?n=f:(c&&c.enter(),n=i(f),c&&(c.exit(),e=!0)),n===t.promise?u(O("Promise-chain cycle")):(r=N(n))?r.call(n,o,u):o(n)):u(f)}catch(t){c&&!e&&c.exit(),u(t)}};e.length>t;)n(e[t++]);s._c=[],s._n=!1,r&&!s._h&&k(s)})}},k=function(o){g.call(f,function(){var t,n,r,e=o._v,i=R(o);if(i&&(t=m(function(){F?E.emit("unhandledRejection",e,o):(n=f.onunhandledrejection)?n({promise:o,reason:e}):(r=f.console)&&r.error&&r.error("Unhandled promise rejection",e)}),o._h=F||R(o)?2:1),o._a=void 0,i&&t.e)throw t.v})},R=function(t){return 1!==t._h&&0===(t._a||t._c).length},C=function(n){g.call(f,function(){var t;F?E.emit("rejectionHandled",n):(t=f.onrejectionhandled)&&t({promise:n,reason:n._v})})},D=function(t){var n=this;n._d||(n._d=!0,(n=n._w||n)._v=t,n._s=2,n._a||(n._a=n._c.slice()),T(n,!0))},G=function(t){var r,e=this;if(!e._d){e._d=!0,e=e._w||e;try{if(e===t)throw O("Promise can't be resolved itself");(r=N(t))?b(function(){var n={_w:e,_d:!1};try{r.call(t,a(G,n,1),a(D,n,1))}catch(t){D.call(n,t)}}):(e._v=t,e._s=1,T(e,!1))}catch(t){D.call({_w:e,_d:!1},t)}}};L||(j=function(t){p(this,j,_,"_h"),v(t),e.call(this);try{t(a(G,this,1),a(D,this,1))}catch(t){D.call(this,t)}},(e=function(t){this._c=[],this._a=void 0,this._s=0,this._d=!1,this._v=void 0,this._h=0,this._n=!1}).prototype=r(76)(j.prototype,{then:function(t,n){var r=I(y(this,j));return r.ok="function"!=typeof t||t,r.fail="function"==typeof n&&n,r.domain=F?E.domain:void 0,this._c.push(r),this._a&&this._a.push(r),this._s&&T(this,!1),r.promise},catch:function(t){return this.then(void 0,t)}}),o=function(){var t=new e;this.promise=t,this.resolve=a(G,t,1),this.reject=a(D,t,1)},x.f=I=function(t){return t===j||t===u?new o(t):i(t)}),l(l.G+l.W+l.F*!L,{Promise:j}),r(83)(j,_),r(77)(_),u=r(46)[_],l(l.S+l.F*!L,_,{reject:function(t){var n=I(this);return(0,n.reject)(t),n.promise}}),l(l.S+l.F*(c||!L),_,{resolve:function(t){return w(c&&this===u?j:this,t)}}),l(l.S+l.F*!(L&&r(125)(function(t){j.all(t).catch(A)})),_,{all:function(t){var u=this,n=I(u),c=n.resolve,f=n.reject,r=m(function(){var e=[],i=0,o=1;d(t,!1,function(t){var n=i++,r=!1;e.push(void 0),o++,u.resolve(t).then(function(t){r||(r=!0,e[n]=t,--o||c(e))},f)}),--o||c(e)});return r.e&&f(r.v),n.promise},race:function(t){var n=this,r=I(n),e=r.reject,i=m(function(){d(t,!1,function(t){n.resolve(t).then(r.resolve,e)})});return i.e&&e(i.v),r.promise}})},function(t,n,r){var e=r(1),o=r(21),u=r(2),c=(r(3).Reflect||{}).apply,f=Function.apply;e(e.S+e.F*!r(4)(function(){c(function(){})}),"Reflect",{apply:function(t,n,r){var e=o(t),i=u(r);return c?c(e,n,i):f.call(e,n,i)}})},function(t,n,r){var e=r(1),c=r(72),f=r(21),a=r(2),s=r(5),i=r(4),l=r(169),h=(r(3).Reflect||{}).construct,v=i(function(){function t(){}return!(h(function(){},[],t)instanceof t)}),p=!i(function(){h(function(){})});e(e.S+e.F*(v||p),"Reflect",{construct:function(t,n){f(t),a(n);var r=arguments.length<3?t:f(arguments[2]);if(p&&!v)return h(t,n,r);if(t==r){switch(n.length){case 0:return new t;case 1:return new t(n[0]);case 2:return new t(n[0],n[1]);case 3:return new t(n[0],n[1],n[2]);case 4:return new t(n[0],n[1],n[2],n[3])}var e=[null];return e.push.apply(e,n),new(l.apply(t,e))}var i=r.prototype,o=c(s(i)?i:Object.prototype),u=Function.apply.call(t,o,n);return s(u)?u:o}})},function(t,n,r){var e=r(11),i=r(1),o=r(2),u=r(53);i(i.S+i.F*r(4)(function(){Reflect.defineProperty(e.f({},1,{value:1}),1,{value:2})}),"Reflect",{defineProperty:function(t,n,r){o(t),n=u(n,!0),o(r);try{return e.f(t,n,r),!0}catch(t){return!1}}})},function(t,n,r){var e=r(1),i=r(31).f,o=r(2);e(e.S,"Reflect",{deleteProperty:function(t,n){var r=i(o(t),n);return!(r&&!r.configurable)&&delete t[n]}})},function(t,n,r){"use strict";var e=r(1),i=r(2),o=function(t){this._t=i(t),this._i=0;var n,r=this._k=[];for(n in t)r.push(n)};r(146)(o,"Object",function(){var t,n=this._k;do{if(this._i>=n.length)return{value:void 0,done:!0}}while(!((t=n[this._i++])in this._t));return{value:t,done:!1}}),e(e.S,"Reflect",{enumerate:function(t){return new o(t)}})},function(t,n,r){var e=r(31),i=r(1),o=r(2);i(i.S,"Reflect",{getOwnPropertyDescriptor:function(t,n){return e.f(o(t),n)}})},function(t,n,r){var e=r(1),i=r(32),o=r(2);e(e.S,"Reflect",{getPrototypeOf:function(t){return i(o(t))}})},function(t,n,r){var u=r(31),c=r(32),f=r(30),e=r(1),a=r(5),s=r(2);e(e.S,"Reflect",{get:function t(n,r){var e,i,o=arguments.length<3?n:arguments[2];return s(n)===o?n[r]:(e=u.f(n,r))?f(e,"value")?e.value:void 0!==e.get?e.get.call(o):void 0:a(i=c(n))?t(i,r,o):void 0}})},function(t,n,r){var e=r(1);e(e.S,"Reflect",{has:function(t,n){return n in t}})},function(t,n,r){var e=r(1),i=r(2),o=Object.isExtensible;e(e.S,"Reflect",{isExtensible:function(t){return i(t),!o||o(t)}})},function(t,n,r){var e=r(1);e(e.S,"Reflect",{ownKeys:r(187)})},function(t,n,r){var e=r(1),i=r(2),o=Object.preventExtensions;e(e.S,"Reflect",{preventExtensions:function(t){i(t);try{return o&&o(t),!0}catch(t){return!1}}})},function(t,n,r){var e=r(1),i=r(153);i&&e(e.S,"Reflect",{setPrototypeOf:function(t,n){i.check(t,n);try{return i.set(t,n),!0}catch(t){return!1}}})},function(t,n,r){var f=r(11),a=r(31),s=r(32),l=r(30),e=r(1),h=r(75),v=r(2),p=r(5);e(e.S,"Reflect",{set:function t(n,r,e){var i,o,u=arguments.length<4?n:arguments[3],c=a.f(v(n),r);if(!c){if(p(o=s(n)))return t(o,r,e,u);c=h(0)}if(l(c,"value")){if(!1===c.writable||!p(u))return!1;if(i=a.f(u,r)){if(i.get||i.set||!1===i.writable)return!1;i.value=e,f.f(u,r,i)}else f.f(u,r,h(0,e));return!0}return void 0!==c.set&&(c.set.call(u,e),!0)}})},function(t,n,r){var e=r(3),o=r(144),i=r(11).f,u=r(73).f,c=r(124),f=r(115),a=e.RegExp,s=a,l=a.prototype,h=/a/g,v=/a/g,p=new a(h)!==h;if(r(10)&&(!p||r(4)(function(){return v[r(7)("match")]=!1,a(h)!=h||a(v)==v||"/a/i"!=a(h,"i")}))){a=function(t,n){var r=this instanceof a,e=c(t),i=void 0===n;return!r&&e&&t.constructor===a&&i?t:o(p?new s(e&&!i?t.source:t,n):s((e=t instanceof a)?t.source:t,e&&i?f.call(t):n),r?this:l,a)};for(var d=function(n){n in a||i(a,n,{configurable:!0,get:function(){return s[n]},set:function(t){s[n]=t}})},y=u(s),g=0;y.length>g;)d(y[g++]);(l.constructor=a).prototype=l,r(27)(e,"RegExp",a)}r(77)("RegExp")},function(t,n,r){"use strict";var l=r(2),h=r(8),v=r(136),p=r(128);r(122)("match",1,function(e,i,a,s){return[function(t){var n=e(this),r=null==t?void 0:t[i];return void 0!==r?r.call(t,n):new RegExp(t)[i](String(n))},function(t){var n=s(a,t,this);if(n.done)return n.value;var r=l(t),e=String(this);if(!r.global)return p(r,e);for(var i,o=r.unicode,u=[],c=r.lastIndex=0;null!==(i=p(r,e));){var f=String(i[0]);""===(u[c]=f)&&(r.lastIndex=v(e,h(r.lastIndex),o)),c++}return 0===c?null:u}]})},function(t,n,r){"use strict";var O=r(2),e=r(17),E=r(8),M=r(49),P=r(136),j=r(128),F=Math.max,A=Math.min,h=Math.floor,v=/\$([$&`']|\d\d?|<[^>]*>)/g,p=/\$([$&`']|\d\d?)/g;r(122)("replace",2,function(i,o,S,w){function _(o,u,c,f,a,t){var s=c+o.length,l=f.length,n=p;return void 0!==a&&(a=e(a),n=v),S.call(t,n,function(t,n){var r;switch(n.charAt(0)){case"$":return"$";case"&":return o;case"`":return u.slice(0,c);case"'":return u.slice(s);case"<":r=a[n.slice(1,-1)];break;default:var e=+n;if(0===e)return t;if(l<e){var i=h(e/10);return 0===i?t:i<=l?void 0===f[i-1]?n.charAt(1):f[i-1]+n.charAt(1):t}r=f[e-1]}return void 0===r?"":r})}return[function(t,n){var r=i(this),e=null==t?void 0:t[o];return void 0!==e?e.call(t,r,n):S.call(String(r),t,n)},function(t,n){var r=w(S,t,this,n);if(r.done)return r.value;var e=O(t),i=String(this),o="function"==typeof n;o||(n=String(n));var u,c=e.global;if(c){var f=e.unicode;e.lastIndex=0}for(var a=[];;){var s=j(e,i);if(null===s)break;if(a.push(s),!c)break;""===String(s[0])&&(e.lastIndex=P(i,E(e.lastIndex),f))}for(var l="",h=0,v=0;v<a.length;v++){s=a[v];for(var p=String(s[0]),d=F(A(M(s.index),i.length),0),y=[],g=1;g<s.length;g++)y.push(void 0===(u=s[g])?u:String(u));var b=s.groups;if(o){var x=[p].concat(y,d,i);void 0!==b&&x.push(b);var m=String(n.apply(void 0,x))}else m=_(p,i,d,y,b,n);h<=d&&(l+=i.slice(h,d)+m,h=d+p.length)}return l+i.slice(h)}]})},function(t,n,r){"use strict";var f=r(2),a=r(192),s=r(128);r(122)("search",1,function(e,i,u,c){return[function(t){var n=e(this),r=null==t?void 0:t[i];return void 0!==r?r.call(t,n):new RegExp(t)[i](String(n))},function(t){var n=c(u,t,this);if(n.done)return n.value;var r=f(t),e=String(this),i=r.lastIndex;a(i,0)||(r.lastIndex=0);var o=s(r,e);return a(r.lastIndex,i)||(r.lastIndex=i),null===o?-1:o.index}]})},function(t,n,r){"use strict";var l=r(124),x=r(2),m=r(119),S=r(136),w=r(8),_=r(128),h=r(152),e=r(4),O=Math.min,v=[].push,u="split",p="length",d="lastIndex",E=4294967295,M=!e(function(){RegExp(E,"y")});r(122)("split",2,function(i,o,y,g){var b;return b="c"=="abbc"[u](/(b)*/)[1]||4!="test"[u](/(?:)/,-1)[p]||2!="ab"[u](/(?:ab)*/)[p]||4!="."[u](/(.?)(.?)/)[p]||1<"."[u](/()()/)[p]||""[u](/.?/)[p]?function(t,n){var r=String(this);if(void 0===t&&0===n)return[];if(!l(t))return y.call(r,t,n);for(var e,i,o,u=[],c=(t.ignoreCase?"i":"")+(t.multiline?"m":"")+(t.unicode?"u":"")+(t.sticky?"y":""),f=0,a=void 0===n?E:n>>>0,s=new RegExp(t.source,c+"g");(e=h.call(s,r))&&!(f<(i=s[d])&&(u.push(r.slice(f,e.index)),1<e[p]&&e.index<r[p]&&v.apply(u,e.slice(1)),o=e[0][p],f=i,u[p]>=a));)s[d]===e.index&&s[d]++;return f===r[p]?!o&&s.test("")||u.push(""):u.push(r.slice(f)),u[p]>a?u.slice(0,a):u}:"0"[u](void 0,0)[p]?function(t,n){return void 0===t&&0===n?[]:y.call(this,t,n)}:y,[function(t,n){var r=i(this),e=null==t?void 0:t[o];return void 0!==e?e.call(t,r,n):b.call(String(r),t,n)},function(t,n){var r=g(b,t,this,n,b!==y);if(r.done)return r.value;var e=x(t),i=String(this),o=m(e,RegExp),u=e.unicode,c=(e.ignoreCase?"i":"")+(e.multiline?"m":"")+(e.unicode?"u":"")+(M?"y":"g"),f=new o(M?e:"^(?:"+e.source+")",c),a=void 0===n?E:n>>>0;if(0===a)return[];if(0===i.length)return null===_(f,i)?[i]:[];for(var s=0,l=0,h=[];l<i.length;){f.lastIndex=M?l:0;var v,p=_(f,M?i:i.slice(l));if(null===p||(v=O(w(f.lastIndex+(M?0:l)),i.length))===s)l=S(i,l,u);else{if(h.push(i.slice(s,l)),h.length===a)return h;for(var d=1;d<=p.length-1;d++)if(h.push(p[d]),h.length===a)return h;l=s=v}}return h.push(i.slice(s)),h}]})},function(t,n,r){"use strict";r(198);var e=r(2),i=r(115),o=r(10),u="toString",c=/./[u],f=function(t){r(27)(RegExp.prototype,u,t,!0)};r(4)(function(){return"/a/b"!=c.call({source:"a",flags:"b"})})?f(function(){var t=e(this);return"/".concat(t.source,"/","flags"in t?t.flags:!o&&t instanceof RegExp?i.call(t):void 0)}):c.name!=u&&f(function(){return c.call(this)})},function(t,n,r){"use strict";r(28)("anchor",function(n){return function(t){return n(this,"a","name",t)}})},function(t,n,r){"use strict";r(28)("big",function(t){return function(){return t(this,"big","","")}})},function(t,n,r){"use strict";r(28)("blink",function(t){return function(){return t(this,"blink","","")}})},function(t,n,r){"use strict";r(28)("bold",function(t){return function(){return t(this,"b","","")}})},function(t,n,r){"use strict";var e=r(1),i=r(131)(!1);e(e.P,"String",{codePointAt:function(t){return i(this,t)}})},function(t,n,r){"use strict";var e=r(1),u=r(8),c=r(155),f="endsWith",a=""[f];e(e.P+e.F*r(142)(f),"String",{endsWith:function(t){var n=c(this,t,f),r=1<arguments.length?arguments[1]:void 0,e=u(n.length),i=void 0===r?e:Math.min(u(r),e),o=String(t);return a?a.call(n,o,i):n.slice(i-o.length,i)===o}})},function(t,n,r){"use strict";r(28)("fixed",function(t){return function(){return t(this,"tt","","")}})},function(t,n,r){"use strict";r(28)("fontcolor",function(n){return function(t){return n(this,"font","color",t)}})},function(t,n,r){"use strict";r(28)("fontsize",function(n){return function(t){return n(this,"font","size",t)}})},function(t,n,r){var e=r(1),o=r(78),u=String.fromCharCode,i=String.fromCodePoint;e(e.S+e.F*(!!i&&1!=i.length),"String",{fromCodePoint:function(t){for(var n,r=[],e=arguments.length,i=0;i<e;){if(n=+arguments[i++],o(n,1114111)!==n)throw RangeError(n+" is not a valid code point");r.push(n<65536?u(n):u(55296+((n-=65536)>>10),n%1024+56320))}return r.join("")}})},function(t,n,r){"use strict";var e=r(1),i=r(155);e(e.P+e.F*r(142)("includes"),"String",{includes:function(t){return!!~i(this,t,"includes").indexOf(t,1<arguments.length?arguments[1]:void 0)}})},function(t,n,r){"use strict";r(28)("italics",function(t){return function(){return t(this,"i","","")}})},function(t,n,r){"use strict";var e=r(131)(!0);r(147)(String,"String",function(t){this._t=String(t),this._i=0},function(){var t,n=this._t,r=this._i;return r>=n.length?{value:void 0,done:!0}:(t=e(n,r),this._i+=t.length,{value:t,done:!1})})},function(t,n,r){"use strict";r(28)("link",function(n){return function(t){return n(this,"a","href",t)}})},function(t,n,r){var e=r(1),u=r(33),c=r(8);e(e.S,"String",{raw:function(t){for(var n=u(t.raw),r=c(n.length),e=arguments.length,i=[],o=0;o<r;)i.push(String(n[o++])),o<e&&i.push(String(arguments[o]));return i.join("")}})},function(t,n,r){var e=r(1);e(e.P,"String",{repeat:r(156)})},function(t,n,r){"use strict";r(28)("small",function(t){return function(){return t(this,"small","","")}})},function(t,n,r){"use strict";var e=r(1),i=r(8),o=r(155),u="startsWith",c=""[u];e(e.P+e.F*r(142)(u),"String",{startsWith:function(t){var n=o(this,t,u),r=i(Math.min(1<arguments.length?arguments[1]:void 0,n.length)),e=String(t);return c?c.call(n,e,r):n.slice(r,r+e.length)===e}})},function(t,n,r){"use strict";r(28)("strike",function(t){return function(){return t(this,"strike","","")}})},function(t,n,r){"use strict";r(28)("sub",function(t){return function(){return t(this,"sub","","")}})},function(t,n,r){"use strict";r(28)("sup",function(t){return function(){return t(this,"sup","","")}})},function(t,n,r){"use strict";r(84)("trim",function(t){return function(){return t(this,3)}})},function(t,n,r){"use strict";var e=r(3),u=r(30),i=r(10),o=r(1),c=r(27),f=r(69).KEY,a=r(4),s=r(118),l=r(83),h=r(79),v=r(7),p=r(195),d=r(160),y=r(218),g=r(123),b=r(2),x=r(5),m=r(17),S=r(33),w=r(53),_=r(75),O=r(72),E=r(184),M=r(31),P=r(127),j=r(11),F=r(74),A=M.f,I=j.f,L=E.f,N=e.Symbol,T=e.JSON,k=T&&T.stringify,R="prototype",C=v("_hidden"),D=v("toPrimitive"),G={}.propertyIsEnumerable,W=s("symbol-registry"),U=s("symbols"),V=s("op-symbols"),B=Object[R],q="function"==typeof N&&!!P.f,z=e.QObject,K=!z||!z[R]||!z[R].findChild,H=i&&a(function(){return 7!=O(I({},"a",{get:function(){return I(this,"a",{value:7}).a}})).a})?function(t,n,r){var e=A(B,n);e&&delete B[n],I(t,n,r),e&&t!==B&&I(B,n,e)}:I,J=function(t){var n=U[t]=O(N[R]);return n._k=t,n},$=q&&"symbol"==typeof N.iterator?function(t){return"symbol"==typeof t}:function(t){return t instanceof N},Y=function(t,n,r){return t===B&&Y(V,n,r),b(t),n=w(n,!0),b(r),u(U,n)?(r.enumerable?(u(t,C)&&t[C][n]&&(t[C][n]=!1),r=O(r,{enumerable:_(0,!1)})):(u(t,C)||I(t,C,_(1,{})),t[C][n]=!0),H(t,n,r)):I(t,n,r)},X=function(t,n){b(t);for(var r,e=y(n=S(n)),i=0,o=e.length;i<o;)Y(t,r=e[i++],n[r]);return t},Q=function(t){var n=G.call(this,t=w(t,!0));return!(this===B&&u(U,t)&&!u(V,t))&&(!(n||!u(this,t)||!u(U,t)||u(this,C)&&this[C][t])||n)},Z=function(t,n){if(t=S(t),n=w(n,!0),t!==B||!u(U,n)||u(V,n)){var r=A(t,n);return!r||!u(U,n)||u(t,C)&&t[C][n]||(r.enumerable=!0),r}},tt=function(t){for(var n,r=L(S(t)),e=[],i=0;r.length>i;)u(U,n=r[i++])||n==C||n==f||e.push(n);return e},nt=function(t){for(var n,r=t===B,e=L(r?V:S(t)),i=[],o=0;e.length>o;)!u(U,n=e[o++])||r&&!u(B,n)||i.push(U[n]);return i};q||(c((N=function(){if(this instanceof N)throw TypeError("Symbol is not a constructor!");var n=h(0<arguments.length?arguments[0]:void 0),r=function(t){this===B&&r.call(V,t),u(this,C)&&u(this[C],n)&&(this[C][n]=!1),H(this,n,_(1,t))};return i&&K&&H(B,n,{configurable:!0,set:r}),J(n)})[R],"toString",function(){return this._k}),M.f=Z,j.f=Y,r(73).f=E.f=tt,r(117).f=Q,P.f=nt,i&&!r(68)&&c(B,"propertyIsEnumerable",Q,!0),p.f=function(t){return J(v(t))}),o(o.G+o.W+o.F*!q,{Symbol:N});for(var rt="hasInstance,isConcatSpreadable,iterator,match,replace,search,species,split,toPrimitive,toStringTag,unscopables".split(","),et=0;rt.length>et;)v(rt[et++]);for(var it=F(v.store),ot=0;it.length>ot;)d(it[ot++]);o(o.S+o.F*!q,"Symbol",{for:function(t){return u(W,t+="")?W[t]:W[t]=N(t)},keyFor:function(t){if(!$(t))throw TypeError(t+" is not a symbol!");for(var n in W)if(W[n]===t)return n},useSetter:function(){K=!0},useSimple:function(){K=!1}}),o(o.S+o.F*!q,"Object",{create:function(t,n){return void 0===n?O(t):X(O(t),n)},defineProperty:Y,defineProperties:X,getOwnPropertyDescriptor:Z,getOwnPropertyNames:tt,getOwnPropertySymbols:nt});var ut=a(function(){P.f(1)});o(o.S+o.F*ut,"Object",{getOwnPropertySymbols:function(t){return P.f(m(t))}}),T&&o(o.S+o.F*(!q||a(function(){var t=N();return"[null]"!=k([t])||"{}"!=k({a:t})||"{}"!=k(Object(t))})),"JSON",{stringify:function(t){for(var n,r,e=[t],i=1;arguments.length>i;)e.push(arguments[i++]);if(r=n=e[1],(x(n)||void 0!==t)&&!$(t))return g(n)||(n=function(t,n){if("function"==typeof r&&(n=r.call(this,t,n)),!$(n))return n}),e[1]=n,k.apply(T,e)}}),N[R][D]||r(26)(N[R],D,N[R].valueOf),l(N,"Symbol"),l(Math,"Math",!0),l(e.JSON,"JSON",!0)},function(t,n,r){"use strict";var e=r(1),i=r(132),o=r(159),a=r(2),s=r(78),l=r(8),u=r(5),c=r(3).ArrayBuffer,h=r(119),v=o.ArrayBuffer,p=o.DataView,f=i.ABV&&c.isView,d=v.prototype.slice,y=i.VIEW,g="ArrayBuffer";e(e.G+e.W+e.F*(c!==v),{ArrayBuffer:v}),e(e.S+e.F*!i.CONSTR,g,{isView:function(t){return f&&f(t)||u(t)&&y in t}}),e(e.P+e.U+e.F*r(4)(function(){return!new v(2).slice(1,void 0).byteLength}),g,{slice:function(t,n){if(void 0!==d&&void 0===n)return d.call(a(this),t);for(var r=a(this).byteLength,e=s(t,r),i=s(void 0===n?r:n,r),o=new(h(this,v))(l(i-e)),u=new p(this),c=new p(o),f=0;e<i;)c.setUint8(f++,u.getUint8(e++));return o}}),r(77)(g)},function(t,n,r){var e=r(1);e(e.G+e.W+e.F*!r(132).ABV,{DataView:r(159).DataView})},function(t,n,r){r(57)("Float32",4,function(e){return function(t,n,r){return e(this,t,n,r)}})},function(t,n,r){r(57)("Float64",8,function(e){return function(t,n,r){return e(this,t,n,r)}})},function(t,n,r){r(57)("Int16",2,function(e){return function(t,n,r){return e(this,t,n,r)}})},function(t,n,r){r(57)("Int32",4,function(e){return function(t,n,r){return e(this,t,n,r)}})},function(t,n,r){r(57)("Int8",1,function(e){return function(t,n,r){return e(this,t,n,r)}})},function(t,n,r){r(57)("Uint16",2,function(e){return function(t,n,r){return e(this,t,n,r)}})},function(t,n,r){r(57)("Uint32",4,function(e){return function(t,n,r){return e(this,t,n,r)}})},function(t,n,r){r(57)("Uint8",1,function(e){return function(t,n,r){return e(this,t,n,r)}})},function(t,n,r){r(57)("Uint8",1,function(e){return function(t,n,r){return e(this,t,n,r)}},!0)},function(t,n,r){"use strict";var e=r(172),i=r(80);r(121)("WeakSet",function(t){return function(){return t(this,0<arguments.length?arguments[0]:void 0)}},{add:function(t){return e.def(i(this,"WeakSet"),t,!0)}},e,!1,!0)},function(t,n,r){"use strict";var e=r(1),i=r(173),o=r(17),u=r(8),c=r(21),f=r(138);e(e.P,"Array",{flatMap:function(t){var n,r,e=o(this);return c(t),n=u(e.length),r=f(e,0),i(r,e,e,n,0,1,t,arguments[1]),r}}),r(67)("flatMap")},function(t,n,r){"use strict";var e=r(1),i=r(173),o=r(17),u=r(8),c=r(49),f=r(138);e(e.P,"Array",{flatten:function(){var t=arguments[0],n=o(this),r=u(n.length),e=f(n,0);return i(e,n,n,r,0,void 0===t?1:c(t)),e}}),r(67)("flatten")},function(t,n,r){"use strict";var e=r(1),i=r(120)(!0);e(e.P,"Array",{includes:function(t){return i(this,t,1<arguments.length?arguments[1]:void 0)}}),r(67)("includes")},function(t,n,r){var e=r(1),i=r(150)(),o=r(3).process,u="process"==r(45)(o);e(e.G,{asap:function(t){var n=u&&o.domain;i(n?n.bind(t):t)}})},function(t,n,r){var e=r(1),i=r(45);e(e.S,"Error",{isError:function(t){return"Error"===i(t)}})},function(t,n,r){var e=r(1);e(e.G,{global:r(3)})},function(t,n,r){r(129)("Map")},function(t,n,r){r(130)("Map")},function(t,n,r){var e=r(1);e(e.P+e.R,"Map",{toJSON:r(171)("Map")})},function(t,n,r){var e=r(1);e(e.S,"Math",{clamp:function(t,n,r){return Math.min(r,Math.max(n,t))}})},function(t,n,r){var e=r(1);e(e.S,"Math",{DEG_PER_RAD:Math.PI/180})},function(t,n,r){var e=r(1),i=180/Math.PI;e(e.S,"Math",{degrees:function(t){return t*i}})},function(t,n,r){var e=r(1),o=r(181),u=r(179);e(e.S,"Math",{fscale:function(t,n,r,e,i){return u(o(t,n,r,e,i))}})},function(t,n,r){var e=r(1);e(e.S,"Math",{iaddh:function(t,n,r,e){var i=t>>>0,o=r>>>0;return(n>>>0)+(e>>>0)+((i&o|(i|o)&~(i+o>>>0))>>>31)|0}})},function(t,n,r){var e=r(1);e(e.S,"Math",{imulh:function(t,n){var r=+t,e=+n,i=65535&r,o=65535&e,u=r>>16,c=e>>16,f=(u*o>>>0)+(i*o>>>16);return u*c+(f>>16)+((i*c>>>0)+(65535&f)>>16)}})},function(t,n,r){var e=r(1);e(e.S,"Math",{isubh:function(t,n,r,e){var i=t>>>0,o=r>>>0;return(n>>>0)-(e>>>0)-((~i&o|~(i^o)&i-o>>>0)>>>31)|0}})},function(t,n,r){var e=r(1);e(e.S,"Math",{RAD_PER_DEG:180/Math.PI})},function(t,n,r){var e=r(1),i=Math.PI/180;e(e.S,"Math",{radians:function(t){return t*i}})},function(t,n,r){var e=r(1);e(e.S,"Math",{scale:r(181)})},function(t,n,r){var e=r(1);e(e.S,"Math",{signbit:function(t){return(t=+t)!=t?t:0==t?1/t==1/0:0<t}})},function(t,n,r){var e=r(1);e(e.S,"Math",{umulh:function(t,n){var r=+t,e=+n,i=65535&r,o=65535&e,u=r>>>16,c=e>>>16,f=(u*o>>>0)+(i*o>>>16);return u*c+(f>>>16)+((i*c>>>0)+(65535&f)>>>16)}})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(21),u=r(11);r(10)&&e(e.P+r(126),"Object",{__defineGetter__:function(t,n){u.f(i(this),t,{get:o(n),enumerable:!0,configurable:!0})}})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(21),u=r(11);r(10)&&e(e.P+r(126),"Object",{__defineSetter__:function(t,n){u.f(i(this),t,{set:o(n),enumerable:!0,configurable:!0})}})},function(t,n,r){var e=r(1),i=r(186)(!0);e(e.S,"Object",{entries:function(t){return i(t)}})},function(t,n,r){var e=r(1),f=r(187),a=r(33),s=r(31),l=r(139);e(e.S,"Object",{getOwnPropertyDescriptors:function(t){for(var n,r,e=a(t),i=s.f,o=f(e),u={},c=0;o.length>c;)void 0!==(r=i(e,n=o[c++]))&&l(u,n,r);return u}})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(53),u=r(32),c=r(31).f;r(10)&&e(e.P+r(126),"Object",{__lookupGetter__:function(t){var n,r=i(this),e=o(t,!0);do{if(n=c(r,e))return n.get}while(r=u(r))}})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(53),u=r(32),c=r(31).f;r(10)&&e(e.P+r(126),"Object",{__lookupSetter__:function(t){var n,r=i(this),e=o(t,!0);do{if(n=c(r,e))return n.set}while(r=u(r))}})},function(t,n,r){var e=r(1),i=r(186)(!1);e(e.S,"Object",{values:function(t){return i(t)}})},function(t,n,r){"use strict";var e=r(1),o=r(3),u=r(46),i=r(150)(),c=r(7)("observable"),f=r(21),a=r(2),s=r(70),l=r(76),h=r(26),v=r(71),p=v.RETURN,d=function(t){return null==t?void 0:f(t)},y=function(t){var n=t._c;n&&(t._c=void 0,n())},g=function(t){return void 0===t._o},b=function(t){g(t)||(t._o=void 0,y(t))},x=function(t,n){a(t),this._c=void 0,this._o=t,t=new m(this);try{var r=n(t),e=r;null!=r&&("function"==typeof r.unsubscribe?r=function(){e.unsubscribe()}:f(r),this._c=r)}catch(n){return void t.error(n)}g(this)&&y(this)};x.prototype=l({},{unsubscribe:function(){b(this)}});var m=function(t){this._s=t};m.prototype=l({},{next:function(t){var n=this._s;if(!g(n)){var r=n._o;try{var e=d(r.next);if(e)return e.call(r,t)}catch(t){try{b(n)}finally{throw t}}}},error:function(t){var n=this._s;if(g(n))throw t;var r=n._o;n._o=void 0;try{var e=d(r.error);if(!e)throw t;t=e.call(r,t)}catch(t){try{y(n)}finally{throw t}}return y(n),t},complete:function(t){var n=this._s;if(!g(n)){var r=n._o;n._o=void 0;try{var e=d(r.complete);t=e?e.call(r,t):void 0}catch(t){try{y(n)}finally{throw t}}return y(n),t}}});var S=function(t){s(this,S,"Observable","_f")._f=f(t)};l(S.prototype,{subscribe:function(t){return new x(t,this._f)},forEach:function(e){var i=this;return new(u.Promise||o.Promise)(function(t,n){f(e);var r=i.subscribe({next:function(t){try{return e(t)}catch(t){n(t),r.unsubscribe()}},error:n,complete:t})})}}),l(S,{from:function(t){var n="function"==typeof this?this:S,r=d(a(t)[c]);if(r){var e=a(r.call(t));return e.constructor===n?e:new n(function(t){return e.subscribe(t)})}return new n(function(n){var r=!1;return i(function(){if(!r){try{if(v(t,!1,function(t){if(n.next(t),r)return p})===p)return}catch(t){if(r)throw t;return void n.error(t)}n.complete()}}),function(){r=!0}})},of:function(){for(var t=0,n=arguments.length,e=new Array(n);t<n;)e[t]=arguments[t++];return new("function"==typeof this?this:S)(function(n){var r=!1;return i(function(){if(!r){for(var t=0;t<e.length;++t)if(n.next(e[t]),r)return;n.complete()}}),function(){r=!0}})}}),h(S.prototype,c,function(){return this}),e(e.G,{Observable:S}),r(77)("Observable")},function(t,n,r){"use strict";var e=r(1),i=r(46),o=r(3),u=r(119),c=r(191);e(e.P+e.R,"Promise",{finally:function(n){var r=u(this,i.Promise||o.Promise),t="function"==typeof n;return this.then(t?function(t){return c(r,n()).then(function(){return t})}:n,t?function(t){return c(r,n()).then(function(){throw t})}:n)}})},function(t,n,r){"use strict";var e=r(1),i=r(151),o=r(190);e(e.S,"Promise",{try:function(t){var n=i.f(this),r=o(t);return(r.e?n.reject:n.resolve)(r.v),n.promise}})},function(t,n,r){var e=r(56),i=r(2),o=e.key,u=e.set;e.exp({defineMetadata:function(t,n,r,e){u(t,n,i(r),o(e))}})},function(t,n,r){var e=r(56),o=r(2),u=e.key,c=e.map,f=e.store;e.exp({deleteMetadata:function(t,n){var r=arguments.length<3?void 0:u(arguments[2]),e=c(o(n),r,!1);if(void 0===e||!e.delete(t))return!1;if(e.size)return!0;var i=f.get(n);return i.delete(r),!!i.size||f.delete(n)}})},function(t,n,r){var o=r(199),u=r(167),e=r(56),i=r(2),c=r(32),f=e.keys,a=e.key,s=function(t,n){var r=f(t,n),e=c(t);if(null===e)return r;var i=s(e,n);return i.length?r.length?u(new o(r.concat(i))):i:r};e.exp({getMetadataKeys:function(t){return s(i(t),arguments.length<2?void 0:a(arguments[1]))}})},function(t,n,r){var e=r(56),i=r(2),o=r(32),u=e.has,c=e.get,f=e.key,a=function(t,n,r){if(u(t,n,r))return c(t,n,r);var e=o(n);return null!==e?a(t,e,r):void 0};e.exp({getMetadata:function(t,n){return a(t,i(n),arguments.length<3?void 0:f(arguments[2]))}})},function(t,n,r){var e=r(56),i=r(2),o=e.keys,u=e.key;e.exp({getOwnMetadataKeys:function(t){return o(i(t),arguments.length<2?void 0:u(arguments[1]))}})},function(t,n,r){var e=r(56),i=r(2),o=e.get,u=e.key;e.exp({getOwnMetadata:function(t,n){return o(t,i(n),arguments.length<3?void 0:u(arguments[2]))}})},function(t,n,r){var e=r(56),i=r(2),o=r(32),u=e.has,c=e.key,f=function(t,n,r){if(u(t,n,r))return!0;var e=o(n);return null!==e&&f(t,e,r)};e.exp({hasMetadata:function(t,n){return f(t,i(n),arguments.length<3?void 0:c(arguments[2]))}})},function(t,n,r){var e=r(56),i=r(2),o=e.has,u=e.key;e.exp({hasOwnMetadata:function(t,n){return o(t,i(n),arguments.length<3?void 0:u(arguments[2]))}})},function(t,n,r){var e=r(56),i=r(2),o=r(21),u=e.key,c=e.set;e.exp({metadata:function(r,e){return function(t,n){c(r,e,(void 0!==n?i:o)(t),u(n))}}})},function(t,n,r){r(129)("Set")},function(t,n,r){r(130)("Set")},function(t,n,r){var e=r(1);e(e.P+e.R,"Set",{toJSON:r(171)("Set")})},function(t,n,r){"use strict";var e=r(1),i=r(131)(!0);e(e.P,"String",{at:function(t){return i(this,t)}})},function(t,n,r){"use strict";var e=r(1),i=r(51),o=r(8),u=r(124),c=r(115),f=RegExp.prototype,a=function(t,n){this._r=t,this._s=n};r(146)(a,"RegExp String",function(){var t=this._r.exec(this._s);return{value:t,done:null===t}}),e(e.P,"String",{matchAll:function(t){if(i(this),!u(t))throw TypeError(t+" is not a regexp!");var n=String(this),r="flags"in f?String(t.flags):c.call(t),e=new RegExp(t.source,~r.indexOf("g")?r:"g"+r);return e.lastIndex=o(t.lastIndex),new a(e,n)}})},function(t,n,r){"use strict";var e=r(1),i=r(193),o=r(133),u=/Version\/10\.\d+(\.\d+)?( Mobile\/\w+)? Safari\//.test(o);e(e.P+e.F*u,"String",{padEnd:function(t){return i(this,t,1<arguments.length?arguments[1]:void 0,!1)}})},function(t,n,r){"use strict";var e=r(1),i=r(193),o=r(133),u=/Version\/10\.\d+(\.\d+)?( Mobile\/\w+)? Safari\//.test(o);e(e.P+e.F*u,"String",{padStart:function(t){return i(this,t,1<arguments.length?arguments[1]:void 0,!0)}})},function(t,n,r){"use strict";r(84)("trimLeft",function(t){return function(){return t(this,1)}},"trimStart")},function(t,n,r){"use strict";r(84)("trimRight",function(t){return function(){return t(this,2)}},"trimEnd")},function(t,n,r){r(160)("asyncIterator")},function(t,n,r){r(160)("observable")},function(t,n,r){var e=r(1);e(e.S,"System",{global:r(3)})},function(t,n,r){r(129)("WeakMap")},function(t,n,r){r(130)("WeakMap")},function(t,n,r){r(129)("WeakSet")},function(t,n,r){r(130)("WeakSet")},function(t,n,r){for(var e=r(162),i=r(74),o=r(27),u=r(3),c=r(26),f=r(82),a=r(7),s=a("iterator"),l=a("toStringTag"),h=f.Array,v={CSSRuleList:!0,CSSStyleDeclaration:!1,CSSValueList:!1,ClientRectList:!1,DOMRectList:!1,DOMStringList:!1,DOMTokenList:!0,DataTransferItemList:!1,FileList:!1,HTMLAllCollection:!1,HTMLCollection:!1,HTMLFormElement:!1,HTMLSelectElement:!1,MediaList:!0,MimeTypeArray:!1,NamedNodeMap:!1,NodeList:!0,PaintRequestList:!1,Plugin:!1,PluginArray:!1,SVGLengthList:!1,SVGNumberList:!1,SVGPathSegList:!1,SVGPointList:!1,SVGStringList:!1,SVGTransformList:!1,SourceBufferList:!1,StyleSheetList:!0,TextTrackCueList:!1,TextTrackList:!1,TouchList:!1},p=i(v),d=0;d<p.length;d++){var y,g=p[d],b=v[g],x=u[g],m=x&&x.prototype;if(m&&(m[s]||c(m,s,h),m[l]||c(m,l,g),f[g]=h,b))for(y in e)m[y]||o(m,y,e[y],!0)}},function(t,n,r){var e=r(1),i=r(158);e(e.G+e.B,{setImmediate:i.set,clearImmediate:i.clear})},function(t,n,r){var e=r(3),i=r(1),o=r(133),u=[].slice,c=/MSIE .\./.test(o),f=function(i){return function(t,n){var r=2<arguments.length,e=!!r&&u.call(arguments,2);return i(r?function(){("function"==typeof t?t:Function(t)).apply(this,e)}:t,n)}};i(i.G+i.B+i.F*c,{setTimeout:f(e.setTimeout),setInterval:f(e.setInterval)})},function(t,n,r){r(341),r(280),r(282),r(281),r(284),r(286),r(291),r(285),r(283),r(293),r(292),r(288),r(289),r(287),r(279),r(290),r(294),r(295),r(247),r(249),r(248),r(297),r(296),r(267),r(277),r(278),r(268),r(269),r(270),r(271),r(272),r(273),r(274),r(275),r(276),r(250),r(251),r(252),r(253),r(254),r(255),r(256),r(257),r(258),r(259),r(260),r(261),r(262),r(263),r(264),r(265),r(266),r(328),r(333),r(340),r(331),r(323),r(324),r(329),r(334),r(336),r(319),r(320),r(321),r(322),r(325),r(326),r(327),r(330),r(332),r(335),r(337),r(338),r(339),r(242),r(244),r(243),r(246),r(245),r(231),r(229),r(235),r(232),r(238),r(240),r(228),r(234),r(225),r(239),r(223),r(237),r(236),r(230),r(233),r(222),r(224),r(227),r(226),r(241),r(162),r(313),r(197),r(318),r(198),r(314),r(315),r(316),r(317),r(298),r(196),r(199),r(200),r(353),r(342),r(343),r(348),r(351),r(352),r(346),r(349),r(347),r(350),r(344),r(345),r(299),r(300),r(301),r(302),r(303),r(306),r(304),r(305),r(307),r(308),r(309),r(310),r(312),r(311),r(356),r(354),r(355),r(397),r(400),r(399),r(401),r(402),r(398),r(403),r(404),r(378),r(381),r(377),r(375),r(376),r(379),r(380),r(362),r(396),r(361),r(395),r(407),r(409),r(360),r(394),r(406),r(408),r(359),r(405),r(358),r(363),r(364),r(365),r(366),r(367),r(369),r(368),r(370),r(371),r(372),r(374),r(373),r(383),r(384),r(385),r(386),r(388),r(387),r(390),r(389),r(391),r(392),r(393),r(357),r(382),r(412),r(411),r(410),t.exports=r(46)},function(t,n){t.exports=function(t,n){if("string"==typeof n)return t.insertAdjacentHTML("afterend",n);var r=t.nextSibling;return r?t.parentNode.insertBefore(n,r):t.parentNode.appendChild(n)}}])</script><script src="/./main.a5fda8.js"></script><script>!function(){var e,t;e="/slider.27463f.js",t=document.createElement("script"),document.getElementsByTagName("body")[0].appendChild(t),t.setAttribute("src",e)}()</script>



<!--  -->

<script>
  /* 标签页标题切换 */
  var originTitle = document.title;
  var titleTime;
  document.addEventListener("visibilitychange", function () {
    if (document.hidden) {
      document.title = "(つェ⊂) 我藏好了哦~ " + originTitle;
      clearTimeout(titleTime);
    } else {
      document.title = "(*´∇｀*) 被你发现啦~ " + originTitle;
      titleTime = setTimeout(function () {
        document.title = originTitle;
      }, 2000);
    }
  });
</script>


    
<div class="tools-col" q-class="show:isShow,hide:isShow|isFalse" q-on="click:stop(e)">
  <div class="tools-nav header-menu">
    
    
      
      
      
    
      
      
      
    
      
      
      
    
    

    <ul style="width: 70%">
    
    
      
      <li style="width: 33.333333333333336%" q-on="click: openSlider(e, 'innerArchive')"><a href="javascript:void(0)" q-class="active:innerArchive">所有文章</a></li>
      
        
      
      <li style="width: 33.333333333333336%" q-on="click: openSlider(e, 'friends')"><a href="javascript:void(0)" q-class="active:friends">友链</a></li>
      
        
      
      <li style="width: 33.333333333333336%" q-on="click: openSlider(e, 'aboutme')"><a href="javascript:void(0)" q-class="active:aboutme">关于我</a></li>
      
        
    </ul>
  </div>
  <div class="tools-wrap">
    
    	<section class="tools-section tools-section-all" q-show="innerArchive">
        <div class="search-wrap">
          <input class="search-ipt" q-model="search" type="text" placeholder="find something…">
          <i class="icon-search icon" q-show="search|isEmptyStr"></i>
          <i class="icon-close icon" q-show="search|isNotEmptyStr" q-on="click:clearChose(e)"></i>
        </div>
        <div class="widget tagcloud search-tag">
          <p class="search-tag-wording">tag:</p>
          <label class="search-switch">
            <input type="checkbox" q-on="click:toggleTag(e)" q-attr="checked:showTags">
          </label>
          <ul class="article-tag-list" q-show="showTags">
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color3">BioNews</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color2">BioLearning</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color5">文献分析报告：干细胞基因编辑</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color2">BioProtocol</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color1">Paper</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color1">计算机随笔</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color3">BioInfo</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color4">Bio-News</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color3">多组学数据分析</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color2">类器官之纤毛</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color2">类器官之肝脏</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color4">转座子</a>
              </li>
             
              <li class="article-tag-list-item">
                <a href="javascript:void(0)" class="js-tag color5">类器官之胚胎干细胞</a>
              </li>
            
            <div class="clearfix"></div>
          </ul>
        </div>
        <ul class="search-ul">
          <p q-show="jsonFail" style="padding: 20px; font-size: 12px;">
            缺失模块。<br/>1、请确保node版本大于6.2<br/>2、在博客根目录（注意不是yilia-plus根目录）执行以下命令：<br/> npm i hexo-generator-json-content --save<br/><br/>
            3、在根目录_config.yml里添加配置：
<pre style="font-size: 12px;" q-show="jsonFail">
  jsonContent:
    meta: false
    pages: false
    posts:
      title: true
      date: true
      path: true
      text: false
      raw: false
      content: false
      slug: false
      updated: false
      comments: false
      link: false
      permalink: false
      excerpt: false
      categories: false
      tags: true
</pre>
          </p>
          <li class="search-li" q-repeat="items" q-show="isShow">
            <a q-attr="href:path|urlformat" class="search-title"><i class="icon-quo-left icon"></i><span q-text="title"></span></a>
            <p class="search-time">
              <i class="icon-calendar icon"></i>
              <span q-text="date|dateformat"></span>
            </p>
            <p class="search-tag">
              <i class="icon-price-tags icon"></i>
              <span q-repeat="tags" q-on="click:choseTag(e, name)" q-text="name|tagformat"></span>
            </p>
          </li>
        </ul>
    	</section>
    

    
    	<section class="tools-section tools-section-friends" q-show="friends">
  		
        <ul class="search-ul">
          
            <li class="search-li">
              <a href="https://zhousiwei.gitee.io/ibooks/" target="_blank" 
                class="search-title" title="记录工作和学习过程中的笔记：Java、前端开发、Hexo博客、聚合支付、Linux笔记、ElasticSearch、ELK日志分析">
                
                  <img src="https://zhousiwei.gitee.io/ibooks/favicon.ico">
                
                技术笔记
              </a>
            </li>
          
            <li class="search-li">
              <a href="https://github.com/JoeyBling" target="_blank" 
                class="search-title" >
                
                  <i class="icon-link icon"></i>
                
                GitHub
              </a>
            </li>
          
            <li class="search-li">
              <a href="https://gitee.com/zhousiwei" target="_blank" 
                class="search-title" >
                
                  <i class="icon-link icon"></i>
                
                码云
              </a>
            </li>
          
            <li class="search-li">
              <a href="https://www.jianshu.com/u/02cbf31a043a" target="_blank" 
                class="search-title" >
                
                  <i class="icon-link icon"></i>
                
                简书
              </a>
            </li>
          
            <li class="search-li">
              <a href="https://blog.csdn.net/qq_30930805" target="_blank" 
                class="search-title" >
                
                  <i class="icon-link icon"></i>
                
                CSDN
              </a>
            </li>
          
        </ul>
  		
    	</section>
    

    
    	<section class="tools-section tools-section-me" q-show="aboutme">
        
          
  	  		<div class="aboutme-wrap" id="aboutme">主要涉及技术：<br>python R开发、开源爱好者、Linux<br><br>联系QQ:17769680069<br><br>志存高远<br><br>心怀善意<br></div>
  	  	
    	</section>
    
  </div>
  
</div>
    <!-- Root element of PhotoSwipe. Must have class pswp. -->
<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

    <!-- Background of PhotoSwipe. 
         It's a separate element as animating opacity is faster than rgba(). -->
    <div class="pswp__bg"></div>

    <!-- Slides wrapper with overflow:hidden. -->
    <div class="pswp__scroll-wrap">

        <!-- Container that holds slides. 
            PhotoSwipe keeps only 3 of them in the DOM to save memory.
            Don't modify these 3 pswp__item elements, data is added later on. -->
        <div class="pswp__container">
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
        </div>

        <!-- Default (PhotoSwipeUI_Default) interface on top of sliding area. Can be changed. -->
        <div class="pswp__ui pswp__ui--hidden">

            <div class="pswp__top-bar">

                <!--  Controls are self-explanatory. Order can be changed. -->

                <div class="pswp__counter"></div>

                <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>

                <button class="pswp__button pswp__button--share" style="display:none" title="Share"></button>

                <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>

                <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>

                <!-- Preloader demo http://codepen.io/dimsemenov/pen/yyBWoR -->
                <!-- element will get class pswp__preloader--active when preloader is running -->
                <div class="pswp__preloader">
                    <div class="pswp__preloader__icn">
                      <div class="pswp__preloader__cut">
                        <div class="pswp__preloader__donut"></div>
                      </div>
                    </div>
                </div>
            </div>

            <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
                <div class="pswp__share-tooltip"></div> 
            </div>

            <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
            </button>

            <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
            </button>

            <div class="pswp__caption">
                <div class="pswp__caption__center"></div>
            </div>

        </div>

    </div>

</div>
  </div>

  
  
<script type="text/javascript" src="/plugins/activate-power-mode/activate-power-mode.js"></script>
<script>
  POWERMODE.colorful = true; // make power mode colorful
  POWERMODE.shake = false; // turn off shake
  document.body.addEventListener('input', POWERMODE);
</script>

  
  <!-- <script async type="text/javascript" size="90" alpha="0.2" zIndex="0" src="/plugins/ribbon.js/ribbon.min.js"></script> -->
  
  
    <script type="text/javascript" src="/plugins/live2d-widget.js/L2Dwidget.min.js"></script>
<script type="text/javascript">
  /**
   * 完整配置请参考：https://l2dwidget.js.org/docs/class/src/index.js~L2Dwidget.html#instance-method-init
   */
  var pluginModelPath = "assets/", // 模型的文件路径，无需更改
    pluginRootPath = "/live2d_models/"; // 插件的路径，无需更改
  //特定的模型json文件名称
  var modelPathJson = {
    "epsilon2_1": "Epsilon2.1",
    "gf": "Gantzert_Felixander",
    "haru/01": "haru01",
    "haru/02": "haru02",
    "nietzsche": "nietzche"
  }

  
    var modelName = "wanko";
  

  var modelJsonPath = "";
  if (modelPathJson[modelName]) {
    modelJsonPath = pluginRootPath + modelName + "/" + pluginModelPath + modelPathJson[modelName] +
      ".model.json";
  } else {
    modelJsonPath = pluginRootPath + modelName + "/" + pluginModelPath + modelName + ".model.json";
  }
  L2Dwidget.init({
    model: {
      jsonPath: modelJsonPath
    },
    display: {
       position: "right", 
       width: 145, 
       height: 315, 
       hOffset: 50, 
      
    },
    mobile: {
       show: false, 
       scale: 0.6 
    },
    react: {
       opacity: 0.7 
    }
  })

</script>

  
  
  <script type="text/javascript" src="/lib/snow.js"></script>
  <script type="text/javascript" src="/lib/jquery-2.1.4.min.js"></script>
  <script>
    snow.down();
    $(window).resize(function() {
      $("canvas").css("z-index","500").remove();
      snow.down();
    });
  </script>
  
</body>

</html>
